{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2997e421",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import dask.dataframe as dd\n",
    "import pandas as pd \n",
    "pd.get_option(\"display.max_columns\", None)\n",
    "pd.options.display.max_columns = None\n",
    "import numpy as np\n",
    "import glob\n",
    "import json\n",
    "from IPython.display import HTML\n",
    "os.chdir(\"/mnt/BioHome/jreyna/jreyna/projects/dchallenge/\")\n",
    "outdir = 'results/main/sgls/combined/'\n",
    "os.makedirs(outdir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "54972544",
   "metadata": {},
   "outputs": [],
   "source": [
    "major_cols = [\n",
    " 'gwas_source',\n",
    " 'ge_source',\n",
    " 'loop_source',\n",
    " 'sid',\n",
    " 'rsid',\n",
    " 'geneid',\n",
    " 'gene_name',\n",
    " 'chrom',\n",
    " 'snp_pos',\n",
    " 'tss_start',\n",
    " 'tss_end',\n",
    " 'is_eqtl_pair',\n",
    " 'is_coloc_pair',\n",
    " 'is_closest_gene',\n",
    " 'has_fithichip_loop',\n",
    " 'eqtl_pval',\n",
    " 'eqtl_fdr',\n",
    " 'dist',\n",
    " 'ppH4',\n",
    " 'gene_start',\n",
    " 'gene_end',\n",
    " 'ref',\n",
    " 'alt',\n",
    " 'AC',\n",
    " 'AF',\n",
    " 'AN',\n",
    " 'gwas_pval',\n",
    " 'gene_strand',\n",
    " 'eqtl_source']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e09c3e78",
   "metadata": {},
   "source": [
    "## Loading GENCODE information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a7eddbea",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_info = pd.read_table('results/refs/gencode/v30/gencode.v30.annotation.bed', header=None)\n",
    "gene_dict = pd.concat([gene_info.iloc[:, [5,6]], gene_info.iloc[:, [6,5]]])\n",
    "#{sr[5]: sr[6] for i, sr in gene_info.iterrows()}\n",
    "gene_dict.drop_duplicates(inplace=True)\n",
    "gene_dict.set_index(5, inplace=True)\n",
    "gene_dict = gene_dict.squeeze()\n",
    "gene_dict = gene_dict.to_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "434c4499",
   "metadata": {},
   "source": [
    "## Checking the samplesheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c87d1274",
   "metadata": {},
   "outputs": [],
   "source": [
    "samplesheet = pd.read_table('config/sgl.samplesheet.tsv.v2')\n",
    "template = 'results/main/sgls/T1D_34012112_Gaulton/{}/{}'\n",
    "output_exists = []\n",
    "for i, sr in samplesheet.iterrows():\n",
    "    test = os.path.exists(template.format(sr.eqtl_db, sr.eqtl_origin, sr.loop_origin))\n",
    "    output_exists.append(test)\n",
    "samplesheet['analyzed'] = output_exists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2a1832ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "samplesheet.sort_values(['analyzed', 'eqtl_db', 'eqtl_origin', 'loop_origin'], \\\n",
    "                       ascending=[False, True, True, True], inplace=True)\n",
    "\n",
    "samplesheet.reset_index(drop=True, inplace=True)\n",
    "samplesheet.columns = ['GWAS Source', 'eQTL Source', 'GE Source', 'Loop Source', 'Analyzed?']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7a21a068",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GWAS Source</th>\n",
       "      <th>eQTL Source</th>\n",
       "      <th>GE Source</th>\n",
       "      <th>Loop Source</th>\n",
       "      <th>Analyzed?</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T1D_34012112_Gaulton</td>\n",
       "      <td>Quach_2016</td>\n",
       "      <td>monocyte_IAV</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "      <td>Quach_2016</td>\n",
       "      <td>monocyte_IAV</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>Quach_2016</td>\n",
       "      <td>monocyte_IAV</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T1D_25751624</td>\n",
       "      <td>Quach_2016</td>\n",
       "      <td>monocyte_IAV</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T1D_34012112_Gaulton</td>\n",
       "      <td>Quach_2016</td>\n",
       "      <td>monocyte_Pam3CSK4</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "      <td>Quach_2016</td>\n",
       "      <td>monocyte_Pam3CSK4</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>Quach_2016</td>\n",
       "      <td>monocyte_Pam3CSK4</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>T1D_25751624</td>\n",
       "      <td>Quach_2016</td>\n",
       "      <td>monocyte_Pam3CSK4</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>T1D_34012112_Gaulton</td>\n",
       "      <td>Quach_2016</td>\n",
       "      <td>monocyte_R848</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "      <td>Quach_2016</td>\n",
       "      <td>monocyte_R848</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>Quach_2016</td>\n",
       "      <td>monocyte_R848</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>T1D_25751624</td>\n",
       "      <td>Quach_2016</td>\n",
       "      <td>monocyte_R848</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>T1D_34012112_Gaulton</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>B-cell_naive</td>\n",
       "      <td>B-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>B-cell_naive</td>\n",
       "      <td>B-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>B-cell_naive</td>\n",
       "      <td>B-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T1D_25751624</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>B-cell_naive</td>\n",
       "      <td>B-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>T1D_34012112_Gaulton</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD4_T-cell_anti-CD3-CD28</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD4_T-cell_anti-CD3-CD28</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD4_T-cell_anti-CD3-CD28</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>T1D_25751624</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD4_T-cell_anti-CD3-CD28</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>T1D_34012112_Gaulton</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>T1D_25751624</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>T1D_34012112_Gaulton</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD8_T-cell_anti-CD3-CD28</td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD8_T-cell_anti-CD3-CD28</td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD8_T-cell_anti-CD3-CD28</td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>T1D_25751624</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD8_T-cell_anti-CD3-CD28</td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>T1D_34012112_Gaulton</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>T1D_25751624</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>T1D_34012112_Gaulton</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>NK-cell_naive</td>\n",
       "      <td>NK-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>NK-cell_naive</td>\n",
       "      <td>NK-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>NK-cell_naive</td>\n",
       "      <td>NK-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>T1D_25751624</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>NK-cell_naive</td>\n",
       "      <td>NK-cell_naive</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>T1D_34012112_Gaulton</td>\n",
       "      <td>Quach_2016</td>\n",
       "      <td>monocyte_LPS</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "      <td>Quach_2016</td>\n",
       "      <td>monocyte_LPS</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>Quach_2016</td>\n",
       "      <td>monocyte_LPS</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>T1D_25751624</td>\n",
       "      <td>Quach_2016</td>\n",
       "      <td>monocyte_LPS</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>T1D_34012112_Gaulton</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>monocyte_CD16_naive</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>monocyte_CD16_naive</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>monocyte_CD16_naive</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>T1D_25751624</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>monocyte_CD16_naive</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>T1D_34012112_Gaulton</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>T1D_25751624</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  GWAS Source     eQTL Source                 GE Source  \\\n",
       "0        T1D_34012112_Gaulton      Quach_2016              monocyte_IAV   \n",
       "1   T1D_34594039_GCST90018925      Quach_2016              monocyte_IAV   \n",
       "2                T1D_32005708      Quach_2016              monocyte_IAV   \n",
       "3                T1D_25751624      Quach_2016              monocyte_IAV   \n",
       "4        T1D_34012112_Gaulton      Quach_2016         monocyte_Pam3CSK4   \n",
       "5   T1D_34594039_GCST90018925      Quach_2016         monocyte_Pam3CSK4   \n",
       "6                T1D_32005708      Quach_2016         monocyte_Pam3CSK4   \n",
       "7                T1D_25751624      Quach_2016         monocyte_Pam3CSK4   \n",
       "8        T1D_34012112_Gaulton      Quach_2016             monocyte_R848   \n",
       "9   T1D_34594039_GCST90018925      Quach_2016             monocyte_R848   \n",
       "10               T1D_32005708      Quach_2016             monocyte_R848   \n",
       "11               T1D_25751624      Quach_2016             monocyte_R848   \n",
       "12       T1D_34012112_Gaulton  Schmiedel_2018              B-cell_naive   \n",
       "13  T1D_34594039_GCST90018925  Schmiedel_2018              B-cell_naive   \n",
       "14               T1D_32005708  Schmiedel_2018              B-cell_naive   \n",
       "15               T1D_25751624  Schmiedel_2018              B-cell_naive   \n",
       "16       T1D_34012112_Gaulton  Schmiedel_2018  CD4_T-cell_anti-CD3-CD28   \n",
       "17  T1D_34594039_GCST90018925  Schmiedel_2018  CD4_T-cell_anti-CD3-CD28   \n",
       "18               T1D_32005708  Schmiedel_2018  CD4_T-cell_anti-CD3-CD28   \n",
       "19               T1D_25751624  Schmiedel_2018  CD4_T-cell_anti-CD3-CD28   \n",
       "20       T1D_34012112_Gaulton  Schmiedel_2018          CD4_T-cell_naive   \n",
       "21  T1D_34594039_GCST90018925  Schmiedel_2018          CD4_T-cell_naive   \n",
       "22               T1D_32005708  Schmiedel_2018          CD4_T-cell_naive   \n",
       "23               T1D_25751624  Schmiedel_2018          CD4_T-cell_naive   \n",
       "24       T1D_34012112_Gaulton  Schmiedel_2018  CD8_T-cell_anti-CD3-CD28   \n",
       "25  T1D_34594039_GCST90018925  Schmiedel_2018  CD8_T-cell_anti-CD3-CD28   \n",
       "26               T1D_32005708  Schmiedel_2018  CD8_T-cell_anti-CD3-CD28   \n",
       "27               T1D_25751624  Schmiedel_2018  CD8_T-cell_anti-CD3-CD28   \n",
       "28       T1D_34012112_Gaulton  Schmiedel_2018          CD8_T-cell_naive   \n",
       "29  T1D_34594039_GCST90018925  Schmiedel_2018          CD8_T-cell_naive   \n",
       "30               T1D_32005708  Schmiedel_2018          CD8_T-cell_naive   \n",
       "31               T1D_25751624  Schmiedel_2018          CD8_T-cell_naive   \n",
       "32       T1D_34012112_Gaulton  Schmiedel_2018             NK-cell_naive   \n",
       "33  T1D_34594039_GCST90018925  Schmiedel_2018             NK-cell_naive   \n",
       "34               T1D_32005708  Schmiedel_2018             NK-cell_naive   \n",
       "35               T1D_25751624  Schmiedel_2018             NK-cell_naive   \n",
       "36       T1D_34012112_Gaulton      Quach_2016              monocyte_LPS   \n",
       "37  T1D_34594039_GCST90018925      Quach_2016              monocyte_LPS   \n",
       "38               T1D_32005708      Quach_2016              monocyte_LPS   \n",
       "39               T1D_25751624      Quach_2016              monocyte_LPS   \n",
       "40       T1D_34012112_Gaulton  Schmiedel_2018       monocyte_CD16_naive   \n",
       "41  T1D_34594039_GCST90018925  Schmiedel_2018       monocyte_CD16_naive   \n",
       "42               T1D_32005708  Schmiedel_2018       monocyte_CD16_naive   \n",
       "43               T1D_25751624  Schmiedel_2018       monocyte_CD16_naive   \n",
       "44       T1D_34012112_Gaulton  Schmiedel_2018            monocyte_naive   \n",
       "45  T1D_34594039_GCST90018925  Schmiedel_2018            monocyte_naive   \n",
       "46               T1D_32005708  Schmiedel_2018            monocyte_naive   \n",
       "47               T1D_25751624  Schmiedel_2018            monocyte_naive   \n",
       "\n",
       "         Loop Source  Analyzed?  \n",
       "0     monocyte_naive       True  \n",
       "1     monocyte_naive       True  \n",
       "2     monocyte_naive       True  \n",
       "3     monocyte_naive       True  \n",
       "4     monocyte_naive       True  \n",
       "5     monocyte_naive       True  \n",
       "6     monocyte_naive       True  \n",
       "7     monocyte_naive       True  \n",
       "8     monocyte_naive       True  \n",
       "9     monocyte_naive       True  \n",
       "10    monocyte_naive       True  \n",
       "11    monocyte_naive       True  \n",
       "12      B-cell_naive       True  \n",
       "13      B-cell_naive       True  \n",
       "14      B-cell_naive       True  \n",
       "15      B-cell_naive       True  \n",
       "16  CD4_T-cell_naive       True  \n",
       "17  CD4_T-cell_naive       True  \n",
       "18  CD4_T-cell_naive       True  \n",
       "19  CD4_T-cell_naive       True  \n",
       "20  CD4_T-cell_naive       True  \n",
       "21  CD4_T-cell_naive       True  \n",
       "22  CD4_T-cell_naive       True  \n",
       "23  CD4_T-cell_naive       True  \n",
       "24  CD8_T-cell_naive       True  \n",
       "25  CD8_T-cell_naive       True  \n",
       "26  CD8_T-cell_naive       True  \n",
       "27  CD8_T-cell_naive       True  \n",
       "28  CD8_T-cell_naive       True  \n",
       "29  CD8_T-cell_naive       True  \n",
       "30  CD8_T-cell_naive       True  \n",
       "31  CD8_T-cell_naive       True  \n",
       "32     NK-cell_naive       True  \n",
       "33     NK-cell_naive       True  \n",
       "34     NK-cell_naive       True  \n",
       "35     NK-cell_naive       True  \n",
       "36    monocyte_naive      False  \n",
       "37    monocyte_naive      False  \n",
       "38    monocyte_naive      False  \n",
       "39    monocyte_naive      False  \n",
       "40    monocyte_naive      False  \n",
       "41    monocyte_naive      False  \n",
       "42    monocyte_naive      False  \n",
       "43    monocyte_naive      False  \n",
       "44    monocyte_naive      False  \n",
       "45    monocyte_naive      False  \n",
       "46    monocyte_naive      False  \n",
       "47    monocyte_naive      False  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samplesheet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49e05840",
   "metadata": {},
   "source": [
    "## Combining the other datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "5d506079",
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading and concat all the data\n",
    "data = []\n",
    "sgls_annos = [x for x in glob.glob('results/main/sgls/*/*/*/*/master.tsv') if 'sgls_20220222' in x]\n",
    "sgls_annos = [x for x in glob.glob('results/main/sgls/*/*/*/*/master.tsv') if 'sgls_20220222' not in x]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "bef79137",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['results/main/sgls/T1D_32005708/Schmiedel_2018/CD4_T-cell_anti-CD3-CD28/CD4_T-cell_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_32005708/Schmiedel_2018/CD8_T-cell_naive/CD8_T-cell_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_32005708/Schmiedel_2018/NK-cell_naive/NK-cell_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_32005708/Schmiedel_2018/CD8_T-cell_anti-CD3-CD28/CD8_T-cell_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_32005708/Schmiedel_2018/B-cell_naive/B-cell_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_32005708/Schmiedel_2018/monocyte_CD16_naive/monocyte_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_32005708/Schmiedel_2018/monocyte_naive/monocyte_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_32005708/Quach_2016/monocyte_IAV/monocyte_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_25751624/Schmiedel_2018/CD4_T-cell_naive/CD4_T-cell_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_25751624/Schmiedel_2018/monocyte_naive/monocyte_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_25751624/Schmiedel_2018/CD4_T-cell_anti-CD3-CD28/CD4_T-cell_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_25751624/Schmiedel_2018/monocyte_CD16_naive/monocyte_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_25751624/Schmiedel_2018/CD8_T-cell_anti-CD3-CD28/CD8_T-cell_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_25751624/Schmiedel_2018/B-cell_naive/B-cell_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_25751624/Schmiedel_2018/NK-cell_naive/NK-cell_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_25751624/Quach_2016/monocyte_R848/monocyte_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_25751624/Quach_2016/monocyte_IAV/monocyte_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_25751624/Quach_2016/monocyte_LPS/monocyte_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_25751624/Quach_2016/monocyte_Pam3CSK4/monocyte_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_34012112_Gaulton/Schmiedel_2018/CD8_T-cell_naive/CD8_T-cell_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_34012112_Gaulton/Schmiedel_2018/CD8_T-cell_anti-CD3-CD28/CD8_T-cell_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_34012112_Gaulton/Schmiedel_2018/NK-cell_naive/NK-cell_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_34012112_Gaulton/Schmiedel_2018/CD4_T-cell_anti-CD3-CD28/CD4_T-cell_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_34012112_Gaulton/Schmiedel_2018/CD4_T-cell_naive/CD4_T-cell_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_34012112_Gaulton/Schmiedel_2018/B-cell_naive/B-cell_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_34012112_Gaulton/Quach_2016/monocyte_Pam3CSK4/monocyte_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_34012112_Gaulton/Quach_2016/monocyte_R848/monocyte_naive/master.tsv',\n",
       " 'results/main/sgls/T1D_34012112_Gaulton/Quach_2016/monocyte_IAV/monocyte_naive/master.tsv']"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sgls_annos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0b622e5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for fn in sgls_annos:\n",
    "        \n",
    "    gwas_source = fn.split('/')[3]\n",
    "    eqtl_source = fn.split('/')[4]\n",
    "    ge_source = fn.split('/')[5]\n",
    "    loop_source = fn.split('/')[6]    \n",
    "    \n",
    "    df = pd.read_table(fn, header=0)\n",
    "    df['gwas_source'] = gwas_source\n",
    "    df['eqtl_source'] = eqtl_source\n",
    "    df['ge_source'] = ge_source\n",
    "    df['loop_source'] = loop_source\n",
    "    \n",
    "    # combine the columns gene_id and geneid\n",
    "    #df.loc[~df['gene_id'].isna(), 'geneid'] = df.loc[~df['gene_id'].isna(), 'gene_id']\n",
    "    #df.loc[:, 'gene_name'] = df.loc[:, 'geneid'].apply(lambda x: gene_dict[x])\n",
    "    df = df.loc[(df.is_coloc_pair == 1) & (df.has_fithichip_loop == 1)]\n",
    "    \n",
    "    if df.shape[0] > 0: \n",
    "        data.append(df)\n",
    "        \n",
    "data = pd.concat(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "eead3300",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sid</th>\n",
       "      <th>rsid</th>\n",
       "      <th>geneid</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp_pos</th>\n",
       "      <th>tss_start</th>\n",
       "      <th>tss_end</th>\n",
       "      <th>is_eqtl_pair</th>\n",
       "      <th>is_coloc_pair</th>\n",
       "      <th>is_closest_gene</th>\n",
       "      <th>has_fithichip_loop</th>\n",
       "      <th>has_colocSNP_anchor</th>\n",
       "      <th>eqtl_pval</th>\n",
       "      <th>eqtl_beta</th>\n",
       "      <th>dist</th>\n",
       "      <th>ppH0</th>\n",
       "      <th>ppH1</th>\n",
       "      <th>ppH2</th>\n",
       "      <th>ppH3</th>\n",
       "      <th>ppH4</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>AC</th>\n",
       "      <th>AF</th>\n",
       "      <th>AN</th>\n",
       "      <th>gwas_slope</th>\n",
       "      <th>gwas_slope_se</th>\n",
       "      <th>gwas_pval</th>\n",
       "      <th>gene_start.1</th>\n",
       "      <th>gene_end.1</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>gwas_source</th>\n",
       "      <th>eqtl_source</th>\n",
       "      <th>ge_source</th>\n",
       "      <th>loop_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>10:6068912</td>\n",
       "      <td>rs7090530</td>\n",
       "      <td>ENSG00000213994</td>\n",
       "      <td>AL157395.1</td>\n",
       "      <td>chr10</td>\n",
       "      <td>6068912</td>\n",
       "      <td>6202692</td>\n",
       "      <td>6202693</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2.746830e-02</td>\n",
       "      <td>0.338513</td>\n",
       "      <td>133781.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.833528e-03</td>\n",
       "      <td>0.070616</td>\n",
       "      <td>0.924551</td>\n",
       "      <td>6197612</td>\n",
       "      <td>6202693</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>3112.0</td>\n",
       "      <td>0.621406</td>\n",
       "      <td>5008.0</td>\n",
       "      <td>0.185500</td>\n",
       "      <td>0.024500</td>\n",
       "      <td>4.026000e-14</td>\n",
       "      <td>6197612</td>\n",
       "      <td>6202693</td>\n",
       "      <td>-</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>21:42405613</td>\n",
       "      <td>rs9981624</td>\n",
       "      <td>ENSG00000160183</td>\n",
       "      <td>TMPRSS3</td>\n",
       "      <td>chr21</td>\n",
       "      <td>42405613</td>\n",
       "      <td>42396845</td>\n",
       "      <td>42396846</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6.365800e-11</td>\n",
       "      <td>0.870668</td>\n",
       "      <td>8767.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.053392e-02</td>\n",
       "      <td>0.120225</td>\n",
       "      <td>0.819241</td>\n",
       "      <td>42371837</td>\n",
       "      <td>42396846</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1175.0</td>\n",
       "      <td>0.234625</td>\n",
       "      <td>5008.0</td>\n",
       "      <td>0.152034</td>\n",
       "      <td>0.019004</td>\n",
       "      <td>1.244192e-15</td>\n",
       "      <td>42371837</td>\n",
       "      <td>42396846</td>\n",
       "      <td>-</td>\n",
       "      <td>T1D_25751624</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>2:203875100</td>\n",
       "      <td>rs231724</td>\n",
       "      <td>ENSG00000163599</td>\n",
       "      <td>CTLA4</td>\n",
       "      <td>chr2</td>\n",
       "      <td>203875100</td>\n",
       "      <td>203867770</td>\n",
       "      <td>203867771</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.846690e-02</td>\n",
       "      <td>-0.482733</td>\n",
       "      <td>7329.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.471615e-09</td>\n",
       "      <td>0.103521</td>\n",
       "      <td>0.896479</td>\n",
       "      <td>203867771</td>\n",
       "      <td>203873965</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2094.0</td>\n",
       "      <td>0.418131</td>\n",
       "      <td>5008.0</td>\n",
       "      <td>0.174961</td>\n",
       "      <td>0.018996</td>\n",
       "      <td>3.245297e-20</td>\n",
       "      <td>203867771</td>\n",
       "      <td>203873965</td>\n",
       "      <td>+</td>\n",
       "      <td>T1D_25751624</td>\n",
       "      <td>Schmiedel_2018</td>\n",
       "      <td>monocyte_CD16_naive</td>\n",
       "      <td>monocyte_naive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            sid       rsid           geneid   gene_name  chrom    snp_pos  \\\n",
       "42   10:6068912  rs7090530  ENSG00000213994  AL157395.1  chr10    6068912   \n",
       "27  21:42405613  rs9981624  ENSG00000160183     TMPRSS3  chr21   42405613   \n",
       "22  2:203875100   rs231724  ENSG00000163599       CTLA4   chr2  203875100   \n",
       "\n",
       "    tss_start    tss_end  is_eqtl_pair  is_coloc_pair  is_closest_gene  \\\n",
       "42    6202692    6202693             0              1                0   \n",
       "27   42396845   42396846             1              1                0   \n",
       "22  203867770  203867771             0              1                1   \n",
       "\n",
       "    has_fithichip_loop  has_colocSNP_anchor     eqtl_pval  eqtl_beta  \\\n",
       "42                   1                    0  2.746830e-02   0.338513   \n",
       "27                   1                    0  6.365800e-11   0.870668   \n",
       "22                   1                    0  1.846690e-02  -0.482733   \n",
       "\n",
       "        dist  ppH0  ppH1          ppH2      ppH3      ppH4  gene_start  \\\n",
       "42  133781.0   0.0   0.0  4.833528e-03  0.070616  0.924551     6197612   \n",
       "27    8767.0   0.0   0.0  6.053392e-02  0.120225  0.819241    42371837   \n",
       "22    7329.0   0.0   0.0  3.471615e-09  0.103521  0.896479   203867771   \n",
       "\n",
       "     gene_end ref alt      AC        AF      AN  gwas_slope  gwas_slope_se  \\\n",
       "42    6202693   C   A  3112.0  0.621406  5008.0    0.185500       0.024500   \n",
       "27   42396846   G   C  1175.0  0.234625  5008.0    0.152034       0.019004   \n",
       "22  203873965   A   G  2094.0  0.418131  5008.0    0.174961       0.018996   \n",
       "\n",
       "       gwas_pval  gene_start.1  gene_end.1 gene_strand   gwas_source  \\\n",
       "42  4.026000e-14       6197612     6202693           -  T1D_32005708   \n",
       "27  1.244192e-15      42371837    42396846           -  T1D_25751624   \n",
       "22  3.245297e-20     203867771   203873965           +  T1D_25751624   \n",
       "\n",
       "       eqtl_source            ge_source       loop_source  \n",
       "42  Schmiedel_2018     CD8_T-cell_naive  CD8_T-cell_naive  \n",
       "27  Schmiedel_2018     CD4_T-cell_naive  CD4_T-cell_naive  \n",
       "22  Schmiedel_2018  monocyte_CD16_naive    monocyte_naive  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bcd628a",
   "metadata": {},
   "source": [
    "### Checking the uniq set of genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "ac004bee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "      <th>genename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000213994</td>\n",
       "      <td>AL157395.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000160183</td>\n",
       "      <td>TMPRSS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000163599</td>\n",
       "      <td>CTLA4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            geneid    genename\n",
       "0  ENSG00000213994  AL157395.1\n",
       "1  ENSG00000160183     TMPRSS3\n",
       "2  ENSG00000163599       CTLA4"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_genes = data.geneid.unique()\n",
    "unique_genes = [x for x in unique_genes if type(x) == str]\n",
    "unique_gnames = [gene_dict[x] if x in gene_dict else x for x in unique_genes]\n",
    "unique_genes_df = pd.DataFrame([unique_genes, unique_gnames]).T\n",
    "unique_genes_df.columns = ['geneid', 'genename']\n",
    "unique_genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "169ab3d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find the unique SNPs\n",
    "uniq_snps_by_cells = data.groupby('ge_source').sid.nunique()\n",
    "uniq_snps_by_cells = uniq_snps_by_cells.to_frame()\n",
    "\n",
    "# find the unique genes \n",
    "uniq_genes_by_cells = data.groupby('ge_source').geneid.nunique()\n",
    "uniq_genes_by_cells = uniq_genes_by_cells.to_frame()\n",
    "\n",
    "# merge snps and genes\n",
    "uniq_counts_by_cells = pd.merge(uniq_snps_by_cells, uniq_genes_by_cells, left_index=True, right_index=True)\n",
    "uniq_counts_by_cells.columns = ['Number of Unique SNPs', 'Number of Unique Genes']\n",
    "uniq_counts_by_cells.index.name = 'Cell Line'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "c7067546",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the file\n",
    "excel_analysis = os.path.join(outdir, 'Unique_Counts_By_Cell_Line.xlsx')\n",
    "uniq_counts_by_cells.to_excel(excel_analysis, sheet_name='pieqtls')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "975c6e63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ge_source</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD4_T-cell_naive</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8_T-cell_naive</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>monocyte_CD16_naive</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     geneid\n",
       "ge_source                  \n",
       "CD4_T-cell_naive          1\n",
       "CD8_T-cell_naive          1\n",
       "monocyte_CD16_naive       1"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniq_genes_by_cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "71a378db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'results/main/sgls/combined/Unique_Counts_By_Cell_Line.xlsx'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "excel_analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49166bd3",
   "metadata": {},
   "source": [
    "#### Write the gene list as well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "39ab2750",
   "metadata": {},
   "outputs": [],
   "source": [
    "gh_list = sorted(unique_genes_df.geneid.unique().tolist())\n",
    "gh_fn = os.path.join(outdir, 'gene_list.txt')\n",
    "with open(gh_fn, 'w') as fw:\n",
    "    for x in gh_list:\n",
    "        fw.write('{}\\n'.format(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e0d245c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'results/main/sgls/combined/gene_list.txt'"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gh_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "753de50a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a24b7a41",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8e6a68a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fadd83cc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c34fe11f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d3a4a75",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30e4ffa2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cdd7393",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40b4ee4a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9e8134a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f26ddba0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28fc19fb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "19fe9be8",
   "metadata": {},
   "source": [
    "### Analyzing the Number of eQTLs, loops, colocaled SNP-Gene (per DataSet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "22fd9194",
   "metadata": {},
   "outputs": [
    {
     "ename": "MemoryError",
     "evalue": "Unable to allocate 1.42 GiB for an array with shape (28, 6788669) and data type float64",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mMemoryError\u001b[0m                               Traceback (most recent call last)",
      "Input \u001b[0;32mIn [14]\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0m cells_with_loops \u001b[38;5;241m=\u001b[39m \u001b[43mdata\u001b[49m\u001b[43m[\u001b[49m\u001b[43mdata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhas_fithichip_loop\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m==\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m      2\u001b[0m uniq_cells \u001b[38;5;241m=\u001b[39m cells_with_loops[[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124meqtl_source\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mge_source\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mloop_source\u001b[39m\u001b[38;5;124m'\u001b[39m]]\u001b[38;5;241m.\u001b[39mvalues\u001b[38;5;241m.\u001b[39mtolist()\n\u001b[1;32m      3\u001b[0m uniq_cells \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m([\u001b[38;5;28mtuple\u001b[39m(x) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m uniq_cells])\n",
      "File \u001b[0;32m~/software/anaconda3/hichip-db/lib/python3.10/site-packages/pandas/core/frame.py:3497\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3495\u001b[0m \u001b[38;5;66;03m# Do we have a (boolean) 1d indexer?\u001b[39;00m\n\u001b[1;32m   3496\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m com\u001b[38;5;241m.\u001b[39mis_bool_indexer(key):\n\u001b[0;32m-> 3497\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_bool_array\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3499\u001b[0m \u001b[38;5;66;03m# We are left with two options: a single key, and a collection of keys,\u001b[39;00m\n\u001b[1;32m   3500\u001b[0m \u001b[38;5;66;03m# We interpret tuples as collections only for non-MultiIndex\u001b[39;00m\n\u001b[1;32m   3501\u001b[0m is_single_key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28misinstance\u001b[39m(key, \u001b[38;5;28mtuple\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_list_like(key)\n",
      "File \u001b[0;32m~/software/anaconda3/hichip-db/lib/python3.10/site-packages/pandas/core/frame.py:3552\u001b[0m, in \u001b[0;36mDataFrame._getitem_bool_array\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3550\u001b[0m key \u001b[38;5;241m=\u001b[39m check_bool_indexer(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mindex, key)\n\u001b[1;32m   3551\u001b[0m indexer \u001b[38;5;241m=\u001b[39m key\u001b[38;5;241m.\u001b[39mnonzero()[\u001b[38;5;241m0\u001b[39m]\n\u001b[0;32m-> 3552\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_take_with_is_copy\u001b[49m\u001b[43m(\u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/software/anaconda3/hichip-db/lib/python3.10/site-packages/pandas/core/generic.py:3728\u001b[0m, in \u001b[0;36mNDFrame._take_with_is_copy\u001b[0;34m(self, indices, axis)\u001b[0m\n\u001b[1;32m   3720\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_take_with_is_copy\u001b[39m(\u001b[38;5;28mself\u001b[39m: NDFrameT, indices, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m NDFrameT:\n\u001b[1;32m   3721\u001b[0m     \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   3722\u001b[0m \u001b[38;5;124;03m    Internal version of the `take` method that sets the `_is_copy`\u001b[39;00m\n\u001b[1;32m   3723\u001b[0m \u001b[38;5;124;03m    attribute to keep track of the parent dataframe (using in indexing\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   3726\u001b[0m \u001b[38;5;124;03m    See the docstring of `take` for full explanation of the parameters.\u001b[39;00m\n\u001b[1;32m   3727\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 3728\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtake\u001b[49m\u001b[43m(\u001b[49m\u001b[43mindices\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindices\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3729\u001b[0m     \u001b[38;5;66;03m# Maybe set copy if we didn't actually change the index.\u001b[39;00m\n\u001b[1;32m   3730\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m result\u001b[38;5;241m.\u001b[39m_get_axis(axis)\u001b[38;5;241m.\u001b[39mequals(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_axis(axis)):\n",
      "File \u001b[0;32m~/software/anaconda3/hichip-db/lib/python3.10/site-packages/pandas/core/generic.py:3713\u001b[0m, in \u001b[0;36mNDFrame.take\u001b[0;34m(self, indices, axis, is_copy, **kwargs)\u001b[0m\n\u001b[1;32m   3704\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m   3705\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mis_copy is deprecated and will be removed in a future version. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   3706\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtake\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m always returns a copy, so there is no need to specify this.\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m   3707\u001b[0m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[1;32m   3708\u001b[0m         stacklevel\u001b[38;5;241m=\u001b[39mfind_stack_level(),\n\u001b[1;32m   3709\u001b[0m     )\n\u001b[1;32m   3711\u001b[0m nv\u001b[38;5;241m.\u001b[39mvalidate_take((), kwargs)\n\u001b[0;32m-> 3713\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_consolidate_inplace\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3715\u001b[0m new_data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_mgr\u001b[38;5;241m.\u001b[39mtake(\n\u001b[1;32m   3716\u001b[0m     indices, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_block_manager_axis(axis), verify\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m   3717\u001b[0m )\n\u001b[1;32m   3718\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_constructor(new_data)\u001b[38;5;241m.\u001b[39m__finalize__(\u001b[38;5;28mself\u001b[39m, method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtake\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/software/anaconda3/hichip-db/lib/python3.10/site-packages/pandas/core/generic.py:5661\u001b[0m, in \u001b[0;36mNDFrame._consolidate_inplace\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   5658\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mf\u001b[39m():\n\u001b[1;32m   5659\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_mgr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_mgr\u001b[38;5;241m.\u001b[39mconsolidate()\n\u001b[0;32m-> 5661\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_protect_consolidate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/software/anaconda3/hichip-db/lib/python3.10/site-packages/pandas/core/generic.py:5649\u001b[0m, in \u001b[0;36mNDFrame._protect_consolidate\u001b[0;34m(self, f)\u001b[0m\n\u001b[1;32m   5647\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m f()\n\u001b[1;32m   5648\u001b[0m blocks_before \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_mgr\u001b[38;5;241m.\u001b[39mblocks)\n\u001b[0;32m-> 5649\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[43mf\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   5650\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_mgr\u001b[38;5;241m.\u001b[39mblocks) \u001b[38;5;241m!=\u001b[39m blocks_before:\n\u001b[1;32m   5651\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_clear_item_cache()\n",
      "File \u001b[0;32m~/software/anaconda3/hichip-db/lib/python3.10/site-packages/pandas/core/generic.py:5659\u001b[0m, in \u001b[0;36mNDFrame._consolidate_inplace.<locals>.f\u001b[0;34m()\u001b[0m\n\u001b[1;32m   5658\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mf\u001b[39m():\n\u001b[0;32m-> 5659\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_mgr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_mgr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconsolidate\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/software/anaconda3/hichip-db/lib/python3.10/site-packages/pandas/core/internals/managers.py:631\u001b[0m, in \u001b[0;36mBaseBlockManager.consolidate\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    629\u001b[0m bm \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mtype\u001b[39m(\u001b[38;5;28mself\u001b[39m)(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mblocks, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxes, verify_integrity\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m    630\u001b[0m bm\u001b[38;5;241m.\u001b[39m_is_consolidated \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[0;32m--> 631\u001b[0m \u001b[43mbm\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_consolidate_inplace\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    632\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m bm\n",
      "File \u001b[0;32m~/software/anaconda3/hichip-db/lib/python3.10/site-packages/pandas/core/internals/managers.py:1685\u001b[0m, in \u001b[0;36mBlockManager._consolidate_inplace\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1683\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_consolidate_inplace\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1684\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mis_consolidated():\n\u001b[0;32m-> 1685\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mblocks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mtuple\u001b[39m(\u001b[43m_consolidate\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mblocks\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m   1686\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_is_consolidated \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m   1687\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_known_consolidated \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m~/software/anaconda3/hichip-db/lib/python3.10/site-packages/pandas/core/internals/managers.py:2084\u001b[0m, in \u001b[0;36m_consolidate\u001b[0;34m(blocks)\u001b[0m\n\u001b[1;32m   2082\u001b[0m new_blocks: \u001b[38;5;28mlist\u001b[39m[Block] \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m   2083\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m (_can_consolidate, dtype), group_blocks \u001b[38;5;129;01min\u001b[39;00m grouper:\n\u001b[0;32m-> 2084\u001b[0m     merged_blocks \u001b[38;5;241m=\u001b[39m \u001b[43m_merge_blocks\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   2085\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mlist\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mgroup_blocks\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcan_consolidate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m_can_consolidate\u001b[49m\n\u001b[1;32m   2086\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2087\u001b[0m     new_blocks \u001b[38;5;241m=\u001b[39m extend_blocks(merged_blocks, new_blocks)\n\u001b[1;32m   2088\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m new_blocks\n",
      "File \u001b[0;32m~/software/anaconda3/hichip-db/lib/python3.10/site-packages/pandas/core/internals/managers.py:2118\u001b[0m, in \u001b[0;36m_merge_blocks\u001b[0;34m(blocks, dtype, can_consolidate)\u001b[0m\n\u001b[1;32m   2115\u001b[0m     new_values \u001b[38;5;241m=\u001b[39m bvals2[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m.\u001b[39m_concat_same_type(bvals2, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)\n\u001b[1;32m   2117\u001b[0m argsort \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39margsort(new_mgr_locs)\n\u001b[0;32m-> 2118\u001b[0m new_values \u001b[38;5;241m=\u001b[39m \u001b[43mnew_values\u001b[49m\u001b[43m[\u001b[49m\u001b[43margsort\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m   2119\u001b[0m new_mgr_locs \u001b[38;5;241m=\u001b[39m new_mgr_locs[argsort]\n\u001b[1;32m   2121\u001b[0m bp \u001b[38;5;241m=\u001b[39m BlockPlacement(new_mgr_locs)\n",
      "\u001b[0;31mMemoryError\u001b[0m: Unable to allocate 1.42 GiB for an array with shape (28, 6788669) and data type float64"
     ]
    }
   ],
   "source": [
    "cells_with_loops = data[data.has_fithichip_loop == 1]\n",
    "uniq_cells = cells_with_loops[['eqtl_source', 'ge_source', 'loop_source']].values.tolist()\n",
    "uniq_cells = set([tuple(x) for x in uniq_cells])\n",
    "uniq_cells = list(uniq_cells)\n",
    "uniq_cells = pd.DataFrame(uniq_cells)\n",
    "uniq_cells.columns = ['eqtl_source', 'ge_source', 'loop_source']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7adb581",
   "metadata": {},
   "outputs": [],
   "source": [
    "uniq_cells.sort_values('ge_source')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f7a4d64",
   "metadata": {},
   "outputs": [],
   "source": [
    "# group by the dataset combos\n",
    "eqtl_ge_grps = data.groupby(['eqtl_source', 'ge_source', 'loop_source'])\n",
    "\n",
    "# calculate the number of sg pairs \n",
    "def count_uniq(x):\n",
    "    v = x[['sid', 'geneid']].values.tolist()\n",
    "    v = [tuple(x) for x in v]\n",
    "    v = set(v)\n",
    "    return(len(v))\n",
    "eqtl_ge_sg_pairs = eqtl_ge_grps.apply(count_uniq).to_frame()\n",
    "eqtl_ge_sg_pairs.columns = ['num_sg_pairs']\n",
    "\n",
    "# calculate the number of sg pairs with loops\n",
    "eqtl_ge_loops = eqtl_ge_grps.has_fithichip_loop.sum().to_frame()\n",
    "\n",
    "# calculate the number of sg pairs with significant eQTL signal\n",
    "eqtl_ge_eqtl = eqtl_ge_grps.is_eqtl_pair.sum().to_frame()\n",
    "\n",
    "# calculate the number of sg pairs with significant coloc \n",
    "eqtl_ge_coloc = eqtl_ge_grps.is_coloc_pair.sum().to_frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8b61d4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_ge_master = eqtl_ge_sg_pairs.merge(eqtl_ge_coloc, left_index=True, right_index=True)\n",
    "eqtl_ge_master = eqtl_ge_master.merge(eqtl_ge_eqtl, left_index=True, right_index=True)\n",
    "eqtl_ge_master = eqtl_ge_master.merge(eqtl_ge_loops, left_index=True, right_index=True)\n",
    "\n",
    "eqtl_ge_master.columns = ['Number of\\\\nSNP-Gene Pairs',\n",
    "                          'Number of\\\\nColoc Pairs',\n",
    "                          'Number of\\\\neQTL Pairs',\n",
    "                          'Number of\\\\nPairs with a Loop']\n",
    "eqtl_ge_master.index.names = ['eQTL Source', 'GE Source', 'Loop Source']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f0ed392",
   "metadata": {},
   "outputs": [],
   "source": [
    "HTML(eqtl_ge_master.to_html().replace('\\\\n', '<br>'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "759269d9",
   "metadata": {},
   "source": [
    "#### Extract all pairs with a significant eQTL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cbc1c8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sg_with_eqtl = data.loc[(data.is_eqtl_pair == 1), major_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57cc5b6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('There are {} SNP-Gene pairs with an eQTL.'.format(len(sg_with_eqtl)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c515cfae",
   "metadata": {},
   "outputs": [],
   "source": [
    "sg_with_eqtl"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10a08a57",
   "metadata": {},
   "source": [
    "#### Extract all pairs with a colocalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c551040a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sg_with_coloc = data.loc[(data.is_coloc_pair == 1), major_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9a43239",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('There are {} SNP-Gene pairs with a coloc.'.format(len(sg_with_coloc)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3af25581",
   "metadata": {},
   "outputs": [],
   "source": [
    "sg_with_coloc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d600b3f",
   "metadata": {},
   "source": [
    "#### Extract all pairs with a FitHiChIP Loop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37116dc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "sg_with_loops = data.loc[(data.has_fithichip_loop == 1), major_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fefc9c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('There are {} SNP-Gene pairs with a loop.'.format(len(sg_with_loops)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a072cc14",
   "metadata": {},
   "outputs": [],
   "source": [
    "sg_with_loops"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1c964b2",
   "metadata": {},
   "source": [
    "## Finalizing the super master table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec786699",
   "metadata": {},
   "outputs": [],
   "source": [
    "#data = data.loc[:, new_order]\n",
    "data.drop_duplicates(subset=['sid', 'geneid', 'eqtl_source', 'ge_source', 'loop_source'], inplace=True)\n",
    "master_fn = os.path.join(outdir, 'super_master.snp_gene_loop.analysis.tsv')\n",
    "data[major_cols].to_csv(master_fn, sep='\\t', index=False, na_rep='nan')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74e3a543",
   "metadata": {},
   "outputs": [],
   "source": [
    "master_fn = os.path.join(outdir, 'super_master.snp_gene_loop.analysis.xlsx')\n",
    "xdata = data.sort_values(['rsid', 'gwas_source','eqtl_source', 'ge_source', 'loop_source', 'geneid'])\n",
    "xdata = xdata[major_cols].set_index(['rsid', 'gwas_source','eqtl_source', 'ge_source', 'loop_source', 'geneid'])\n",
    "xdata.to_excel(master_fn, na_rep='nan')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a601330",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06ac6070",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26396b70",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d25e9dfb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cef6821",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5aa157aa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0e28fa3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbbfe0dc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2aae39f5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28e6dc10",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3db32419",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a9c15bc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "547a7996",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fd019d3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71ef4cf0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2b3bad5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ff275eb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "HiChIP DB",
   "language": "python",
   "name": "hichip-db"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
