{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7b5ad180",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import glob\n",
    "import pybedtools as pbt \n",
    "from IPython.display import HTML\n",
    "\n",
    "pbt.set_bedtools_path('/mnt/BioHome/jreyna/software/anaconda3/envs/hic_tls/bin/')\n",
    "os.chdir('/mnt/BioHome/jreyna/jreyna/projects/dchallenge/')\n",
    "\n",
    "gsizes = 'results/refs/hg19/hg19.chrom.sizes'\n",
    "res = 10000\n",
    "\n",
    "# make the directory to save our data\n",
    "outdir = 'results/main/gwas_pieqtls/2021_chiou_et_al/2021_chandra_et_al/'\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "bedpe_cols = ['chrA', 'startA', 'endA', 'chrB', 'startB', 'endB']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25ec296e",
   "metadata": {},
   "source": [
    "## Load Fine Mapped GWAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "46298a0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas = 'results/main/finemapping/T1D_34012112_Gaulton/GRCh37/offset_1000000/Summary/sss/FINAL_top_snp_credible_set.txt'\n",
    "gwas_df = pd.read_table(gwas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "60271124",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_66823/167340614.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gwas_bed['start'] = gwas_bed['start'] - 1\n"
     ]
    }
   ],
   "source": [
    "gwas_bed = gwas_df.iloc[:, [4,5,5]]\n",
    "gwas_bed.columns = ['chrom', 'start', 'end']\n",
    "gwas_bed['start'] = gwas_bed['start'] - 1\n",
    "gwas_pbt = pbt.BedTool.from_dataframe(gwas_bed)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4b1e34a",
   "metadata": {},
   "source": [
    "## Intersect Fine Mapped GWAS and loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b49da1dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_seB(x): \n",
    "    s,e = x.split(':')[1].split('-')\n",
    "    e = e.split(',')[0]\n",
    "    return((s,e))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6cd23b57",
   "metadata": {},
   "outputs": [],
   "source": [
    "loops = 'results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/'\n",
    "loops += '*/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz'\n",
    "loops = glob.glob(loops)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1c387b45",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/TREGMEM/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n",
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/TH2/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n",
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/NCM/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n",
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/CD4N/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n",
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/TREGNAIVE/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n",
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/TH1/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n",
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/CD8N/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n",
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/THSTAR/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n",
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/NB/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n",
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/TFH/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n",
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/CM/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n",
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/NK/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n",
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/TH17/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n"
     ]
    }
   ],
   "source": [
    "loop_data = []\n",
    "for loop in loops:\n",
    "    print(loop)\n",
    "    \n",
    "    cline = loop.split('/')[5]\n",
    "    df = pd.read_table(loop, header=None)    \n",
    "    df.columns = ['chrom', 'startA', 'endA', 'seB', 'e1', 'e2']\n",
    "    \n",
    "    df['chrom'] = df['chrom'].str.replace('chr', '')\n",
    "    \n",
    "    df['startB'], df['endB'] = zip(*df['seB'].apply(parse_seB))\n",
    "    df['startB'] = df['startB'].astype(int)\n",
    "    \n",
    "    df['startA'] = df['startA'] + 1 \n",
    "    df['endA'] = df['startA'] + res\n",
    "\n",
    "    df['startB'] = df['startB'] + 1 \n",
    "    df['endB'] = df['startB'] + res\n",
    "    \n",
    "    # re-organize the data into bedpe-like\n",
    "    df = df.iloc[:, [0,1,2,0,6,7,3,4,5]]\n",
    "\n",
    "    # add cell type\n",
    "    df['cline'] = cline\n",
    "\n",
    "    loop_data.append(df) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "541b020f",
   "metadata": {},
   "outputs": [],
   "source": [
    "loop_df = pd.concat(loop_data)\n",
    "loop_bed = loop_df.iloc[:, [0,1,2,3,4,5,-1]]\n",
    "loop_pbt = pbt.BedTool.from_dataframe(loop_bed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c7b3cadc",
   "metadata": {},
   "outputs": [],
   "source": [
    "intersect_pbt = loop_pbt.pair_to_bed(gwas_pbt, type='either')\n",
    "#intersect_pbt = loop_pbt.pair_to_bed(gwas_pbt.slop(b=100000, g=gsizes), type='either')\n",
    "gwas_hichip = intersect_pbt.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b850dcad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chrom               10\n",
       "start         89622500\n",
       "end           89632500\n",
       "name                10\n",
       "score         90312500\n",
       "strand        90322500\n",
       "thickStart     TREGMEM\n",
       "thickEnd            10\n",
       "itemRgb       90319219\n",
       "blockCount    90319220\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_hichip.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1ff50d07",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_hichip = gwas_hichip.iloc[:, [7,8,9,0,1,2,3,4,5,6]]\n",
    "loop_cols = ['{}_loop'.format(x) for x in bedpe_cols]\n",
    "gwas_hichip.columns = ['chr_snp', 'start_snp', 'end_snp'] + loop_cols + ['cline_loop']\n",
    "gwas_hichip = gwas_hichip.merge(gwas_df, left_on=['chr_snp', 'end_snp'], right_on=['chromosome', 'position'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "afbd3a5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_hichip.drop('start_snp', axis=1, inplace=True)\n",
    "gwas_hichip.rename(columns={'end_snp': 'position_snp'}, inplace=True)\n",
    "gwas_hichip['sid'] = 'chr' +  gwas_hichip['chr_snp'].astype(str) + ':' + gwas_hichip['position_snp'].astype(str)\n",
    "\n",
    "# add loop ids\n",
    "def make_lid(sr, cols):\n",
    "    lid = sr[cols].tolist()\n",
    "    lid = [str(x) for x in lid]\n",
    "    lid = ':'.join(lid)\n",
    "    return(lid)\n",
    "\n",
    "lid_cols = [2,3,4,5,6,7]\n",
    "lids = []\n",
    "for sr in gwas_hichip.values: \n",
    "    new_lid = make_lid(sr, lid_cols)\n",
    "    lids.append(new_lid)\n",
    "gwas_hichip['loop_id'] = lids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2defcbde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_snp</th>\n",
       "      <th>position_snp</th>\n",
       "      <th>chrA_loop</th>\n",
       "      <th>startA_loop</th>\n",
       "      <th>endA_loop</th>\n",
       "      <th>chrB_loop</th>\n",
       "      <th>startB_loop</th>\n",
       "      <th>endB_loop</th>\n",
       "      <th>cline_loop</th>\n",
       "      <th>regionID</th>\n",
       "      <th>...</th>\n",
       "      <th>z</th>\n",
       "      <th>prob</th>\n",
       "      <th>log10bf</th>\n",
       "      <th>mean</th>\n",
       "      <th>sd</th>\n",
       "      <th>mean_incl</th>\n",
       "      <th>sd_incl</th>\n",
       "      <th>pval</th>\n",
       "      <th>sid</th>\n",
       "      <th>loop_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10</td>\n",
       "      <td>90319220</td>\n",
       "      <td>10</td>\n",
       "      <td>89622500</td>\n",
       "      <td>89632500</td>\n",
       "      <td>10</td>\n",
       "      <td>90312500</td>\n",
       "      <td>90322500</td>\n",
       "      <td>TREGMEM</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>1.47094</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.5355</td>\n",
       "      <td>0.187599</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.187599</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.070654</td>\n",
       "      <td>chr10:90319220</td>\n",
       "      <td>10:89622500:89632500:10:90312500:90322500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>90319220</td>\n",
       "      <td>10</td>\n",
       "      <td>89622500</td>\n",
       "      <td>89632500</td>\n",
       "      <td>10</td>\n",
       "      <td>90317500</td>\n",
       "      <td>90327500</td>\n",
       "      <td>TH2</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>1.47094</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.5355</td>\n",
       "      <td>0.187599</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.187599</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.070654</td>\n",
       "      <td>chr10:90319220</td>\n",
       "      <td>10:89622500:89632500:10:90317500:90327500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>90319220</td>\n",
       "      <td>10</td>\n",
       "      <td>89622500</td>\n",
       "      <td>89632500</td>\n",
       "      <td>10</td>\n",
       "      <td>90312500</td>\n",
       "      <td>90322500</td>\n",
       "      <td>CD4N</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>1.47094</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.5355</td>\n",
       "      <td>0.187599</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.187599</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.070654</td>\n",
       "      <td>chr10:90319220</td>\n",
       "      <td>10:89622500:89632500:10:90312500:90322500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "      <td>90319220</td>\n",
       "      <td>10</td>\n",
       "      <td>89622500</td>\n",
       "      <td>89632500</td>\n",
       "      <td>10</td>\n",
       "      <td>90317500</td>\n",
       "      <td>90327500</td>\n",
       "      <td>CD4N</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>1.47094</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.5355</td>\n",
       "      <td>0.187599</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.187599</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.070654</td>\n",
       "      <td>chr10:90319220</td>\n",
       "      <td>10:89622500:89632500:10:90317500:90327500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>90319220</td>\n",
       "      <td>10</td>\n",
       "      <td>89622500</td>\n",
       "      <td>89632500</td>\n",
       "      <td>10</td>\n",
       "      <td>90317500</td>\n",
       "      <td>90327500</td>\n",
       "      <td>TREGNAIVE</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>1.47094</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.5355</td>\n",
       "      <td>0.187599</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.187599</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.070654</td>\n",
       "      <td>chr10:90319220</td>\n",
       "      <td>10:89622500:89632500:10:90317500:90327500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr_snp  position_snp  chrA_loop  startA_loop  endA_loop  chrB_loop  \\\n",
       "0       10      90319220         10     89622500   89632500         10   \n",
       "1       10      90319220         10     89622500   89632500         10   \n",
       "2       10      90319220         10     89622500   89632500         10   \n",
       "3       10      90319220         10     89622500   89632500         10   \n",
       "4       10      90319220         10     89622500   89632500         10   \n",
       "\n",
       "   startB_loop  endB_loop cline_loop  regionID  ...        z  prob  log10bf  \\\n",
       "0     90312500   90322500    TREGMEM        16  ...  1.47094   1.0  13.5355   \n",
       "1     90317500   90327500        TH2        16  ...  1.47094   1.0  13.5355   \n",
       "2     90312500   90322500       CD4N        16  ...  1.47094   1.0  13.5355   \n",
       "3     90317500   90327500       CD4N        16  ...  1.47094   1.0  13.5355   \n",
       "4     90317500   90327500  TREGNAIVE        16  ...  1.47094   1.0  13.5355   \n",
       "\n",
       "       mean        sd mean_incl   sd_incl      pval             sid  \\\n",
       "0  0.187599  0.000003  0.187599  0.000003  0.070654  chr10:90319220   \n",
       "1  0.187599  0.000003  0.187599  0.000003  0.070654  chr10:90319220   \n",
       "2  0.187599  0.000003  0.187599  0.000003  0.070654  chr10:90319220   \n",
       "3  0.187599  0.000003  0.187599  0.000003  0.070654  chr10:90319220   \n",
       "4  0.187599  0.000003  0.187599  0.000003  0.070654  chr10:90319220   \n",
       "\n",
       "                                     loop_id  \n",
       "0  10:89622500:89632500:10:90312500:90322500  \n",
       "1  10:89622500:89632500:10:90317500:90327500  \n",
       "2  10:89622500:89632500:10:90312500:90322500  \n",
       "3  10:89622500:89632500:10:90317500:90327500  \n",
       "4  10:89622500:89632500:10:90317500:90327500  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_hichip.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a87355e2",
   "metadata": {},
   "source": [
    "## Summarize the SNPs, Loops and Intersection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "822d1071",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_gwas = gwas_df.shape[0]\n",
    "# # summarize the total number of GWAS loops per cell # DOESN't MAKE SENSE\n",
    "# cell_summary['total_gwas'] = gwas_hichip.groupby('cline_loop').nunique('sid')['chr_snp']\n",
    "# cell_summary['total_gwas'] = cell_summary['total_gwas'].to_frame()\n",
    "# cell_summary['total_gwas'].columns = ['Total GWAS SNPs']\n",
    "# cell_summary['total_gwas']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1e91d3cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_summary = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6452a2b8",
   "metadata": {},
   "source": [
    "### Summarize the Number of Loops per Cell (pre-intersection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2b34797f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total_hichip</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cline</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD4N</th>\n",
       "      <td>114421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8N</th>\n",
       "      <td>84599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM</th>\n",
       "      <td>84298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NB</th>\n",
       "      <td>128288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NCM</th>\n",
       "      <td>103342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK</th>\n",
       "      <td>129890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TFH</th>\n",
       "      <td>46172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH1</th>\n",
       "      <td>63241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH17</th>\n",
       "      <td>76270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH2</th>\n",
       "      <td>58115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>THSTAR</th>\n",
       "      <td>50757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TREGMEM</th>\n",
       "      <td>56567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TREGNAIVE</th>\n",
       "      <td>51613</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           total_hichip\n",
       "cline                  \n",
       "CD4N             114421\n",
       "CD8N              84599\n",
       "CM                84298\n",
       "NB               128288\n",
       "NCM              103342\n",
       "NK               129890\n",
       "TFH               46172\n",
       "TH1               63241\n",
       "TH17              76270\n",
       "TH2               58115\n",
       "THSTAR            50757\n",
       "TREGMEM           56567\n",
       "TREGNAIVE         51613"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_summary['total_loops'] = loop_df.groupby('cline').count()['startA'].to_frame()\n",
    "cell_summary['total_loops'].columns = ['total_hichip']\n",
    "cell_summary['total_loops']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08ae84d7",
   "metadata": {},
   "source": [
    "### Summarize the Number of SNP-Loop (SL) Pairs per Cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1d116fc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sl_pairs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NB</th>\n",
       "      <td>316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK</th>\n",
       "      <td>237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8N</th>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH1</th>\n",
       "      <td>154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH17</th>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4N</th>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH2</th>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TREGMEM</th>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TREGNAIVE</th>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NCM</th>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>THSTAR</th>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TFH</th>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM</th>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           sl_pairs\n",
       "NB              316\n",
       "NK              237\n",
       "CD8N            161\n",
       "TH1             154\n",
       "TH17            152\n",
       "CD4N            145\n",
       "TH2             141\n",
       "TREGMEM         138\n",
       "TREGNAIVE       112\n",
       "NCM              95\n",
       "THSTAR           78\n",
       "TFH              77\n",
       "CM               69"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_summary['sl_pairs'] = gwas_hichip['cline_loop'].value_counts().to_frame()\n",
    "cell_summary['sl_pairs'].columns = ['sl_pairs']\n",
    "cell_summary['sl_pairs']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c468bc58",
   "metadata": {},
   "source": [
    "### Summarize the Number of Unique GWAS SNPs which Overlap a HiChIP Loop Cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "42cc7348",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniq_gwas_in_slpairs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cline_loop</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD4N</th>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8N</th>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM</th>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NB</th>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NCM</th>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK</th>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TFH</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH1</th>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH17</th>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH2</th>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>THSTAR</th>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TREGMEM</th>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TREGNAIVE</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            uniq_gwas_in_slpairs\n",
       "cline_loop                      \n",
       "CD4N                          25\n",
       "CD8N                          28\n",
       "CM                            30\n",
       "NB                            58\n",
       "NCM                           28\n",
       "NK                            38\n",
       "TFH                           16\n",
       "TH1                           32\n",
       "TH17                          30\n",
       "TH2                           33\n",
       "THSTAR                        18\n",
       "TREGMEM                       28\n",
       "TREGNAIVE                     23"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_summary['uniq_gwas'] = gwas_hichip.groupby('cline_loop')['sid'].nunique().to_frame()\n",
    "cell_summary['uniq_gwas'].columns = ['uniq_gwas_in_slpairs']\n",
    "cell_summary['uniq_gwas']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eec14a11",
   "metadata": {},
   "source": [
    "### Summarize the Number of Loops with GWAS Overlaps (per cell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3a8f01e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniq_loops_in_slpairs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cline_loop</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD4N</th>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8N</th>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM</th>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NB</th>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NCM</th>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK</th>\n",
       "      <td>168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TFH</th>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH1</th>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH17</th>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH2</th>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>THSTAR</th>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TREGMEM</th>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TREGNAIVE</th>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            uniq_loops_in_slpairs\n",
       "cline_loop                       \n",
       "CD4N                          110\n",
       "CD8N                           99\n",
       "CM                             46\n",
       "NB                            192\n",
       "NCM                            64\n",
       "NK                            168\n",
       "TFH                            58\n",
       "TH1                            99\n",
       "TH17                          108\n",
       "TH2                            80\n",
       "THSTAR                         58\n",
       "TREGMEM                        87\n",
       "TREGNAIVE                      72"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loop_cols = ['chrA_loop', 'startA_loop', 'endA_loop', 'chrB_loop', 'startB_loop', 'endB_loop']\n",
    "cell_summary['uniq_loops'] = gwas_hichip.groupby('cline_loop')['loop_id'].nunique().to_frame()\n",
    "cell_summary['uniq_loops'].columns = ['uniq_loops_in_slpairs']\n",
    "cell_summary['uniq_loops']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "091f4b50",
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_list = [cell_summary['total_loops'], cell_summary['sl_pairs'],\n",
    "               cell_summary['uniq_gwas'], cell_summary['uniq_loops']]\n",
    "summary = pd.concat(concat_list, axis=1)\n",
    "summary['pct_uniq_gwas_in_slpairs'] = summary['uniq_gwas_in_slpairs'] / total_gwas * 100\n",
    "summary['pct_uniq_loops_in_slpairs'] = summary['uniq_loops_in_slpairs'] / summary['total_hichip'] * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "371814c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total_hichip</th>\n",
       "      <th>sl_pairs</th>\n",
       "      <th>uniq_gwas_in_slpairs</th>\n",
       "      <th>uniq_loops_in_slpairs</th>\n",
       "      <th>pct_uniq_gwas_in_slpairs</th>\n",
       "      <th>pct_uniq_loops_in_slpairs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD4N</th>\n",
       "      <td>114421</td>\n",
       "      <td>145</td>\n",
       "      <td>25</td>\n",
       "      <td>110</td>\n",
       "      <td>25.252525</td>\n",
       "      <td>0.096136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8N</th>\n",
       "      <td>84599</td>\n",
       "      <td>161</td>\n",
       "      <td>28</td>\n",
       "      <td>99</td>\n",
       "      <td>28.282828</td>\n",
       "      <td>0.117023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM</th>\n",
       "      <td>84298</td>\n",
       "      <td>69</td>\n",
       "      <td>30</td>\n",
       "      <td>46</td>\n",
       "      <td>30.303030</td>\n",
       "      <td>0.054568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NB</th>\n",
       "      <td>128288</td>\n",
       "      <td>316</td>\n",
       "      <td>58</td>\n",
       "      <td>192</td>\n",
       "      <td>58.585859</td>\n",
       "      <td>0.149663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NCM</th>\n",
       "      <td>103342</td>\n",
       "      <td>95</td>\n",
       "      <td>28</td>\n",
       "      <td>64</td>\n",
       "      <td>28.282828</td>\n",
       "      <td>0.061930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK</th>\n",
       "      <td>129890</td>\n",
       "      <td>237</td>\n",
       "      <td>38</td>\n",
       "      <td>168</td>\n",
       "      <td>38.383838</td>\n",
       "      <td>0.129340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TFH</th>\n",
       "      <td>46172</td>\n",
       "      <td>77</td>\n",
       "      <td>16</td>\n",
       "      <td>58</td>\n",
       "      <td>16.161616</td>\n",
       "      <td>0.125617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH1</th>\n",
       "      <td>63241</td>\n",
       "      <td>154</td>\n",
       "      <td>32</td>\n",
       "      <td>99</td>\n",
       "      <td>32.323232</td>\n",
       "      <td>0.156544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH17</th>\n",
       "      <td>76270</td>\n",
       "      <td>152</td>\n",
       "      <td>30</td>\n",
       "      <td>108</td>\n",
       "      <td>30.303030</td>\n",
       "      <td>0.141602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH2</th>\n",
       "      <td>58115</td>\n",
       "      <td>141</td>\n",
       "      <td>33</td>\n",
       "      <td>80</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>0.137658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>THSTAR</th>\n",
       "      <td>50757</td>\n",
       "      <td>78</td>\n",
       "      <td>18</td>\n",
       "      <td>58</td>\n",
       "      <td>18.181818</td>\n",
       "      <td>0.114270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TREGMEM</th>\n",
       "      <td>56567</td>\n",
       "      <td>138</td>\n",
       "      <td>28</td>\n",
       "      <td>87</td>\n",
       "      <td>28.282828</td>\n",
       "      <td>0.153800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TREGNAIVE</th>\n",
       "      <td>51613</td>\n",
       "      <td>112</td>\n",
       "      <td>23</td>\n",
       "      <td>72</td>\n",
       "      <td>23.232323</td>\n",
       "      <td>0.139500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           total_hichip  sl_pairs  uniq_gwas_in_slpairs  \\\n",
       "CD4N             114421       145                    25   \n",
       "CD8N              84599       161                    28   \n",
       "CM                84298        69                    30   \n",
       "NB               128288       316                    58   \n",
       "NCM              103342        95                    28   \n",
       "NK               129890       237                    38   \n",
       "TFH               46172        77                    16   \n",
       "TH1               63241       154                    32   \n",
       "TH17              76270       152                    30   \n",
       "TH2               58115       141                    33   \n",
       "THSTAR            50757        78                    18   \n",
       "TREGMEM           56567       138                    28   \n",
       "TREGNAIVE         51613       112                    23   \n",
       "\n",
       "           uniq_loops_in_slpairs  pct_uniq_gwas_in_slpairs  \\\n",
       "CD4N                         110                 25.252525   \n",
       "CD8N                          99                 28.282828   \n",
       "CM                            46                 30.303030   \n",
       "NB                           192                 58.585859   \n",
       "NCM                           64                 28.282828   \n",
       "NK                           168                 38.383838   \n",
       "TFH                           58                 16.161616   \n",
       "TH1                           99                 32.323232   \n",
       "TH17                         108                 30.303030   \n",
       "TH2                           80                 33.333333   \n",
       "THSTAR                        58                 18.181818   \n",
       "TREGMEM                       87                 28.282828   \n",
       "TREGNAIVE                     72                 23.232323   \n",
       "\n",
       "           pct_uniq_loops_in_slpairs  \n",
       "CD4N                        0.096136  \n",
       "CD8N                        0.117023  \n",
       "CM                          0.054568  \n",
       "NB                          0.149663  \n",
       "NCM                         0.061930  \n",
       "NK                          0.129340  \n",
       "TFH                         0.125617  \n",
       "TH1                         0.156544  \n",
       "TH17                        0.141602  \n",
       "TH2                         0.137658  \n",
       "THSTAR                      0.114270  \n",
       "TREGMEM                     0.153800  \n",
       "TREGNAIVE                   0.139500  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bf5dbdb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_summary = summary.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "cbb2cb48",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_colnames = ['Total\\\\nHiChIP Loops', \n",
    "                  'Number of\\\\nGWAS-Loop Pairs',\n",
    "                  'Number of\\\\nUnique GWAS SNPs in GL Pairs', \n",
    "                  'Number of\\\\nUnique loops in GL Pairs',\n",
    "                  'Percentage of\\\\nUnique GWAS SNPs in GL Pairs', \n",
    "                  'Percentage of\\\\nUnique loops in GL Pairs']\n",
    "final_colnames = ['Total HiChIP Loops', \n",
    "                  'Number of GWAS-Loop Pairs',\n",
    "                  'Number of Unique GWAS SNPs in GL Pairs', \n",
    "                  'Number of Unique loops in GL Pairs',\n",
    "                  'Percentage of Unique GWAS SNPs in GL Pairs', \n",
    "                  'Percentage of Unique loops in GL Pairs']\n",
    "final_summary.columns = final_colnames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a2180e2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Total HiChIP Loops</th>\n",
       "      <th>Number of GWAS-Loop Pairs</th>\n",
       "      <th>Number of Unique GWAS SNPs in GL Pairs</th>\n",
       "      <th>Number of Unique loops in GL Pairs</th>\n",
       "      <th>Percentage of Unique GWAS SNPs in GL Pairs</th>\n",
       "      <th>Percentage of Unique loops in GL Pairs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD4N</th>\n",
       "      <td>114421</td>\n",
       "      <td>145</td>\n",
       "      <td>25</td>\n",
       "      <td>110</td>\n",
       "      <td>25.252525</td>\n",
       "      <td>0.096136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8N</th>\n",
       "      <td>84599</td>\n",
       "      <td>161</td>\n",
       "      <td>28</td>\n",
       "      <td>99</td>\n",
       "      <td>28.282828</td>\n",
       "      <td>0.117023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM</th>\n",
       "      <td>84298</td>\n",
       "      <td>69</td>\n",
       "      <td>30</td>\n",
       "      <td>46</td>\n",
       "      <td>30.303030</td>\n",
       "      <td>0.054568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NB</th>\n",
       "      <td>128288</td>\n",
       "      <td>316</td>\n",
       "      <td>58</td>\n",
       "      <td>192</td>\n",
       "      <td>58.585859</td>\n",
       "      <td>0.149663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NCM</th>\n",
       "      <td>103342</td>\n",
       "      <td>95</td>\n",
       "      <td>28</td>\n",
       "      <td>64</td>\n",
       "      <td>28.282828</td>\n",
       "      <td>0.061930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK</th>\n",
       "      <td>129890</td>\n",
       "      <td>237</td>\n",
       "      <td>38</td>\n",
       "      <td>168</td>\n",
       "      <td>38.383838</td>\n",
       "      <td>0.129340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TFH</th>\n",
       "      <td>46172</td>\n",
       "      <td>77</td>\n",
       "      <td>16</td>\n",
       "      <td>58</td>\n",
       "      <td>16.161616</td>\n",
       "      <td>0.125617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH1</th>\n",
       "      <td>63241</td>\n",
       "      <td>154</td>\n",
       "      <td>32</td>\n",
       "      <td>99</td>\n",
       "      <td>32.323232</td>\n",
       "      <td>0.156544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH17</th>\n",
       "      <td>76270</td>\n",
       "      <td>152</td>\n",
       "      <td>30</td>\n",
       "      <td>108</td>\n",
       "      <td>30.303030</td>\n",
       "      <td>0.141602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TH2</th>\n",
       "      <td>58115</td>\n",
       "      <td>141</td>\n",
       "      <td>33</td>\n",
       "      <td>80</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>0.137658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>THSTAR</th>\n",
       "      <td>50757</td>\n",
       "      <td>78</td>\n",
       "      <td>18</td>\n",
       "      <td>58</td>\n",
       "      <td>18.181818</td>\n",
       "      <td>0.114270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TREGMEM</th>\n",
       "      <td>56567</td>\n",
       "      <td>138</td>\n",
       "      <td>28</td>\n",
       "      <td>87</td>\n",
       "      <td>28.282828</td>\n",
       "      <td>0.153800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TREGNAIVE</th>\n",
       "      <td>51613</td>\n",
       "      <td>112</td>\n",
       "      <td>23</td>\n",
       "      <td>72</td>\n",
       "      <td>23.232323</td>\n",
       "      <td>0.139500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(HTML(final_summary.to_html().replace(\"\\\\n\",\"<br>\")))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "HiChIP DB",
   "language": "python",
   "name": "hichip-db"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
