{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 246,
   "id": "7a4b0657",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import requests\n",
    "import numpy as np\n",
    "\n",
    "os.chdir('/mnt/BioHome/jreyna/jreyna/projects/dchallenge/')\n",
    "outdir = 'results/notebooks/gwas_catalog_disease_traits'\n",
    "os.makedirs(outdir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "id": "e1be389c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_meta = 'results/refs/gwas_catalog/gwas-catalog-v1.0.3-studies-r2022-02-02.tsv'\n",
    "gwas_meta = pd.read_table(gwas_meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "id": "a0b663ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATE ADDED TO CATALOG</th>\n",
       "      <th>PUBMED ID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>LINK</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>...</th>\n",
       "      <th>MAPPED_TRAIT</th>\n",
       "      <th>MAPPED_TRAIT_URI</th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>GENOTYPING TECHNOLOGY</th>\n",
       "      <th>SUMMARY STATS LOCATION</th>\n",
       "      <th>SUBMISSION DATE</th>\n",
       "      <th>STATISTICAL MODEL</th>\n",
       "      <th>BACKGROUND TRAIT</th>\n",
       "      <th>MAPPED BACKGROUND TRAIT</th>\n",
       "      <th>MAPPED BACKGROUND TRAIT URI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2017-08-04</td>\n",
       "      <td>27989323</td>\n",
       "      <td>Ahola-Olli AV</td>\n",
       "      <td>2016-12-13</td>\n",
       "      <td>Am J Hum Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/27989323</td>\n",
       "      <td>Genome-wide Association Study Identifies 27 Lo...</td>\n",
       "      <td>beta-nerve growth factor levels</td>\n",
       "      <td>3,531 Finnish ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>beta-nerve growth factor measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0008035</td>\n",
       "      <td>GCST004421</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2017-08-04</td>\n",
       "      <td>27989323</td>\n",
       "      <td>Ahola-Olli AV</td>\n",
       "      <td>2016-12-13</td>\n",
       "      <td>Am J Hum Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/27989323</td>\n",
       "      <td>Genome-wide Association Study Identifies 27 Lo...</td>\n",
       "      <td>Vascular endothelial growth factor levels</td>\n",
       "      <td>7,118 Finnish ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>vascular endothelial growth factor measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004762</td>\n",
       "      <td>GCST004422</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2017-08-04</td>\n",
       "      <td>27989323</td>\n",
       "      <td>Ahola-Olli AV</td>\n",
       "      <td>2016-12-13</td>\n",
       "      <td>Am J Hum Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/27989323</td>\n",
       "      <td>Genome-wide Association Study Identifies 27 Lo...</td>\n",
       "      <td>CTACK levels</td>\n",
       "      <td>3,631 Finnish ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>chemokine (C-C motif) ligand 27 measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0008082</td>\n",
       "      <td>GCST004420</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2017-08-04</td>\n",
       "      <td>27989323</td>\n",
       "      <td>Ahola-Olli AV</td>\n",
       "      <td>2016-12-13</td>\n",
       "      <td>Am J Hum Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/27989323</td>\n",
       "      <td>Genome-wide Association Study Identifies 27 Lo...</td>\n",
       "      <td>Eotaxin levels</td>\n",
       "      <td>8,153 Finnish ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>eotaxin measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0008122</td>\n",
       "      <td>GCST004460</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2017-08-04</td>\n",
       "      <td>27989323</td>\n",
       "      <td>Ahola-Olli AV</td>\n",
       "      <td>2016-12-13</td>\n",
       "      <td>Am J Hum Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/27989323</td>\n",
       "      <td>Genome-wide Association Study Identifies 27 Lo...</td>\n",
       "      <td>Fibroblast growth factor basic levels</td>\n",
       "      <td>7,565 Finnish ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>fibroblast growth factor basic measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0008130</td>\n",
       "      <td>GCST004459</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATE ADDED TO CATALOG  PUBMED ID   FIRST AUTHOR        DATE         JOURNAL  \\\n",
       "0            2017-08-04   27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
       "1            2017-08-04   27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
       "2            2017-08-04   27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
       "3            2017-08-04   27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
       "4            2017-08-04   27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
       "\n",
       "                                   LINK  \\\n",
       "0  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
       "1  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
       "2  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
       "3  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
       "4  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
       "\n",
       "                                               STUDY  \\\n",
       "0  Genome-wide Association Study Identifies 27 Lo...   \n",
       "1  Genome-wide Association Study Identifies 27 Lo...   \n",
       "2  Genome-wide Association Study Identifies 27 Lo...   \n",
       "3  Genome-wide Association Study Identifies 27 Lo...   \n",
       "4  Genome-wide Association Study Identifies 27 Lo...   \n",
       "\n",
       "                               DISEASE/TRAIT  \\\n",
       "0            beta-nerve growth factor levels   \n",
       "1  Vascular endothelial growth factor levels   \n",
       "2                               CTACK levels   \n",
       "3                             Eotaxin levels   \n",
       "4      Fibroblast growth factor basic levels   \n",
       "\n",
       "                  INITIAL SAMPLE SIZE REPLICATION SAMPLE SIZE  ...  \\\n",
       "0  3,531 Finnish ancestry individuals                     NaN  ...   \n",
       "1  7,118 Finnish ancestry individuals                     NaN  ...   \n",
       "2  3,631 Finnish ancestry individuals                     NaN  ...   \n",
       "3  8,153 Finnish ancestry individuals                     NaN  ...   \n",
       "4  7,565 Finnish ancestry individuals                     NaN  ...   \n",
       "\n",
       "                                     MAPPED_TRAIT  \\\n",
       "0            beta-nerve growth factor measurement   \n",
       "1  vascular endothelial growth factor measurement   \n",
       "2     chemokine (C-C motif) ligand 27 measurement   \n",
       "3                             eotaxin measurement   \n",
       "4      fibroblast growth factor basic measurement   \n",
       "\n",
       "                       MAPPED_TRAIT_URI STUDY ACCESSION  \\\n",
       "0  http://www.ebi.ac.uk/efo/EFO_0008035      GCST004421   \n",
       "1  http://www.ebi.ac.uk/efo/EFO_0004762      GCST004422   \n",
       "2  http://www.ebi.ac.uk/efo/EFO_0008082      GCST004420   \n",
       "3  http://www.ebi.ac.uk/efo/EFO_0008122      GCST004460   \n",
       "4  http://www.ebi.ac.uk/efo/EFO_0008130      GCST004459   \n",
       "\n",
       "          GENOTYPING TECHNOLOGY SUMMARY STATS LOCATION SUBMISSION DATE  \\\n",
       "0  Genome-wide genotyping array                    NaN             NaN   \n",
       "1  Genome-wide genotyping array                    NaN             NaN   \n",
       "2  Genome-wide genotyping array                    NaN             NaN   \n",
       "3  Genome-wide genotyping array                    NaN             NaN   \n",
       "4  Genome-wide genotyping array                    NaN             NaN   \n",
       "\n",
       "   STATISTICAL MODEL  BACKGROUND TRAIT  MAPPED BACKGROUND TRAIT  \\\n",
       "0                NaN               NaN                      NaN   \n",
       "1                NaN               NaN                      NaN   \n",
       "2                NaN               NaN                      NaN   \n",
       "3                NaN               NaN                      NaN   \n",
       "4                NaN               NaN                      NaN   \n",
       "\n",
       "   MAPPED BACKGROUND TRAIT URI  \n",
       "0                          NaN  \n",
       "1                          NaN  \n",
       "2                          NaN  \n",
       "3                          NaN  \n",
       "4                          NaN  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9027a3b3",
   "metadata": {},
   "source": [
    "## Getting all the OLS Link IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "id": "ddb589c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "link_ids = []\n",
    "for uris in gwas_meta['MAPPED_TRAIT_URI']:\n",
    "    if type(uris) == str:\n",
    "        for uri in uris.split(','):\n",
    "            link_ids.append(uri.strip())\n",
    "link_ids = sorted(set(link_ids))\n",
    "num_files = np.ceil(len(link_ids) / 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "id": "d0978af0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# split the IDs into different files\n",
    "split_dir = os.path.join(outdir, 'splits/')\n",
    "os.makedirs(split_dir, exist_ok=True)\n",
    "curr_bin = 0 \n",
    "curr_fhandle = open(os.path.join(split_dir, 'split.ols_id.{}.txt'.format(curr_bin)), 'w')\n",
    "\n",
    "for i, link_id in enumerate(link_ids):    \n",
    "    if np.floor(i / 100) != curr_bin:\n",
    "        curr_fhandle.close()\n",
    "        curr_bin += 1 \n",
    "        curr_fhandle = open(os.path.join(split_dir, 'split.ols_id.{}.txt'.format(curr_bin)), 'w')\n",
    "    curr_fhandle.write(link_id + '\\n')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "id": "94e0c3f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# close the last handle\n",
    "curr_fhandle.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "HiChIP-DB",
   "language": "python",
   "name": "hichip-db"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
