{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7b5ad180",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import glob\n",
    "import pybedtools as pbt \n",
    "from IPython.display import HTML\n",
    "pd.set_option('display.min_rows', 100) \n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "pbt.set_bedtools_path('/mnt/BioHome/jreyna/software/anaconda3/envs/hic_tls/bin/')\n",
    "os.chdir('/mnt/BioHome/jreyna/jreyna/projects/dchallenge/')\n",
    "\n",
    "gsizes = 'results/refs/hg19/hg19.chrom.sizes'\n",
    "res = 5000\n",
    "\n",
    "# make the directory to save our data\n",
    "outdir = 'results/main/Intersect_T1D_Finemap_GWAS_SNPs_with_HiChIP/'\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "bedpe_cols = ['chrA', 'startA', 'endA', 'chrB', 'startB', 'endB']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25ec296e",
   "metadata": {},
   "source": [
    "## Load Fine Mapped GWAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5268fb2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "major_gwas = ['T1D_32005708', 'T1D_34594039_GCST90018925', 'T1D_34012112_Gaulton']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "46c41d31",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_glob = 'results/main/finemapping/*/GRCh37/offset_1000000/Summary/sss/FINAL_top_snp_credible_set.txt'\n",
    "gwas_glob = glob.glob(gwas_glob)\n",
    "data = []\n",
    "for fn in gwas_glob:\n",
    "    \n",
    "    # get meta data from the path\n",
    "    path_info = fn.split('/')\n",
    "    \n",
    "    if path_info[3] not in major_gwas:\n",
    "        print('skipped: {}'.format(path_info[3]))\n",
    "        continue\n",
    "    \n",
    "    # get the bin coordinates\n",
    "    df = pd.read_table(fn)\n",
    "    df.loc[:, 'bin_start'] = np.floor(df.loc[:, 'position'] / res).astype(int) * res\n",
    "    df.loc[:, 'bin_end'] = df.loc[:, 'bin_start'] + res\n",
    "    df = df.loc[(df.allele1.str.len() == 1 ) & (df.allele2.str.len() == 1)]\n",
    "    df.loc[:, 'gwas_source'] = path_info[3]\n",
    "    data.append(df)\n",
    "\n",
    "gwas_df = pd.concat(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0ce6ac3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regionID</th>\n",
       "      <th>GWASLoci</th>\n",
       "      <th>index</th>\n",
       "      <th>rsid</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>maf</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>z</th>\n",
       "      <th>prob</th>\n",
       "      <th>log10bf</th>\n",
       "      <th>mean</th>\n",
       "      <th>sd</th>\n",
       "      <th>mean_incl</th>\n",
       "      <th>sd_incl</th>\n",
       "      <th>pval</th>\n",
       "      <th>bin_start</th>\n",
       "      <th>bin_end</th>\n",
       "      <th>gwas_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1:113310083-115099755</td>\n",
       "      <td>2336</td>\n",
       "      <td>1:114089649</td>\n",
       "      <td>1</td>\n",
       "      <td>114089649</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.016251</td>\n",
       "      <td>-0.1324</td>\n",
       "      <td>0.0749</td>\n",
       "      <td>-1.767690</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.5586</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>9.614436e-01</td>\n",
       "      <td>114085000</td>\n",
       "      <td>114090000</td>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1:113310083-115099755</td>\n",
       "      <td>2829</td>\n",
       "      <td>1:114270326</td>\n",
       "      <td>1</td>\n",
       "      <td>114270326</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.260388</td>\n",
       "      <td>-0.1634</td>\n",
       "      <td>0.0209</td>\n",
       "      <td>-7.818180</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.5586</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>114270000</td>\n",
       "      <td>114275000</td>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1:113310083-115099755</td>\n",
       "      <td>5046</td>\n",
       "      <td>1:114909703</td>\n",
       "      <td>1</td>\n",
       "      <td>114909703</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.131631</td>\n",
       "      <td>-0.0046</td>\n",
       "      <td>0.0261</td>\n",
       "      <td>-0.176245</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.5586</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>5.699493e-01</td>\n",
       "      <td>114905000</td>\n",
       "      <td>114910000</td>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1:113310083-115099755</td>\n",
       "      <td>3035</td>\n",
       "      <td>1:114377568</td>\n",
       "      <td>1</td>\n",
       "      <td>114377568</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.114168</td>\n",
       "      <td>-0.4287</td>\n",
       "      <td>0.0286</td>\n",
       "      <td>-14.989500</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.5586</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>114375000</td>\n",
       "      <td>114380000</td>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1:113310083-115099755</td>\n",
       "      <td>3131</td>\n",
       "      <td>1:114420328</td>\n",
       "      <td>1</td>\n",
       "      <td>114420328</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.325922</td>\n",
       "      <td>0.1320</td>\n",
       "      <td>0.0189</td>\n",
       "      <td>6.984130</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.5586</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>1.433164e-12</td>\n",
       "      <td>114420000</td>\n",
       "      <td>114425000</td>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   regionID                  GWASLoci  index         rsid  chromosome  \\\n",
       "0         1  chr1:113310083-115099755   2336  1:114089649           1   \n",
       "1         1  chr1:113310083-115099755   2829  1:114270326           1   \n",
       "2         1  chr1:113310083-115099755   5046  1:114909703           1   \n",
       "3         1  chr1:113310083-115099755   3035  1:114377568           1   \n",
       "4         1  chr1:113310083-115099755   3131  1:114420328           1   \n",
       "\n",
       "    position allele1 allele2       maf    beta      se          z  prob  \\\n",
       "0  114089649       A       G  0.016251 -0.1324  0.0749  -1.767690   1.0   \n",
       "1  114270326       A       C  0.260388 -0.1634  0.0209  -7.818180   1.0   \n",
       "2  114909703       T       C  0.131631 -0.0046  0.0261  -0.176245   1.0   \n",
       "3  114377568       G       A  0.114168 -0.4287  0.0286 -14.989500   1.0   \n",
       "4  114420328       T       C  0.325922  0.1320  0.0189   6.984130   1.0   \n",
       "\n",
       "   log10bf      mean        sd  mean_incl   sd_incl          pval  bin_start  \\\n",
       "0  13.5586  0.001264  0.000008   0.001264  0.000008  9.614436e-01  114085000   \n",
       "1  13.5586  0.001264  0.000008   0.001264  0.000008  1.000000e+00  114270000   \n",
       "2  13.5586  0.001264  0.000008   0.001264  0.000008  5.699493e-01  114905000   \n",
       "3  13.5586  0.001264  0.000008   0.001264  0.000008  1.000000e+00  114375000   \n",
       "4  13.5586  0.001264  0.000008   0.001264  0.000008  1.433164e-12  114420000   \n",
       "\n",
       "     bin_end                gwas_source  \n",
       "0  114090000  T1D_34594039_GCST90018925  \n",
       "1  114275000  T1D_34594039_GCST90018925  \n",
       "2  114910000  T1D_34594039_GCST90018925  \n",
       "3  114380000  T1D_34594039_GCST90018925  \n",
       "4  114425000  T1D_34594039_GCST90018925  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "46298a0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a pybedtools for finemap data\n",
    "gwas_bed = gwas_df.loc[:, ['chromosome','bin_start','bin_end', 'position', 'gwas_source']]\n",
    "gwas_pbt = pbt.BedTool.from_dataframe(gwas_bed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3756334b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>bin_start</th>\n",
       "      <th>bin_end</th>\n",
       "      <th>position</th>\n",
       "      <th>gwas_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>114085000</td>\n",
       "      <td>114090000</td>\n",
       "      <td>114089649</td>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>114270000</td>\n",
       "      <td>114275000</td>\n",
       "      <td>114270326</td>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>114905000</td>\n",
       "      <td>114910000</td>\n",
       "      <td>114909703</td>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>114375000</td>\n",
       "      <td>114380000</td>\n",
       "      <td>114377568</td>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>114420000</td>\n",
       "      <td>114425000</td>\n",
       "      <td>114420328</td>\n",
       "      <td>T1D_34594039_GCST90018925</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome  bin_start    bin_end   position                gwas_source\n",
       "0           1  114085000  114090000  114089649  T1D_34594039_GCST90018925\n",
       "1           1  114270000  114275000  114270326  T1D_34594039_GCST90018925\n",
       "2           1  114905000  114910000  114909703  T1D_34594039_GCST90018925\n",
       "3           1  114375000  114380000  114377568  T1D_34594039_GCST90018925\n",
       "4           1  114420000  114425000  114420328  T1D_34594039_GCST90018925"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_bed.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b1e3652",
   "metadata": {},
   "source": [
    "## Load HiChIP Loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b49da1dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_seB(x): \n",
    "    s,e = x.split(':')[1].split('-')\n",
    "    e = e.split(',')[0]\n",
    "    return((s,e))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6cd23b57",
   "metadata": {},
   "outputs": [],
   "source": [
    "loops = 'results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/'\n",
    "loops += '*/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz'\n",
    "loops = glob.glob(loops)\n",
    "\n",
    "# only analyze loop data from main cell types \n",
    "loops = ['results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/CD4N/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz',\n",
    " 'results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/CD8N/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz',\n",
    " 'results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/NB/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz',\n",
    " 'results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/CM/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz',\n",
    " 'results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/NK/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1c387b45",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/CD4N/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n",
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/CD8N/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n",
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/NB/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n",
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/CM/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n",
      "results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/NK/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz\n"
     ]
    }
   ],
   "source": [
    "loop_data = []\n",
    "for loop in loops:\n",
    "    print(loop)\n",
    "    \n",
    "    # extract cell line\n",
    "    cline = loop.split('/')[5]\n",
    "    \n",
    "    # load and parse the data\n",
    "    df = pd.read_table(loop, header=None)    \n",
    "    df.columns = ['chrom', 'startA', 'endA', 'seB', 'e1', 'e2']\n",
    "    df['chrom'] = df['chrom'].str.replace('chr', '')\n",
    "    df['startB'], df['endB'] = zip(*df['seB'].apply(parse_seB))\n",
    "    df['startB'] = df['startB'].astype(int)\n",
    "    df['startA'] = df['startA'] + 1 - int(res / 2)\n",
    "    df['endA'] = df['startA'] + res\n",
    "    df['startB'] = df['startB'] + 1 - int(res / 2)\n",
    "    df['endB'] = df['startB'] + res\n",
    "    \n",
    "    # re-organize the data into bedpe-like\n",
    "    df = df.iloc[:, [0,1,2,0,6,7,3,4,5]]\n",
    "\n",
    "    # add cell type\n",
    "    df['cline'] = cline\n",
    "\n",
    "    loop_data.append(df) \n",
    "loop_df = pd.concat(loop_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "541b020f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a pybedtools for the looping data\n",
    "loop_bed = loop_df.iloc[:, [0,1,2,3,4,5,-1]]\n",
    "loop_pbt = pbt.BedTool.from_dataframe(loop_bed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cb73e3db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>startA</th>\n",
       "      <th>endA</th>\n",
       "      <th>chrom</th>\n",
       "      <th>startB</th>\n",
       "      <th>endB</th>\n",
       "      <th>cline</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>710000</td>\n",
       "      <td>715000</td>\n",
       "      <td>1</td>\n",
       "      <td>1305000</td>\n",
       "      <td>1310000</td>\n",
       "      <td>CD4N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>710000</td>\n",
       "      <td>715000</td>\n",
       "      <td>1</td>\n",
       "      <td>755000</td>\n",
       "      <td>760000</td>\n",
       "      <td>CD4N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>710000</td>\n",
       "      <td>715000</td>\n",
       "      <td>1</td>\n",
       "      <td>760000</td>\n",
       "      <td>765000</td>\n",
       "      <td>CD4N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>710000</td>\n",
       "      <td>715000</td>\n",
       "      <td>1</td>\n",
       "      <td>775000</td>\n",
       "      <td>780000</td>\n",
       "      <td>CD4N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>710000</td>\n",
       "      <td>715000</td>\n",
       "      <td>1</td>\n",
       "      <td>805000</td>\n",
       "      <td>810000</td>\n",
       "      <td>CD4N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  startA    endA chrom   startB     endB cline\n",
       "0     1  710000  715000     1  1305000  1310000  CD4N\n",
       "1     1  710000  715000     1   755000   760000  CD4N\n",
       "2     1  710000  715000     1   760000   765000  CD4N\n",
       "3     1  710000  715000     1   775000   780000  CD4N\n",
       "4     1  710000  715000     1   805000   810000  CD4N"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loop_bed.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bf47849",
   "metadata": {},
   "source": [
    "## Intersect Fine Mapped GWAS and loops"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca1bd83c",
   "metadata": {},
   "source": [
    "#### Perform the intersection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "09fb40ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "intersect_pbt = loop_pbt.pair_to_bed(gwas_pbt, type='either')\n",
    "gwas_hichip = intersect_pbt.to_dataframe(header=None, disable_auto_names=True)\n",
    "gwas_hichip = gwas_hichip.iloc[:, [7,8,9,10,0,1,2,3,4,5,6,11]]\n",
    "loop_cols = ['{}_loop'.format(x) for x in bedpe_cols]\n",
    "gwas_hichip.columns = ['chr_snp', 'bin_start', 'bin_end', 'pos'] + loop_cols + ['cline_loop', 'gwas_source']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6153a64",
   "metadata": {},
   "source": [
    "#### Add back fields from the original gwas data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "67a931f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_hichip = gwas_hichip.merge(gwas_df.drop('gwas_source', axis=1),\n",
    "                                left_on=['chr_snp', 'pos'],\n",
    "                                right_on=['chromosome', 'position'])\n",
    "# add the sid\n",
    "gwas_hichip['sid'] = 'chr' +  gwas_hichip['chr_snp'].astype(str) + ':' + gwas_hichip['position'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e9f1b24f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_snp</th>\n",
       "      <th>bin_start_x</th>\n",
       "      <th>bin_end_x</th>\n",
       "      <th>pos</th>\n",
       "      <th>chrA_loop</th>\n",
       "      <th>startA_loop</th>\n",
       "      <th>endA_loop</th>\n",
       "      <th>chrB_loop</th>\n",
       "      <th>startB_loop</th>\n",
       "      <th>endB_loop</th>\n",
       "      <th>cline_loop</th>\n",
       "      <th>gwas_source</th>\n",
       "      <th>regionID</th>\n",
       "      <th>GWASLoci</th>\n",
       "      <th>index</th>\n",
       "      <th>rsid</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>maf</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>z</th>\n",
       "      <th>prob</th>\n",
       "      <th>log10bf</th>\n",
       "      <th>mean</th>\n",
       "      <th>sd</th>\n",
       "      <th>mean_incl</th>\n",
       "      <th>sd_incl</th>\n",
       "      <th>pval</th>\n",
       "      <th>bin_start_y</th>\n",
       "      <th>bin_end_y</th>\n",
       "      <th>sid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19535000</td>\n",
       "      <td>19540000</td>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>CD4N</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:19579228-20579228</td>\n",
       "      <td>982</td>\n",
       "      <td>1:19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19972330</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0843</td>\n",
       "      <td>0.1618</td>\n",
       "      <td>0.0419</td>\n",
       "      <td>3.86158</td>\n",
       "      <td>0.04914</td>\n",
       "      <td>2.07697</td>\n",
       "      <td>-0.001938</td>\n",
       "      <td>0.06338</td>\n",
       "      <td>-0.039446</td>\n",
       "      <td>0.283315</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>chr1:19972330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19715000</td>\n",
       "      <td>19720000</td>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>CD4N</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:19579228-20579228</td>\n",
       "      <td>982</td>\n",
       "      <td>1:19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19972330</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0843</td>\n",
       "      <td>0.1618</td>\n",
       "      <td>0.0419</td>\n",
       "      <td>3.86158</td>\n",
       "      <td>0.04914</td>\n",
       "      <td>2.07697</td>\n",
       "      <td>-0.001938</td>\n",
       "      <td>0.06338</td>\n",
       "      <td>-0.039446</td>\n",
       "      <td>0.283315</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>chr1:19972330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19810000</td>\n",
       "      <td>19815000</td>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>CD4N</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:19579228-20579228</td>\n",
       "      <td>982</td>\n",
       "      <td>1:19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19972330</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0843</td>\n",
       "      <td>0.1618</td>\n",
       "      <td>0.0419</td>\n",
       "      <td>3.86158</td>\n",
       "      <td>0.04914</td>\n",
       "      <td>2.07697</td>\n",
       "      <td>-0.001938</td>\n",
       "      <td>0.06338</td>\n",
       "      <td>-0.039446</td>\n",
       "      <td>0.283315</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>chr1:19972330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19920000</td>\n",
       "      <td>19925000</td>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>CD4N</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:19579228-20579228</td>\n",
       "      <td>982</td>\n",
       "      <td>1:19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19972330</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0843</td>\n",
       "      <td>0.1618</td>\n",
       "      <td>0.0419</td>\n",
       "      <td>3.86158</td>\n",
       "      <td>0.04914</td>\n",
       "      <td>2.07697</td>\n",
       "      <td>-0.001938</td>\n",
       "      <td>0.06338</td>\n",
       "      <td>-0.039446</td>\n",
       "      <td>0.283315</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>chr1:19972330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19925000</td>\n",
       "      <td>19930000</td>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>CD4N</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:19579228-20579228</td>\n",
       "      <td>982</td>\n",
       "      <td>1:19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19972330</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0843</td>\n",
       "      <td>0.1618</td>\n",
       "      <td>0.0419</td>\n",
       "      <td>3.86158</td>\n",
       "      <td>0.04914</td>\n",
       "      <td>2.07697</td>\n",
       "      <td>-0.001938</td>\n",
       "      <td>0.06338</td>\n",
       "      <td>-0.039446</td>\n",
       "      <td>0.283315</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>chr1:19972330</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr_snp  bin_start_x  bin_end_x       pos  chrA_loop  startA_loop  \\\n",
       "0        1     19970000   19975000  19972330          1     19535000   \n",
       "1        1     19970000   19975000  19972330          1     19715000   \n",
       "2        1     19970000   19975000  19972330          1     19810000   \n",
       "3        1     19970000   19975000  19972330          1     19920000   \n",
       "4        1     19970000   19975000  19972330          1     19925000   \n",
       "\n",
       "   endA_loop  chrB_loop  startB_loop  endB_loop cline_loop   gwas_source  \\\n",
       "0   19540000          1     19970000   19975000       CD4N  T1D_32005708   \n",
       "1   19720000          1     19970000   19975000       CD4N  T1D_32005708   \n",
       "2   19815000          1     19970000   19975000       CD4N  T1D_32005708   \n",
       "3   19925000          1     19970000   19975000       CD4N  T1D_32005708   \n",
       "4   19930000          1     19970000   19975000       CD4N  T1D_32005708   \n",
       "\n",
       "   regionID                GWASLoci  index        rsid  chromosome  position  \\\n",
       "0         1  chr1:19579228-20579228    982  1:19972330           1  19972330   \n",
       "1         1  chr1:19579228-20579228    982  1:19972330           1  19972330   \n",
       "2         1  chr1:19579228-20579228    982  1:19972330           1  19972330   \n",
       "3         1  chr1:19579228-20579228    982  1:19972330           1  19972330   \n",
       "4         1  chr1:19579228-20579228    982  1:19972330           1  19972330   \n",
       "\n",
       "  allele1 allele2     maf    beta      se        z     prob  log10bf  \\\n",
       "0       A       G  0.0843  0.1618  0.0419  3.86158  0.04914  2.07697   \n",
       "1       A       G  0.0843  0.1618  0.0419  3.86158  0.04914  2.07697   \n",
       "2       A       G  0.0843  0.1618  0.0419  3.86158  0.04914  2.07697   \n",
       "3       A       G  0.0843  0.1618  0.0419  3.86158  0.04914  2.07697   \n",
       "4       A       G  0.0843  0.1618  0.0419  3.86158  0.04914  2.07697   \n",
       "\n",
       "       mean       sd  mean_incl   sd_incl      pval  bin_start_y  bin_end_y  \\\n",
       "0 -0.001938  0.06338  -0.039446  0.283315  0.000056     19970000   19975000   \n",
       "1 -0.001938  0.06338  -0.039446  0.283315  0.000056     19970000   19975000   \n",
       "2 -0.001938  0.06338  -0.039446  0.283315  0.000056     19970000   19975000   \n",
       "3 -0.001938  0.06338  -0.039446  0.283315  0.000056     19970000   19975000   \n",
       "4 -0.001938  0.06338  -0.039446  0.283315  0.000056     19970000   19975000   \n",
       "\n",
       "             sid  \n",
       "0  chr1:19972330  \n",
       "1  chr1:19972330  \n",
       "2  chr1:19972330  \n",
       "3  chr1:19972330  \n",
       "4  chr1:19972330  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_hichip.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7570dec9",
   "metadata": {},
   "source": [
    "#### Add loop ids which are used for unique set analysis downstream"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ac73d56c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_lid(sr, cols):\n",
    "    lid = sr[cols].tolist()\n",
    "    lid = [str(x) for x in lid]\n",
    "    lid = ':'.join(lid)\n",
    "    return(lid)\n",
    "lid_cols = [2,3,4,5,6,7]\n",
    "lids = []\n",
    "for sr in gwas_hichip.values: \n",
    "    new_lid = make_lid(sr, lid_cols)\n",
    "    lids.append(new_lid)\n",
    "gwas_hichip['loop_id'] = lids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2defcbde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_snp</th>\n",
       "      <th>bin_start_x</th>\n",
       "      <th>bin_end_x</th>\n",
       "      <th>pos</th>\n",
       "      <th>chrA_loop</th>\n",
       "      <th>startA_loop</th>\n",
       "      <th>endA_loop</th>\n",
       "      <th>chrB_loop</th>\n",
       "      <th>startB_loop</th>\n",
       "      <th>endB_loop</th>\n",
       "      <th>cline_loop</th>\n",
       "      <th>gwas_source</th>\n",
       "      <th>regionID</th>\n",
       "      <th>GWASLoci</th>\n",
       "      <th>index</th>\n",
       "      <th>rsid</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>maf</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>z</th>\n",
       "      <th>prob</th>\n",
       "      <th>log10bf</th>\n",
       "      <th>mean</th>\n",
       "      <th>sd</th>\n",
       "      <th>mean_incl</th>\n",
       "      <th>sd_incl</th>\n",
       "      <th>pval</th>\n",
       "      <th>bin_start_y</th>\n",
       "      <th>bin_end_y</th>\n",
       "      <th>sid</th>\n",
       "      <th>loop_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19535000</td>\n",
       "      <td>19540000</td>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>CD4N</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:19579228-20579228</td>\n",
       "      <td>982</td>\n",
       "      <td>1:19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19972330</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0843</td>\n",
       "      <td>0.1618</td>\n",
       "      <td>0.0419</td>\n",
       "      <td>3.86158</td>\n",
       "      <td>0.04914</td>\n",
       "      <td>2.07697</td>\n",
       "      <td>-0.001938</td>\n",
       "      <td>0.06338</td>\n",
       "      <td>-0.039446</td>\n",
       "      <td>0.283315</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>chr1:19972330</td>\n",
       "      <td>19975000:19972330:1:19535000:19540000:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19715000</td>\n",
       "      <td>19720000</td>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>CD4N</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:19579228-20579228</td>\n",
       "      <td>982</td>\n",
       "      <td>1:19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19972330</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0843</td>\n",
       "      <td>0.1618</td>\n",
       "      <td>0.0419</td>\n",
       "      <td>3.86158</td>\n",
       "      <td>0.04914</td>\n",
       "      <td>2.07697</td>\n",
       "      <td>-0.001938</td>\n",
       "      <td>0.06338</td>\n",
       "      <td>-0.039446</td>\n",
       "      <td>0.283315</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>chr1:19972330</td>\n",
       "      <td>19975000:19972330:1:19715000:19720000:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19810000</td>\n",
       "      <td>19815000</td>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>CD4N</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:19579228-20579228</td>\n",
       "      <td>982</td>\n",
       "      <td>1:19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19972330</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0843</td>\n",
       "      <td>0.1618</td>\n",
       "      <td>0.0419</td>\n",
       "      <td>3.86158</td>\n",
       "      <td>0.04914</td>\n",
       "      <td>2.07697</td>\n",
       "      <td>-0.001938</td>\n",
       "      <td>0.06338</td>\n",
       "      <td>-0.039446</td>\n",
       "      <td>0.283315</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>chr1:19972330</td>\n",
       "      <td>19975000:19972330:1:19810000:19815000:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19920000</td>\n",
       "      <td>19925000</td>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>CD4N</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:19579228-20579228</td>\n",
       "      <td>982</td>\n",
       "      <td>1:19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19972330</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0843</td>\n",
       "      <td>0.1618</td>\n",
       "      <td>0.0419</td>\n",
       "      <td>3.86158</td>\n",
       "      <td>0.04914</td>\n",
       "      <td>2.07697</td>\n",
       "      <td>-0.001938</td>\n",
       "      <td>0.06338</td>\n",
       "      <td>-0.039446</td>\n",
       "      <td>0.283315</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>chr1:19972330</td>\n",
       "      <td>19975000:19972330:1:19920000:19925000:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19925000</td>\n",
       "      <td>19930000</td>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>CD4N</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:19579228-20579228</td>\n",
       "      <td>982</td>\n",
       "      <td>1:19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19972330</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0843</td>\n",
       "      <td>0.1618</td>\n",
       "      <td>0.0419</td>\n",
       "      <td>3.86158</td>\n",
       "      <td>0.04914</td>\n",
       "      <td>2.07697</td>\n",
       "      <td>-0.001938</td>\n",
       "      <td>0.06338</td>\n",
       "      <td>-0.039446</td>\n",
       "      <td>0.283315</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>chr1:19972330</td>\n",
       "      <td>19975000:19972330:1:19925000:19930000:1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr_snp  bin_start_x  bin_end_x       pos  chrA_loop  startA_loop  \\\n",
       "0        1     19970000   19975000  19972330          1     19535000   \n",
       "1        1     19970000   19975000  19972330          1     19715000   \n",
       "2        1     19970000   19975000  19972330          1     19810000   \n",
       "3        1     19970000   19975000  19972330          1     19920000   \n",
       "4        1     19970000   19975000  19972330          1     19925000   \n",
       "\n",
       "   endA_loop  chrB_loop  startB_loop  endB_loop cline_loop   gwas_source  \\\n",
       "0   19540000          1     19970000   19975000       CD4N  T1D_32005708   \n",
       "1   19720000          1     19970000   19975000       CD4N  T1D_32005708   \n",
       "2   19815000          1     19970000   19975000       CD4N  T1D_32005708   \n",
       "3   19925000          1     19970000   19975000       CD4N  T1D_32005708   \n",
       "4   19930000          1     19970000   19975000       CD4N  T1D_32005708   \n",
       "\n",
       "   regionID                GWASLoci  index        rsid  chromosome  position  \\\n",
       "0         1  chr1:19579228-20579228    982  1:19972330           1  19972330   \n",
       "1         1  chr1:19579228-20579228    982  1:19972330           1  19972330   \n",
       "2         1  chr1:19579228-20579228    982  1:19972330           1  19972330   \n",
       "3         1  chr1:19579228-20579228    982  1:19972330           1  19972330   \n",
       "4         1  chr1:19579228-20579228    982  1:19972330           1  19972330   \n",
       "\n",
       "  allele1 allele2     maf    beta      se        z     prob  log10bf  \\\n",
       "0       A       G  0.0843  0.1618  0.0419  3.86158  0.04914  2.07697   \n",
       "1       A       G  0.0843  0.1618  0.0419  3.86158  0.04914  2.07697   \n",
       "2       A       G  0.0843  0.1618  0.0419  3.86158  0.04914  2.07697   \n",
       "3       A       G  0.0843  0.1618  0.0419  3.86158  0.04914  2.07697   \n",
       "4       A       G  0.0843  0.1618  0.0419  3.86158  0.04914  2.07697   \n",
       "\n",
       "       mean       sd  mean_incl   sd_incl      pval  bin_start_y  bin_end_y  \\\n",
       "0 -0.001938  0.06338  -0.039446  0.283315  0.000056     19970000   19975000   \n",
       "1 -0.001938  0.06338  -0.039446  0.283315  0.000056     19970000   19975000   \n",
       "2 -0.001938  0.06338  -0.039446  0.283315  0.000056     19970000   19975000   \n",
       "3 -0.001938  0.06338  -0.039446  0.283315  0.000056     19970000   19975000   \n",
       "4 -0.001938  0.06338  -0.039446  0.283315  0.000056     19970000   19975000   \n",
       "\n",
       "             sid                                  loop_id  \n",
       "0  chr1:19972330  19975000:19972330:1:19535000:19540000:1  \n",
       "1  chr1:19972330  19975000:19972330:1:19715000:19720000:1  \n",
       "2  chr1:19972330  19975000:19972330:1:19810000:19815000:1  \n",
       "3  chr1:19972330  19975000:19972330:1:19920000:19925000:1  \n",
       "4  chr1:19972330  19975000:19972330:1:19925000:19930000:1  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_hichip.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "74c4bd66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2240, 35)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_hichip.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ec0240f",
   "metadata": {},
   "source": [
    "## Integrate genes "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1804679b",
   "metadata": {},
   "source": [
    "### Load the gene data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4777d5a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Load the gene data\n",
      "# make a genes pbt for intersection\n",
      "   chrom  start    end        gname          gene_id strand  bin_start  \\\n",
      "0      1  11868  11869      DDX11L1  ENSG00000223972      +      10000   \n",
      "12     1  29569  29570       WASH7P  ENSG00000227232      -      25000   \n",
      "25     1  17435  17436    MIR6859-1  ENSG00000278267      -      15000   \n",
      "28     1  29553  29554  MIR1302-2HG  ENSG00000243485      +      25000   \n",
      "36     1  30365  30366    MIR1302-2  ENSG00000284332      +      30000   \n",
      "\n",
      "    bin_end  \n",
      "0     15000  \n",
      "12    30000  \n",
      "25    20000  \n",
      "28    30000  \n",
      "36    35000  \n",
      "There are 58825 genes in this GTF-derived file.\n"
     ]
    }
   ],
   "source": [
    "print('# Load the gene data')\n",
    "\n",
    "genes_fn = 'results/refs/gencode/v30/gencode.v30.annotation.bed'\n",
    "\n",
    "# load the gencode coords\n",
    "cols = ['chrom', 'start', 'end', 'strand', 'type', 'gene_id', 'gname']\n",
    "gencode = pd.read_table(genes_fn, header=None, names=cols)\n",
    "\n",
    "# extract just the genes\n",
    "genes_df = gencode.loc[gencode['type'].isin(['gene'])]\n",
    "genes_df = genes_df.loc[~genes_df.duplicated(subset='gene_id'), :]\n",
    "genes_df.loc[:, 'chrom'] = genes_df['chrom'].astype(str)\n",
    "genes_df = genes_df.iloc[:, [0,1,2,6,5,3]]\n",
    "\n",
    "# create a copy of the original gene bed before coordinate shrinking\n",
    "orig_genes_df = genes_df.copy()\n",
    "\n",
    "# convert the start/end position into start/end for the TSS\n",
    "# if the gene is + then the start is uses as the tss otherwise\n",
    "# the end is used as the tss\n",
    "genes_df.loc[(genes_df.strand == '+'), 'end'] = genes_df.loc[(genes_df.strand == '+'), 'start']\n",
    "genes_df.loc[(genes_df.strand == '+'), 'start'] = genes_df.loc[(genes_df.strand == '+'), 'start'] - 1\n",
    "genes_df.loc[(genes_df.strand == '-'), 'end'] = genes_df.loc[(genes_df.strand == '-'), 'end']\n",
    "genes_df.loc[(genes_df.strand == '-'), 'start'] = genes_df.loc[(genes_df.strand == '-'), 'end'] - 1\n",
    "genes_df.loc[:, 'chrom'] = genes_df.loc[:, 'chrom'].str.replace('chr', '')\n",
    "genes_df.loc[:, 'bin_start'] = (np.floor(genes_df.loc[:, 'start'] / res) * res).astype(int)\n",
    "genes_df.loc[:, 'bin_end'] = genes_df.loc[:, 'bin_start'] + res\n",
    "\n",
    "# make a genes pbt for intersection\n",
    "print(\"# make a genes pbt for intersection\")\n",
    "print(genes_df.head())\n",
    "genes_pbt = pbt.BedTool.from_dataframe(genes_df).sort()\n",
    "\n",
    "print('There are {} genes in this GTF-derived file.'.format(genes_df.shape[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0c92c29",
   "metadata": {},
   "source": [
    "### Determine the which anchor the SNP falls into"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0a887324",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_anchor = []\n",
    "for i, sr in gwas_hichip.iterrows():\n",
    "    if (sr.startA_loop <= sr.position) & (sr.position <= sr.endA_loop):\n",
    "        snp_anchor.append('AnchorA')\n",
    "    elif (sr.startB_loop <= sr.position) & (sr.position <= sr.endB_loop):\n",
    "        snp_anchor.append('AnchorB')\n",
    "    else:\n",
    "        snp_anchor.append('bug')\n",
    "        print('bug')\n",
    "        break\n",
    "gwas_hichip.loc[:, 'snp_anchor'] = snp_anchor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f822f928",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNP anchor designation: ['AnchorB', 'AnchorA']\n"
     ]
    }
   ],
   "source": [
    "print('SNP anchor designation:', gwas_hichip['snp_anchor'].unique().tolist())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "300491b7",
   "metadata": {},
   "source": [
    "### Extract anchors opposite of a SNP anchor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5fe2e11d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# using a basic serial id for merging post bedtools intersection\n",
    "gwas_hichip['gh_id'] = range(gwas_hichip.shape[0])\n",
    "\n",
    "anchor_cols = ['chrB_loop', 'startB_loop', 'endB_loop', 'gh_id']\n",
    "nonsnp_anchorsA = gwas_hichip.loc[gwas_hichip['snp_anchor'] == 'AnchorA', anchor_cols]\n",
    "anchor_cols =  ['chrA_loop', 'startA_loop', 'endA_loop', 'gh_id']\n",
    "nonsnp_anchorsB = gwas_hichip.loc[gwas_hichip['snp_anchor'] == 'AnchorB', anchor_cols]\n",
    "\n",
    "nonsnp_anchorsA.columns = ['chr', 'start', 'end', 'gh_id']\n",
    "nonsnp_anchorsB.columns = ['chr', 'start', 'end', 'gh_id']\n",
    "nonsnp_anchors = pd.concat([nonsnp_anchorsA, nonsnp_anchorsB], axis=0)\n",
    "nonsnp_anchors_pbt = pbt.BedTool.from_dataframe(nonsnp_anchors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b020a064",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gh_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1</td>\n",
       "      <td>36020000</td>\n",
       "      <td>36025000</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>36020000</td>\n",
       "      <td>36025000</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1</td>\n",
       "      <td>59760000</td>\n",
       "      <td>59765000</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1</td>\n",
       "      <td>65530000</td>\n",
       "      <td>65535000</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1</td>\n",
       "      <td>63925000</td>\n",
       "      <td>63930000</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr     start       end  gh_id\n",
       "22    1  36020000  36025000     22\n",
       "24    1  36020000  36025000     24\n",
       "34    1  59760000  59765000     34\n",
       "35    1  65530000  65535000     35\n",
       "36    1  63925000  63930000     36"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nonsnp_anchors.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17a16801",
   "metadata": {},
   "source": [
    "### Intersecting genes on anchors opposing a SNP anchor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "29f9a59d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_overlaps = nonsnp_anchors_pbt.intersect(genes_pbt, wa=True, wb=True)\n",
    "gene_overlaps = gene_overlaps.to_dataframe(header=None, disable_auto_names=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "da4c79c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of anchor gene overlaps is: (375, 12)\n"
     ]
    }
   ],
   "source": [
    "print('The number of anchor gene overlaps is:', gene_overlaps.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "71cfb7d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>115640000</td>\n",
       "      <td>115645000</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>115641969</td>\n",
       "      <td>115641970</td>\n",
       "      <td>VANGL1</td>\n",
       "      <td>ENSG00000173218</td>\n",
       "      <td>+</td>\n",
       "      <td>115640000</td>\n",
       "      <td>115645000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>7830000</td>\n",
       "      <td>7835000</td>\n",
       "      <td>151</td>\n",
       "      <td>10</td>\n",
       "      <td>7833956</td>\n",
       "      <td>7833957</td>\n",
       "      <td>AL353754.1</td>\n",
       "      <td>ENSG00000233990</td>\n",
       "      <td>-</td>\n",
       "      <td>7830000</td>\n",
       "      <td>7835000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>42660000</td>\n",
       "      <td>42665000</td>\n",
       "      <td>166</td>\n",
       "      <td>10</td>\n",
       "      <td>42660727</td>\n",
       "      <td>42660728</td>\n",
       "      <td>VN1R54P</td>\n",
       "      <td>ENSG00000232109</td>\n",
       "      <td>+</td>\n",
       "      <td>42660000</td>\n",
       "      <td>42665000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "      <td>42660000</td>\n",
       "      <td>42665000</td>\n",
       "      <td>173</td>\n",
       "      <td>10</td>\n",
       "      <td>42660727</td>\n",
       "      <td>42660728</td>\n",
       "      <td>VN1R54P</td>\n",
       "      <td>ENSG00000232109</td>\n",
       "      <td>+</td>\n",
       "      <td>42660000</td>\n",
       "      <td>42665000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>42660000</td>\n",
       "      <td>42665000</td>\n",
       "      <td>178</td>\n",
       "      <td>10</td>\n",
       "      <td>42660727</td>\n",
       "      <td>42660728</td>\n",
       "      <td>VN1R54P</td>\n",
       "      <td>ENSG00000232109</td>\n",
       "      <td>+</td>\n",
       "      <td>42660000</td>\n",
       "      <td>42665000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0          1          2    3   4          5          6           7   \\\n",
       "0   1  115640000  115645000   68   1  115641969  115641970      VANGL1   \n",
       "1  10    7830000    7835000  151  10    7833956    7833957  AL353754.1   \n",
       "2  10   42660000   42665000  166  10   42660727   42660728     VN1R54P   \n",
       "3  10   42660000   42665000  173  10   42660727   42660728     VN1R54P   \n",
       "4  10   42660000   42665000  178  10   42660727   42660728     VN1R54P   \n",
       "\n",
       "                8  9          10         11  \n",
       "0  ENSG00000173218  +  115640000  115645000  \n",
       "1  ENSG00000233990  -    7830000    7835000  \n",
       "2  ENSG00000232109  +   42660000   42665000  \n",
       "3  ENSG00000232109  +   42660000   42665000  \n",
       "4  ENSG00000232109  +   42660000   42665000  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_overlaps.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac2f0f0d",
   "metadata": {},
   "source": [
    "### Add gene overlaps to SNP-Loop Pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a5584a32",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_overlaps.columns = ['chrSNP', 'startSNP', 'endSNP', 'gh_id',\n",
    "                         'chrGene', 'startGene', 'endGene',\n",
    "                         'genename', 'geneid', 'strand', 'bin_start', 'bin_end']\n",
    "gwas_hichip_genes = gwas_hichip.merge(gene_overlaps,\n",
    "                                      on=['gh_id'],\n",
    "                                      how='left')\n",
    "gwas_hichip_genes = gwas_hichip_genes.loc[~gwas_hichip_genes.chrSNP.isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "19c08612",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_snp</th>\n",
       "      <th>bin_start_x</th>\n",
       "      <th>bin_end_x</th>\n",
       "      <th>pos</th>\n",
       "      <th>chrA_loop</th>\n",
       "      <th>startA_loop</th>\n",
       "      <th>endA_loop</th>\n",
       "      <th>chrB_loop</th>\n",
       "      <th>startB_loop</th>\n",
       "      <th>endB_loop</th>\n",
       "      <th>cline_loop</th>\n",
       "      <th>gwas_source</th>\n",
       "      <th>regionID</th>\n",
       "      <th>GWASLoci</th>\n",
       "      <th>index</th>\n",
       "      <th>rsid</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>maf</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>z</th>\n",
       "      <th>prob</th>\n",
       "      <th>log10bf</th>\n",
       "      <th>mean</th>\n",
       "      <th>sd</th>\n",
       "      <th>mean_incl</th>\n",
       "      <th>sd_incl</th>\n",
       "      <th>pval</th>\n",
       "      <th>bin_start_y</th>\n",
       "      <th>bin_end_y</th>\n",
       "      <th>sid</th>\n",
       "      <th>loop_id</th>\n",
       "      <th>snp_anchor</th>\n",
       "      <th>gh_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19535000</td>\n",
       "      <td>19540000</td>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>CD4N</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:19579228-20579228</td>\n",
       "      <td>982</td>\n",
       "      <td>1:19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19972330</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0843</td>\n",
       "      <td>0.1618</td>\n",
       "      <td>0.0419</td>\n",
       "      <td>3.86158</td>\n",
       "      <td>0.04914</td>\n",
       "      <td>2.07697</td>\n",
       "      <td>-0.001938</td>\n",
       "      <td>0.06338</td>\n",
       "      <td>-0.039446</td>\n",
       "      <td>0.283315</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>chr1:19972330</td>\n",
       "      <td>19975000:19972330:1:19535000:19540000:1</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19715000</td>\n",
       "      <td>19720000</td>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>CD4N</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:19579228-20579228</td>\n",
       "      <td>982</td>\n",
       "      <td>1:19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19972330</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0843</td>\n",
       "      <td>0.1618</td>\n",
       "      <td>0.0419</td>\n",
       "      <td>3.86158</td>\n",
       "      <td>0.04914</td>\n",
       "      <td>2.07697</td>\n",
       "      <td>-0.001938</td>\n",
       "      <td>0.06338</td>\n",
       "      <td>-0.039446</td>\n",
       "      <td>0.283315</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>chr1:19972330</td>\n",
       "      <td>19975000:19972330:1:19715000:19720000:1</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19810000</td>\n",
       "      <td>19815000</td>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>CD4N</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:19579228-20579228</td>\n",
       "      <td>982</td>\n",
       "      <td>1:19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19972330</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0843</td>\n",
       "      <td>0.1618</td>\n",
       "      <td>0.0419</td>\n",
       "      <td>3.86158</td>\n",
       "      <td>0.04914</td>\n",
       "      <td>2.07697</td>\n",
       "      <td>-0.001938</td>\n",
       "      <td>0.06338</td>\n",
       "      <td>-0.039446</td>\n",
       "      <td>0.283315</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>chr1:19972330</td>\n",
       "      <td>19975000:19972330:1:19810000:19815000:1</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19920000</td>\n",
       "      <td>19925000</td>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>CD4N</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:19579228-20579228</td>\n",
       "      <td>982</td>\n",
       "      <td>1:19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19972330</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0843</td>\n",
       "      <td>0.1618</td>\n",
       "      <td>0.0419</td>\n",
       "      <td>3.86158</td>\n",
       "      <td>0.04914</td>\n",
       "      <td>2.07697</td>\n",
       "      <td>-0.001938</td>\n",
       "      <td>0.06338</td>\n",
       "      <td>-0.039446</td>\n",
       "      <td>0.283315</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>chr1:19972330</td>\n",
       "      <td>19975000:19972330:1:19920000:19925000:1</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19925000</td>\n",
       "      <td>19930000</td>\n",
       "      <td>1</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>CD4N</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:19579228-20579228</td>\n",
       "      <td>982</td>\n",
       "      <td>1:19972330</td>\n",
       "      <td>1</td>\n",
       "      <td>19972330</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0843</td>\n",
       "      <td>0.1618</td>\n",
       "      <td>0.0419</td>\n",
       "      <td>3.86158</td>\n",
       "      <td>0.04914</td>\n",
       "      <td>2.07697</td>\n",
       "      <td>-0.001938</td>\n",
       "      <td>0.06338</td>\n",
       "      <td>-0.039446</td>\n",
       "      <td>0.283315</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>19970000</td>\n",
       "      <td>19975000</td>\n",
       "      <td>chr1:19972330</td>\n",
       "      <td>19975000:19972330:1:19925000:19930000:1</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr_snp  bin_start_x  bin_end_x       pos  chrA_loop  startA_loop  \\\n",
       "0        1     19970000   19975000  19972330          1     19535000   \n",
       "1        1     19970000   19975000  19972330          1     19715000   \n",
       "2        1     19970000   19975000  19972330          1     19810000   \n",
       "3        1     19970000   19975000  19972330          1     19920000   \n",
       "4        1     19970000   19975000  19972330          1     19925000   \n",
       "\n",
       "   endA_loop  chrB_loop  startB_loop  endB_loop cline_loop   gwas_source  \\\n",
       "0   19540000          1     19970000   19975000       CD4N  T1D_32005708   \n",
       "1   19720000          1     19970000   19975000       CD4N  T1D_32005708   \n",
       "2   19815000          1     19970000   19975000       CD4N  T1D_32005708   \n",
       "3   19925000          1     19970000   19975000       CD4N  T1D_32005708   \n",
       "4   19930000          1     19970000   19975000       CD4N  T1D_32005708   \n",
       "\n",
       "   regionID                GWASLoci  index        rsid  chromosome  position  \\\n",
       "0         1  chr1:19579228-20579228    982  1:19972330           1  19972330   \n",
       "1         1  chr1:19579228-20579228    982  1:19972330           1  19972330   \n",
       "2         1  chr1:19579228-20579228    982  1:19972330           1  19972330   \n",
       "3         1  chr1:19579228-20579228    982  1:19972330           1  19972330   \n",
       "4         1  chr1:19579228-20579228    982  1:19972330           1  19972330   \n",
       "\n",
       "  allele1 allele2     maf    beta      se        z     prob  log10bf  \\\n",
       "0       A       G  0.0843  0.1618  0.0419  3.86158  0.04914  2.07697   \n",
       "1       A       G  0.0843  0.1618  0.0419  3.86158  0.04914  2.07697   \n",
       "2       A       G  0.0843  0.1618  0.0419  3.86158  0.04914  2.07697   \n",
       "3       A       G  0.0843  0.1618  0.0419  3.86158  0.04914  2.07697   \n",
       "4       A       G  0.0843  0.1618  0.0419  3.86158  0.04914  2.07697   \n",
       "\n",
       "       mean       sd  mean_incl   sd_incl      pval  bin_start_y  bin_end_y  \\\n",
       "0 -0.001938  0.06338  -0.039446  0.283315  0.000056     19970000   19975000   \n",
       "1 -0.001938  0.06338  -0.039446  0.283315  0.000056     19970000   19975000   \n",
       "2 -0.001938  0.06338  -0.039446  0.283315  0.000056     19970000   19975000   \n",
       "3 -0.001938  0.06338  -0.039446  0.283315  0.000056     19970000   19975000   \n",
       "4 -0.001938  0.06338  -0.039446  0.283315  0.000056     19970000   19975000   \n",
       "\n",
       "             sid                                  loop_id snp_anchor  gh_id  \n",
       "0  chr1:19972330  19975000:19972330:1:19535000:19540000:1    AnchorB      0  \n",
       "1  chr1:19972330  19975000:19972330:1:19715000:19720000:1    AnchorB      1  \n",
       "2  chr1:19972330  19975000:19972330:1:19810000:19815000:1    AnchorB      2  \n",
       "3  chr1:19972330  19975000:19972330:1:19920000:19925000:1    AnchorB      3  \n",
       "4  chr1:19972330  19975000:19972330:1:19925000:19930000:1    AnchorB      4  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_hichip.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "6c055db1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2240, 37)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_hichip.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "6979fcca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['T1D_32005708', 'T1D_34594039_GCST90018925', 'T1D_34012112_Gaulton']"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_hichip_genes.gwas_source.unique().tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "597c9cd5",
   "metadata": {},
   "source": [
    "#### Make a table of uniq SNPs and Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "66d8b2f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find the unique SNPs\n",
    "uniq_snps_by_cells = gwas_hichip_genes.groupby('cline_loop').sid.nunique()\n",
    "uniq_snps_by_cells = uniq_snps_by_cells.to_frame()\n",
    "\n",
    "# find the unique genes \n",
    "uniq_genes_by_cells = gwas_hichip_genes.groupby('cline_loop').geneid.nunique()\n",
    "uniq_genes_by_cells = uniq_genes_by_cells.to_frame()\n",
    "\n",
    "# merge snps and genes\n",
    "uniq_counts_by_cells = pd.merge(uniq_snps_by_cells, uniq_genes_by_cells, left_index=True, right_index=True)\n",
    "uniq_counts_by_cells.columns = ['Number of Unique SNPs', 'Number of Unique Genes']\n",
    "uniq_counts_by_cells.index.name = 'Cell Line'\n",
    "\n",
    "# save the file\n",
    "excel_analysis = os.path.join(outdir, 'Unique_Counts_By_Cell_Line.xlsx')\n",
    "uniq_counts_by_cells.to_excel(excel_analysis, sheet_name='finemapping')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d5f0762f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'results/main/Intersect_T1D_Finemap_GWAS_SNPs_with_HiChIP/Unique_Counts_By_Cell_Line.xlsx'"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "excel_analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "121499ee",
   "metadata": {},
   "source": [
    "#### Write the gene list as well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "d5a69fa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "gh_list = gwas_hichip_genes.geneid.unique()\n",
    "gh_fn = os.path.join(outdir, 'gene_list.txt')\n",
    "with open(gh_fn, 'w') as fw:\n",
    "    for x in gh_list:\n",
    "        fw.write('{}\\n'.format(x))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6408a24",
   "metadata": {},
   "source": [
    "#### Get the unique genes per cell type "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a5faabba",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_by_cell = gwas_hichip_genes[['cline_loop', 'geneid']].drop_duplicates()\n",
    "genes_by_cell.sort_values(['cline_loop', 'geneid'], inplace=True)\n",
    "genes_by_cell['source'] = 'finemap_with_hichip'\n",
    "genes_by_cell.columns = ['cline', 'geneid', 'source']\n",
    "fn = os.path.join(outdir, 'genes_by_cell.xlsx')\n",
    "genes_by_cell.to_excel(fn, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a87355e2",
   "metadata": {},
   "source": [
    "## Summarize the SNPs, Loops and Intersection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "822d1071",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_gwas = gwas_df.shape[0]\n",
    "# # summarize the total number of GWAS loops per cell # DOESN't MAKE SENSE\n",
    "# cell_summary['total_gwas'] = gwas_hichip.groupby('cline_loop').nunique('sid')['chr_snp']\n",
    "# cell_summary['total_gwas'] = cell_summary['total_gwas'].to_frame()\n",
    "# cell_summary['total_gwas'].columns = ['Total GWAS SNPs']\n",
    "# cell_summary['total_gwas']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "1e91d3cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_summary = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6452a2b8",
   "metadata": {},
   "source": [
    "### Summarize the Number of Loops per Cell (pre-intersection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "2b34797f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total_hichip</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cline</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD4N</th>\n",
       "      <td>114421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8N</th>\n",
       "      <td>84599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM</th>\n",
       "      <td>84298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NB</th>\n",
       "      <td>128288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK</th>\n",
       "      <td>129890</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       total_hichip\n",
       "cline              \n",
       "CD4N         114421\n",
       "CD8N          84599\n",
       "CM            84298\n",
       "NB           128288\n",
       "NK           129890"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_summary['total_loops'] = loop_df.groupby('cline').count()['startA'].to_frame()\n",
    "cell_summary['total_loops'].columns = ['total_hichip']\n",
    "cell_summary['total_loops']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08ae84d7",
   "metadata": {},
   "source": [
    "### Summarize the Number of SNP-Loop (SL) Pairs per Cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "1d116fc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sl_pairs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NB</th>\n",
       "      <td>712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK</th>\n",
       "      <td>529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4N</th>\n",
       "      <td>411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM</th>\n",
       "      <td>295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8N</th>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sl_pairs\n",
       "NB         712\n",
       "NK         529\n",
       "CD4N       411\n",
       "CM         295\n",
       "CD8N       293"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_summary['sl_pairs'] = gwas_hichip['cline_loop'].value_counts().to_frame()\n",
    "cell_summary['sl_pairs'].columns = ['sl_pairs']\n",
    "cell_summary['sl_pairs']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c468bc58",
   "metadata": {},
   "source": [
    "### Summarize the Number of Unique GWAS SNPs which Overlap a HiChIP Loop Cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "42cc7348",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniq_gwas_in_slpairs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cline_loop</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD4N</th>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8N</th>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM</th>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NB</th>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK</th>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            uniq_gwas_in_slpairs\n",
       "cline_loop                      \n",
       "CD4N                          91\n",
       "CD8N                          74\n",
       "CM                            79\n",
       "NB                           147\n",
       "NK                           111"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_summary['uniq_gwas'] = gwas_hichip.groupby('cline_loop')['sid'].nunique().to_frame()\n",
    "cell_summary['uniq_gwas'].columns = ['uniq_gwas_in_slpairs']\n",
    "cell_summary['uniq_gwas']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eec14a11",
   "metadata": {},
   "source": [
    "### Summarize the Number of Loops with GWAS Overlaps (per cell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "3a8f01e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniq_loops_in_slpairs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cline_loop</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD4N</th>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8N</th>\n",
       "      <td>211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM</th>\n",
       "      <td>226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NB</th>\n",
       "      <td>470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK</th>\n",
       "      <td>342</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            uniq_loops_in_slpairs\n",
       "cline_loop                       \n",
       "CD4N                          287\n",
       "CD8N                          211\n",
       "CM                            226\n",
       "NB                            470\n",
       "NK                            342"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loop_cols = ['chrA_loop', 'startA_loop', 'endA_loop', 'chrB_loop', 'startB_loop', 'endB_loop']\n",
    "cell_summary['uniq_loops'] = gwas_hichip.groupby('cline_loop')['loop_id'].nunique().to_frame()\n",
    "cell_summary['uniq_loops'].columns = ['uniq_loops_in_slpairs']\n",
    "cell_summary['uniq_loops']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "091f4b50",
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_list = [cell_summary['total_loops'], cell_summary['sl_pairs'],\n",
    "               cell_summary['uniq_gwas'], cell_summary['uniq_loops']]\n",
    "summary = pd.concat(concat_list, axis=1)\n",
    "summary['pct_uniq_gwas_in_slpairs'] = summary['uniq_gwas_in_slpairs'] / total_gwas * 100\n",
    "summary['pct_uniq_loops_in_slpairs'] = summary['uniq_loops_in_slpairs'] / summary['total_hichip'] * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "371814c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total_hichip</th>\n",
       "      <th>sl_pairs</th>\n",
       "      <th>uniq_gwas_in_slpairs</th>\n",
       "      <th>uniq_loops_in_slpairs</th>\n",
       "      <th>pct_uniq_gwas_in_slpairs</th>\n",
       "      <th>pct_uniq_loops_in_slpairs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD4N</th>\n",
       "      <td>114421</td>\n",
       "      <td>411</td>\n",
       "      <td>91</td>\n",
       "      <td>287</td>\n",
       "      <td>12.816901</td>\n",
       "      <td>0.250828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8N</th>\n",
       "      <td>84599</td>\n",
       "      <td>293</td>\n",
       "      <td>74</td>\n",
       "      <td>211</td>\n",
       "      <td>10.422535</td>\n",
       "      <td>0.249412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM</th>\n",
       "      <td>84298</td>\n",
       "      <td>295</td>\n",
       "      <td>79</td>\n",
       "      <td>226</td>\n",
       "      <td>11.126761</td>\n",
       "      <td>0.268097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NB</th>\n",
       "      <td>128288</td>\n",
       "      <td>712</td>\n",
       "      <td>147</td>\n",
       "      <td>470</td>\n",
       "      <td>20.704225</td>\n",
       "      <td>0.366363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK</th>\n",
       "      <td>129890</td>\n",
       "      <td>529</td>\n",
       "      <td>111</td>\n",
       "      <td>342</td>\n",
       "      <td>15.633803</td>\n",
       "      <td>0.263300</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total_hichip  sl_pairs  uniq_gwas_in_slpairs  uniq_loops_in_slpairs  \\\n",
       "CD4N        114421       411                    91                    287   \n",
       "CD8N         84599       293                    74                    211   \n",
       "CM           84298       295                    79                    226   \n",
       "NB          128288       712                   147                    470   \n",
       "NK          129890       529                   111                    342   \n",
       "\n",
       "      pct_uniq_gwas_in_slpairs  pct_uniq_loops_in_slpairs  \n",
       "CD4N                 12.816901                   0.250828  \n",
       "CD8N                 10.422535                   0.249412  \n",
       "CM                   11.126761                   0.268097  \n",
       "NB                   20.704225                   0.366363  \n",
       "NK                   15.633803                   0.263300  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "bf5dbdb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_summary = summary.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "cbb2cb48",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_colnames = ['Total\\\\nHiChIP Loops', \n",
    "                  'Number of\\\\nGWAS-Loop Pairs',\n",
    "                  'Number of\\\\nUnique GWAS SNPs in GL Pairs', \n",
    "                  'Number of\\\\nUnique loops in GL Pairs',\n",
    "                  'Percentage of\\\\nUnique GWAS SNPs in GL Pairs', \n",
    "                  'Percentage of\\\\nUnique loops in GL Pairs']\n",
    "final_colnames = ['Total HiChIP Loops', \n",
    "                  'Number of GWAS-Loop Pairs',\n",
    "                  'Number of Unique GWAS SNPs in GL Pairs', \n",
    "                  'Number of Unique loops in GL Pairs',\n",
    "                  'Percentage of Unique GWAS SNPs in GL Pairs', \n",
    "                  'Percentage of Unique loops in GL Pairs']\n",
    "final_summary.columns = final_colnames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "a2180e2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Total HiChIP Loops</th>\n",
       "      <th>Number of GWAS-Loop Pairs</th>\n",
       "      <th>Number of Unique GWAS SNPs in GL Pairs</th>\n",
       "      <th>Number of Unique loops in GL Pairs</th>\n",
       "      <th>Percentage of Unique GWAS SNPs in GL Pairs</th>\n",
       "      <th>Percentage of Unique loops in GL Pairs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD4N</th>\n",
       "      <td>114421</td>\n",
       "      <td>411</td>\n",
       "      <td>91</td>\n",
       "      <td>287</td>\n",
       "      <td>12.816901</td>\n",
       "      <td>0.250828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8N</th>\n",
       "      <td>84599</td>\n",
       "      <td>293</td>\n",
       "      <td>74</td>\n",
       "      <td>211</td>\n",
       "      <td>10.422535</td>\n",
       "      <td>0.249412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM</th>\n",
       "      <td>84298</td>\n",
       "      <td>295</td>\n",
       "      <td>79</td>\n",
       "      <td>226</td>\n",
       "      <td>11.126761</td>\n",
       "      <td>0.268097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NB</th>\n",
       "      <td>128288</td>\n",
       "      <td>712</td>\n",
       "      <td>147</td>\n",
       "      <td>470</td>\n",
       "      <td>20.704225</td>\n",
       "      <td>0.366363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK</th>\n",
       "      <td>129890</td>\n",
       "      <td>529</td>\n",
       "      <td>111</td>\n",
       "      <td>342</td>\n",
       "      <td>15.633803</td>\n",
       "      <td>0.263300</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(HTML(final_summary.to_html().replace(\"\\\\n\",\"<br>\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54b8ec78",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c2110fd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21b40f9f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d155cb24",
   "metadata": {},
   "source": [
    "## Investigate "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "f8c6d9ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_snp</th>\n",
       "      <th>bin_start_x</th>\n",
       "      <th>bin_end_x</th>\n",
       "      <th>pos</th>\n",
       "      <th>chrA_loop</th>\n",
       "      <th>startA_loop</th>\n",
       "      <th>endA_loop</th>\n",
       "      <th>chrB_loop</th>\n",
       "      <th>startB_loop</th>\n",
       "      <th>endB_loop</th>\n",
       "      <th>cline_loop</th>\n",
       "      <th>gwas_source</th>\n",
       "      <th>regionID</th>\n",
       "      <th>GWASLoci</th>\n",
       "      <th>index</th>\n",
       "      <th>rsid</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>maf</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>z</th>\n",
       "      <th>prob</th>\n",
       "      <th>log10bf</th>\n",
       "      <th>mean</th>\n",
       "      <th>sd</th>\n",
       "      <th>mean_incl</th>\n",
       "      <th>sd_incl</th>\n",
       "      <th>pval</th>\n",
       "      <th>bin_start_y</th>\n",
       "      <th>bin_end_y</th>\n",
       "      <th>sid</th>\n",
       "      <th>loop_id</th>\n",
       "      <th>snp_anchor</th>\n",
       "      <th>gh_id</th>\n",
       "      <th>chrSNP</th>\n",
       "      <th>startSNP</th>\n",
       "      <th>endSNP</th>\n",
       "      <th>chrGene</th>\n",
       "      <th>startGene</th>\n",
       "      <th>endGene</th>\n",
       "      <th>genename</th>\n",
       "      <th>geneid</th>\n",
       "      <th>strand</th>\n",
       "      <th>bin_start</th>\n",
       "      <th>bin_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chr_snp, bin_start_x, bin_end_x, pos, chrA_loop, startA_loop, endA_loop, chrB_loop, startB_loop, endB_loop, cline_loop, gwas_source, regionID, GWASLoci, index, rsid, chromosome, position, allele1, allele2, maf, beta, se, z, prob, log10bf, mean, sd, mean_incl, sd_incl, pval, bin_start_y, bin_end_y, sid, loop_id, snp_anchor, gh_id, chrSNP, startSNP, endSNP, chrGene, startGene, endGene, genename, geneid, strand, bin_start, bin_end]\n",
       "Index: []"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_hichip_genes.loc[gwas_hichip_genes.genename == 'ZKSCAN4']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab4085da",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f66bfe3d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb4d6f5b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9578328e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "HiChIP-DB",
   "language": "python",
   "name": "hichip-db"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
