{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7b5ad180",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import glob\n",
    "import pybedtools as pbt \n",
    "from IPython.display import HTML\n",
    "pd.set_option('display.max_rows', 200) \n",
    "pd.set_option('display.max_columns', None)\n",
    "import upsetplot\n",
    "\n",
    "import seaborn as sns \n",
    "import matplotlib.pyplot as plt \n",
    "\n",
    "pbt.set_bedtools_path('/mnt/BioHome/jreyna/software/anaconda3/envs/hic_tls/bin/')\n",
    "#pbt.set_tempdir('/mnt/hpcscratch/jreyna/')\n",
    "os.chdir('/mnt/bioadhoc-temp/Groups/vd-ay/jreyna/projects/t1d-loop-catalog/')\n",
    "\n",
    "gsizes = 'results/hg38/refs/hg38/hg38.chrom.sizes'\n",
    "res = 5000\n",
    "\n",
    "# make the directory to save our data\n",
    "outdir = 'results/hg38/finemapping/sgl_intersect/'\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "bedpe_cols = ['chrA', 'startA', 'endA', 'chrB', 'startB', 'endB']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25ec296e",
   "metadata": {},
   "source": [
    "## Load Fine Mapped GWAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5268fb2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "fn = 'results/hg38/external_studies/chiou_2021/processing/finemapping/hg38.finemapping.basic.bed'\n",
    "# get the bin coordinates\n",
    "gwas_df = pd.read_table(fn)\n",
    "# df.loc[:, 'bin_start'] = np.floor(df.loc[:, 'position'] / res).astype(int) * res\n",
    "# df.loc[:, 'bin_end'] = df.loc[:, 'bin_start'] + res\n",
    "# df = df.loc[(df.allele1.str.len() == 1 ) & (df.allele2.str.len() == 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b3d61541",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46738, 5)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ef5ebec5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter for PIP > 0.05\n",
    "gwas_df = gwas_df.loc[gwas_df.ppa > 0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0ce6ac3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>rsid</th>\n",
       "      <th>ppa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>1</td>\n",
       "      <td>35571620</td>\n",
       "      <td>35571621</td>\n",
       "      <td>rs6425948</td>\n",
       "      <td>0.130213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>1</td>\n",
       "      <td>35622059</td>\n",
       "      <td>35622060</td>\n",
       "      <td>rs574384</td>\n",
       "      <td>0.195293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>1</td>\n",
       "      <td>35641908</td>\n",
       "      <td>35641909</td>\n",
       "      <td>rs676614</td>\n",
       "      <td>0.186331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>1</td>\n",
       "      <td>35666944</td>\n",
       "      <td>35666945</td>\n",
       "      <td>rs596471</td>\n",
       "      <td>0.084043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>1</td>\n",
       "      <td>37879348</td>\n",
       "      <td>37879349</td>\n",
       "      <td>rs4585948</td>\n",
       "      <td>0.064023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end       rsid       ppa\n",
       "105      1  35571620  35571621  rs6425948  0.130213\n",
       "112      1  35622059  35622060   rs574384  0.195293\n",
       "115      1  35641908  35641909   rs676614  0.186331\n",
       "117      1  35666944  35666945   rs596471  0.084043\n",
       "236      1  37879348  37879349  rs4585948  0.064023"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "391985fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(490, 5)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "46298a0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a pybedtools for finemap data\n",
    "gwas_pbt = pbt.BedTool.from_dataframe(gwas_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b1e3652",
   "metadata": {},
   "source": [
    "## Load HiChIP Loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6cd23b57",
   "metadata": {},
   "outputs": [],
   "source": [
    "# only analyze loop data from main cell types \n",
    "loop_fn = 'results/hg38/loops/hichip/chip-seq/macs2/loose/Jurkat.GSE99519.Homo_Sapiens.YY1.b1.25000.interactions_FitHiC_Q0.01.bed'\n",
    "loop_df = pd.read_table(loop_fn)\n",
    "loop_df['-log10_qval'] =  loop_df['Q-Value_Bias'].apply(lambda x: -np.log(x))\n",
    "loop_df['chr1'] = loop_df['chr1'].str.replace('chr', '') \n",
    "loop_df['chr2'] = loop_df['chr2'].str.replace('chr', '') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6b3d12cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19844, 27)\n"
     ]
    }
   ],
   "source": [
    "print(loop_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "541b020f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a dataframe in bed format which filters for significant\n",
    "# SNPs only p-val < 0.05 (or -log10(p-val) > 1.3)\n",
    "loop_bed = loop_df.iloc[:, [0,1,2,3,4,5,-1]]\n",
    "\n",
    "# create a pybedtools for the looping data\n",
    "loop_pbt = pbt.BedTool.from_dataframe(loop_bed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cb73e3db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr1</th>\n",
       "      <th>s1</th>\n",
       "      <th>e1</th>\n",
       "      <th>chr2</th>\n",
       "      <th>s2</th>\n",
       "      <th>e2</th>\n",
       "      <th>-log10_qval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>975000</td>\n",
       "      <td>1000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1300000</td>\n",
       "      <td>1325000</td>\n",
       "      <td>7.087591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1000000</td>\n",
       "      <td>1025000</td>\n",
       "      <td>1</td>\n",
       "      <td>1225000</td>\n",
       "      <td>1250000</td>\n",
       "      <td>6.426370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1000000</td>\n",
       "      <td>1025000</td>\n",
       "      <td>1</td>\n",
       "      <td>1250000</td>\n",
       "      <td>1275000</td>\n",
       "      <td>19.420019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1000000</td>\n",
       "      <td>1025000</td>\n",
       "      <td>1</td>\n",
       "      <td>1275000</td>\n",
       "      <td>1300000</td>\n",
       "      <td>7.210015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1025000</td>\n",
       "      <td>1050000</td>\n",
       "      <td>1</td>\n",
       "      <td>1400000</td>\n",
       "      <td>1425000</td>\n",
       "      <td>6.051496</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr1       s1       e1 chr2       s2       e2  -log10_qval\n",
       "0    1   975000  1000000    1  1300000  1325000     7.087591\n",
       "1    1  1000000  1025000    1  1225000  1250000     6.426370\n",
       "2    1  1000000  1025000    1  1250000  1275000    19.420019\n",
       "3    1  1000000  1025000    1  1275000  1300000     7.210015\n",
       "4    1  1025000  1050000    1  1400000  1425000     6.051496"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loop_bed.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bf47849",
   "metadata": {},
   "source": [
    "## Intersect Fine Mapped GWAS and loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a49920c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>35571620</td>\n",
       "      <td>35571621</td>\n",
       "      <td>rs6425948</td>\n",
       "      <td>0.130213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>35622059</td>\n",
       "      <td>35622060</td>\n",
       "      <td>rs574384</td>\n",
       "      <td>0.195293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>35641908</td>\n",
       "      <td>35641909</td>\n",
       "      <td>rs676614</td>\n",
       "      <td>0.186331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>35666944</td>\n",
       "      <td>35666945</td>\n",
       "      <td>rs596471</td>\n",
       "      <td>0.084043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>37879348</td>\n",
       "      <td>37879349</td>\n",
       "      <td>rs4585948</td>\n",
       "      <td>0.064023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>9</td>\n",
       "      <td>4283136</td>\n",
       "      <td>4283137</td>\n",
       "      <td>rs1574285</td>\n",
       "      <td>0.213968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>9</td>\n",
       "      <td>4289049</td>\n",
       "      <td>4289050</td>\n",
       "      <td>rs7034200</td>\n",
       "      <td>0.186268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487</th>\n",
       "      <td>9</td>\n",
       "      <td>4289195</td>\n",
       "      <td>4289196</td>\n",
       "      <td>rs10814914</td>\n",
       "      <td>0.089392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488</th>\n",
       "      <td>9</td>\n",
       "      <td>4292082</td>\n",
       "      <td>4292083</td>\n",
       "      <td>rs10758593</td>\n",
       "      <td>0.069463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>489</th>\n",
       "      <td>9</td>\n",
       "      <td>133060937</td>\n",
       "      <td>133060938</td>\n",
       "      <td>rs541856133</td>\n",
       "      <td>0.999987</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>490 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end         name     score\n",
       "0        1   35571620   35571621    rs6425948  0.130213\n",
       "1        1   35622059   35622060     rs574384  0.195293\n",
       "2        1   35641908   35641909     rs676614  0.186331\n",
       "3        1   35666944   35666945     rs596471  0.084043\n",
       "4        1   37879348   37879349    rs4585948  0.064023\n",
       "..     ...        ...        ...          ...       ...\n",
       "485      9    4283136    4283137    rs1574285  0.213968\n",
       "486      9    4289049    4289050    rs7034200  0.186268\n",
       "487      9    4289195    4289196   rs10814914  0.089392\n",
       "488      9    4292082    4292083   rs10758593  0.069463\n",
       "489      9  133060937  133060938  rs541856133  0.999987\n",
       "\n",
       "[490 rows x 5 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_pbt.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca1bd83c",
   "metadata": {},
   "source": [
    "#### Perform the intersection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "09fb40ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# intersecting the GWAS SNPs with loops \n",
    "# loop anchor doesn't matter therefore type='either'\n",
    "intersect_pbt = loop_pbt.pair_to_bed(gwas_pbt, type='either')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "00c9b837",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # convert the intersection into a dataframe and rename columns \n",
    "gwas_hichip = intersect_pbt.to_dataframe(header=None, disable_auto_names=True)\n",
    "\n",
    "if len(gwas_hichip) == 0:\n",
    "    print('No overlap between GWAS and HiChIP loops')\n",
    "    \n",
    "gwas_hichip.columns = ['chrA_loop', 'startA_loop', 'endA_loop',\n",
    "                       'chrB_loop', 'startB_loop', 'endB_loop', \n",
    "                       '-log10_qval', 'chr_snp', 'start_snp', 'end_snp', 'rsid', 'ppa_snp']\n",
    "\n",
    "# gwas_hichip = gwas_hichip.iloc[:, [7,8,9,10,0,1,2,3,4,5,6,11]]\n",
    "# loop_cols = ['{}_loop'.format(x) for x in bedpe_cols]\n",
    "# gwas_hichip.columns = ['chr_snp', 'bin_start', 'bin_end', 'pos'] + loop_cols + ['cline_loop', 'gwas_source']\n",
    "\n",
    "# # add back fields from the original gwas data\n",
    "# # some data was lost in the pybedtool process\n",
    "# gwas_hichip = gwas_hichip.merge(gwas_df.drop('gwas_source', axis=1),\n",
    "#                                 left_on=['chr_snp', 'pos'],\n",
    "#                                 right_on=['chromosome', 'position'])\n",
    "# # add the sid as the variant id\n",
    "# gwas_hichip['sid'] = 'chr' +  gwas_hichip['chr_snp'].astype(str) + ':' + gwas_hichip['position'].astype(str)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ec0240f",
   "metadata": {},
   "source": [
    "## Integrate genes and located SGLs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa3498fe",
   "metadata": {},
   "source": [
    "For the types of SGLs through this approach we have to find identify loops where one anchor overlaps a SNP and the other one overlaps a gene promoter. As such, we used the previously intersected data to find loop anchors without a SNP overlap and we will intersect these anchors with gene promoters next. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1804679b",
   "metadata": {},
   "source": [
    "### Load the gene data and build promoter based pybedtool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4777d5a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Load the gene data\n"
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mCanceled future for execute_request message before replies were done"
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the the current cell or a previous cell. Please review the code in the cell(s) to identify a possible cause of the failure. Click <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. View Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "print('# Load the gene data')\n",
    "\n",
    "genes_fn = 'results/hg38/refs/gencode/v30/gencode.v30.annotation.bed'\n",
    "\n",
    "# load the gencode coords\n",
    "cols = ['chrom', 'start', 'end', 'strand', 'type', 'gene_id', 'gname']\n",
    "gencode = pd.read_table(genes_fn, header=None, names=cols)\n",
    "\n",
    "# extract just the genes\n",
    "genes_df = gencode.loc[gencode['type'].isin(['gene'])]\n",
    "genes_df = genes_df.loc[~genes_df.duplicated(subset='gene_id'), :]\n",
    "genes_df.loc[:, 'chrom'] = genes_df['chrom'].astype(str)\n",
    "genes_df = genes_df.iloc[:, [0,1,2,6,5,3]]\n",
    "\n",
    "# create a copy of the original gene bed before coordinate shrinking\n",
    "orig_genes_df = genes_df.copy()\n",
    "\n",
    "# convert the start/end position into start/end for the TSS\n",
    "# if the gene is + then the start is uses as the tss otherwise\n",
    "# the end is used as the tss\n",
    "genes_df.loc[(genes_df.strand == '+'), 'end'] = genes_df.loc[(genes_df.strand == '+'), 'start']\n",
    "genes_df.loc[(genes_df.strand == '+'), 'start'] = genes_df.loc[(genes_df.strand == '+'), 'start'] - 1\n",
    "genes_df.loc[(genes_df.strand == '-'), 'end'] = genes_df.loc[(genes_df.strand == '-'), 'end']\n",
    "genes_df.loc[(genes_df.strand == '-'), 'start'] = genes_df.loc[(genes_df.strand == '-'), 'end'] - 1\n",
    "genes_df.loc[:, 'chrom'] = genes_df.loc[:, 'chrom'].str.replace('chr', '')\n",
    "genes_df.loc[:, 'bin_start'] = (np.floor(genes_df.loc[:, 'start'] / res) * res).astype(int)\n",
    "genes_df.loc[:, 'bin_end'] = genes_df.loc[:, 'bin_start'] + res\n",
    "\n",
    "# make a genes pbt for intersection\n",
    "print(\"# make a genes pbt for intersection\")\n",
    "print(genes_df.head())\n",
    "genes_pbt = pbt.BedTool.from_dataframe(genes_df).sort()\n",
    "\n",
    "print('There are {} genes in this GTF-derived file.'.format(genes_df.shape[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0c92c29",
   "metadata": {},
   "source": [
    "### Determine which anchor the SNP falls into"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a887324",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_anchor = []\n",
    "for i, sr in gwas_hichip.iterrows():\n",
    "    if (sr.startA_loop <= sr.end_snp) & (sr.end_snp <= sr.endA_loop):\n",
    "        snp_anchor.append('AnchorA')\n",
    "    elif (sr.startB_loop <= sr.end_snp) & (sr.end_snp <= sr.endB_loop):\n",
    "        snp_anchor.append('AnchorB')\n",
    "    else:\n",
    "        snp_anchor.append('bug')\n",
    "        print('bug')\n",
    "        break\n",
    "gwas_hichip.loc[:, 'snp_anchor'] = snp_anchor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f822f928",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNP anchor designation: ['AnchorB', 'AnchorA']\n"
     ]
    }
   ],
   "source": [
    "print('SNP anchor designation:', gwas_hichip['snp_anchor'].unique().tolist())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "300491b7",
   "metadata": {},
   "source": [
    "### Extract anchors opposite of a SNP anchor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fe2e11d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# using a basic serial id for merging post bedtools intersection\n",
    "gwas_hichip['gh_id'] = range(gwas_hichip.shape[0])\n",
    "\n",
    "anchor_cols = ['chrB_loop', 'startB_loop', 'endB_loop', 'gh_id']\n",
    "nonsnp_anchorsA = gwas_hichip.loc[gwas_hichip['snp_anchor'] == 'AnchorA', anchor_cols]\n",
    "anchor_cols =  ['chrA_loop', 'startA_loop', 'endA_loop', 'gh_id']\n",
    "nonsnp_anchorsB = gwas_hichip.loc[gwas_hichip['snp_anchor'] == 'AnchorB', anchor_cols]\n",
    "\n",
    "nonsnp_anchorsA.columns = ['chr', 'start', 'end', 'gh_id']\n",
    "nonsnp_anchorsB.columns = ['chr', 'start', 'end', 'gh_id']\n",
    "nonsnp_anchors = pd.concat([nonsnp_anchorsA, nonsnp_anchorsB], axis=0)\n",
    "nonsnp_anchors_pbt = pbt.BedTool.from_dataframe(nonsnp_anchors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b020a064",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gh_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>35625000</td>\n",
       "      <td>35650000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>35700000</td>\n",
       "      <td>35725000</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>35725000</td>\n",
       "      <td>35750000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>35750000</td>\n",
       "      <td>35775000</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>35775000</td>\n",
       "      <td>35800000</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr     start       end  gh_id\n",
       "2    1  35625000  35650000      2\n",
       "4    1  35700000  35725000      4\n",
       "5    1  35725000  35750000      5\n",
       "6    1  35750000  35775000      6\n",
       "7    1  35775000  35800000      7"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nonsnp_anchors.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17a16801",
   "metadata": {},
   "source": [
    "### Intersecting genes on anchors opposing a SNP anchor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29f9a59d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_overlaps = nonsnp_anchors_pbt.intersect(genes_pbt, wa=True, wb=True)\n",
    "gene_overlaps = gene_overlaps.to_dataframe(header=None, disable_auto_names=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da4c79c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of anchor gene overlaps is: (798, 12)\n"
     ]
    }
   ],
   "source": [
    "print('The number of anchor gene overlaps is:', gene_overlaps.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71cfb7d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>35625000</td>\n",
       "      <td>35650000</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>35641525</td>\n",
       "      <td>35641526</td>\n",
       "      <td>PSMB2</td>\n",
       "      <td>ENSG00000126067</td>\n",
       "      <td>-</td>\n",
       "      <td>35640000</td>\n",
       "      <td>35645000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>35700000</td>\n",
       "      <td>35725000</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>35718893</td>\n",
       "      <td>35718894</td>\n",
       "      <td>C1orf216</td>\n",
       "      <td>ENSG00000142686</td>\n",
       "      <td>-</td>\n",
       "      <td>35715000</td>\n",
       "      <td>35720000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>35725000</td>\n",
       "      <td>35750000</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>35739388</td>\n",
       "      <td>35739389</td>\n",
       "      <td>AL354864.1</td>\n",
       "      <td>ENSG00000232335</td>\n",
       "      <td>+</td>\n",
       "      <td>35735000</td>\n",
       "      <td>35740000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>35750000</td>\n",
       "      <td>35775000</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>35769977</td>\n",
       "      <td>35769978</td>\n",
       "      <td>CLSPN</td>\n",
       "      <td>ENSG00000092853</td>\n",
       "      <td>-</td>\n",
       "      <td>35765000</td>\n",
       "      <td>35770000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>37975000</td>\n",
       "      <td>38000000</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>37990074</td>\n",
       "      <td>37990075</td>\n",
       "      <td>SF3A3</td>\n",
       "      <td>ENSG00000183431</td>\n",
       "      <td>-</td>\n",
       "      <td>37990000</td>\n",
       "      <td>37995000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0         1         2   3   4         5         6           7   \\\n",
       "0   1  35625000  35650000   2   1  35641525  35641526       PSMB2   \n",
       "1   1  35700000  35725000   4   1  35718893  35718894    C1orf216   \n",
       "2   1  35725000  35750000   5   1  35739388  35739389  AL354864.1   \n",
       "3   1  35750000  35775000   6   1  35769977  35769978       CLSPN   \n",
       "4   1  37975000  38000000   8   1  37990074  37990075       SF3A3   \n",
       "\n",
       "                8  9         10        11  \n",
       "0  ENSG00000126067  -  35640000  35645000  \n",
       "1  ENSG00000142686  -  35715000  35720000  \n",
       "2  ENSG00000232335  +  35735000  35740000  \n",
       "3  ENSG00000092853  -  35765000  35770000  \n",
       "4  ENSG00000183431  -  37990000  37995000  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_overlaps.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac2f0f0d",
   "metadata": {},
   "source": [
    "### Add gene overlaps to SNP-Loop Pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d1ec78d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_overlaps.columns = ['chr_anchor', 'start_anchor', 'end_anchor', 'gh_id',\n",
    "                         'chr_gene', 'start_gene', 'end_gene',\n",
    "                         'genename', 'geneid', 'strand', 'bin_start', 'bin_end']\n",
    "gwas_hichip_genes = gwas_hichip.merge(gene_overlaps.drop(['', '', ''], axis=1), on=['gh_id'], how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1785ba34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 798 SGLs.\n"
     ]
    }
   ],
   "source": [
    "print('There are {} SGLs.'.format(len(gwas_hichip_genes)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f9e9313",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are rsid    184\n",
      "dtype: int64 unique SNPs within an SGL\n"
     ]
    }
   ],
   "source": [
    "print('There are {} unique SNPs within an SGL'.format(gwas_hichip_genes[['rsid']].nunique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbac1a19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrA_loop</th>\n",
       "      <th>startA_loop</th>\n",
       "      <th>endA_loop</th>\n",
       "      <th>chrB_loop</th>\n",
       "      <th>startB_loop</th>\n",
       "      <th>endB_loop</th>\n",
       "      <th>-log10_qval</th>\n",
       "      <th>chr_snp</th>\n",
       "      <th>start_snp</th>\n",
       "      <th>end_snp</th>\n",
       "      <th>rsid</th>\n",
       "      <th>ppa_snp</th>\n",
       "      <th>snp_anchor</th>\n",
       "      <th>gh_id</th>\n",
       "      <th>chr_anchor</th>\n",
       "      <th>start_anchor</th>\n",
       "      <th>end_anchor</th>\n",
       "      <th>chr_gene</th>\n",
       "      <th>start_gene</th>\n",
       "      <th>end_gene</th>\n",
       "      <th>genename</th>\n",
       "      <th>geneid</th>\n",
       "      <th>strand</th>\n",
       "      <th>bin_start</th>\n",
       "      <th>bin_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>35550000</td>\n",
       "      <td>35575000</td>\n",
       "      <td>1</td>\n",
       "      <td>35625000</td>\n",
       "      <td>35650000</td>\n",
       "      <td>4.919412</td>\n",
       "      <td>1</td>\n",
       "      <td>35571620</td>\n",
       "      <td>35571621</td>\n",
       "      <td>rs6425948</td>\n",
       "      <td>0.130213</td>\n",
       "      <td>AnchorA</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>35625000</td>\n",
       "      <td>35650000</td>\n",
       "      <td>1</td>\n",
       "      <td>35641525</td>\n",
       "      <td>35641526</td>\n",
       "      <td>PSMB2</td>\n",
       "      <td>ENSG00000126067</td>\n",
       "      <td>-</td>\n",
       "      <td>35640000</td>\n",
       "      <td>35645000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>35550000</td>\n",
       "      <td>35575000</td>\n",
       "      <td>1</td>\n",
       "      <td>35625000</td>\n",
       "      <td>35650000</td>\n",
       "      <td>4.919412</td>\n",
       "      <td>1</td>\n",
       "      <td>35641908</td>\n",
       "      <td>35641909</td>\n",
       "      <td>rs676614</td>\n",
       "      <td>0.186331</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>35550000</td>\n",
       "      <td>35575000</td>\n",
       "      <td>1</td>\n",
       "      <td>35557472</td>\n",
       "      <td>35557473</td>\n",
       "      <td>NCDN</td>\n",
       "      <td>ENSG00000020129</td>\n",
       "      <td>+</td>\n",
       "      <td>35555000</td>\n",
       "      <td>35560000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>35550000</td>\n",
       "      <td>35575000</td>\n",
       "      <td>1</td>\n",
       "      <td>35625000</td>\n",
       "      <td>35650000</td>\n",
       "      <td>4.919412</td>\n",
       "      <td>1</td>\n",
       "      <td>35641908</td>\n",
       "      <td>35641909</td>\n",
       "      <td>rs676614</td>\n",
       "      <td>0.186331</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>35550000</td>\n",
       "      <td>35575000</td>\n",
       "      <td>1</td>\n",
       "      <td>35557949</td>\n",
       "      <td>35557950</td>\n",
       "      <td>KIAA0319L</td>\n",
       "      <td>ENSG00000142687</td>\n",
       "      <td>-</td>\n",
       "      <td>35555000</td>\n",
       "      <td>35560000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>35550000</td>\n",
       "      <td>35575000</td>\n",
       "      <td>1</td>\n",
       "      <td>35625000</td>\n",
       "      <td>35650000</td>\n",
       "      <td>4.919412</td>\n",
       "      <td>1</td>\n",
       "      <td>35641908</td>\n",
       "      <td>35641909</td>\n",
       "      <td>rs676614</td>\n",
       "      <td>0.186331</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>35550000</td>\n",
       "      <td>35575000</td>\n",
       "      <td>1</td>\n",
       "      <td>35573313</td>\n",
       "      <td>35573314</td>\n",
       "      <td>TFAP2E</td>\n",
       "      <td>ENSG00000116819</td>\n",
       "      <td>+</td>\n",
       "      <td>35570000</td>\n",
       "      <td>35575000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>35550000</td>\n",
       "      <td>35575000</td>\n",
       "      <td>1</td>\n",
       "      <td>35700000</td>\n",
       "      <td>35725000</td>\n",
       "      <td>38.252608</td>\n",
       "      <td>1</td>\n",
       "      <td>35571620</td>\n",
       "      <td>35571621</td>\n",
       "      <td>rs6425948</td>\n",
       "      <td>0.130213</td>\n",
       "      <td>AnchorA</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>35700000</td>\n",
       "      <td>35725000</td>\n",
       "      <td>1</td>\n",
       "      <td>35718893</td>\n",
       "      <td>35718894</td>\n",
       "      <td>C1orf216</td>\n",
       "      <td>ENSG00000142686</td>\n",
       "      <td>-</td>\n",
       "      <td>35715000</td>\n",
       "      <td>35720000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793</th>\n",
       "      <td>8</td>\n",
       "      <td>140500000</td>\n",
       "      <td>140525000</td>\n",
       "      <td>8</td>\n",
       "      <td>140600000</td>\n",
       "      <td>140625000</td>\n",
       "      <td>8.287003</td>\n",
       "      <td>8</td>\n",
       "      <td>140608622</td>\n",
       "      <td>140608623</td>\n",
       "      <td>rs884934</td>\n",
       "      <td>0.086075</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>661</td>\n",
       "      <td>8</td>\n",
       "      <td>140500000</td>\n",
       "      <td>140525000</td>\n",
       "      <td>8</td>\n",
       "      <td>140511310</td>\n",
       "      <td>140511311</td>\n",
       "      <td>CHRAC1</td>\n",
       "      <td>ENSG00000104472</td>\n",
       "      <td>+</td>\n",
       "      <td>140510000</td>\n",
       "      <td>140515000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>794</th>\n",
       "      <td>8</td>\n",
       "      <td>140500000</td>\n",
       "      <td>140525000</td>\n",
       "      <td>8</td>\n",
       "      <td>140600000</td>\n",
       "      <td>140625000</td>\n",
       "      <td>8.287003</td>\n",
       "      <td>8</td>\n",
       "      <td>140612790</td>\n",
       "      <td>140612791</td>\n",
       "      <td>rs6998901</td>\n",
       "      <td>0.063517</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>662</td>\n",
       "      <td>8</td>\n",
       "      <td>140500000</td>\n",
       "      <td>140525000</td>\n",
       "      <td>8</td>\n",
       "      <td>140508042</td>\n",
       "      <td>140508043</td>\n",
       "      <td>AC107375.1</td>\n",
       "      <td>ENSG00000259891</td>\n",
       "      <td>-</td>\n",
       "      <td>140505000</td>\n",
       "      <td>140510000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>795</th>\n",
       "      <td>8</td>\n",
       "      <td>140500000</td>\n",
       "      <td>140525000</td>\n",
       "      <td>8</td>\n",
       "      <td>140600000</td>\n",
       "      <td>140625000</td>\n",
       "      <td>8.287003</td>\n",
       "      <td>8</td>\n",
       "      <td>140612790</td>\n",
       "      <td>140612791</td>\n",
       "      <td>rs6998901</td>\n",
       "      <td>0.063517</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>662</td>\n",
       "      <td>8</td>\n",
       "      <td>140500000</td>\n",
       "      <td>140525000</td>\n",
       "      <td>8</td>\n",
       "      <td>140511310</td>\n",
       "      <td>140511311</td>\n",
       "      <td>CHRAC1</td>\n",
       "      <td>ENSG00000104472</td>\n",
       "      <td>+</td>\n",
       "      <td>140510000</td>\n",
       "      <td>140515000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>796</th>\n",
       "      <td>8</td>\n",
       "      <td>140500000</td>\n",
       "      <td>140525000</td>\n",
       "      <td>8</td>\n",
       "      <td>140625000</td>\n",
       "      <td>140650000</td>\n",
       "      <td>14.203714</td>\n",
       "      <td>8</td>\n",
       "      <td>140630634</td>\n",
       "      <td>140630635</td>\n",
       "      <td>rs13248602</td>\n",
       "      <td>0.056641</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>663</td>\n",
       "      <td>8</td>\n",
       "      <td>140500000</td>\n",
       "      <td>140525000</td>\n",
       "      <td>8</td>\n",
       "      <td>140508042</td>\n",
       "      <td>140508043</td>\n",
       "      <td>AC107375.1</td>\n",
       "      <td>ENSG00000259891</td>\n",
       "      <td>-</td>\n",
       "      <td>140505000</td>\n",
       "      <td>140510000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>797</th>\n",
       "      <td>8</td>\n",
       "      <td>140500000</td>\n",
       "      <td>140525000</td>\n",
       "      <td>8</td>\n",
       "      <td>140625000</td>\n",
       "      <td>140650000</td>\n",
       "      <td>14.203714</td>\n",
       "      <td>8</td>\n",
       "      <td>140630634</td>\n",
       "      <td>140630635</td>\n",
       "      <td>rs13248602</td>\n",
       "      <td>0.056641</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>663</td>\n",
       "      <td>8</td>\n",
       "      <td>140500000</td>\n",
       "      <td>140525000</td>\n",
       "      <td>8</td>\n",
       "      <td>140511310</td>\n",
       "      <td>140511311</td>\n",
       "      <td>CHRAC1</td>\n",
       "      <td>ENSG00000104472</td>\n",
       "      <td>+</td>\n",
       "      <td>140510000</td>\n",
       "      <td>140515000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>798 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrA_loop  startA_loop  endA_loop  chrB_loop  startB_loop  endB_loop  \\\n",
       "0            1     35550000   35575000          1     35625000   35650000   \n",
       "1            1     35550000   35575000          1     35625000   35650000   \n",
       "2            1     35550000   35575000          1     35625000   35650000   \n",
       "3            1     35550000   35575000          1     35625000   35650000   \n",
       "4            1     35550000   35575000          1     35700000   35725000   \n",
       "..         ...          ...        ...        ...          ...        ...   \n",
       "793          8    140500000  140525000          8    140600000  140625000   \n",
       "794          8    140500000  140525000          8    140600000  140625000   \n",
       "795          8    140500000  140525000          8    140600000  140625000   \n",
       "796          8    140500000  140525000          8    140625000  140650000   \n",
       "797          8    140500000  140525000          8    140625000  140650000   \n",
       "\n",
       "     -log10_qval  chr_snp  start_snp    end_snp        rsid   ppa_snp  \\\n",
       "0       4.919412        1   35571620   35571621   rs6425948  0.130213   \n",
       "1       4.919412        1   35641908   35641909    rs676614  0.186331   \n",
       "2       4.919412        1   35641908   35641909    rs676614  0.186331   \n",
       "3       4.919412        1   35641908   35641909    rs676614  0.186331   \n",
       "4      38.252608        1   35571620   35571621   rs6425948  0.130213   \n",
       "..           ...      ...        ...        ...         ...       ...   \n",
       "793     8.287003        8  140608622  140608623    rs884934  0.086075   \n",
       "794     8.287003        8  140612790  140612791   rs6998901  0.063517   \n",
       "795     8.287003        8  140612790  140612791   rs6998901  0.063517   \n",
       "796    14.203714        8  140630634  140630635  rs13248602  0.056641   \n",
       "797    14.203714        8  140630634  140630635  rs13248602  0.056641   \n",
       "\n",
       "    snp_anchor  gh_id  chr_anchor  start_anchor  end_anchor  chr_gene  \\\n",
       "0      AnchorA      2           1      35625000    35650000         1   \n",
       "1      AnchorB      3           1      35550000    35575000         1   \n",
       "2      AnchorB      3           1      35550000    35575000         1   \n",
       "3      AnchorB      3           1      35550000    35575000         1   \n",
       "4      AnchorA      4           1      35700000    35725000         1   \n",
       "..         ...    ...         ...           ...         ...       ...   \n",
       "793    AnchorB    661           8     140500000   140525000         8   \n",
       "794    AnchorB    662           8     140500000   140525000         8   \n",
       "795    AnchorB    662           8     140500000   140525000         8   \n",
       "796    AnchorB    663           8     140500000   140525000         8   \n",
       "797    AnchorB    663           8     140500000   140525000         8   \n",
       "\n",
       "     start_gene   end_gene    genename           geneid strand  bin_start  \\\n",
       "0      35641525   35641526       PSMB2  ENSG00000126067      -   35640000   \n",
       "1      35557472   35557473        NCDN  ENSG00000020129      +   35555000   \n",
       "2      35557949   35557950   KIAA0319L  ENSG00000142687      -   35555000   \n",
       "3      35573313   35573314      TFAP2E  ENSG00000116819      +   35570000   \n",
       "4      35718893   35718894    C1orf216  ENSG00000142686      -   35715000   \n",
       "..          ...        ...         ...              ...    ...        ...   \n",
       "793   140511310  140511311      CHRAC1  ENSG00000104472      +  140510000   \n",
       "794   140508042  140508043  AC107375.1  ENSG00000259891      -  140505000   \n",
       "795   140511310  140511311      CHRAC1  ENSG00000104472      +  140510000   \n",
       "796   140508042  140508043  AC107375.1  ENSG00000259891      -  140505000   \n",
       "797   140511310  140511311      CHRAC1  ENSG00000104472      +  140510000   \n",
       "\n",
       "       bin_end  \n",
       "0     35645000  \n",
       "1     35560000  \n",
       "2     35560000  \n",
       "3     35575000  \n",
       "4     35720000  \n",
       "..         ...  \n",
       "793  140515000  \n",
       "794  140510000  \n",
       "795  140515000  \n",
       "796  140510000  \n",
       "797  140515000  \n",
       "\n",
       "[798 rows x 25 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_hichip_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a51e0228",
   "metadata": {},
   "source": [
    "## Save SGLs for Further Analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "813846b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "finemap_sgls_fn = os.path.join(outdir, 'finemap_sgls.tsv')\n",
    "gwas_hichip_genes.to_csv(finemap_sgls_fn,  sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d2fd436",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrA_loop</th>\n",
       "      <th>startA_loop</th>\n",
       "      <th>endA_loop</th>\n",
       "      <th>chrB_loop</th>\n",
       "      <th>startB_loop</th>\n",
       "      <th>endB_loop</th>\n",
       "      <th>-log10_qval</th>\n",
       "      <th>chr_snp</th>\n",
       "      <th>start_snp</th>\n",
       "      <th>end_snp</th>\n",
       "      <th>rsid</th>\n",
       "      <th>ppa_snp</th>\n",
       "      <th>snp_anchor</th>\n",
       "      <th>gh_id</th>\n",
       "      <th>chrSNP</th>\n",
       "      <th>startSNP</th>\n",
       "      <th>endSNP</th>\n",
       "      <th>chrGene</th>\n",
       "      <th>startGene</th>\n",
       "      <th>endGene</th>\n",
       "      <th>genename</th>\n",
       "      <th>geneid</th>\n",
       "      <th>strand</th>\n",
       "      <th>bin_start</th>\n",
       "      <th>bin_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>35550000</td>\n",
       "      <td>35575000</td>\n",
       "      <td>1</td>\n",
       "      <td>35625000</td>\n",
       "      <td>35650000</td>\n",
       "      <td>4.919412</td>\n",
       "      <td>1</td>\n",
       "      <td>35571620</td>\n",
       "      <td>35571621</td>\n",
       "      <td>rs6425948</td>\n",
       "      <td>0.130213</td>\n",
       "      <td>AnchorA</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>35625000.0</td>\n",
       "      <td>35650000.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>35641525.0</td>\n",
       "      <td>35641526.0</td>\n",
       "      <td>PSMB2</td>\n",
       "      <td>ENSG00000126067</td>\n",
       "      <td>-</td>\n",
       "      <td>35640000.0</td>\n",
       "      <td>35645000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>35550000</td>\n",
       "      <td>35575000</td>\n",
       "      <td>1</td>\n",
       "      <td>35625000</td>\n",
       "      <td>35650000</td>\n",
       "      <td>4.919412</td>\n",
       "      <td>1</td>\n",
       "      <td>35641908</td>\n",
       "      <td>35641909</td>\n",
       "      <td>rs676614</td>\n",
       "      <td>0.186331</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>35550000.0</td>\n",
       "      <td>35575000.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>35557472.0</td>\n",
       "      <td>35557473.0</td>\n",
       "      <td>NCDN</td>\n",
       "      <td>ENSG00000020129</td>\n",
       "      <td>+</td>\n",
       "      <td>35555000.0</td>\n",
       "      <td>35560000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>35550000</td>\n",
       "      <td>35575000</td>\n",
       "      <td>1</td>\n",
       "      <td>35625000</td>\n",
       "      <td>35650000</td>\n",
       "      <td>4.919412</td>\n",
       "      <td>1</td>\n",
       "      <td>35641908</td>\n",
       "      <td>35641909</td>\n",
       "      <td>rs676614</td>\n",
       "      <td>0.186331</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>35550000.0</td>\n",
       "      <td>35575000.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>35557949.0</td>\n",
       "      <td>35557950.0</td>\n",
       "      <td>KIAA0319L</td>\n",
       "      <td>ENSG00000142687</td>\n",
       "      <td>-</td>\n",
       "      <td>35555000.0</td>\n",
       "      <td>35560000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>35550000</td>\n",
       "      <td>35575000</td>\n",
       "      <td>1</td>\n",
       "      <td>35625000</td>\n",
       "      <td>35650000</td>\n",
       "      <td>4.919412</td>\n",
       "      <td>1</td>\n",
       "      <td>35641908</td>\n",
       "      <td>35641909</td>\n",
       "      <td>rs676614</td>\n",
       "      <td>0.186331</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>35550000.0</td>\n",
       "      <td>35575000.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>35573313.0</td>\n",
       "      <td>35573314.0</td>\n",
       "      <td>TFAP2E</td>\n",
       "      <td>ENSG00000116819</td>\n",
       "      <td>+</td>\n",
       "      <td>35570000.0</td>\n",
       "      <td>35575000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>35550000</td>\n",
       "      <td>35575000</td>\n",
       "      <td>1</td>\n",
       "      <td>35700000</td>\n",
       "      <td>35725000</td>\n",
       "      <td>38.252608</td>\n",
       "      <td>1</td>\n",
       "      <td>35571620</td>\n",
       "      <td>35571621</td>\n",
       "      <td>rs6425948</td>\n",
       "      <td>0.130213</td>\n",
       "      <td>AnchorA</td>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>35700000.0</td>\n",
       "      <td>35725000.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>35718893.0</td>\n",
       "      <td>35718894.0</td>\n",
       "      <td>C1orf216</td>\n",
       "      <td>ENSG00000142686</td>\n",
       "      <td>-</td>\n",
       "      <td>35715000.0</td>\n",
       "      <td>35720000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1097</th>\n",
       "      <td>8</td>\n",
       "      <td>140500000</td>\n",
       "      <td>140525000</td>\n",
       "      <td>8</td>\n",
       "      <td>140600000</td>\n",
       "      <td>140625000</td>\n",
       "      <td>8.287003</td>\n",
       "      <td>8</td>\n",
       "      <td>140608622</td>\n",
       "      <td>140608623</td>\n",
       "      <td>rs884934</td>\n",
       "      <td>0.086075</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>661</td>\n",
       "      <td>8.0</td>\n",
       "      <td>140500000.0</td>\n",
       "      <td>140525000.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>140511310.0</td>\n",
       "      <td>140511311.0</td>\n",
       "      <td>CHRAC1</td>\n",
       "      <td>ENSG00000104472</td>\n",
       "      <td>+</td>\n",
       "      <td>140510000.0</td>\n",
       "      <td>140515000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1098</th>\n",
       "      <td>8</td>\n",
       "      <td>140500000</td>\n",
       "      <td>140525000</td>\n",
       "      <td>8</td>\n",
       "      <td>140600000</td>\n",
       "      <td>140625000</td>\n",
       "      <td>8.287003</td>\n",
       "      <td>8</td>\n",
       "      <td>140612790</td>\n",
       "      <td>140612791</td>\n",
       "      <td>rs6998901</td>\n",
       "      <td>0.063517</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>662</td>\n",
       "      <td>8.0</td>\n",
       "      <td>140500000.0</td>\n",
       "      <td>140525000.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>140508042.0</td>\n",
       "      <td>140508043.0</td>\n",
       "      <td>AC107375.1</td>\n",
       "      <td>ENSG00000259891</td>\n",
       "      <td>-</td>\n",
       "      <td>140505000.0</td>\n",
       "      <td>140510000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1099</th>\n",
       "      <td>8</td>\n",
       "      <td>140500000</td>\n",
       "      <td>140525000</td>\n",
       "      <td>8</td>\n",
       "      <td>140600000</td>\n",
       "      <td>140625000</td>\n",
       "      <td>8.287003</td>\n",
       "      <td>8</td>\n",
       "      <td>140612790</td>\n",
       "      <td>140612791</td>\n",
       "      <td>rs6998901</td>\n",
       "      <td>0.063517</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>662</td>\n",
       "      <td>8.0</td>\n",
       "      <td>140500000.0</td>\n",
       "      <td>140525000.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>140511310.0</td>\n",
       "      <td>140511311.0</td>\n",
       "      <td>CHRAC1</td>\n",
       "      <td>ENSG00000104472</td>\n",
       "      <td>+</td>\n",
       "      <td>140510000.0</td>\n",
       "      <td>140515000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1100</th>\n",
       "      <td>8</td>\n",
       "      <td>140500000</td>\n",
       "      <td>140525000</td>\n",
       "      <td>8</td>\n",
       "      <td>140625000</td>\n",
       "      <td>140650000</td>\n",
       "      <td>14.203714</td>\n",
       "      <td>8</td>\n",
       "      <td>140630634</td>\n",
       "      <td>140630635</td>\n",
       "      <td>rs13248602</td>\n",
       "      <td>0.056641</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>663</td>\n",
       "      <td>8.0</td>\n",
       "      <td>140500000.0</td>\n",
       "      <td>140525000.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>140508042.0</td>\n",
       "      <td>140508043.0</td>\n",
       "      <td>AC107375.1</td>\n",
       "      <td>ENSG00000259891</td>\n",
       "      <td>-</td>\n",
       "      <td>140505000.0</td>\n",
       "      <td>140510000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1101</th>\n",
       "      <td>8</td>\n",
       "      <td>140500000</td>\n",
       "      <td>140525000</td>\n",
       "      <td>8</td>\n",
       "      <td>140625000</td>\n",
       "      <td>140650000</td>\n",
       "      <td>14.203714</td>\n",
       "      <td>8</td>\n",
       "      <td>140630634</td>\n",
       "      <td>140630635</td>\n",
       "      <td>rs13248602</td>\n",
       "      <td>0.056641</td>\n",
       "      <td>AnchorB</td>\n",
       "      <td>663</td>\n",
       "      <td>8.0</td>\n",
       "      <td>140500000.0</td>\n",
       "      <td>140525000.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>140511310.0</td>\n",
       "      <td>140511311.0</td>\n",
       "      <td>CHRAC1</td>\n",
       "      <td>ENSG00000104472</td>\n",
       "      <td>+</td>\n",
       "      <td>140510000.0</td>\n",
       "      <td>140515000.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>798 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrA_loop  startA_loop  endA_loop  chrB_loop  startB_loop  endB_loop  \\\n",
       "2             1     35550000   35575000          1     35625000   35650000   \n",
       "3             1     35550000   35575000          1     35625000   35650000   \n",
       "4             1     35550000   35575000          1     35625000   35650000   \n",
       "5             1     35550000   35575000          1     35625000   35650000   \n",
       "6             1     35550000   35575000          1     35700000   35725000   \n",
       "...         ...          ...        ...        ...          ...        ...   \n",
       "1097          8    140500000  140525000          8    140600000  140625000   \n",
       "1098          8    140500000  140525000          8    140600000  140625000   \n",
       "1099          8    140500000  140525000          8    140600000  140625000   \n",
       "1100          8    140500000  140525000          8    140625000  140650000   \n",
       "1101          8    140500000  140525000          8    140625000  140650000   \n",
       "\n",
       "      -log10_qval  chr_snp  start_snp    end_snp        rsid   ppa_snp  \\\n",
       "2        4.919412        1   35571620   35571621   rs6425948  0.130213   \n",
       "3        4.919412        1   35641908   35641909    rs676614  0.186331   \n",
       "4        4.919412        1   35641908   35641909    rs676614  0.186331   \n",
       "5        4.919412        1   35641908   35641909    rs676614  0.186331   \n",
       "6       38.252608        1   35571620   35571621   rs6425948  0.130213   \n",
       "...           ...      ...        ...        ...         ...       ...   \n",
       "1097     8.287003        8  140608622  140608623    rs884934  0.086075   \n",
       "1098     8.287003        8  140612790  140612791   rs6998901  0.063517   \n",
       "1099     8.287003        8  140612790  140612791   rs6998901  0.063517   \n",
       "1100    14.203714        8  140630634  140630635  rs13248602  0.056641   \n",
       "1101    14.203714        8  140630634  140630635  rs13248602  0.056641   \n",
       "\n",
       "     snp_anchor  gh_id  chrSNP     startSNP       endSNP  chrGene  \\\n",
       "2       AnchorA      2     1.0   35625000.0   35650000.0      1.0   \n",
       "3       AnchorB      3     1.0   35550000.0   35575000.0      1.0   \n",
       "4       AnchorB      3     1.0   35550000.0   35575000.0      1.0   \n",
       "5       AnchorB      3     1.0   35550000.0   35575000.0      1.0   \n",
       "6       AnchorA      4     1.0   35700000.0   35725000.0      1.0   \n",
       "...         ...    ...     ...          ...          ...      ...   \n",
       "1097    AnchorB    661     8.0  140500000.0  140525000.0      8.0   \n",
       "1098    AnchorB    662     8.0  140500000.0  140525000.0      8.0   \n",
       "1099    AnchorB    662     8.0  140500000.0  140525000.0      8.0   \n",
       "1100    AnchorB    663     8.0  140500000.0  140525000.0      8.0   \n",
       "1101    AnchorB    663     8.0  140500000.0  140525000.0      8.0   \n",
       "\n",
       "        startGene      endGene    genename           geneid strand  \\\n",
       "2      35641525.0   35641526.0       PSMB2  ENSG00000126067      -   \n",
       "3      35557472.0   35557473.0        NCDN  ENSG00000020129      +   \n",
       "4      35557949.0   35557950.0   KIAA0319L  ENSG00000142687      -   \n",
       "5      35573313.0   35573314.0      TFAP2E  ENSG00000116819      +   \n",
       "6      35718893.0   35718894.0    C1orf216  ENSG00000142686      -   \n",
       "...           ...          ...         ...              ...    ...   \n",
       "1097  140511310.0  140511311.0      CHRAC1  ENSG00000104472      +   \n",
       "1098  140508042.0  140508043.0  AC107375.1  ENSG00000259891      -   \n",
       "1099  140511310.0  140511311.0      CHRAC1  ENSG00000104472      +   \n",
       "1100  140508042.0  140508043.0  AC107375.1  ENSG00000259891      -   \n",
       "1101  140511310.0  140511311.0      CHRAC1  ENSG00000104472      +   \n",
       "\n",
       "        bin_start      bin_end  \n",
       "2      35640000.0   35645000.0  \n",
       "3      35555000.0   35560000.0  \n",
       "4      35555000.0   35560000.0  \n",
       "5      35570000.0   35575000.0  \n",
       "6      35715000.0   35720000.0  \n",
       "...           ...          ...  \n",
       "1097  140510000.0  140515000.0  \n",
       "1098  140505000.0  140510000.0  \n",
       "1099  140510000.0  140515000.0  \n",
       "1100  140505000.0  140510000.0  \n",
       "1101  140510000.0  140515000.0  \n",
       "\n",
       "[798 rows x 25 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_hichip_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74a0ec7f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "(hichip-db)",
   "language": "python",
   "name": "hichip-db"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
