{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5aa817b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import pandas as pd\n",
    "import pybedtools as pbt\n",
    "import subprocess as sp\n",
    "os.chdir('/mnt/BioHome/jreyna/jreyna/projects/dchallenge')\n",
    "pbt.helpers.set_tempdir('/mnt/BioHome/jreyna/tmp/')\n",
    "pbt.set_bedtools_path('/mnt/BioApps/bedtools/bin/')\n",
    "bgzip = '/mnt/BioApps/tabix/tabix-0.2.6/bgzip'\n",
    "tabix = '/mnt/BioApps/tabix/tabix-0.2.6/tabix'\n",
    "outdir = 'results/refs/washu/'\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "from biolib import washu"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aacd3040",
   "metadata": {},
   "source": [
    "## Load the Gencode Coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c5b32abe",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode = pd.read_table('results/refs/gencode/v30/gencode.v30.annotation.grch37.bed', header=None)\n",
    "gencode.columns = ['chr', 'start', 'end', 'strand', 'type', 'geneid', 'genename']\n",
    "gene_coords = gencode.loc[gencode.type == 'gene']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b9ba77d",
   "metadata": {},
   "source": [
    "### Make the WashU file for gene coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f17b3fb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "washu_genes = gene_coords.copy()\n",
    "\n",
    "# setting very basic coordinates for transcript, translation and exon products\n",
    "washu_genes['transcript_start'] = washu_genes['start']\n",
    "washu_genes['transcript_stop'] = washu_genes['end']\n",
    "washu_genes['translation_start'] = washu_genes['start']\n",
    "washu_genes['translation_stop'] = washu_genes['end']\n",
    "washu_genes['transcript_id'] = washu_genes['genename']\n",
    "washu_genes['exon_start'] = washu_genes['start']\n",
    "washu_genes['exon_stops'] = washu_genes['end']\n",
    "washu_genes['type'] = 'coding'\n",
    "\n",
    "# reordering the dataframe for the WashU browser\n",
    "refcols = ['chr', 'transcript_start', 'transcript_stop',\n",
    "           'translation_start', 'translation_stop',\n",
    "           'strand', 'genename', 'transcript_id', 'type',\n",
    "           'exon_start', 'exon_stops', 'geneid']\n",
    "washu_genes = washu_genes.loc[:, refcols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "bce63f73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_stop</th>\n",
       "      <th>translation_start</th>\n",
       "      <th>translation_stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>genename</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>type</th>\n",
       "      <th>exon_start</th>\n",
       "      <th>exon_stops</th>\n",
       "      <th>geneid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>coding</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>-</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>coding</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>+</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>coding</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr1</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>-</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>coding</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>ENSG00000237613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>chr1</td>\n",
       "      <td>52473</td>\n",
       "      <td>53312</td>\n",
       "      <td>52473</td>\n",
       "      <td>53312</td>\n",
       "      <td>+</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>coding</td>\n",
       "      <td>52473</td>\n",
       "      <td>53312</td>\n",
       "      <td>ENSG00000268020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2782597</th>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>137995</td>\n",
       "      <td>141363</td>\n",
       "      <td>137995</td>\n",
       "      <td>141363</td>\n",
       "      <td>+</td>\n",
       "      <td>CU638689.1</td>\n",
       "      <td>CU638689.1</td>\n",
       "      <td>coding</td>\n",
       "      <td>137995</td>\n",
       "      <td>141363</td>\n",
       "      <td>ENSG00000278878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2782611</th>\n",
       "      <td>GL000220.1</td>\n",
       "      <td>153803</td>\n",
       "      <td>154725</td>\n",
       "      <td>153803</td>\n",
       "      <td>154725</td>\n",
       "      <td>-</td>\n",
       "      <td>FP236383.3</td>\n",
       "      <td>FP236383.3</td>\n",
       "      <td>coding</td>\n",
       "      <td>153803</td>\n",
       "      <td>154725</td>\n",
       "      <td>ENSG00000281181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2782614</th>\n",
       "      <td>GL000205.1</td>\n",
       "      <td>24619</td>\n",
       "      <td>28363</td>\n",
       "      <td>24619</td>\n",
       "      <td>28363</td>\n",
       "      <td>+</td>\n",
       "      <td>CU633904.2</td>\n",
       "      <td>CU633904.2</td>\n",
       "      <td>coding</td>\n",
       "      <td>24619</td>\n",
       "      <td>28363</td>\n",
       "      <td>ENSG00000279647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2782625</th>\n",
       "      <td>GL000220.1</td>\n",
       "      <td>153803</td>\n",
       "      <td>154725</td>\n",
       "      <td>153803</td>\n",
       "      <td>154725</td>\n",
       "      <td>-</td>\n",
       "      <td>FP671120.5</td>\n",
       "      <td>FP671120.5</td>\n",
       "      <td>coding</td>\n",
       "      <td>153803</td>\n",
       "      <td>154725</td>\n",
       "      <td>ENSG00000281383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2782630</th>\n",
       "      <td>GL000204.1</td>\n",
       "      <td>43956</td>\n",
       "      <td>49839</td>\n",
       "      <td>43956</td>\n",
       "      <td>49839</td>\n",
       "      <td>-</td>\n",
       "      <td>AC233701.1</td>\n",
       "      <td>AC233701.1</td>\n",
       "      <td>coding</td>\n",
       "      <td>43956</td>\n",
       "      <td>49839</td>\n",
       "      <td>ENSG00000263278</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60627 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                chr  transcript_start  transcript_stop  translation_start  \\\n",
       "0              chr1             11869            14409              11869   \n",
       "12             chr1             14404            29570              14404   \n",
       "25             chr1             29554            31109              29554   \n",
       "33             chr1             34554            36081              34554   \n",
       "41             chr1             52473            53312              52473   \n",
       "...             ...               ...              ...                ...   \n",
       "2782597  GL000195.1            137995           141363             137995   \n",
       "2782611  GL000220.1            153803           154725             153803   \n",
       "2782614  GL000205.1             24619            28363              24619   \n",
       "2782625  GL000220.1            153803           154725             153803   \n",
       "2782630  GL000204.1             43956            49839              43956   \n",
       "\n",
       "         translation_stop strand     genename transcript_id    type  \\\n",
       "0                   14409      +      DDX11L1       DDX11L1  coding   \n",
       "12                  29570      -       WASH7P        WASH7P  coding   \n",
       "25                  31109      +  MIR1302-2HG   MIR1302-2HG  coding   \n",
       "33                  36081      -      FAM138A       FAM138A  coding   \n",
       "41                  53312      +       OR4G4P        OR4G4P  coding   \n",
       "...                   ...    ...          ...           ...     ...   \n",
       "2782597            141363      +   CU638689.1    CU638689.1  coding   \n",
       "2782611            154725      -   FP236383.3    FP236383.3  coding   \n",
       "2782614             28363      +   CU633904.2    CU633904.2  coding   \n",
       "2782625            154725      -   FP671120.5    FP671120.5  coding   \n",
       "2782630             49839      -   AC233701.1    AC233701.1  coding   \n",
       "\n",
       "         exon_start  exon_stops           geneid  \n",
       "0             11869       14409  ENSG00000223972  \n",
       "12            14404       29570  ENSG00000227232  \n",
       "25            29554       31109  ENSG00000243485  \n",
       "33            34554       36081  ENSG00000237613  \n",
       "41            52473       53312  ENSG00000268020  \n",
       "...             ...         ...              ...  \n",
       "2782597      137995      141363  ENSG00000278878  \n",
       "2782611      153803      154725  ENSG00000281181  \n",
       "2782614       24619       28363  ENSG00000279647  \n",
       "2782625      153803      154725  ENSG00000281383  \n",
       "2782630       43956       49839  ENSG00000263278  \n",
       "\n",
       "[60627 rows x 12 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "washu_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "afdba6b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running: /mnt/BioApps/tabix/tabix-0.2.6/bgzip -f results/refs/washu/gencode.washu.refbed\n",
      "Created the gzfn: results/refs/washu/gencode.washu.refbed.gz\n",
      "Running:  /mnt/BioApps/tabix/tabix-0.2.6/tabix results/refs/washu/gencode.washu.refbed.gz\n",
      "Created the tabix: results/refs/washu/gencode.washu.refbed.gz.tbi\n"
     ]
    }
   ],
   "source": [
    "# writing the dataframe to refbed format with an index\n",
    "washu_fn = os.path.join(outdir, 'gencode.washu.refbed')\n",
    "washu.dataframe_to_WashU_refbed(washu_genes, washu_fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcef672e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "HiChIP-DB",
   "language": "python",
   "name": "hichip-db"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
