{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9f38e15f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import pandas as pd\n",
    "import pybedtools as pbt\n",
    "import numpy as np\n",
    "import glob \n",
    "import json\n",
    "os.chdir('/mnt/BioHome/jreyna/jreyna/projects/dchallenge')\n",
    "pbt.helpers.set_tempdir('/mnt/BioHome/jreyna/tmp/')\n",
    "pbt.set_bedtools_path('/mnt/BioApps/bedtools/bin/')\n",
    "res = 5000\n",
    "from biolib import washu\n",
    "from biolib import coordinates\n",
    "import importlib\n",
    "importlib.reload(washu)\n",
    "importlib.reload(coordinates)\n",
    "\n",
    "outdir = 'results/main/sgls/washu/'\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "tracks_dir = os.path.join(outdir, 'tracks/')\n",
    "os.makedirs(tracks_dir, exist_ok=True)\n",
    "hubs_dir = os.path.join(outdir, 'hubs/')\n",
    "os.makedirs(tracks_dir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1b2f22e",
   "metadata": {},
   "source": [
    "### Make longrange files for SGLs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4c8a3fe4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'snp_start_bin'] = snp_bins[:, 0]\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:16: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'snp_end_bin'] = snp_bins[:, 1]\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'tss_start_bin'] = tss_start_bin\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:22: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'tss_end_bin'] = tss_start_bin + res\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:26: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls_washu.loc[:, 'dummy_pval'] = 0.00001\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running: /mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_34012112_Gaulton.Quach_2016.monocyte_Pam3CSK4.monocyte_naive.script_version.longrange.bed\n",
      "Created the gzfn: results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_34012112_Gaulton.Quach_2016.monocyte_Pam3CSK4.monocyte_naive.script_version.longrange.bed.gz\n",
      "Running: /mnt/BioApps/tabix/tabix-0.2.6/tabix -f results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_34012112_Gaulton.Quach_2016.monocyte_Pam3CSK4.monocyte_naive.script_version.longrange.bed.gz\n",
      "Created the tabix: results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_34012112_Gaulton.Quach_2016.monocyte_Pam3CSK4.monocyte_naive.script_version.longrange.bed.gz.tbi\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'snp_start_bin'] = snp_bins[:, 0]\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:16: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'snp_end_bin'] = snp_bins[:, 1]\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'tss_start_bin'] = tss_start_bin\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:22: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'tss_end_bin'] = tss_start_bin + res\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:26: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls_washu.loc[:, 'dummy_pval'] = 0.00001\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running: /mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_34012112_Gaulton.Quach_2016.monocyte_R848.monocyte_naive.script_version.longrange.bed\n",
      "Created the gzfn: results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_34012112_Gaulton.Quach_2016.monocyte_R848.monocyte_naive.script_version.longrange.bed.gz\n",
      "Running: /mnt/BioApps/tabix/tabix-0.2.6/tabix -f results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_34012112_Gaulton.Quach_2016.monocyte_R848.monocyte_naive.script_version.longrange.bed.gz\n",
      "Created the tabix: results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_34012112_Gaulton.Quach_2016.monocyte_R848.monocyte_naive.script_version.longrange.bed.gz.tbi\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'snp_start_bin'] = snp_bins[:, 0]\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:16: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'snp_end_bin'] = snp_bins[:, 1]\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'tss_start_bin'] = tss_start_bin\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:22: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'tss_end_bin'] = tss_start_bin + res\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:26: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls_washu.loc[:, 'dummy_pval'] = 0.00001\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running: /mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_32005708.Quach_2016.monocyte_IAV.monocyte_naive.script_version.longrange.bed\n",
      "Created the gzfn: results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_32005708.Quach_2016.monocyte_IAV.monocyte_naive.script_version.longrange.bed.gz\n",
      "Running: /mnt/BioApps/tabix/tabix-0.2.6/tabix -f results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_32005708.Quach_2016.monocyte_IAV.monocyte_naive.script_version.longrange.bed.gz\n",
      "Created the tabix: results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_32005708.Quach_2016.monocyte_IAV.monocyte_naive.script_version.longrange.bed.gz.tbi\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'snp_start_bin'] = snp_bins[:, 0]\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:16: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'snp_end_bin'] = snp_bins[:, 1]\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'tss_start_bin'] = tss_start_bin\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:22: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'tss_end_bin'] = tss_start_bin + res\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:26: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls_washu.loc[:, 'dummy_pval'] = 0.00001\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running: /mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_25751624.Quach_2016.monocyte_IAV.monocyte_naive.script_version.longrange.bed\n",
      "Created the gzfn: results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_25751624.Quach_2016.monocyte_IAV.monocyte_naive.script_version.longrange.bed.gz\n",
      "Running: /mnt/BioApps/tabix/tabix-0.2.6/tabix -f results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_25751624.Quach_2016.monocyte_IAV.monocyte_naive.script_version.longrange.bed.gz\n",
      "Created the tabix: results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_25751624.Quach_2016.monocyte_IAV.monocyte_naive.script_version.longrange.bed.gz.tbi\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'snp_start_bin'] = snp_bins[:, 0]\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:16: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'snp_end_bin'] = snp_bins[:, 1]\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'tss_start_bin'] = tss_start_bin\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:22: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'tss_end_bin'] = tss_start_bin + res\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:26: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls_washu.loc[:, 'dummy_pval'] = 0.00001\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running: /mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_25751624.Quach_2016.monocyte_R848.monocyte_naive.script_version.longrange.bed\n",
      "Created the gzfn: results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_25751624.Quach_2016.monocyte_R848.monocyte_naive.script_version.longrange.bed.gz\n",
      "Running: /mnt/BioApps/tabix/tabix-0.2.6/tabix -f results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_25751624.Quach_2016.monocyte_R848.monocyte_naive.script_version.longrange.bed.gz\n",
      "Created the tabix: results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_25751624.Quach_2016.monocyte_R848.monocyte_naive.script_version.longrange.bed.gz.tbi\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'snp_start_bin'] = snp_bins[:, 0]\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:16: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'snp_end_bin'] = snp_bins[:, 1]\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'tss_start_bin'] = tss_start_bin\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:22: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls.loc[:, 'tss_end_bin'] = tss_start_bin + res\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_124145/2612498160.py:26: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sgls_washu.loc[:, 'dummy_pval'] = 0.00001\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running: /mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_25751624.Quach_2016.monocyte_Pam3CSK4.monocyte_naive.script_version.longrange.bed\n",
      "Created the gzfn: results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_25751624.Quach_2016.monocyte_Pam3CSK4.monocyte_naive.script_version.longrange.bed.gz\n",
      "Running: /mnt/BioApps/tabix/tabix-0.2.6/tabix -f results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_25751624.Quach_2016.monocyte_Pam3CSK4.monocyte_naive.script_version.longrange.bed.gz\n",
      "Created the tabix: results/main/sgls/washu/tracks/GRCh37.sgls.ldpairs.T1D_25751624.Quach_2016.monocyte_Pam3CSK4.monocyte_naive.script_version.longrange.bed.gz.tbi\n"
     ]
    }
   ],
   "source": [
    "# load sgl annotations \n",
    "fn = 'results/main/sgls/ldpairs/T1D_34012112_Gaulton/GRCh37/Quach_2016/monocyte_IAV/monocyte_naive/master.tsv'\n",
    "fn = 'results/main/sgls/ldpairs/T1D_34012112_Gaulton/GRCh37/Quach_2016/monocyte_R848/monocyte_naive/master.tsv'\n",
    "tpl = 'results/main/GRCh37/sgls/ldpairs/*/*/*/*/script_version/master.tsv'\n",
    "for fn in glob.glob(tpl):\n",
    "        \n",
    "    sgl_annos = pd.read_table(fn)\n",
    "\n",
    "    # extract SNP-gene pairs which form an SGL\n",
    "    sgls = sgl_annos.loc[(sgl_annos.has_fithichip_loop == 1)]\n",
    "\n",
    "    # add SNP bin location\n",
    "    snp_bins = sgls.snp_pos.apply(coordinates.get_snp_bin, res=res).tolist()\n",
    "    snp_bins = np.array(snp_bins)\n",
    "    sgls.loc[:, 'snp_start_bin'] = snp_bins[:, 0]\n",
    "    sgls.loc[:, 'snp_end_bin'] = snp_bins[:, 1]\n",
    "\n",
    "    # addd TSS bin location\n",
    "    tss_start_bin = sgls.tss_start.apply(coordinates.get_closest_bin_start_end, res=res)\n",
    "    tss_end_bin = sgls.tss_end.apply(coordinates.get_closest_bin_start_end, res=res)\n",
    "    sgls.loc[:, 'tss_start_bin'] = tss_start_bin\n",
    "    sgls.loc[:, 'tss_end_bin'] = tss_start_bin + res\n",
    "\n",
    "    sgls_washu = sgls[['chrom', 'snp_start_bin', 'snp_end_bin', \n",
    "                      'chrom', 'tss_start_bin', 'tss_end_bin']]\n",
    "    sgls_washu.loc[:, 'dummy_pval'] = 0.00001\n",
    "\n",
    "    basename = fn.split('/')[2:-1]\n",
    "    basename = '.'.join(basename)\n",
    "    basename = '{}.longrange.bed'.format(basename)\n",
    "    washu_fn = os.path.join(tracks_dir, basename)\n",
    "    washu.bedpe_to_WashU_longrange(sgls_washu, washu_fn)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e8c2f94",
   "metadata": {},
   "source": [
    "### Make a hub for all these "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "90934c50",
   "metadata": {},
   "outputs": [],
   "source": [
    "hub_jsons = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "051a7fa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "curr_hub = os.path.join(hubs_dir, 'loops/')\n",
    "os.makedirs(curr_hub, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7d1f6b8",
   "metadata": {},
   "source": [
    "#### Add Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "04e5a3bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_refbed = 'results/refs/washu/gencode.washu.refbed.gz'\n",
    "gene_refbed_basename = os.path.basename(gene_refbed)\n",
    "gene_refbed_link = os.path.join(curr_hub, gene_refbed_basename)\n",
    "if not os.path.exists(gene_refbed_link):\n",
    "    os.link(gene_refbed, gene_refbed_link)\n",
    "    os.link(gene_refbed + '.tbi', gene_refbed_link + '.tbi')\n",
    "\n",
    "gene_refbed_json = {'type': 'refbed',\n",
    "                  'filename': gene_refbed_basename, \n",
    "                  'name': 'Gencode V30 - GRCh37',\n",
    "                  'options': {'height': 100}}\n",
    "hub_jsons.append(gene_refbed_json)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7fc9db8",
   "metadata": {},
   "source": [
    "#### Add LongRanges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "139863ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "tpl = os.path.join(tracks_dir, '*.longrange.bed.gz')\n",
    "for track_fn in glob.glob(tpl):   \n",
    "    basename = os.path.basename(track_fn)\n",
    "    name = '.'.join(basename.split('.')[2:7])\n",
    "    json_str = {'type': 'longrange',\n",
    "                'filename': os.path.basename(track_fn),\n",
    "                'name': 'SGL Pairs for {}'.format(name),\n",
    "                'options': {'displayMode': 'arc', 'color':'purple', 'height': 200}}\n",
    "    \n",
    "    # symlink the track over\n",
    "    track_link = os.path.join(curr_hub, basename)\n",
    "    if not os.path.exists(track_link):\n",
    "        os.link(track_fn, track_link)\n",
    "        os.link(track_fn + '.tbi', track_link + '.tbi')\n",
    "\n",
    "        \n",
    "    hub_jsons.append(json_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "772a6b86",
   "metadata": {},
   "outputs": [],
   "source": [
    "ignore_loop_sources = ['TFH', 'TH1', 'TH2', 'TH17', 'THSTAR', 'TREGMEM', 'TREGNAIVE']\n",
    "\n",
    "tpl = 'results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/*/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01_WashU.bed.gz'\n",
    "for track_fn in glob.glob(tpl): \n",
    "    basename = os.path.basename(track_fn)\n",
    "    loop_source = track_fn.split('/')[5]\n",
    "    new_basename = '{}.{}'.format(loop_source, basename)\n",
    "    \n",
    "    if loop_source in ignore_loop_sources:\n",
    "        continue\n",
    "    \n",
    "    # symlink the track over\n",
    "    track_link = os.path.join(curr_hub, new_basename)\n",
    "    if not os.path.exists(track_link):\n",
    "        os.link(track_fn, track_link)\n",
    "        os.link(track_fn + '.tbi', track_link + '.tbi')    \n",
    "    \n",
    "    json_str = {'type': 'longrange',\n",
    "                'filename': os.path.basename(track_link),\n",
    "                'name': 'Loops from {}'.format(loop_source),\n",
    "                'options': {'displayMode': 'arc', 'color':'red', 'height': 200}}\n",
    "    \n",
    "\n",
    "    hub_jsons.append(json_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "38cf21f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "hub_json_fn = os.path.join(curr_hub, 'hub.config.json')\n",
    "with open(hub_json_fn, 'w') as f:\n",
    "    f.write(json.dumps(hub_jsons, indent=4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a978d538",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41a03ec4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df3a63bf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b281d807",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8063f9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "##gwas, cline = coloc_fn.split('/')[5:7]\n",
    "#\n",
    "#print(\"# make the hub json file\")\n",
    "#\n",
    "#gene_refbed_json = {'type': 'refbed',\n",
    "#                   'filename': os.path.basename(sg_genes_fn) + '.gz',\n",
    "#                   'name': 'Gencode V19',\n",
    "#                   'showOnHubLoad': True\n",
    "#                  }\n",
    "#\n",
    "#orig_loops_json = {'type': 'longrange',\n",
    "#                  'filename': os.path.basename(loop_gz_link),\n",
    "#                  'name': 'Original Loops',\n",
    "#                  'options': {'displayMode': 'arc', 'color':'red'},\n",
    "#                  'showOnHubLoad': True\n",
    "#               }\n",
    "#\n",
    "## fivekb_snp_gene_pairs_json = {'type': 'longrange',\n",
    "##                    'filename': os.path.basename(fivekb_snp_gene_pairs_fn) + '.gz',\n",
    "##                    'name': '5kb SNP-Gene Pairs',\n",
    "##                    'options': {'displayMode': 'arc', 'color':'purple', 'height': 200},\n",
    "##                    'showOnHubLoad': False\n",
    "##                 }\n",
    "#\n",
    "## fivekb_snp_gene_loops_json = {'type': 'longrange',\n",
    "##                    'filename': os.path.basename(fivekb_snp_gene_loops_fn) + '.gz',\n",
    "##                    'name': '5kb SNP-Gene Loops',\n",
    "##                    'options': {'displayMode': 'arc', 'color':'red', 'height': 200},\n",
    "##                    'showOnHubLoad': False\n",
    "##                 }\n",
    "#\n",
    "#\n",
    "#eqtl_snp_gene_pairs_json = {'type': 'longrange',\n",
    "#                'filename': os.path.basename(eqtl_snp_gene_pairs_fn) + '.gz',\n",
    "#                'name': 'eQTL SNP-Gene Pairs',\n",
    "#                'options': {'displayMode': 'arc', 'color':'purple', 'height': 200},\n",
    "#                'showOnHubLoad': True\n",
    "#               }\n",
    "#\n",
    "#\n",
    "#eqtl_snps_only_json = {'type': 'bed',\n",
    "#                'filename': os.path.basename(eqtl_snps_only_fn) + '.gz',\n",
    "#                'name': 'eQTL SNPs only',\n",
    "#                'options': {'color':'purple'},\n",
    "#                'showOnHubLoad': True\n",
    "#\n",
    "#               }\n",
    "#\n",
    "#\n",
    "#eqtl_genes_only_json = {'type': 'bed',\n",
    "#                'filename': os.path.basename(eqtl_genes_only_fn) + '.gz',\n",
    "#                'name': 'eQTL genes only',\n",
    "#                'options': {'color':'purple'},\n",
    "#                'showOnHubLoad': True\n",
    "#               }\n",
    "#\n",
    "#\n",
    "#coloc_snp_gene_pairs_json = {'type': 'longrange',\n",
    "#                'filename': os.path.basename(coloc_snp_gene_pairs_fn) + '.gz',\n",
    "#                'name': 'coloc SNP-Gene Pairs',\n",
    "#                'options': {'displayMode': 'arc', 'color':'purple', 'height': 200},\n",
    "#                'showOnHubLoad': True\n",
    "#               }\n",
    "#\n",
    "#coloc_snp_gene_loops_json = {'type': 'longrange',\n",
    "#                'filename': os.path.basename(coloc_snp_gene_loops_fn) + '.gz',\n",
    "#                'name': 'coloc SNP-Gene Loops',\n",
    "#                'options': {'displayMode': 'arc', 'color':'red', 'height': 200},\n",
    "#                'showOnHubLoad': True\n",
    "#               }\n",
    "#\n",
    "#coloc_snps_only_json = {'type': 'bed',\n",
    "#                'filename': os.path.basename(coloc_snps_only_fn) + '.gz',\n",
    "#                'name': 'coloc SNPs only',\n",
    "#                'options': {'color':'purple'},\n",
    "#                'showOnHubLoad': False\n",
    "#               }\n",
    "#\n",
    "#\n",
    "#coloc_genes_only_json = {'type': 'bed',\n",
    "#                'filename': os.path.basename(coloc_genes_only_fn) + '.gz',\n",
    "#                'name': 'coloc genes only',\n",
    "#                'options': {'color':'purple'},\n",
    "#                'showOnHubLoad': False\n",
    "#               }\n",
    "#\n",
    "#\n",
    "#coloc_anchors_only_json = {'type': 'bed',\n",
    "#                'filename': os.path.basename(coloc_anchors_fn) + '.gz',\n",
    "#                'name': 'coloc loop anchors only',\n",
    "#                'options': {'color':'red'},\n",
    "#                'showOnHubLoad': False\n",
    "#               }\n",
    "#\n",
    "#\n",
    "##hub_json = [orig_loops_json, sg_pairs_json, sg_loops_json, sg_snps_json, sg_genes_json]\n",
    "#hub_json = [gene_refbed_json, \n",
    "#           #fivekb_snp_gene_pairs_json,\n",
    "#           #fivekb_snp_gene_loops_json,\n",
    "#           eqtl_snps_only_json,\n",
    "#           eqtl_genes_only_json,\n",
    "#           eqtl_snp_gene_pairs_json,\n",
    "#           coloc_snps_only_json,\n",
    "#           coloc_genes_only_json,\n",
    "#           coloc_snp_gene_pairs_json, \n",
    "#           coloc_anchors_only_json,\n",
    "#           coloc_snp_gene_loops_json,\n",
    "#           orig_loops_json, \n",
    "#          ]\n",
    "#\n",
    "#hub_json_fn = os.path.join(outdir, 'hub.config.json')\n",
    "#with open(hub_json_fn, 'w') as f:\n",
    "#    f.write(json.dumps(hub_json, indent=4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84df1cda",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bc1bd77",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d21c8c41",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6e81f55",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85b4b602",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adcf7a95",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fbf90fb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb5ff753",
   "metadata": {},
   "outputs": [],
   "source": [
    "#loop_gz = os.path.abspath(loop_fn.replace('.bed', '_WashU.bed.gz'))\n",
    "#loop_tbi = os.path.abspath(loop_fn.replace('.bed', '_WashU.bed.gz.tbi'))\n",
    "#loop_gz_link = os.path.join(outdir, os.path.basename(loop_gz))\n",
    "#loop_tbi_link = os.path.join(outdir, os.path.basename(loop_tbi))\n",
    "#if not os.path.exists(loop_gz_link):\n",
    "#    os.link(loop_gz, loop_gz_link)\n",
    "#    os.link(loop_tbi, loop_tbi_link)# ## All 5kb Washu Files\n",
    "\n",
    "## make the longrange link for snp-gene fivekb pairs (status: running)\n",
    "##fivekb_lrange = fivekb_lrange[fivekb_lrange.rsid.notna()].reset_index(drop=True)\n",
    "#fivekb_lrange = fivekb_genes.copy()\n",
    "\n",
    "## convert full for viz\n",
    "#fivekb_lrange = fivekb_lrange[['chrA', 'startA', 'endA', 'chrB', 'startB', 'endB']]\n",
    "#fivekb_lrange.iloc[:, 1] -= 1 \n",
    "#fivekb_lrange['score'] = 0.01\n",
    "\n",
    "#fivekb_snp_gene_pairs_fn = os.path.join(outdir, '5kb.snp_gene_pairs.bed')\n",
    "#bedpe_to_WashU_longrange(fivekb_snp_gene_pairs_fn, fivekb_lrange)\n",
    "\n",
    "## plot all snp-gene pairs with a loop\n",
    "#fivekb_gloops_lrange = fivekb_gloops.iloc[:, [11,12,13,14,15,16,18]]\n",
    "#\n",
    "#fivekb_snp_gene_loops_fn = os.path.join(outdir, '5kb.snp_gene_loops.bed')\n",
    "#bedpe_to_WashU_longrange(fivekb_snp_gene_loops_fn, fivekb_gloops_lrange)# ## eQTL WashU Files\n",
    "#\n",
    "## make the longrange link for snp-gene eQTL pairs (status: running)\n",
    "#eqtl_lrange = master.loc[master.is_eqtl_pair == 1]\n",
    "#eqtl_lrange = eqtl_lrange[eqtl_lrange.rsid.notna()].reset_index(drop=True)\n",
    "#\n",
    "## convert full for viz\n",
    "#eqtl_snp_gene_pairs = eqtl_lrange[['chrom', 'snp_pos', 'snp_pos', 'chrom', 'tss_start', 'tss_end', 'genename']]\n",
    "#eqtl_snp_gene_pairs.iloc[:, 1] -= 1 \n",
    "#eqtl_snp_gene_pairs['score'] = 0.01\n",
    "#\n",
    "#eqtl_snp_gene_pairs_fn = os.path.join(outdir, 'eqtl.snp_gene_pairs.bed')\n",
    "#bedpe_to_WashU_longrange(eqtl_snp_gene_pairs_fn, eqtl_snp_gene_pairs.iloc[:, [0,1,2,3,4,5,7]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0c4012d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7c5a318",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6d5a6d3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "aadc22fa",
   "metadata": {},
   "source": [
    "### make the bed for egenes only (status: running)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0155547e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#eqtl_genes = eqtl_lrange.loc[:, ['chrom', 'gene_start', 'gene_end', 'gene_name']]\n",
    "#eqtl_genes_only_fn = os.path.join(outdir, 'eqtl.genes_only.bed')\n",
    "#bed_WashU_bedgz(eqtl_genes_only_fn, eqtl_genes)# make the bed for eSNPs only (status: running)\n",
    "#eqtl_snps = eqtl_lrange.loc[:, ['chrom', 'snp_pos', 'snp_pos', 'rsid']]\n",
    "#eqtl_snps.iloc[:, 1] -= 1\n",
    "#eqtl_snps = pbt.BedTool.from_dataframe(eqtl_snps)\n",
    "#eqtl_snps = eqtl_snps.slop(b=500, g=gs_fn).to_dataframe()\n",
    "#eqtl_snps_only_fn = os.path.join(outdir, 'eqtl.snps_only.bed')\n",
    "#bed_WashU_bedgz(eqtl_snps_only_fn, eqtl_snps)\n",
    "#"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e69cdc0",
   "metadata": {},
   "source": [
    "### make the longrange link for snp-gene eQTL loops (status: didn't add loop coordinates to master.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a41bf8cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "## eqtl_loops = master.loc[master.is_eqtl_pair == 1]# ## Coloc WashU Files\n",
    "#\n",
    "## make the bed for colocalized SNPs (status: running)\n",
    "#final_snps = coloc_sig_df.copy()\n",
    "#final_snps = final_snps.loc[~final_snps.duplicated()]\n",
    "#final_snps = pbt.BedTool.from_dataframe(final_snps)\n",
    "#final_snps = final_snps.slop(b=500, g=gs_fn)\n",
    "#final_snps = final_snps.to_dataframe()[0:4]\n",
    "#\n",
    "#coloc_snps_only_fn = os.path.join(outdir, 'coloc.snps_only.bed')\n",
    "#bed_WashU_bedgz(coloc_snps_only_fn, final_snps)\n",
    "#"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0d1b3cf",
   "metadata": {},
   "source": [
    "### make the bed for colocalized genes (status: running)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "470a2353",
   "metadata": {},
   "outputs": [],
   "source": [
    "#final_coloc_snp_genes_pairs = master[(master.is_coloc_pair == 1)]\n",
    "#final_coloc_genes = final_coloc_snp_genes_pairs[['chrom', 'gene_start' ,'gene_end', 'gene_name']]\n",
    "#coloc_genes_only_fn = os.path.join(outdir, 'coloc.genes_only.bed')\n",
    "#bed_WashU_bedgz(coloc_genes_only_fn, final_coloc_genes)# make the longrange for colocalized snp-gene pairs (status: running)\n",
    "#final_coloc_snp_genes_pairs_out = final_coloc_snp_genes_pairs[['chrom', 'snp_pos', 'snp_pos', 'chrom', \n",
    "#                                                              'tss_start', 'tss_end', 'gene_name']]\n",
    "#final_coloc_snp_genes_pairs_out.iloc[:, 1] -= 1\n",
    "#final_coloc_snp_genes_pairs_out.iloc[:, 6] = 0.01\n",
    "#\n",
    "#coloc_snp_gene_pairs_fn = os.path.join(outdir, 'coloc.snp_gene_pairs.bed')\n",
    "#bedpe_to_WashU_longrange(coloc_snp_gene_pairs_fn, final_coloc_snp_genes_pairs_out)# make the longrange for colocalized snp-gene loops (status: running)\n",
    "#final_coloc_snp_genes_loops = fivekb_gloops[(fivekb_gloops[10].isin(final_coloc_snp_genes_pairs['rsid'])) & \n",
    "#             (fivekb_gloops[12].isin(final_coloc_snp_genes_pairs['geneid']))]\n",
    "#final_coloc_snp_genes_loops_out = final_coloc_snp_genes_loops.iloc[:, [14,15,16,17,18,19,21]]\n",
    "#\n",
    "#coloc_snp_gene_loops_fn = os.path.join(outdir, 'coloc.snp_gene_loops.bed')\n",
    "#bedpe_to_WashU_longrange(coloc_snp_gene_loops_fn, final_coloc_snp_genes_loops_out)\n",
    "#\n",
    "#final_coloc_snp_genes_loops_out# make the loop anchors as bed files (status: developing)\n",
    "#left = final_coloc_snp_genes_loops[[14,15,16,6]].T.reset_index(drop=True).T\n",
    "#left[3] = 'L-' + left[3] \n",
    "#right = final_coloc_snp_genes_loops[[17,18,19,6]].T.reset_index(drop=True).T\n",
    "#right[3] = 'R-' + right[3] \n",
    "#anchors = pd.concat([left, right], ignore_index=True, axis=0)coloc_anchors_fn = os.path.join(outdir, 'coloc.anchors_only.bed')\n",
    "#bed_WashU_bedgz(coloc_anchors_fn, anchors)\n",
    "#"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0bbac8e",
   "metadata": {},
   "source": [
    "### # make the hub json file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a2943e40",
   "metadata": {},
   "outputs": [],
   "source": [
    "##gwas, cline = coloc_fn.split('/')[5:7]\n",
    "#\n",
    "#print(\"# make the hub json file\")\n",
    "#\n",
    "#gene_refbed_json = {'type': 'refbed',\n",
    "#                   'filename': os.path.basename(sg_genes_fn) + '.gz',\n",
    "#                   'name': 'Gencode V19',\n",
    "#                   'showOnHubLoad': True\n",
    "#                  }\n",
    "#\n",
    "#orig_loops_json = {'type': 'longrange',\n",
    "#                  'filename': os.path.basename(loop_gz_link),\n",
    "#                  'name': 'Original Loops',\n",
    "#                  'options': {'displayMode': 'arc', 'color':'red'},\n",
    "#                  'showOnHubLoad': True\n",
    "#               }\n",
    "#\n",
    "## fivekb_snp_gene_pairs_json = {'type': 'longrange',\n",
    "##                    'filename': os.path.basename(fivekb_snp_gene_pairs_fn) + '.gz',\n",
    "##                    'name': '5kb SNP-Gene Pairs',\n",
    "##                    'options': {'displayMode': 'arc', 'color':'purple', 'height': 200},\n",
    "##                    'showOnHubLoad': False\n",
    "##                 }\n",
    "#\n",
    "## fivekb_snp_gene_loops_json = {'type': 'longrange',\n",
    "##                    'filename': os.path.basename(fivekb_snp_gene_loops_fn) + '.gz',\n",
    "##                    'name': '5kb SNP-Gene Loops',\n",
    "##                    'options': {'displayMode': 'arc', 'color':'red', 'height': 200},\n",
    "##                    'showOnHubLoad': False\n",
    "##                 }\n",
    "#\n",
    "#\n",
    "#eqtl_snp_gene_pairs_json = {'type': 'longrange',\n",
    "#                'filename': os.path.basename(eqtl_snp_gene_pairs_fn) + '.gz',\n",
    "#                'name': 'eQTL SNP-Gene Pairs',\n",
    "#                'options': {'displayMode': 'arc', 'color':'purple', 'height': 200},\n",
    "#                'showOnHubLoad': True\n",
    "#               }\n",
    "#\n",
    "#\n",
    "#eqtl_snps_only_json = {'type': 'bed',\n",
    "#                'filename': os.path.basename(eqtl_snps_only_fn) + '.gz',\n",
    "#                'name': 'eQTL SNPs only',\n",
    "#                'options': {'color':'purple'},\n",
    "#                'showOnHubLoad': True\n",
    "#\n",
    "#               }\n",
    "#\n",
    "#\n",
    "#eqtl_genes_only_json = {'type': 'bed',\n",
    "#                'filename': os.path.basename(eqtl_genes_only_fn) + '.gz',\n",
    "#                'name': 'eQTL genes only',\n",
    "#                'options': {'color':'purple'},\n",
    "#                'showOnHubLoad': True\n",
    "#               }\n",
    "#\n",
    "#\n",
    "#coloc_snp_gene_pairs_json = {'type': 'longrange',\n",
    "#                'filename': os.path.basename(coloc_snp_gene_pairs_fn) + '.gz',\n",
    "#                'name': 'coloc SNP-Gene Pairs',\n",
    "#                'options': {'displayMode': 'arc', 'color':'purple', 'height': 200},\n",
    "#                'showOnHubLoad': True\n",
    "#               }\n",
    "#\n",
    "#coloc_snp_gene_loops_json = {'type': 'longrange',\n",
    "#                'filename': os.path.basename(coloc_snp_gene_loops_fn) + '.gz',\n",
    "#                'name': 'coloc SNP-Gene Loops',\n",
    "#                'options': {'displayMode': 'arc', 'color':'red', 'height': 200},\n",
    "#                'showOnHubLoad': True\n",
    "#               }\n",
    "#\n",
    "#coloc_snps_only_json = {'type': 'bed',\n",
    "#                'filename': os.path.basename(coloc_snps_only_fn) + '.gz',\n",
    "#                'name': 'coloc SNPs only',\n",
    "#                'options': {'color':'purple'},\n",
    "#                'showOnHubLoad': False\n",
    "#               }\n",
    "#\n",
    "#\n",
    "#coloc_genes_only_json = {'type': 'bed',\n",
    "#                'filename': os.path.basename(coloc_genes_only_fn) + '.gz',\n",
    "#                'name': 'coloc genes only',\n",
    "#                'options': {'color':'purple'},\n",
    "#                'showOnHubLoad': False\n",
    "#               }\n",
    "#\n",
    "#\n",
    "#coloc_anchors_only_json = {'type': 'bed',\n",
    "#                'filename': os.path.basename(coloc_anchors_fn) + '.gz',\n",
    "#                'name': 'coloc loop anchors only',\n",
    "#                'options': {'color':'red'},\n",
    "#                'showOnHubLoad': False\n",
    "#               }\n",
    "#\n",
    "#\n",
    "##hub_json = [orig_loops_json, sg_pairs_json, sg_loops_json, sg_snps_json, sg_genes_json]\n",
    "#hub_json = [gene_refbed_json, \n",
    "#           #fivekb_snp_gene_pairs_json,\n",
    "#           #fivekb_snp_gene_loops_json,\n",
    "#           eqtl_snps_only_json,\n",
    "#           eqtl_genes_only_json,\n",
    "#           eqtl_snp_gene_pairs_json,\n",
    "#           coloc_snps_only_json,\n",
    "#           coloc_genes_only_json,\n",
    "#           coloc_snp_gene_pairs_json, \n",
    "#           coloc_anchors_only_json,\n",
    "#           coloc_snp_gene_loops_json,\n",
    "#           orig_loops_json, \n",
    "#          ]\n",
    "#\n",
    "#hub_json_fn = os.path.join(outdir, 'hub.config.json')\n",
    "#with open(hub_json_fn, 'w') as f:\n",
    "#    f.write(json.dumps(hub_json, indent=4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43d00ca5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "HiChIP DB",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
