{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6f173ae9",
   "metadata": {},
   "outputs": [],
   "source": [
    "######## snakemake preamble start (automatically inserted, do not edit) ########\n",
    "import sys; sys.path.extend(['/mnt/BioAdHoc/Groups/vd-ay/jreyna/software/mambaforge/envs/hichip-db/lib/python3.9/site-packages', '/mnt/BioAdHoc/Groups/vd-ay/jreyna/projects/dchallenge/workflow/rules/../scripts/sgls']); import pickle; snakemake = pickle.loads(b'\\x80\\x04\\x95>\\x0f\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x10snakemake.script\\x94\\x8c\\tSnakemake\\x94\\x93\\x94)\\x81\\x94}\\x94(\\x8c\\x05input\\x94\\x8c\\x0csnakemake.io\\x94\\x8c\\nInputFiles\\x94\\x93\\x94)\\x81\\x94(\\x8c[results/main/sgls/T1D_34012112_Gaulton/Quach_2016/monocyte_R848/eqtls.coloc_filtered.tsv.gz\\x94\\x8cWresults/main/coloc/Results/eQTL_Catalogue/T1D_34012112_Gaulton/Quach_2016/monocyte_R848\\x94\\x8cpresults/main/coloc/eQTL_Catalogue/T1D_34012112_Gaulton/GRCh37/Quach_2016/monocyte_R848/ldpairs/coloc_ld_snps.txt\\x94\\x8ceresults/main/h3k27ac_hichip/monocyte_naive/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01.grch38.bed\\x94\\x8c\"results/refs/hg38/hg38.chrom.sizes\\x94\\x8c3results/refs/gencode/v30/gencode.v30.annotation.bed\\x94e}\\x94(\\x8c\\x06_names\\x94}\\x94(\\x8c\\x04eqtl\\x94K\\x00K\\x01\\x86\\x94\\x8c\\tcoloc_dir\\x94K\\x01N\\x86\\x94\\x8c\\x05coloc\\x94K\\x02N\\x86\\x94\\x8c\\x05loops\\x94K\\x03N\\x86\\x94\\x8c\\x0cgenome_sizes\\x94K\\x04N\\x86\\x94\\x8c\\x07gencode\\x94K\\x05N\\x86\\x94u\\x8c\\x12_allowed_overrides\\x94]\\x94(\\x8c\\x05index\\x94\\x8c\\x04sort\\x94eh!\\x8c\\tfunctools\\x94\\x8c\\x07partial\\x94\\x93\\x94h\\x06\\x8c\\x19Namedlist._used_attribute\\x94\\x93\\x94\\x85\\x94R\\x94(h\\')}\\x94\\x8c\\x05_name\\x94h!sNt\\x94bh\"h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h\"sNt\\x94bh\\x13h\\x06\\x8c\\tNamedlist\\x94\\x93\\x94)\\x81\\x94h\\na}\\x94(h\\x11}\\x94h\\x1f]\\x94(h!h\"eh!h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h!sNt\\x94bh\"h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h\"sNt\\x94bubh\\x15h\\x0bh\\x17h\\x0ch\\x19h\\rh\\x1bh\\x0eh\\x1dh\\x0fub\\x8c\\x06output\\x94h\\x06\\x8c\\x0bOutputFiles\\x94\\x93\\x94)\\x81\\x94\\x8c]results/main/sgls/ldpairs/T1D_34012112_Gaulton/GRCh37/Quach_2016/monocyte_R848/monocyte_naive\\x94a}\\x94(h\\x11}\\x94\\x8c\\x06outdir\\x94K\\x00N\\x86\\x94sh\\x1f]\\x94(h!h\"eh!h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h!sNt\\x94bh\"h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h\"sNt\\x94bhFhCub\\x8c\\x06params\\x94h\\x06\\x8c\\x06Params\\x94\\x93\\x94)\\x81\\x94M\\xa8aa}\\x94(h\\x11}\\x94\\x8c\\tloop_slop\\x94K\\x00N\\x86\\x94sh\\x1f]\\x94(h!h\"eh!h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h!sNt\\x94bh\"h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h\"sNt\\x94bhWM\\xa8aub\\x8c\\twildcards\\x94h\\x06\\x8c\\tWildcards\\x94\\x93\\x94)\\x81\\x94(\\x8c\\x14T1D_34012112_Gaulton\\x94\\x8c\\x06GRCh37\\x94\\x8c\\nQuach_2016\\x94\\x8c\\rmonocyte_R848\\x94\\x8c\\x0emonocyte_naive\\x94e}\\x94(h\\x11}\\x94(\\x8c\\x0bgwas_source\\x94K\\x00N\\x86\\x94\\x8c\\nref_genome\\x94K\\x01N\\x86\\x94\\x8c\\x0beqtl_source\\x94K\\x02N\\x86\\x94\\x8c\\tge_source\\x94K\\x03N\\x86\\x94\\x8c\\x0bloop_source\\x94K\\x04N\\x86\\x94uh\\x1f]\\x94(h!h\"eh!h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h!sNt\\x94bh\"h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h\"sNt\\x94bhmhfhohghqhhhshihuhjub\\x8c\\x07threads\\x94K\\x01\\x8c\\tresources\\x94h\\x06\\x8c\\tResources\\x94\\x93\\x94)\\x81\\x94(K\\x01K\\x01\\x8c\\x08/scratch\\x94M\\xc0]K\\x01K\\x01e}\\x94(h\\x11}\\x94(\\x8c\\x06_cores\\x94K\\x00N\\x86\\x94\\x8c\\x06_nodes\\x94K\\x01N\\x86\\x94\\x8c\\x06tmpdir\\x94K\\x02N\\x86\\x94\\x8c\\x06mem_mb\\x94K\\x03N\\x86\\x94\\x8c\\x05nodes\\x94K\\x04N\\x86\\x94\\x8c\\x03ppn\\x94K\\x05N\\x86\\x94uh\\x1f]\\x94(h!h\"eh!h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h!sNt\\x94bh\"h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h\"sNt\\x94bh\\x88K\\x01h\\x8aK\\x01h\\x8ch\\x85h\\x8eM\\xc0]h\\x90K\\x01h\\x92K\\x01ub\\x8c\\x03log\\x94h\\x06\\x8c\\x03Log\\x94\\x93\\x94)\\x81\\x94\\x8cnresults/main/sgls/logs/annotate_colocs.T1D_34012112_Gaulton.GRCh37.Quach_2016.monocyte_R848.monocyte_naive.log\\x94a}\\x94(h\\x11}\\x94h\\x1f]\\x94(h!h\"eh!h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h!sNt\\x94bh\"h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h\"sNt\\x94bub\\x8c\\x06config\\x94}\\x94(\\x8c\\x0chichip_db_py\\x94\\x8c;/mnt/BioHome/jreyna/software/anaconda3/hichip-db/bin/python\\x94\\x8c\\nhic_tls_py\\x94\\x8c>/mnt/BioHome/jreyna/software/anaconda3/envs/hic_tls/bin/python\\x94\\x8c\\thic_tls_R\\x94\\x8c\\'/share/apps/R/3.4.3/lib64/R/bin/Rscript\\x94\\x8c\\x02R4\\x94\\x8c2/mnt/BioHome/jreyna/R-Versions/R-4.0.3/bin/Rscript\\x94\\x8c\\x06fithic\\x94\\x8c4/mnt/BioHome/jreyna/software/fithic/fithic/fithic.py\\x94\\x8c\\nfithic_tls\\x94\\x8c8/mnt/BioHome/jreyna/software/fithic/fithic/fithic_tls.py\\x94\\x8c\\rhicpro2fithic\\x94\\x8cA/mnt/BioHome/jreyna/software/fithic/fithic/utils/HiCPro2FitHiC.py\\x94\\x8c\\x0chicpro_utils\\x94\\x8c//mnt/BioHome/jreyna/software/HiC-Pro/bin/utils/\\x94\\x8c\\x0cjuicer_tools\\x94\\x8c{/mnt/BioAdHoc/Groups/vd-ay/abhijit/overflow/proj_overflow/Alternate_Chromosomal_structures/juicer_tools.1.8.9_jcuda.0.8.jar\\x94\\x8c\\x05Rpath\\x94\\x8c!/share/apps/R/3.4.3/lib64/R/bin/R\\x94\\x8c\\x06CLINES\\x94]\\x94(\\x8c\\x04A549\\x94\\x8c\\x05Caki2\\x94\\x8c\\x06HEK293\\x94\\x8c\\x05LNCaP\\x94\\x8c\\x08NCI-H460\\x94\\x8c\\x05Panc1\\x94\\x8c\\x05SKNSH\\x94e\\x8c\\x12DIXON_HIC_OVERLAPS\\x94]\\x94(\\x8c\\x04A549\\x94\\x8c\\x05Caki2\\x94\\x8c\\x05LNCaP\\x94\\x8c\\x08NCI-H460\\x94\\x8c\\x05Panc1\\x94\\x8c\\x05SKNSH\\x94e\\x8c\\x12DIXON_WGS_OVERLAPS\\x94]\\x94(\\x8c\\x04A549\\x94\\x8c\\x05Caki2\\x94\\x8c\\x05LNCaP\\x94\\x8c\\x08NCI-H460\\x94\\x8c\\x05Panc1\\x94e\\x8c\\x08hictrans\\x94\\x8c,/home/jreyna/software/HiCtrans/hictrans.v3.R\\x94\\x8c\\x05hicnv\\x94\\x8c&/home/jreyna/software/HiCnv/hicnv_v2.R\\x94\\x8c\\rhicnv_scripts\\x94\\x8c$/home/jreyna/software/HiCnv/scripts/\\x94\\x8c\\x07python2\\x94\\x8cD/mnt/BioHome/jreyna/software/anaconda3/envs/python_2.7.18/bin/python\\x94\\x8c\\x08mustache\\x94\\x8cA/mnt/BioHome/jreyna/software/anaconda3/envs/mustache/bin/mustache\\x94\\x8c\\x0fmustache_python\\x94\\x8c?/mnt/BioHome/jreyna/software/anaconda3/envs/mustache/bin/python\\x94\\x8c\\x14juicer_tools_1.13.02\\x94\\x8c./home/jreyna/software/juicer_tools_1.13.02.jar\\x94\\x8c\\x06cooler\\x94\\x8c>/mnt/BioHome/jreyna/software/anaconda3/envs/hic_tls/bin/cooler\\x94\\x8c\\x05bgzip\\x94\\x8c=/mnt/BioHome/jreyna/software/anaconda3/envs/hic_tls/bin/bgzip\\x94\\x8c\\x08liftover\\x94\\x8c1/home/jreyna/software/UCSC_Browser_Tools/liftOver\\x94\\x8c\\x10hichip_db_python\\x94\\x8c4/home/jreyna/software/anaconda3/hichip-db/bin/python\\x94u\\x8c\\x04rule\\x94\\x8c\\x19Find_SGS_for_Coloc_and_LD\\x94\\x8c\\x0fbench_iteration\\x94N\\x8c\\tscriptdir\\x94\\x8cT/mnt/BioAdHoc/Groups/vd-ay/jreyna/projects/dchallenge/workflow/rules/../scripts/sgls\\x94ub.'); from snakemake.logging import logger; logger.printshellcmds = False; import os; os.chdir(r'/mnt/BioAdHoc/Groups/vd-ay/jreyna/projects/dchallenge');\n",
    "######## snakemake preamble end #########\n",
    "\n",
    "import os \n",
    "import sys\n",
    "import pybedtools as pbt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess as sp\n",
    "import json\n",
    "from biolib import liftover as liftover_utility\n",
    "from biolib import coordinates as biocoords\n",
    "pbt.helpers.set_tempdir('/mnt/BioHome/jreyna/tmp/')\n",
    "pbt.set_bedtools_path('/mnt/BioApps/bedtools/bin/')\n",
    "bgzip = '/mnt/BioApps/tabix/tabix-0.2.6/bgzip'\n",
    "tabix = '/mnt/BioApps/tabix/tabix-0.2.6/tabix'\n",
    "res = 5000\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "500ed886",
   "metadata": {},
   "outputs": [],
   "source": [
    "bedpe_6cols = ['chrA', 'startA', 'endA', 'chrB', 'startB', 'endB']\n",
    "bedpe_10cols = ['chrA', 'startA', 'endA', 'chrB', 'startB', 'endB', 'name', 'score', 'strand1', 'strand2']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36357d8a",
   "metadata": {},
   "source": [
    "# Parse the Snakemake arguments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "be4c0ba1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# snakemake.input = ['results/main/GRCh37/sgls/T1D_34012112_Gaulton/Quach_2016/monocyte_R848/eqtls.coloc_filtered.tsv.gz',\n",
    "#               'results/main/coloc/eQTL_Catalogue/T1D_34012112_Gaulton/GRCh37/Quach_2016/monocyte_R848/',\n",
    "#               'results/main/GRCh37/coloc/eQTL_Catalogue/T1D_34012112_Gaulton/Quach_2016/monocyte_R848/ldpairs/coloc_ld_snps.txt',\n",
    "#               'results/main/h3k27ac_hichip/monocyte_naive/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01.bed',\n",
    "#               'results/refs/hg19/hg19.chrom.sizes',\n",
    "#               'results/refs/gencode/v30/gencode.v30.annotation.grch37.bed']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "64926b2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_fn = snakemake.input[0]\n",
    "coloc_fn = snakemake.input[2]\n",
    "loop_fn = snakemake.input[3]\n",
    "celltype = snakemake.wildcards[3]\n",
    "gs_fn = snakemake.input[4]\n",
    "genes_fn = snakemake.input[5]\n",
    "loop_slop = snakemake.params[0]\n",
    "outdir = snakemake.output[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "afcb7c4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['T1D_34012112_Gaulton',\n",
       " 'GRCh37',\n",
       " 'Quach_2016',\n",
       " 'monocyte_R848',\n",
       " 'monocyte_naive']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snakemake.wildcards"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "10372205",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'results/main/sgls/T1D_34012112_Gaulton/Quach_2016/monocyte_R848/eqtls.coloc_filtered.tsv.gz'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eqtl_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2cc90668",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'results/main/coloc/eQTL_Catalogue/T1D_34012112_Gaulton/GRCh37/Quach_2016/monocyte_R848/ldpairs/coloc_ld_snps.txt'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coloc_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d900dff2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'results/main/h3k27ac_hichip/monocyte_naive/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01.grch38.bed'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loop_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "72efe3a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'results/refs/hg38/hg38.chrom.sizes'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gs_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "07c638dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'results/refs/gencode/v30/gencode.v30.annotation.bed'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "189c85cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25000"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loop_slop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4cf55dd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'results/main/sgls/ldpairs/T1D_34012112_Gaulton/GRCh37/Quach_2016/monocyte_R848/monocyte_naive'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outdir"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddd631a1",
   "metadata": {},
   "source": [
    "# Load the colocalization data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "263f701d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the colocalization data\n",
    "coloc = pd.read_table(coloc_fn)\n",
    "\n",
    "# rename the columns\n",
    "rename_coloc_cols = {\n",
    "    'chr': 'coloc_chr',\n",
    "    'pos': 'coloc_pos',\n",
    "    'pp_H0_Coloc_Summary': 'ppH0',\n",
    "    'pp_H1_Coloc_Summary': 'ppH1',\n",
    "    'pp_H2_Coloc_Summary': 'ppH2',\n",
    "    'pp_H3_Coloc_Summary': 'ppH3',\n",
    "    'pp_H4_Coloc_Summary': 'ppH4',\n",
    "    'rs_id': 'coloc_rsid',\n",
    "    'variant_id': 'varid',\n",
    "    'geneName': 'geneid',\n",
    "    'dist': 'dist',\n",
    "    'pvalue': 'eqtl_pval',\n",
    "    'FDR': 'eqtl_fdr',\n",
    "    'slope_snp': 'eqtl_slope',\n",
    "    'ref': 'ref',\n",
    "    'alt': 'alt',\n",
    "    'AC': 'AC',\n",
    "    'AF': 'AF',\n",
    "    'AN': 'AN',\n",
    "    'slope_se_snp': 'eqtl_slope_se',\n",
    "    'slope_gwas': 'gwas_slope',\n",
    "    'slope_se_gwas': 'gwas_slope_se',\n",
    "    'pval_nominal': 'gwas_pval',\n",
    "    'SampleSize': 'sample_size',\n",
    "    'rsID': 'coloc_rsid_dup',\n",
    "    'ld_chr': 'chr',\n",
    "    'ld_pos': 'pos',\n",
    "    'ld_rsID': 'rsid'}\n",
    "coloc.rename(columns=rename_coloc_cols, inplace=True)\n",
    "\n",
    "# reorder the columns \n",
    "major_cols = ['chr', 'pos', 'rsid',\n",
    "             'ppH0', 'ppH1', 'ppH2', 'ppH3', 'ppH4',\n",
    "             'varid', 'geneid', 'dist',\n",
    "             'eqtl_pval', 'eqtl_fdr', 'eqtl_slope',\n",
    "             'ref', 'alt', 'AC', 'AF', 'AN',\n",
    "             'eqtl_slope_se', 'gwas_slope', 'gwas_slope_se', 'gwas_pval',\n",
    "             'coloc_chr', 'coloc_pos', 'coloc_rsid', 'LD']\n",
    "coloc = coloc[major_cols]\n",
    "\n",
    "# construct the sid\n",
    "coloc.loc[:, 'sid'] = coloc['chr'].str.replace('chr', '') + ':' + coloc['pos'].astype(str) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3eba1180",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>rsid</th>\n",
       "      <th>ppH0</th>\n",
       "      <th>ppH1</th>\n",
       "      <th>ppH2</th>\n",
       "      <th>ppH3</th>\n",
       "      <th>ppH4</th>\n",
       "      <th>varid</th>\n",
       "      <th>geneid</th>\n",
       "      <th>dist</th>\n",
       "      <th>eqtl_pval</th>\n",
       "      <th>eqtl_fdr</th>\n",
       "      <th>eqtl_slope</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>AC</th>\n",
       "      <th>AF</th>\n",
       "      <th>AN</th>\n",
       "      <th>eqtl_slope_se</th>\n",
       "      <th>gwas_slope</th>\n",
       "      <th>gwas_slope_se</th>\n",
       "      <th>gwas_pval</th>\n",
       "      <th>coloc_chr</th>\n",
       "      <th>coloc_pos</th>\n",
       "      <th>coloc_rsid</th>\n",
       "      <th>LD</th>\n",
       "      <th>sid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9</td>\n",
       "      <td>4285986</td>\n",
       "      <td>rs10758591</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000916</td>\n",
       "      <td>0.192281</td>\n",
       "      <td>0.806803</td>\n",
       "      <td>9:4282536:A:G</td>\n",
       "      <td>ENSG00000107249</td>\n",
       "      <td>65856</td>\n",
       "      <td>0.00644</td>\n",
       "      <td>0.336787</td>\n",
       "      <td>-0.407677</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1754</td>\n",
       "      <td>0.35024</td>\n",
       "      <td>5008</td>\n",
       "      <td>0.005361</td>\n",
       "      <td>-0.121582</td>\n",
       "      <td>0.013995</td>\n",
       "      <td>3.710000e-18</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4282536</td>\n",
       "      <td>rs4380994</td>\n",
       "      <td>0.984588</td>\n",
       "      <td>9:4285986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9</td>\n",
       "      <td>4289196</td>\n",
       "      <td>rs10814914</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000916</td>\n",
       "      <td>0.192281</td>\n",
       "      <td>0.806803</td>\n",
       "      <td>9:4282536:A:G</td>\n",
       "      <td>ENSG00000107249</td>\n",
       "      <td>65856</td>\n",
       "      <td>0.00644</td>\n",
       "      <td>0.336787</td>\n",
       "      <td>-0.407677</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1754</td>\n",
       "      <td>0.35024</td>\n",
       "      <td>5008</td>\n",
       "      <td>0.005361</td>\n",
       "      <td>-0.121582</td>\n",
       "      <td>0.013995</td>\n",
       "      <td>3.710000e-18</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4282536</td>\n",
       "      <td>rs4380994</td>\n",
       "      <td>0.976890</td>\n",
       "      <td>9:4289196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9</td>\n",
       "      <td>4290544</td>\n",
       "      <td>rs10814915</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000916</td>\n",
       "      <td>0.192281</td>\n",
       "      <td>0.806803</td>\n",
       "      <td>9:4282536:A:G</td>\n",
       "      <td>ENSG00000107249</td>\n",
       "      <td>65856</td>\n",
       "      <td>0.00644</td>\n",
       "      <td>0.336787</td>\n",
       "      <td>-0.407677</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1754</td>\n",
       "      <td>0.35024</td>\n",
       "      <td>5008</td>\n",
       "      <td>0.005361</td>\n",
       "      <td>-0.121582</td>\n",
       "      <td>0.013995</td>\n",
       "      <td>3.710000e-18</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4282536</td>\n",
       "      <td>rs4380994</td>\n",
       "      <td>0.854214</td>\n",
       "      <td>9:4290544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9</td>\n",
       "      <td>4293150</td>\n",
       "      <td>rs10814916</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000916</td>\n",
       "      <td>0.192281</td>\n",
       "      <td>0.806803</td>\n",
       "      <td>9:4282536:A:G</td>\n",
       "      <td>ENSG00000107249</td>\n",
       "      <td>65856</td>\n",
       "      <td>0.00644</td>\n",
       "      <td>0.336787</td>\n",
       "      <td>-0.407677</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1754</td>\n",
       "      <td>0.35024</td>\n",
       "      <td>5008</td>\n",
       "      <td>0.005361</td>\n",
       "      <td>-0.121582</td>\n",
       "      <td>0.013995</td>\n",
       "      <td>3.710000e-18</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4282536</td>\n",
       "      <td>rs4380994</td>\n",
       "      <td>0.849530</td>\n",
       "      <td>9:4293150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9</td>\n",
       "      <td>4296430</td>\n",
       "      <td>rs10814917</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000916</td>\n",
       "      <td>0.192281</td>\n",
       "      <td>0.806803</td>\n",
       "      <td>9:4282536:A:G</td>\n",
       "      <td>ENSG00000107249</td>\n",
       "      <td>65856</td>\n",
       "      <td>0.00644</td>\n",
       "      <td>0.336787</td>\n",
       "      <td>-0.407677</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1754</td>\n",
       "      <td>0.35024</td>\n",
       "      <td>5008</td>\n",
       "      <td>0.005361</td>\n",
       "      <td>-0.121582</td>\n",
       "      <td>0.013995</td>\n",
       "      <td>3.710000e-18</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4282536</td>\n",
       "      <td>rs4380994</td>\n",
       "      <td>0.942558</td>\n",
       "      <td>9:4296430</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr      pos        rsid  ppH0  ppH1      ppH2      ppH3      ppH4  \\\n",
       "0  chr9  4285986  rs10758591     0     0  0.000916  0.192281  0.806803   \n",
       "1  chr9  4289196  rs10814914     0     0  0.000916  0.192281  0.806803   \n",
       "2  chr9  4290544  rs10814915     0     0  0.000916  0.192281  0.806803   \n",
       "3  chr9  4293150  rs10814916     0     0  0.000916  0.192281  0.806803   \n",
       "4  chr9  4296430  rs10814917     0     0  0.000916  0.192281  0.806803   \n",
       "\n",
       "           varid           geneid   dist  eqtl_pval  eqtl_fdr  eqtl_slope ref  \\\n",
       "0  9:4282536:A:G  ENSG00000107249  65856    0.00644  0.336787   -0.407677   A   \n",
       "1  9:4282536:A:G  ENSG00000107249  65856    0.00644  0.336787   -0.407677   A   \n",
       "2  9:4282536:A:G  ENSG00000107249  65856    0.00644  0.336787   -0.407677   A   \n",
       "3  9:4282536:A:G  ENSG00000107249  65856    0.00644  0.336787   -0.407677   A   \n",
       "4  9:4282536:A:G  ENSG00000107249  65856    0.00644  0.336787   -0.407677   A   \n",
       "\n",
       "  alt    AC       AF    AN  eqtl_slope_se  gwas_slope  gwas_slope_se  \\\n",
       "0   G  1754  0.35024  5008       0.005361   -0.121582       0.013995   \n",
       "1   G  1754  0.35024  5008       0.005361   -0.121582       0.013995   \n",
       "2   G  1754  0.35024  5008       0.005361   -0.121582       0.013995   \n",
       "3   G  1754  0.35024  5008       0.005361   -0.121582       0.013995   \n",
       "4   G  1754  0.35024  5008       0.005361   -0.121582       0.013995   \n",
       "\n",
       "      gwas_pval coloc_chr  coloc_pos coloc_rsid        LD        sid  \n",
       "0  3.710000e-18      chr9    4282536  rs4380994  0.984588  9:4285986  \n",
       "1  3.710000e-18      chr9    4282536  rs4380994  0.976890  9:4289196  \n",
       "2  3.710000e-18      chr9    4282536  rs4380994  0.854214  9:4290544  \n",
       "3  3.710000e-18      chr9    4282536  rs4380994  0.849530  9:4293150  \n",
       "4  3.710000e-18      chr9    4282536  rs4380994  0.942558  9:4296430  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract the most significant according the H4 \n",
    "coloc_sig_df = coloc.loc[~coloc.duplicated(subset=['chr', 'pos', 'geneid']),]\n",
    "coloc_sig_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e8f174eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18, 28)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coloc_sig_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1ea5a499",
   "metadata": {},
   "outputs": [],
   "source": [
    "# collect the chr and position of the main colocalized snps \n",
    "coloc_snps = coloc.loc[(coloc['chr'] == coloc['coloc_chr']) & (coloc['pos'] == coloc['coloc_pos'])] \n",
    "#coloc_snps = coloc_snps[['chr', 'pos']].values.tolist()\n",
    "#coloc_snps = set([tuple(x) for x in coloc_snps])\n",
    "coloc_snps = coloc_snps['sid'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f49129b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a pybedtool object for intersections\n",
    "coloc_sig_full = coloc_sig_df.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "bbf02849",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the bin start and end coordinates for each snp\n",
    "coloc_bins = coloc_sig_full.pos.apply(biocoords.get_snp_bin, args=(res,))\n",
    "coloc_starts, coloc_ends = list(zip(*coloc_bins.values))\n",
    "coloc_sig_df.loc[:, 'start'] = coloc_starts\n",
    "coloc_sig_df.loc[:, 'end'] = coloc_ends"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "65e29304",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 18 colocalized/LD SNP-gene pairs\n"
     ]
    }
   ],
   "source": [
    "coloc_sig_df = coloc_sig_df[['chr', 'start', 'end', 'sid', 'rsid', 'varid']]\n",
    "coloc_sig_pbt = pbt.BedTool.from_dataframe(coloc_sig_df.iloc[:, 0:5]).sort()\n",
    "\n",
    "print('There are {} colocalized/LD SNP-gene pairs'.format(coloc_sig_df.shape[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "401e0809",
   "metadata": {},
   "source": [
    "# Load the gene data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "94492bea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Load the gene data\n"
     ]
    }
   ],
   "source": [
    "print('# Load the gene data')\n",
    "\n",
    "# load the gencode coords\n",
    "cols = ['chrom', 'start', 'end', 'strand', 'type', 'geneid', 'genename']\n",
    "gencode = pd.read_table(genes_fn, header=None, names=cols)\n",
    "\n",
    "# extract just the genes\n",
    "genes_df = gencode.loc[gencode.type.isin(['gene'])]\n",
    "genes_df = genes_df.loc[~genes_df.duplicated(subset='geneid'), :]\n",
    "genes_df.loc[:, 'chrom'] = genes_df['chrom'].astype(str)\n",
    "genes_df = genes_df.iloc[:, [0,1,2,6,5,3]]\n",
    "\n",
    "# create a copy of the original gene bed before coordinate shrinking\n",
    "orig_genes_df = genes_df.copy()\n",
    "\n",
    "# convert the start/end position into start/end for the TSS\n",
    "# if the gene is + then the start is uses as the tss otherwise\n",
    "# the end is used as the tss\n",
    "genes_df.loc[(genes_df.strand == '+'), 'end'] = genes_df.loc[(genes_df.strand == '+'), 'start']\n",
    "genes_df.loc[(genes_df.strand == '+'), 'start'] = genes_df.loc[(genes_df.strand == '+'), 'start'] - 1\n",
    "genes_df.loc[(genes_df.strand == '-'), 'end'] = genes_df.loc[(genes_df.strand == '-'), 'end']\n",
    "genes_df.loc[(genes_df.strand == '-'), 'start'] = genes_df.loc[(genes_df.strand == '-'), 'end'] - 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "584f4224",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the bin start and end \n",
    "genes_pbt_pre = genes_df.copy()\n",
    "gene_bins = genes_pbt_pre.start.apply(biocoords.get_snp_bin, args=(res,))\n",
    "gene_starts, gene_ends = list(zip(*gene_bins.values))\n",
    "genes_pbt_pre.loc[:, 'start'] = gene_starts \n",
    "genes_pbt_pre.loc[:, 'end'] = gene_ends"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "081a00b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# make a genes pbt for intersection\n",
      "   chrom  start    end     genename           geneid strand\n",
      "0   chr1  11868  11869      DDX11L1  ENSG00000223972      +\n",
      "12  chr1  29569  29570       WASH7P  ENSG00000227232      -\n",
      "25  chr1  17435  17436    MIR6859-1  ENSG00000278267      -\n",
      "28  chr1  29553  29554  MIR1302-2HG  ENSG00000243485      +\n",
      "36  chr1  30365  30366    MIR1302-2  ENSG00000284332      +\n",
      "\n",
      "There are 58825 genes in this GTF-derived file.\n"
     ]
    }
   ],
   "source": [
    "# make a genes pbt for intersection\n",
    "print(\"# make a genes pbt for intersection\")\n",
    "print(genes_df.head())\n",
    "genes_pbt = pbt.BedTool.from_dataframe(genes_pbt_pre).sort()\n",
    "\n",
    "print('\\nThere are {} genes in this GTF-derived file.'.format(genes_df.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "069af174",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>genename</th>\n",
       "      <th>geneid</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10000</td>\n",
       "      <td>15000</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25000</td>\n",
       "      <td>30000</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15000</td>\n",
       "      <td>20000</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>ENSG00000278267</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr1</td>\n",
       "      <td>25000</td>\n",
       "      <td>30000</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30000</td>\n",
       "      <td>35000</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>ENSG00000284332</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2774280</th>\n",
       "      <td>chrM</td>\n",
       "      <td>10000</td>\n",
       "      <td>15000</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2774285</th>\n",
       "      <td>chrM</td>\n",
       "      <td>10000</td>\n",
       "      <td>15000</td>\n",
       "      <td>MT-TE</td>\n",
       "      <td>ENSG00000210194</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2774288</th>\n",
       "      <td>chrM</td>\n",
       "      <td>10000</td>\n",
       "      <td>15000</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>ENSG00000198727</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2774293</th>\n",
       "      <td>chrM</td>\n",
       "      <td>15000</td>\n",
       "      <td>20000</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>ENSG00000210195</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2774296</th>\n",
       "      <td>chrM</td>\n",
       "      <td>15000</td>\n",
       "      <td>20000</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>ENSG00000210196</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58825 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom  start    end     genename           geneid strand\n",
       "0        chr1  10000  15000      DDX11L1  ENSG00000223972      +\n",
       "12       chr1  25000  30000       WASH7P  ENSG00000227232      -\n",
       "25       chr1  15000  20000    MIR6859-1  ENSG00000278267      -\n",
       "28       chr1  25000  30000  MIR1302-2HG  ENSG00000243485      +\n",
       "36       chr1  30000  35000    MIR1302-2  ENSG00000284332      +\n",
       "...       ...    ...    ...          ...              ...    ...\n",
       "2774280  chrM  10000  15000       MT-ND6  ENSG00000198695      -\n",
       "2774285  chrM  10000  15000        MT-TE  ENSG00000210194      -\n",
       "2774288  chrM  10000  15000       MT-CYB  ENSG00000198727      +\n",
       "2774293  chrM  15000  20000        MT-TT  ENSG00000210195      +\n",
       "2774296  chrM  15000  20000        MT-TP  ENSG00000210196      -\n",
       "\n",
       "[58825 rows x 6 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_pbt_pre"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8538168a",
   "metadata": {},
   "source": [
    "# Find all genes +/- 500kb of a colocalized/LD SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e26af376",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Find all genes +/- 500kb of a colocalized/LD SNP\n"
     ]
    }
   ],
   "source": [
    "print('# Find all genes +/- 500kb of a colocalized/LD SNP')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "303e16ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Find all genes +/- 1000kb\n"
     ]
    }
   ],
   "source": [
    "slop = 1000000\n",
    "print(\"# Find all genes +/- {}kb\".format(int(slop/1000)))\n",
    "\n",
    "# get a list of gene names within +- 500kb of the SNPs\n",
    "fivekb_genes = coloc_sig_pbt.slop(b=slop, g=gs_fn)\n",
    "fivekb_genes = fivekb_genes.intersect(genes_pbt, wa=True, wb=True)\n",
    "fivekb_genes = fivekb_genes.to_dataframe()\n",
    "\n",
    "fivekb_genes = fivekb_genes.iloc[:, [0,1,2,5,6,7,3,4,8,9,10]]\n",
    "fivekb_genes.columns = bedpe_6cols + ['sid', 'rsid', 'genename', 'geneid', 'strand2']\n",
    "fivekb_genes['strand1'] = '+'\n",
    "fivekb_genes['name'] = fivekb_genes['rsid'] + '_' + fivekb_genes['genename']\n",
    "fivekb_genes['score'] = '.'\n",
    "\n",
    "# extract only major columns\n",
    "major_cols = bedpe_10cols + ['sid', 'genename', 'geneid']\n",
    "fivekb_genes = fivekb_genes[major_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6411cb3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 738 colocalized snp-gene pairs within +/- 1000kb.\n"
     ]
    }
   ],
   "source": [
    "print('There are {} colocalized snp-gene pairs within +/- 1000kb.'.format(fivekb_genes.shape[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50bfa254",
   "metadata": {},
   "source": [
    "# Find the closest gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e14c1b2e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Find the closest gene\n"
     ]
    }
   ],
   "source": [
    "print('# Find the closest gene')\n",
    "closest_gene = coloc_sig_pbt.closest(genes_pbt, d=True)\n",
    "closest_gene = closest_gene.to_dataframe().iloc[:, [0,1,2,5,6,7,3,4,8,9,11]]\n",
    "closest_gene.columns = bedpe_6cols + ['sid', 'rsid', 'genename', 'geneid', 'dist']\n",
    "closest_gene.set_index(['sid', 'geneid'], inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a314a78d",
   "metadata": {},
   "source": [
    "# Load the H3K27ac HiChIP loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f3e91cb6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Load the H3K27ac HiChIP loops\n"
     ]
    }
   ],
   "source": [
    "print('# Load the H3K27ac HiChIP loops')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "236279f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Get the loops\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/BioHome/jreyna/tmp//ipykernel_122267/1175980432.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  tmp_loops.rename(columns={'p': 'score'}, inplace=True)\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_122267/1175980432.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  tmp_loops.loc[:, 'name'] = '.'\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_122267/1175980432.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  tmp_loops.loc[:, 'score'] = loops['p']\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_122267/1175980432.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  tmp_loops.loc[:, 'strand1'] = '.'\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_122267/1175980432.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  tmp_loops.loc[:, 'strand2'] = '.'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    chr1        s1        e1   chr2        s2        e2 name         score  \\\n",
      "0   chr1    999620   1004620   chr1   1019620   1024620    .  4.851310e-06   \n",
      "1   chr1   1778561   1783561   chr1   3793436   3798436    .  4.646481e-08   \n",
      "2   chr1  26278509  26283509   chr1  26318509  26323509    .  3.598136e-06   \n",
      "3  chr10  44934552  44939552  chr10  44954552  44959552    .  5.488571e-06   \n",
      "4  chr10  44944552  44949552  chr10  44959552  44964552    .  5.173212e-06   \n",
      "\n",
      "  strand1 strand2  \n",
      "0       .       .  \n",
      "1       .       .  \n",
      "2       .       .  \n",
      "3       .       .  \n",
      "4       .       .  \n"
     ]
    }
   ],
   "source": [
    "print('# Get the loops')\n",
    "\n",
    "# load the loop data\n",
    "loops = pd.read_table(loop_fn)\n",
    "tmp_loops = loops[['chr1', 's1', 'e1', 'chr2', 's2', 'e2']]\n",
    "tmp_loops.rename(columns={'p': 'score'}, inplace=True)\n",
    "tmp_loops.loc[:, 'name'] = '.'\n",
    "tmp_loops.loc[:, 'score'] = loops['p']\n",
    "tmp_loops.loc[:, 'strand1'] = '.'\n",
    "tmp_loops.loc[:, 'strand2'] = '.'\n",
    "loops = pbt.BedTool.from_dataframe(tmp_loops)\n",
    "print(tmp_loops.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75c4f3d0",
   "metadata": {},
   "source": [
    "# Find SG pairs overlapping loops (both anchors + single anchor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "6243333c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#print('# Find SNP-Genes overlapping loops (both anchors + single anchor)')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "152f32d5",
   "metadata": {},
   "source": [
    "### Find SNP-Gene overlapping a loop (both anchors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "f914b9cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Find SNP-Gene overlapping a loop (both anchors)\n"
     ]
    }
   ],
   "source": [
    "print('### Find SNP-Gene overlapping a loop (both anchors)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c3055863",
   "metadata": {},
   "outputs": [],
   "source": [
    "# re-arranging to fit bedpe format\n",
    "fivekb_gloops = fivekb_genes.copy()\n",
    "\n",
    "# loading into pbt\n",
    "fivekb_gloops = pbt.BedTool.from_dataframe(fivekb_gloops)\n",
    "fivekb_gloops = fivekb_gloops.pair_to_pair(loops, type='both', slop=loop_slop, **{'is':True})\n",
    "fivekb_gloops = fivekb_gloops.to_dataframe(disable_auto_names=True, header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "50e6773c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3280000</td>\n",
       "      <td>5285000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3525000</td>\n",
       "      <td>3530000</td>\n",
       "      <td>rs10974435_RFX3</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>9:4283682</td>\n",
       "      <td>RFX3</td>\n",
       "      <td>ENSG00000080298</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3505000</td>\n",
       "      <td>3510000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3575000</td>\n",
       "      <td>3580000</td>\n",
       "      <td>.</td>\n",
       "      <td>3.041319e-06</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3280000</td>\n",
       "      <td>5285000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3525000</td>\n",
       "      <td>3530000</td>\n",
       "      <td>rs10974435_RFX3</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>9:4283682</td>\n",
       "      <td>RFX3</td>\n",
       "      <td>ENSG00000080298</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3525000</td>\n",
       "      <td>3530000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4290000</td>\n",
       "      <td>4295000</td>\n",
       "      <td>.</td>\n",
       "      <td>1.384373e-07</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3280000</td>\n",
       "      <td>5285000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3525000</td>\n",
       "      <td>3530000</td>\n",
       "      <td>rs10974435_RFX3</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>9:4283682</td>\n",
       "      <td>RFX3</td>\n",
       "      <td>ENSG00000080298</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3525000</td>\n",
       "      <td>3530000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4675000</td>\n",
       "      <td>4680000</td>\n",
       "      <td>.</td>\n",
       "      <td>8.708686e-08</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3280000</td>\n",
       "      <td>5285000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3525000</td>\n",
       "      <td>3530000</td>\n",
       "      <td>rs10974435_RFX3</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>9:4283682</td>\n",
       "      <td>RFX3</td>\n",
       "      <td>ENSG00000080298</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3525000</td>\n",
       "      <td>3530000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4680000</td>\n",
       "      <td>4685000</td>\n",
       "      <td>.</td>\n",
       "      <td>8.661853e-08</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3280000</td>\n",
       "      <td>5285000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3525000</td>\n",
       "      <td>3530000</td>\n",
       "      <td>rs10974435_RFX3</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>9:4283682</td>\n",
       "      <td>RFX3</td>\n",
       "      <td>ENSG00000080298</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3525000</td>\n",
       "      <td>3530000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4980000</td>\n",
       "      <td>4985000</td>\n",
       "      <td>.</td>\n",
       "      <td>6.738191e-08</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1363</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3295000</td>\n",
       "      <td>5300000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4980000</td>\n",
       "      <td>4985000</td>\n",
       "      <td>rs10814917_JAK2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4296430</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>ENSG00000096968</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4775000</td>\n",
       "      <td>4780000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4985000</td>\n",
       "      <td>4990000</td>\n",
       "      <td>.</td>\n",
       "      <td>6.905469e-07</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1364</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3295000</td>\n",
       "      <td>5300000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4980000</td>\n",
       "      <td>4985000</td>\n",
       "      <td>rs10814917_JAK2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4296430</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>ENSG00000096968</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4790000</td>\n",
       "      <td>4795000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4980000</td>\n",
       "      <td>4985000</td>\n",
       "      <td>.</td>\n",
       "      <td>7.275040e-07</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1365</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3295000</td>\n",
       "      <td>5300000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4980000</td>\n",
       "      <td>4985000</td>\n",
       "      <td>rs10814917_JAK2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4296430</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>ENSG00000096968</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4795000</td>\n",
       "      <td>4800000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4985000</td>\n",
       "      <td>4990000</td>\n",
       "      <td>.</td>\n",
       "      <td>7.086575e-07</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1366</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3295000</td>\n",
       "      <td>5300000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4980000</td>\n",
       "      <td>4985000</td>\n",
       "      <td>rs10814917_JAK2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4296430</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>ENSG00000096968</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4795000</td>\n",
       "      <td>4800000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4980000</td>\n",
       "      <td>4985000</td>\n",
       "      <td>.</td>\n",
       "      <td>3.311919e-06</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1367</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3295000</td>\n",
       "      <td>5300000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4980000</td>\n",
       "      <td>4985000</td>\n",
       "      <td>rs10814917_JAK2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4296430</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>ENSG00000096968</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4935000</td>\n",
       "      <td>4940000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4985000</td>\n",
       "      <td>4990000</td>\n",
       "      <td>.</td>\n",
       "      <td>3.041319e-06</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1368 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        0        1        2     3        4        5                6  7  8   \\\n",
       "0     chr9  3280000  5285000  chr9  3525000  3530000  rs10974435_RFX3  .  +   \n",
       "1     chr9  3280000  5285000  chr9  3525000  3530000  rs10974435_RFX3  .  +   \n",
       "2     chr9  3280000  5285000  chr9  3525000  3530000  rs10974435_RFX3  .  +   \n",
       "3     chr9  3280000  5285000  chr9  3525000  3530000  rs10974435_RFX3  .  +   \n",
       "4     chr9  3280000  5285000  chr9  3525000  3530000  rs10974435_RFX3  .  +   \n",
       "...    ...      ...      ...   ...      ...      ...              ... .. ..   \n",
       "1363  chr9  3295000  5300000  chr9  4980000  4985000  rs10814917_JAK2  .  +   \n",
       "1364  chr9  3295000  5300000  chr9  4980000  4985000  rs10814917_JAK2  .  +   \n",
       "1365  chr9  3295000  5300000  chr9  4980000  4985000  rs10814917_JAK2  .  +   \n",
       "1366  chr9  3295000  5300000  chr9  4980000  4985000  rs10814917_JAK2  .  +   \n",
       "1367  chr9  3295000  5300000  chr9  4980000  4985000  rs10814917_JAK2  .  +   \n",
       "\n",
       "     9          10    11               12    13       14       15    16  \\\n",
       "0     -  9:4283682  RFX3  ENSG00000080298  chr9  3505000  3510000  chr9   \n",
       "1     -  9:4283682  RFX3  ENSG00000080298  chr9  3525000  3530000  chr9   \n",
       "2     -  9:4283682  RFX3  ENSG00000080298  chr9  3525000  3530000  chr9   \n",
       "3     -  9:4283682  RFX3  ENSG00000080298  chr9  3525000  3530000  chr9   \n",
       "4     -  9:4283682  RFX3  ENSG00000080298  chr9  3525000  3530000  chr9   \n",
       "...  ..        ...   ...              ...   ...      ...      ...   ...   \n",
       "1363  +  9:4296430  JAK2  ENSG00000096968  chr9  4775000  4780000  chr9   \n",
       "1364  +  9:4296430  JAK2  ENSG00000096968  chr9  4790000  4795000  chr9   \n",
       "1365  +  9:4296430  JAK2  ENSG00000096968  chr9  4795000  4800000  chr9   \n",
       "1366  +  9:4296430  JAK2  ENSG00000096968  chr9  4795000  4800000  chr9   \n",
       "1367  +  9:4296430  JAK2  ENSG00000096968  chr9  4935000  4940000  chr9   \n",
       "\n",
       "           17       18 19            20 21 22  \n",
       "0     3575000  3580000  .  3.041319e-06  .  .  \n",
       "1     4290000  4295000  .  1.384373e-07  .  .  \n",
       "2     4675000  4680000  .  8.708686e-08  .  .  \n",
       "3     4680000  4685000  .  8.661853e-08  .  .  \n",
       "4     4980000  4985000  .  6.738191e-08  .  .  \n",
       "...       ...      ... ..           ... .. ..  \n",
       "1363  4985000  4990000  .  6.905469e-07  .  .  \n",
       "1364  4980000  4985000  .  7.275040e-07  .  .  \n",
       "1365  4985000  4990000  .  7.086575e-07  .  .  \n",
       "1366  4980000  4985000  .  3.311919e-06  .  .  \n",
       "1367  4985000  4990000  .  3.041319e-06  .  .  \n",
       "\n",
       "[1368 rows x 23 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fivekb_gloops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "dfc9ff74",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 252 SNP-Gene pairs with a loop.\n"
     ]
    }
   ],
   "source": [
    "if len(fivekb_gloops) > 0:\n",
    "    fivekb_gloops_set = fivekb_gloops.iloc[:, [10, 12]]\n",
    "    fivekb_gloops_uniq = set([tuple(x) for x in fivekb_gloops_set.values.tolist()])\n",
    "else:\n",
    "    print('WARNING: Found no overlap between the genes and loops.')\n",
    "    fivekb_gloops_uniq = set()\n",
    "\n",
    "print('There are {} SNP-Gene pairs with a loop.'.format(len(fivekb_gloops_uniq)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a07aaed",
   "metadata": {},
   "source": [
    "### Find coloc-SNP overlapping an anchor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "1531ffbb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Find coloc-SNP overlapping an anchor\n"
     ]
    }
   ],
   "source": [
    "print('### Find coloc-SNP overlapping an anchor')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "5e018a9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "colocSNP_anchors = loops.pairtobed(coloc_sig_pbt.slop(b=loop_slop, g=gs_fn), type='either')\n",
    "colocSNP_anchors = colocSNP_anchors.to_dataframe(disable_auto_names=True, header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "3beb710f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 18 SNPs which overlap a loop anchor.\n"
     ]
    }
   ],
   "source": [
    "if len(colocSNP_anchors) > 0:\n",
    "    colocSNP_anchors_set = colocSNP_anchors.iloc[:, 13]\n",
    "else:\n",
    "    print('WARNING: Found no overlap between coloc SNPs and loop anchors.')\n",
    "    colocSNP_anchors_set = set()\n",
    "\n",
    "print('There are {} SNPs which overlap a loop anchor.'.format(len(colocSNP_anchors_set)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fff0d6de",
   "metadata": {},
   "source": [
    "# Construct the master table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "c8af7cb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Construct master table\n",
      "Master is starting with 738 snp-gene pairs.\n"
     ]
    }
   ],
   "source": [
    "print('# Construct master table')\n",
    "\n",
    "# begin making the master\n",
    "master = fivekb_genes.copy()\n",
    "\n",
    "print('Master is starting with {} snp-gene pairs.'.format(master.shape[0]))\n",
    "\n",
    "# adding back the rsid\n",
    "coord_to_rsid = coloc[['sid', 'rsid']].set_index(['sid']).squeeze().to_dict()\n",
    "master.loc[:, 'rsid'] = master.sid.apply(lambda x: coord_to_rsid[x])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5e3ddef",
   "metadata": {},
   "source": [
    "#### Add eqtl results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "2ea3172d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# #### Add eqtl results\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/BioHome/jreyna/tmp//ipykernel_122267/3796354937.py:27: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  eqtls['sid'] = eqtls['chr'].astype(str) + ':' + eqtls['pos'].astype(str).str.replace('.0', '')\n"
     ]
    }
   ],
   "source": [
    "print(\"# #### Add eqtl results\")\n",
    "\n",
    "# get eQTL's\n",
    "eqtls = pd.read_table(eqtl_fn, header=0)#, nrows=1000)\n",
    "eqtls = eqtls.loc[eqtls.type == 'SNP']\n",
    "eqtls.drop(['molecular_trait_object_id', 'molecular_trait_id', 'type'], axis=1, inplace=True)\n",
    "eqtls_rename_cols = {'chromosome': 'chr',\n",
    "    'position': 'pos',\n",
    "    'ref': 'ref',\n",
    "    'alt': 'alt',\n",
    "    'variant': 'variant',\n",
    "    'ma_samples': 'ma_samples',\n",
    "    'maf': 'maf',\n",
    "    'pvalue': 'eqtl_pval',\n",
    "    'beta': 'eqtl_beta',\n",
    "    'se': 'eqtl_slop_se',\n",
    "    'ac': 'ac',\n",
    "    'an': 'an',\n",
    "    'r2': 'r2',\n",
    "    'gene_id': 'geneid',\n",
    "    'median_tpm': 'median_tpm',\n",
    "    'rsid': 'rsid',\n",
    "    'dist': 'dist',\n",
    "    'fdr': 'eqtl_fdr'}\n",
    "eqtls.rename(columns=eqtls_rename_cols, inplace=True)\n",
    "#eqtls['pos'] = eqtls['pos'].astype(int)\n",
    "eqtls['sid'] = eqtls['chr'].astype(str) + ':' + eqtls['pos'].astype(str).str.replace('.0', '')\n",
    "eqtls['is_eqtl_pair'] = -1 # add column to filter on eqtl snp status\n",
    "eqtls.loc[(eqtls['eqtl_fdr'] <= 0.05), 'is_eqtl_pair'] = 1\n",
    "eqtls.loc[(eqtls['eqtl_fdr'] > 0.05), 'is_eqtl_pair'] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e1b1821e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 866453 eQTLs.\n"
     ]
    }
   ],
   "source": [
    "print('There are {} eQTLs.'.format(eqtls.shape[0]))\n",
    "\n",
    "# need to use outer or else you exclude some eQTL's\n",
    "master = master.merge(eqtls, on=['sid', 'geneid'], how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "c5a68535",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After outer merging with eqtls, master has 738 snp-gene pairs.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/BioHome/jreyna/tmp//ipykernel_122267/3088769827.py:10: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  master.loc[master.chrA.isna(), 'chrA'] = 'chr' + master.loc[master.chrA.isna(), 'sid'].str.replace(':[0-9]+', '')\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_122267/3088769827.py:11: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  master.loc[master.chrB.isna(), 'chrB'] = 'chr' + master.loc[master.chrB.isna(), 'sid'].str.replace(':[0-9]+', '')\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_122267/3088769827.py:12: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  master.loc[master.startA.isna(), 'startA'] = (master.loc[master.startA.isna(), 'sid'].str.replace('[0-9]+:', '')).astype(int)\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_122267/3088769827.py:14: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  master.loc[master.endA.isna(), 'endA'] = master.loc[master.endA.isna(), 'sid'].str.replace('[0-9]+:', '')\n"
     ]
    }
   ],
   "source": [
    "# update eqtl snp status\n",
    "master.loc[master.is_eqtl_pair.isna(), 'is_eqtl_pair' ] = 0 \n",
    "\n",
    "master.loc[:, 'is_eqtl_pair'] = master.loc[:, 'is_eqtl_pair'].astype(int)\n",
    "\n",
    "# add gene names to entries with a missing name (after adding eQTL info)\n",
    "master.loc[master.genename.isna(), 'genename'] = master.loc[master.genename.isna(), 'geneid']\n",
    "\n",
    "# add missing chrA, chrB, startA and endA data for the eQTL rows\n",
    "master.loc[master.chrA.isna(), 'chrA'] = 'chr' + master.loc[master.chrA.isna(), 'sid'].str.replace(':[0-9]+', '')\n",
    "master.loc[master.chrB.isna(), 'chrB'] = 'chr' + master.loc[master.chrB.isna(), 'sid'].str.replace(':[0-9]+', '')\n",
    "master.loc[master.startA.isna(), 'startA'] = (master.loc[master.startA.isna(), 'sid'].str.replace('[0-9]+:', '')).astype(int)\n",
    "master.loc[master.startA.isna(), 'startA'] -= 1     \n",
    "master.loc[master.endA.isna(), 'endA'] = master.loc[master.endA.isna(), 'sid'].str.replace('[0-9]+:', '')\n",
    "\n",
    "print('After outer merging with eqtls, master has {} snp-gene pairs.'.format(master.shape[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f495c29f",
   "metadata": {},
   "source": [
    "### Add the original gene coordinates "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "7bf5b26e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Add the original gene coordinates \n"
     ]
    }
   ],
   "source": [
    "print('### Add the original gene coordinates ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "07dd6889",
   "metadata": {},
   "outputs": [],
   "source": [
    "#master.loc[~master.geneid.isna(), 'geneid'] = master.loc[~master.geneid.isna(), 'geneid']\n",
    "\n",
    "# add back the original gene start and end\n",
    "master = master.merge(orig_genes_df[['start', 'end', 'geneid', 'strand']],\n",
    "                      on='geneid', how='left')\n",
    "\n",
    "# convert the startB/endB position into startB/endB for the TSS\n",
    "# if the gene is + then the startB is uses as the tss otherwise\n",
    "# the endB is used as the tss\n",
    "master.loc[(master.strand == '+'), 'endB'] = master.loc[(master.strand == '+'), 'start']\n",
    "master.loc[(master.strand == '+'), 'startB'] = master.loc[(master.strand == '+'), 'start'] - 1\n",
    "\n",
    "master.loc[(master.strand == '-'), 'endB'] = master.loc[(master.strand == '-'), 'end']\n",
    "master.loc[(master.strand == '-'), 'startB'] = master.loc[(master.strand == '-'), 'end'] - 1\n",
    "\n",
    "# convert the coordinates from floats to ints\n",
    "master.startA = master.startA.astype(int)\n",
    "master.startB = master.startB.astype(int)\n",
    "master.endA = master.endA.astype(int)\n",
    "master.endB = master.endB.astype(int)\n",
    "\n",
    "master.rename(columns={'start': 'gene_start', 'end': 'gene_end', 'strand': 'gene_strand'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "5ed942b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrA</th>\n",
       "      <th>startA</th>\n",
       "      <th>endA</th>\n",
       "      <th>chrB</th>\n",
       "      <th>startB</th>\n",
       "      <th>endB</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand1</th>\n",
       "      <th>strand2</th>\n",
       "      <th>sid</th>\n",
       "      <th>genename</th>\n",
       "      <th>geneid</th>\n",
       "      <th>rsid_x</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variant</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>maf</th>\n",
       "      <th>eqtl_pval</th>\n",
       "      <th>eqtl_beta</th>\n",
       "      <th>eqtl_slop_se</th>\n",
       "      <th>ac</th>\n",
       "      <th>an</th>\n",
       "      <th>r2</th>\n",
       "      <th>median_tpm</th>\n",
       "      <th>rsid_y</th>\n",
       "      <th>dist</th>\n",
       "      <th>eqtl_fdr</th>\n",
       "      <th>is_eqtl_pair</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3280000</td>\n",
       "      <td>5285000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3364791</td>\n",
       "      <td>3364792</td>\n",
       "      <td>rs10974435_AL354941.1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>9:4283682</td>\n",
       "      <td>AL354941.1</td>\n",
       "      <td>ENSG00000215297</td>\n",
       "      <td>rs10974435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3364329</td>\n",
       "      <td>3364792</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3280000</td>\n",
       "      <td>5285000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3526003</td>\n",
       "      <td>3526004</td>\n",
       "      <td>rs10974435_RFX3</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>9:4283682</td>\n",
       "      <td>RFX3</td>\n",
       "      <td>ENSG00000080298</td>\n",
       "      <td>rs10974435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3218297</td>\n",
       "      <td>3526004</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3280000</td>\n",
       "      <td>5285000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3526722</td>\n",
       "      <td>3526723</td>\n",
       "      <td>rs10974435_RFX3-AS1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4283682</td>\n",
       "      <td>RFX3-AS1</td>\n",
       "      <td>ENSG00000232104</td>\n",
       "      <td>rs10974435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3526723</td>\n",
       "      <td>3691814</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3280000</td>\n",
       "      <td>5285000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3647334</td>\n",
       "      <td>3647335</td>\n",
       "      <td>rs10974435_AL354977.2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4283682</td>\n",
       "      <td>AL354977.2</td>\n",
       "      <td>ENSG00000237359</td>\n",
       "      <td>rs10974435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3647335</td>\n",
       "      <td>3648338</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3280000</td>\n",
       "      <td>5285000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3875583</td>\n",
       "      <td>3875584</td>\n",
       "      <td>rs10974435_AL137071.1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4283682</td>\n",
       "      <td>AL137071.1</td>\n",
       "      <td>ENSG00000236724</td>\n",
       "      <td>rs10974435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3875584</td>\n",
       "      <td>3878809</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>733</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3295000</td>\n",
       "      <td>5300000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>5096665</td>\n",
       "      <td>5096666</td>\n",
       "      <td>rs10814917_MTCO1P11</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4296430</td>\n",
       "      <td>MTCO1P11</td>\n",
       "      <td>ENSG00000224083</td>\n",
       "      <td>rs10814917</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>5096666</td>\n",
       "      <td>5098193</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>734</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3295000</td>\n",
       "      <td>5300000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>5193557</td>\n",
       "      <td>5193558</td>\n",
       "      <td>rs10814917_AL133547.1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>9:4296430</td>\n",
       "      <td>AL133547.1</td>\n",
       "      <td>ENSG00000237553</td>\n",
       "      <td>rs10814917</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>5193339</td>\n",
       "      <td>5193558</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>735</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3295000</td>\n",
       "      <td>5300000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4984389</td>\n",
       "      <td>4984390</td>\n",
       "      <td>rs10814917_JAK2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4296430</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>ENSG00000096968</td>\n",
       "      <td>rs10814917</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>4984390</td>\n",
       "      <td>5128183</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3295000</td>\n",
       "      <td>5300000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>5110912</td>\n",
       "      <td>5110913</td>\n",
       "      <td>rs10814917_TCF3P1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4296430</td>\n",
       "      <td>TCF3P1</td>\n",
       "      <td>ENSG00000236567</td>\n",
       "      <td>rs10814917</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>5110913</td>\n",
       "      <td>5112849</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>737</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3295000</td>\n",
       "      <td>5300000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>5113548</td>\n",
       "      <td>5113549</td>\n",
       "      <td>rs10814917_IGHEP2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4296430</td>\n",
       "      <td>IGHEP2</td>\n",
       "      <td>ENSG00000254017</td>\n",
       "      <td>rs10814917</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>5113549</td>\n",
       "      <td>5114804</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>738 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrA   startA     endA  chrB   startB     endB                   name  \\\n",
       "0    chr9  3280000  5285000  chr9  3364791  3364792  rs10974435_AL354941.1   \n",
       "1    chr9  3280000  5285000  chr9  3526003  3526004        rs10974435_RFX3   \n",
       "2    chr9  3280000  5285000  chr9  3526722  3526723    rs10974435_RFX3-AS1   \n",
       "3    chr9  3280000  5285000  chr9  3647334  3647335  rs10974435_AL354977.2   \n",
       "4    chr9  3280000  5285000  chr9  3875583  3875584  rs10974435_AL137071.1   \n",
       "..    ...      ...      ...   ...      ...      ...                    ...   \n",
       "733  chr9  3295000  5300000  chr9  5096665  5096666    rs10814917_MTCO1P11   \n",
       "734  chr9  3295000  5300000  chr9  5193557  5193558  rs10814917_AL133547.1   \n",
       "735  chr9  3295000  5300000  chr9  4984389  4984390        rs10814917_JAK2   \n",
       "736  chr9  3295000  5300000  chr9  5110912  5110913      rs10814917_TCF3P1   \n",
       "737  chr9  3295000  5300000  chr9  5113548  5113549      rs10814917_IGHEP2   \n",
       "\n",
       "    score strand1 strand2        sid    genename           geneid      rsid_x  \\\n",
       "0       .       +       -  9:4283682  AL354941.1  ENSG00000215297  rs10974435   \n",
       "1       .       +       -  9:4283682        RFX3  ENSG00000080298  rs10974435   \n",
       "2       .       +       +  9:4283682    RFX3-AS1  ENSG00000232104  rs10974435   \n",
       "3       .       +       +  9:4283682  AL354977.2  ENSG00000237359  rs10974435   \n",
       "4       .       +       +  9:4283682  AL137071.1  ENSG00000236724  rs10974435   \n",
       "..    ...     ...     ...        ...         ...              ...         ...   \n",
       "733     .       +       +  9:4296430    MTCO1P11  ENSG00000224083  rs10814917   \n",
       "734     .       +       -  9:4296430  AL133547.1  ENSG00000237553  rs10814917   \n",
       "735     .       +       +  9:4296430        JAK2  ENSG00000096968  rs10814917   \n",
       "736     .       +       +  9:4296430      TCF3P1  ENSG00000236567  rs10814917   \n",
       "737     .       +       +  9:4296430      IGHEP2  ENSG00000254017  rs10814917   \n",
       "\n",
       "     chr  pos  ref  alt variant  ma_samples  maf  eqtl_pval  eqtl_beta  \\\n",
       "0    NaN  NaN  NaN  NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "1    NaN  NaN  NaN  NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "2    NaN  NaN  NaN  NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "3    NaN  NaN  NaN  NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "4    NaN  NaN  NaN  NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "..   ...  ...  ...  ...     ...         ...  ...        ...        ...   \n",
       "733  NaN  NaN  NaN  NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "734  NaN  NaN  NaN  NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "735  NaN  NaN  NaN  NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "736  NaN  NaN  NaN  NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "737  NaN  NaN  NaN  NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "\n",
       "     eqtl_slop_se  ac  an  r2  median_tpm rsid_y  dist  eqtl_fdr  \\\n",
       "0             NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "1             NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "2             NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "3             NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "4             NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "..            ...  ..  ..  ..         ...    ...   ...       ...   \n",
       "733           NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "734           NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "735           NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "736           NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "737           NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "\n",
       "     is_eqtl_pair  gene_start  gene_end gene_strand  \n",
       "0               0     3364329   3364792           -  \n",
       "1               0     3218297   3526004           -  \n",
       "2               0     3526723   3691814           +  \n",
       "3               0     3647335   3648338           +  \n",
       "4               0     3875584   3878809           +  \n",
       "..            ...         ...       ...         ...  \n",
       "733             0     5096666   5098193           +  \n",
       "734             0     5193339   5193558           -  \n",
       "735             0     4984390   5128183           +  \n",
       "736             0     5110913   5112849           +  \n",
       "737             0     5113549   5114804           +  \n",
       "\n",
       "[738 rows x 35 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "master"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28e8ca3f",
   "metadata": {},
   "source": [
    "### Add info about closests gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "506b08d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Add info about closests gene\n"
     ]
    }
   ],
   "source": [
    "print('### Add info about closests gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "f2802775",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# #### Add info about closests gene\n",
      "Found 18 closest genes.\n"
     ]
    }
   ],
   "source": [
    "print(\"# #### Add info about closests gene\")\n",
    "\n",
    "# check for the closets gene\n",
    "closets_check = [0] * master.shape[0]\n",
    "for i, sr in master.iterrows():\n",
    "    rs_gene = (sr.sid, sr.geneid)\n",
    "    if rs_gene in closest_gene.index:\n",
    "        closets_check[i] = 1\n",
    "master['is_closest_gene'] = closets_check\n",
    "print('Found {} closest genes.'.format(sum(master['is_closest_gene'])))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77139b44",
   "metadata": {},
   "source": [
    "### Add colocalization data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "a172900e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Add colocalization data\n"
     ]
    }
   ],
   "source": [
    "print('### Add colocalization data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "932b137c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/BioHome/jreyna/tmp//ipykernel_122267/4159951681.py:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  tmp_coloc['sid'] = tmp_coloc['chr'].str.replace('chr', '') + ':' +  tmp_coloc['pos'].astype(str) # NEW\n"
     ]
    }
   ],
   "source": [
    "# add colocalization data for SNP and is_coloc_snp columns\n",
    "tmp_coloc = coloc_sig_full[['chr', 'pos',\n",
    " 'ppH0',\n",
    " 'ppH1',\n",
    " 'ppH2',\n",
    " 'ppH3',\n",
    " 'ppH4',\n",
    " 'rsid',\n",
    " 'geneid',                       \n",
    " 'ref',\n",
    " 'alt',\n",
    " 'AC',\n",
    " 'AF',\n",
    " 'AN',\n",
    " 'gwas_slope',\n",
    " 'gwas_slope_se',\n",
    " 'gwas_pval']]\n",
    "tmp_coloc['sid'] = tmp_coloc['chr'].str.replace('chr', '') + ':' +  tmp_coloc['pos'].astype(str) # NEW\n",
    "master = master.merge(tmp_coloc, on=['sid', 'geneid'], how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "02a67151",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrA</th>\n",
       "      <th>startA</th>\n",
       "      <th>endA</th>\n",
       "      <th>chrB</th>\n",
       "      <th>startB</th>\n",
       "      <th>endB</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand1</th>\n",
       "      <th>strand2</th>\n",
       "      <th>sid</th>\n",
       "      <th>genename</th>\n",
       "      <th>geneid</th>\n",
       "      <th>rsid_x</th>\n",
       "      <th>chr_x</th>\n",
       "      <th>pos_x</th>\n",
       "      <th>ref_x</th>\n",
       "      <th>alt_x</th>\n",
       "      <th>variant</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>maf</th>\n",
       "      <th>eqtl_pval</th>\n",
       "      <th>eqtl_beta</th>\n",
       "      <th>eqtl_slop_se</th>\n",
       "      <th>ac</th>\n",
       "      <th>an</th>\n",
       "      <th>r2</th>\n",
       "      <th>median_tpm</th>\n",
       "      <th>rsid_y</th>\n",
       "      <th>dist</th>\n",
       "      <th>eqtl_fdr</th>\n",
       "      <th>is_eqtl_pair</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>is_closest_gene</th>\n",
       "      <th>chr_y</th>\n",
       "      <th>pos_y</th>\n",
       "      <th>ppH0</th>\n",
       "      <th>ppH1</th>\n",
       "      <th>ppH2</th>\n",
       "      <th>ppH3</th>\n",
       "      <th>ppH4</th>\n",
       "      <th>rsid</th>\n",
       "      <th>ref_y</th>\n",
       "      <th>alt_y</th>\n",
       "      <th>AC</th>\n",
       "      <th>AF</th>\n",
       "      <th>AN</th>\n",
       "      <th>gwas_slope</th>\n",
       "      <th>gwas_slope_se</th>\n",
       "      <th>gwas_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3280000</td>\n",
       "      <td>5285000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3364791</td>\n",
       "      <td>3364792</td>\n",
       "      <td>rs10974435_AL354941.1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>9:4283682</td>\n",
       "      <td>AL354941.1</td>\n",
       "      <td>ENSG00000215297</td>\n",
       "      <td>rs10974435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3364329</td>\n",
       "      <td>3364792</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3280000</td>\n",
       "      <td>5285000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3526003</td>\n",
       "      <td>3526004</td>\n",
       "      <td>rs10974435_RFX3</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>9:4283682</td>\n",
       "      <td>RFX3</td>\n",
       "      <td>ENSG00000080298</td>\n",
       "      <td>rs10974435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3218297</td>\n",
       "      <td>3526004</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3280000</td>\n",
       "      <td>5285000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3526722</td>\n",
       "      <td>3526723</td>\n",
       "      <td>rs10974435_RFX3-AS1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4283682</td>\n",
       "      <td>RFX3-AS1</td>\n",
       "      <td>ENSG00000232104</td>\n",
       "      <td>rs10974435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3526723</td>\n",
       "      <td>3691814</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3280000</td>\n",
       "      <td>5285000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3647334</td>\n",
       "      <td>3647335</td>\n",
       "      <td>rs10974435_AL354977.2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4283682</td>\n",
       "      <td>AL354977.2</td>\n",
       "      <td>ENSG00000237359</td>\n",
       "      <td>rs10974435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3647335</td>\n",
       "      <td>3648338</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3280000</td>\n",
       "      <td>5285000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3875583</td>\n",
       "      <td>3875584</td>\n",
       "      <td>rs10974435_AL137071.1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4283682</td>\n",
       "      <td>AL137071.1</td>\n",
       "      <td>ENSG00000236724</td>\n",
       "      <td>rs10974435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3875584</td>\n",
       "      <td>3878809</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>733</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3295000</td>\n",
       "      <td>5300000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>5096665</td>\n",
       "      <td>5096666</td>\n",
       "      <td>rs10814917_MTCO1P11</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4296430</td>\n",
       "      <td>MTCO1P11</td>\n",
       "      <td>ENSG00000224083</td>\n",
       "      <td>rs10814917</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>5096666</td>\n",
       "      <td>5098193</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>734</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3295000</td>\n",
       "      <td>5300000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>5193557</td>\n",
       "      <td>5193558</td>\n",
       "      <td>rs10814917_AL133547.1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>9:4296430</td>\n",
       "      <td>AL133547.1</td>\n",
       "      <td>ENSG00000237553</td>\n",
       "      <td>rs10814917</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>5193339</td>\n",
       "      <td>5193558</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>735</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3295000</td>\n",
       "      <td>5300000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>4984389</td>\n",
       "      <td>4984390</td>\n",
       "      <td>rs10814917_JAK2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4296430</td>\n",
       "      <td>JAK2</td>\n",
       "      <td>ENSG00000096968</td>\n",
       "      <td>rs10814917</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>4984390</td>\n",
       "      <td>5128183</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3295000</td>\n",
       "      <td>5300000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>5110912</td>\n",
       "      <td>5110913</td>\n",
       "      <td>rs10814917_TCF3P1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4296430</td>\n",
       "      <td>TCF3P1</td>\n",
       "      <td>ENSG00000236567</td>\n",
       "      <td>rs10814917</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>5110913</td>\n",
       "      <td>5112849</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>737</th>\n",
       "      <td>chr9</td>\n",
       "      <td>3295000</td>\n",
       "      <td>5300000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>5113548</td>\n",
       "      <td>5113549</td>\n",
       "      <td>rs10814917_IGHEP2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>9:4296430</td>\n",
       "      <td>IGHEP2</td>\n",
       "      <td>ENSG00000254017</td>\n",
       "      <td>rs10814917</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>5113549</td>\n",
       "      <td>5114804</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>738 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrA   startA     endA  chrB   startB     endB                   name  \\\n",
       "0    chr9  3280000  5285000  chr9  3364791  3364792  rs10974435_AL354941.1   \n",
       "1    chr9  3280000  5285000  chr9  3526003  3526004        rs10974435_RFX3   \n",
       "2    chr9  3280000  5285000  chr9  3526722  3526723    rs10974435_RFX3-AS1   \n",
       "3    chr9  3280000  5285000  chr9  3647334  3647335  rs10974435_AL354977.2   \n",
       "4    chr9  3280000  5285000  chr9  3875583  3875584  rs10974435_AL137071.1   \n",
       "..    ...      ...      ...   ...      ...      ...                    ...   \n",
       "733  chr9  3295000  5300000  chr9  5096665  5096666    rs10814917_MTCO1P11   \n",
       "734  chr9  3295000  5300000  chr9  5193557  5193558  rs10814917_AL133547.1   \n",
       "735  chr9  3295000  5300000  chr9  4984389  4984390        rs10814917_JAK2   \n",
       "736  chr9  3295000  5300000  chr9  5110912  5110913      rs10814917_TCF3P1   \n",
       "737  chr9  3295000  5300000  chr9  5113548  5113549      rs10814917_IGHEP2   \n",
       "\n",
       "    score strand1 strand2        sid    genename           geneid      rsid_x  \\\n",
       "0       .       +       -  9:4283682  AL354941.1  ENSG00000215297  rs10974435   \n",
       "1       .       +       -  9:4283682        RFX3  ENSG00000080298  rs10974435   \n",
       "2       .       +       +  9:4283682    RFX3-AS1  ENSG00000232104  rs10974435   \n",
       "3       .       +       +  9:4283682  AL354977.2  ENSG00000237359  rs10974435   \n",
       "4       .       +       +  9:4283682  AL137071.1  ENSG00000236724  rs10974435   \n",
       "..    ...     ...     ...        ...         ...              ...         ...   \n",
       "733     .       +       +  9:4296430    MTCO1P11  ENSG00000224083  rs10814917   \n",
       "734     .       +       -  9:4296430  AL133547.1  ENSG00000237553  rs10814917   \n",
       "735     .       +       +  9:4296430        JAK2  ENSG00000096968  rs10814917   \n",
       "736     .       +       +  9:4296430      TCF3P1  ENSG00000236567  rs10814917   \n",
       "737     .       +       +  9:4296430      IGHEP2  ENSG00000254017  rs10814917   \n",
       "\n",
       "     chr_x  pos_x ref_x alt_x variant  ma_samples  maf  eqtl_pval  eqtl_beta  \\\n",
       "0      NaN    NaN   NaN   NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "1      NaN    NaN   NaN   NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "2      NaN    NaN   NaN   NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "3      NaN    NaN   NaN   NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "4      NaN    NaN   NaN   NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "..     ...    ...   ...   ...     ...         ...  ...        ...        ...   \n",
       "733    NaN    NaN   NaN   NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "734    NaN    NaN   NaN   NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "735    NaN    NaN   NaN   NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "736    NaN    NaN   NaN   NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "737    NaN    NaN   NaN   NaN     NaN         NaN  NaN        NaN        NaN   \n",
       "\n",
       "     eqtl_slop_se  ac  an  r2  median_tpm rsid_y  dist  eqtl_fdr  \\\n",
       "0             NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "1             NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "2             NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "3             NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "4             NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "..            ...  ..  ..  ..         ...    ...   ...       ...   \n",
       "733           NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "734           NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "735           NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "736           NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "737           NaN NaN NaN NaN         NaN    NaN   NaN       NaN   \n",
       "\n",
       "     is_eqtl_pair  gene_start  gene_end gene_strand  is_closest_gene chr_y  \\\n",
       "0               0     3364329   3364792           -                0   NaN   \n",
       "1               0     3218297   3526004           -                0   NaN   \n",
       "2               0     3526723   3691814           +                0   NaN   \n",
       "3               0     3647335   3648338           +                0   NaN   \n",
       "4               0     3875584   3878809           +                0   NaN   \n",
       "..            ...         ...       ...         ...              ...   ...   \n",
       "733             0     5096666   5098193           +                0   NaN   \n",
       "734             0     5193339   5193558           -                0   NaN   \n",
       "735             0     4984390   5128183           +                0   NaN   \n",
       "736             0     5110913   5112849           +                0   NaN   \n",
       "737             0     5113549   5114804           +                0   NaN   \n",
       "\n",
       "     pos_y  ppH0  ppH1  ppH2  ppH3  ppH4 rsid ref_y alt_y  AC  AF  AN  \\\n",
       "0      NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN   NaN NaN NaN NaN   \n",
       "1      NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN   NaN NaN NaN NaN   \n",
       "2      NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN   NaN NaN NaN NaN   \n",
       "3      NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN   NaN NaN NaN NaN   \n",
       "4      NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN   NaN NaN NaN NaN   \n",
       "..     ...   ...   ...   ...   ...   ...  ...   ...   ...  ..  ..  ..   \n",
       "733    NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN   NaN NaN NaN NaN   \n",
       "734    NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN   NaN NaN NaN NaN   \n",
       "735    NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN   NaN NaN NaN NaN   \n",
       "736    NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN   NaN NaN NaN NaN   \n",
       "737    NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN   NaN NaN NaN NaN   \n",
       "\n",
       "     gwas_slope  gwas_slope_se  gwas_pval  \n",
       "0           NaN            NaN        NaN  \n",
       "1           NaN            NaN        NaN  \n",
       "2           NaN            NaN        NaN  \n",
       "3           NaN            NaN        NaN  \n",
       "4           NaN            NaN        NaN  \n",
       "..          ...            ...        ...  \n",
       "733         NaN            NaN        NaN  \n",
       "734         NaN            NaN        NaN  \n",
       "735         NaN            NaN        NaN  \n",
       "736         NaN            NaN        NaN  \n",
       "737         NaN            NaN        NaN  \n",
       "\n",
       "[738 rows x 52 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "master"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd09d082",
   "metadata": {},
   "source": [
    "### Add column to filter on coloc snp status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "18d88794",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### Add column to filter on coloc snp status\n"
     ]
    }
   ],
   "source": [
    "print('### Add column to filter on coloc snp status')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "429b2c09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After left merging master with the colocalization table there are 738 entries.\n",
      "Checking if I have the correct number of colocalized SNPs: before 18; after 18.\n"
     ]
    }
   ],
   "source": [
    "master['is_coloc_pair'] = (~master['ppH4'].isna()).astype(int)\n",
    "\n",
    "print('After left merging master with the colocalization table there are {} entries.'.format(master.shape[0]))\n",
    "after_cnt = sum(master['is_coloc_pair'])\n",
    "print('Checking if I have the correct number of colocalized SNPs: before {}; after {}.'.\\\n",
    "      format(coloc_sig_df.shape[0], after_cnt))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1733abce",
   "metadata": {},
   "source": [
    "### Add loop data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "c98c11a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# #### Add loop data\n",
      "There are 252 SNP-Gene loops.\n"
     ]
    }
   ],
   "source": [
    "print('# #### Add loop data')\n",
    "\n",
    "# check for the loop gene\n",
    "loop_check = [0] * master.shape[0]\n",
    "for i, sr in master.iterrows():\n",
    "\n",
    "    # check closest gene\n",
    "    rs_gene = (sr.sid, sr.geneid)\n",
    "    if rs_gene in fivekb_gloops_uniq:\n",
    "        loop_check[i] = 1       \n",
    "\n",
    "master['has_fithichip_loop'] = loop_check\n",
    "print('There are {} SNP-Gene loops.'.format(sum(loop_check)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da98c111",
   "metadata": {},
   "source": [
    "### Add coloc-snp boolean column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "8b981083",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Add coloc-snp boolean column\n"
     ]
    }
   ],
   "source": [
    "print('Add coloc-snp boolean column')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "a0f2671a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 738 coloc-SNP anchors.\n"
     ]
    }
   ],
   "source": [
    "master['has_colocSNP_anchor'] = master.sid.isin(colocSNP_anchors_set).astype(int)\n",
    "print('There are {} coloc-SNP anchors.'.format(sum(master['has_colocSNP_anchor'])))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17e50629",
   "metadata": {},
   "source": [
    "### Add LD SNP boolean column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "6239a5f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add a True/False (1/0) column for the presence of a direct coloc snp (non-LD)\n",
    "is_ld_snps = [x not in coloc_snps for x in master.sid]\n",
    "master['is_ld_snp'] = [int(x) for x in is_ld_snps]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fae50508",
   "metadata": {},
   "source": [
    "### Final ordering "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "06c98a84",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/BioHome/jreyna/tmp//ipykernel_122267/466200283.py:40: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  master.rename(columns={'chrA':'chrom',\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_122267/466200283.py:58: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  master.sort_values(['chrom', 'snp_pos', 'tss_start', 'rsid'], inplace=True)\n",
      "/mnt/BioHome/jreyna/tmp//ipykernel_122267/466200283.py:59: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  master.snp_pos = master.snp_pos.astype(int)\n"
     ]
    }
   ],
   "source": [
    "# #### Do the final reordering and saving\n",
    "\n",
    "# Include PC HiC data when it's available\n",
    "final_cols = ['sid', \n",
    "    'rsid',\n",
    "    'geneid',\n",
    "    'genename',\n",
    "    'chrA',\n",
    "    'endA',\n",
    "    'startB',\n",
    "    'endB',\n",
    "    'is_eqtl_pair',\n",
    "    'is_coloc_pair',\n",
    "    'is_closest_gene',\n",
    "    'has_fithichip_loop',\n",
    "    'has_colocSNP_anchor',\n",
    "    'is_ld_snp', \n",
    "    'eqtl_pval', \n",
    "    'eqtl_beta', \n",
    "     #'eqtl_fdr',\n",
    "    'dist',\n",
    "    'ppH0',\n",
    "    'ppH1',\n",
    "    'ppH2',\n",
    "    'ppH3',\n",
    "    'ppH4',\n",
    "    'gene_start',\n",
    "    'gene_end',\n",
    "    'ref_x', \n",
    "    'alt_x',\n",
    "    'AC',\n",
    "    'AF',\n",
    "    'AN',\n",
    "    'gwas_slope',\n",
    "    'gwas_slope_se',\n",
    "    'gwas_pval',\n",
    "    'gene_start', 'gene_end', 'gene_strand']\n",
    "\n",
    "master = master[final_cols]\n",
    "master.rename(columns={'chrA':'chrom',\n",
    "                       'endA': 'snp_pos',\n",
    "                       'ref_x': 'ref',\n",
    "                       'alt_x': 'alt',\n",
    "                       'startB': 'tss_start',\n",
    "                       'endB': 'tss_end',\n",
    "                       'genename': 'gene_name',\n",
    "                       'geneid': 'geneid',\n",
    "                       'pp_H0_Coloc_Summary': 'ppH0',\n",
    "                       'pp_H1_Coloc_Summary': 'ppH1',\n",
    "                       'pp_H2_Coloc_Summary': 'ppH2',\n",
    "                       'pp_H3_Coloc_Summary': 'ppH3',\n",
    "                       'pp_H4_Coloc_Summary': 'ppH4',\n",
    "                       'pvalue': 'eqtl_pval',\n",
    "                       'beta': 'eqtl_beta',\n",
    "                       'fdr': 'eqtl_fdr',\n",
    "                       'gwas_pval_nominal': 'gwas_pval'}, inplace=True)\n",
    "\n",
    "master.sort_values(['chrom', 'snp_pos', 'tss_start', 'rsid'], inplace=True)\n",
    "master.snp_pos = master.snp_pos.astype(int)\n",
    "\n",
    "# write out the master data\n",
    "fn = os.path.join(outdir, 'master.tsv')\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "master.to_csv(fn, sep='\\t', header=True, index=False)\n",
    "\n",
    "#fn = os.path.join(outdir, 'master.xlsx')\n",
    "#excel_master = master.sort_values('rsid').set_index('rsid')\n",
    "#excel_master.to_excel(fn, na_rep='nan')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "f9e5b18e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'results/main/sgls/ldpairs/T1D_34012112_Gaulton/GRCh37/Quach_2016/monocyte_R848/monocyte_naive/master.tsv'"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fn"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "HiChIP-DB",
   "language": "python",
   "name": "hichip-db"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
