{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "912288c2",
   "metadata": {},
   "source": [
    "# Make the master table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d0ab3f6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import sys\n",
    "import pandas as pd\n",
    "import subprocess as sp\n",
    "import glob\n",
    "os.chdir('/mnt/BioHome/jreyna/jreyna/projects/dchallenge/')\n",
    "hic_python='/mnt/BioHome/jreyna/software/anaconda3/envs/hic_tls/bin/python'\n",
    "\n",
    "# default values for the command line\n",
    "# sys.argv = [0] * 8\n",
    "# sys.argv[1] =  'results/main/2021_Nikhil_eQTL/Results/Colocalization/T1D_34012112_Gaulton/'\n",
    "# sys.argv[1] += 'BLUEPRINT_eQTL_Monocyte/FINAL_Summary_Coloc_Gene_SNP_Pairs.bed'\n",
    "# sys.argv[2] = 'results/refs/ensembl/gencode.v19.annotation.bed'\n",
    "# sys.argv[3] = 'results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/CM/FitHiChIP_L/FitHiChIP.interactions_FitHiC_Q0.01.bed'\n",
    "# sys.argv[4] = 'results/refs/spp/SPP_D-Challenge_networks.xlsx'\n",
    "# sys.argv[5] = 'results/refs/hg19/hg19.chrom.sizes'\n",
    "# sys.argv[6] = 'results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_summ_stats/BLUEPRINT_eQTL/Monocyte.txt.gz'\n",
    "# sys.argv[7] = 'results/main/loop_analysis/washU/'\n",
    "\n",
    "# # parsing the commandline arguments\n",
    "# coloc_fn = sys.argv[1]\n",
    "# genes_fn = sys.argv[2]\n",
    "# loop_fn = sys.argv[3]\n",
    "# spp_fn = sys.argv[4]\n",
    "# gs_fn = sys.argv[5]\n",
    "# eqtl_fn = sys.argv[6]\n",
    "# outdir = sys.argv[7]\n",
    "\n",
    "# # setting the output file names\n",
    "# os.makedirs(outdir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "aeca83ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load ./scripts/loop_analysis/Annotation_SNPs_with_Surround_Genes_Loops.sh\n",
    "#!/usr/bin/env python\n",
    "\n",
    "# In[1]:\n",
    "\n",
    "\n",
    "# Commandline Specification\n",
    "# Input\n",
    "# ----------------------------------------------\n",
    "# 1) colocalization file  x\n",
    "# 2) gencode gene annotation file in bed format x\n",
    "# 3) loop data x \n",
    "# 4) spp network data x \n",
    "# 5) reference chrom sizes x \n",
    "# Output\n",
    "# ----------------------------------------------\n",
    "# 1) snp-gene loop summary file\n",
    "# 2) snp-gene pairs longrange file with index\n",
    "# 3) snp-gene loops longrange file with index\n",
    "\n",
    "# default values for the command line\n",
    "#sys.argv = [0] * 7\n",
    "#sys.argv[1] =  'results/main/2021_Nikhil_eQTL/Results/Colocalization/T1D_34012112_Gaulton/'\n",
    "#sys.argv[1] += 'BLUEPRINT_eQTL_Monocyte/FINAL_Summary_Coloc_Gene_SNP_Pairs.bed'\n",
    "#sys.argv[2] = 'results/refs/ensembl/gencode.v19.annotation.bed'\n",
    "#sys.argv[3] = 'results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/CM/FitHiChIP_L/FitHiChIP.interactions_FitHiC_Q0.01.bed'\n",
    "#sys.argv[4] = 'results/refs/spp/SPP_D-Challenge_networks.xlsx'\n",
    "#sys.argv[5] = 'results/refs/hg19/hg19.chrom.sizes'\n",
    "#sys.argv[6] = 'results/main/loop_analysis/washU/'\n",
    "script = \"scripts/loop_analysis/Annotate_Coloc_SNPs_with_Nearby_Genes_Loops.py\"\n",
    "gs = \"results/refs/hg19/hg19.chrom.sizes\"\n",
    "gencode = \"results/refs/ensembl/gencode.v19.annotation.bed\"\n",
    "spp = \"results/refs/spp/SPP_D-Challenge_networks.xlsx\"\n",
    "\n",
    "# template file names for colocalization, hichip, eqtl and outdir\n",
    "coloc_tmp = \"results/main/2021_Nikhil_eQTL/Results/Colocalization/{gwas}/{eqtl}_{cline}/FINAL_Summary_Coloc_Gene_SNP_Pairs.bed\"\n",
    "fithichip_tmp = \"results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/{cline}/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01.bed\"\n",
    "eqtl_tmp = \"results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_summ_stats/{eqtl}/{cline}.txt.gz\"\n",
    "outdir_tmp = \"results/main/loop_analysis/Coloc_Approach/{gwas}/{dice_cline}/{eqtl_study}/{eqtl_cline}/\"\n",
    "\n",
    "# getting a list of DICE clines with HiChIP data\n",
    "fithichip_clines = glob.glob(\"results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/*\")\n",
    "fithichip_clines = [x.split('/')[-1] for x in fithichip_clines]\n",
    "\n",
    "# eqtl mapper\n",
    "eqtl_mapper = pd.read_table('results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_summ_stats/eqtl.cline.mapper.study.tsv')\n",
    "eqtl_mapper.sort_values(['will_analyze', 'study', 'dice_shortname'], ascending=[False, True, True], inplace=True)\n",
    "eqtl_mapper[['study', 'dice_shortname', 'will_analyze', 'eqtl_file']]\n",
    "eqtl_mapper = eqtl_mapper.loc[eqtl_mapper['will_analyze'] == 'Yes']\n",
    "eqtl_mapper = eqtl_mapper.loc[eqtl_mapper.dice_shortname.isin(fithichip_clines)]\n",
    "eqtl_clines = eqtl_mapper['eqtl_file'].str.split('/')\n",
    "eqtl_clines = [x[-1].split('.')[0] for x in eqtl_clines]\n",
    "eqtl_mapper['eqtl_cline'] = eqtl_clines\n",
    "eqtl_mapper = eqtl_mapper[['study', 'dice_shortname', 'eqtl_cline', 'eqtl_file']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c0859bcd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study</th>\n",
       "      <th>dice_shortname</th>\n",
       "      <th>eqtl_cline</th>\n",
       "      <th>eqtl_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BLUEPRINT_eQTL</td>\n",
       "      <td>CM</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BLUEPRINT_eQTL</td>\n",
       "      <td>NCM</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BLUEPRINT_eQTL</td>\n",
       "      <td>TFH</td>\n",
       "      <td>T-cell</td>\n",
       "      <td>results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BLUEPRINT_eQTL</td>\n",
       "      <td>TH1</td>\n",
       "      <td>T-cell</td>\n",
       "      <td>results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BLUEPRINT_eQTL</td>\n",
       "      <td>TH17</td>\n",
       "      <td>T-cell</td>\n",
       "      <td>results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            study dice_shortname eqtl_cline  \\\n",
       "0  BLUEPRINT_eQTL             CM   Monocyte   \n",
       "1  BLUEPRINT_eQTL            NCM   Monocyte   \n",
       "2  BLUEPRINT_eQTL            TFH     T-cell   \n",
       "3  BLUEPRINT_eQTL            TH1     T-cell   \n",
       "4  BLUEPRINT_eQTL           TH17     T-cell   \n",
       "\n",
       "                                           eqtl_file  \n",
       "0  results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...  \n",
       "1  results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...  \n",
       "2  results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...  \n",
       "3  results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...  \n",
       "4  results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eqtl_mapper.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c414df46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get gwas runs\n",
    "gwas_runs = [os.path.basename(x) for x in glob.glob('results/main/2021_Nikhil_eQTL/Data/T1D_GWAS/*')]\n",
    "gwas_runs = gwas_runs[2:3]\n",
    "\n",
    "# get eqtl runs \n",
    "eqtl_runs = [x for x in glob.glob('results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_summ_stats/*/*')]\n",
    "eqtl_runs = [x for x in eqtl_runs if x != 'README' and 'eQTL' in x]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2feed471",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study</th>\n",
       "      <th>dice_shortname</th>\n",
       "      <th>eqtl_cline</th>\n",
       "      <th>eqtl_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BLUEPRINT_eQTL</td>\n",
       "      <td>CM</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BLUEPRINT_eQTL</td>\n",
       "      <td>NCM</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BLUEPRINT_eQTL</td>\n",
       "      <td>TFH</td>\n",
       "      <td>T-cell</td>\n",
       "      <td>results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BLUEPRINT_eQTL</td>\n",
       "      <td>TH1</td>\n",
       "      <td>T-cell</td>\n",
       "      <td>results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BLUEPRINT_eQTL</td>\n",
       "      <td>TH17</td>\n",
       "      <td>T-cell</td>\n",
       "      <td>results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            study dice_shortname eqtl_cline  \\\n",
       "0  BLUEPRINT_eQTL             CM   Monocyte   \n",
       "1  BLUEPRINT_eQTL            NCM   Monocyte   \n",
       "2  BLUEPRINT_eQTL            TFH     T-cell   \n",
       "3  BLUEPRINT_eQTL            TH1     T-cell   \n",
       "4  BLUEPRINT_eQTL           TH17     T-cell   \n",
       "\n",
       "                                           eqtl_file  \n",
       "0  results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...  \n",
       "1  results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...  \n",
       "2  results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...  \n",
       "3  results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...  \n",
       "4  results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_s...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eqtl_mapper.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c3518121",
   "metadata": {},
   "outputs": [],
   "source": [
    "# qscripts_dir = 'results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/qjobs/'\n",
    "# os.makedirs(qscripts_dir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ee0e8af1",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now running: /mnt/BioHome/jreyna/software/anaconda3/envs/hic_tls/bin/python scripts/loop_analysis/Annotate_Coloc_SNPs_with_Nearby_Genes_Loops.py results/main/2021_Nikhil_eQTL/Results/Colocalization/T1D_34012112_Gaulton/BLUEPRINT_eQTL_Monocyte/FINAL_Summary_Coloc_Gene_SNP_Pairs.bed results/refs/ensembl/gencode.v19.annotation.bed results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/CM/FitHiChIP_S/FitHiChIP.interactions_FitHiC_Q0.01.bed results/refs/spp/SPP_D-Challenge_networks.xlsx results/refs/hg19/hg19.chrom.sizes results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_summ_stats/BLUEPRINT_eQTL/Monocyte.txt.gz results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/\n",
      "There are 15 colocalized SNP-gene pairs\n",
      "There are 57774 genes in this GTF-derived file.\n",
      "There are 635 colocalized snp-gene pairs within +/- 5kb.\n",
      "FitHiChIP found 84298 significant loops.\n",
      "There are 35 SNP-Gene pairs with a loop.\n",
      "Master is starting with 635 snp-gene pairs.\n",
      "There are 7716 eQTLs.\n",
      "After outer merging with eqtls master has 8336 snp-gene pairs.\n",
      "After left merging master with the colocalization table there are 8336 entries.\n",
      "There are 36 SNP-Gene loops.\n",
      "/mnt/BioApps/tabix/tabix-0.2.6/bgzip -f results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/gs_genes.bed\n",
      "out: \n",
      "err: \n",
      "/mnt/BioApps/tabix/tabix-0.2.6/tabix results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/gs_genes.bed.gz\n",
      "out: \n",
      "err: \n",
      "Created the gzfn: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/gs_genes.bed.gz\n",
      "Created the tabix: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/gs_genes.bed.gz.tbi\n",
      "/mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/5kb.snp_gene_pairs.bed\n",
      "out: \n",
      "err: \n",
      "/mnt/BioApps/tabix/tabix-0.2.6/tabix -f results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/5kb.snp_gene_pairs.bed.gz\n",
      "out: \n",
      "err: \n",
      "Created the gzfn: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/5kb.snp_gene_pairs.bed.gz\n",
      "Created the tabix: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/5kb.snp_gene_pairs.bed.gz.tbi\n",
      "/mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/5kb.snp_gene_loops.bed\n",
      "out: \n",
      "err: \n",
      "/mnt/BioApps/tabix/tabix-0.2.6/tabix -f results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/5kb.snp_gene_loops.bed.gz\n",
      "out: \n",
      "err: \n",
      "Created the gzfn: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/5kb.snp_gene_loops.bed.gz\n",
      "Created the tabix: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/5kb.snp_gene_loops.bed.gz.tbi\n",
      "/mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/eqtl.snp_gene_pairs.bed\n",
      "out: \n",
      "err: \n",
      "/mnt/BioApps/tabix/tabix-0.2.6/tabix -f results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/eqtl.snp_gene_pairs.bed.gz\n",
      "out: \n",
      "err: \n",
      "Created the gzfn: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/eqtl.snp_gene_pairs.bed.gz\n",
      "Created the tabix: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/eqtl.snp_gene_pairs.bed.gz.tbi\n",
      "/mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/eqtl.genes_only.bed\n",
      "out: \n",
      "err: \n",
      "/mnt/BioApps/tabix/tabix-0.2.6/tabix -f results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/eqtl.genes_only.bed.gz\n",
      "out: \n",
      "err: \n",
      "Created the gzfn: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/eqtl.genes_only.bed.gz\n",
      "Created the tabix: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/eqtl.genes_only.bed.gz.tbi\n",
      "/mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/eqtl.snps_only.bed\n",
      "out: \n",
      "err: \n",
      "/mnt/BioApps/tabix/tabix-0.2.6/tabix -f results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/eqtl.snps_only.bed.gz\n",
      "out: \n",
      "err: \n",
      "Created the gzfn: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/eqtl.snps_only.bed.gz\n",
      "Created the tabix: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/eqtl.snps_only.bed.gz.tbi\n",
      "/mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.snps_only.bed\n",
      "out: \n",
      "err: \n",
      "/mnt/BioApps/tabix/tabix-0.2.6/tabix -f results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.snps_only.bed.gz\n",
      "out: \n",
      "err: \n",
      "Created the gzfn: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.snps_only.bed.gz\n",
      "Created the tabix: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.snps_only.bed.gz.tbi\n",
      "/mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.genes_only.bed\n",
      "out: \n",
      "err: \n",
      "/mnt/BioApps/tabix/tabix-0.2.6/tabix -f results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.genes_only.bed.gz\n",
      "out: \n",
      "err: \n",
      "Created the gzfn: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.genes_only.bed.gz\n",
      "Created the tabix: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.genes_only.bed.gz.tbi\n",
      "/mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.snp_gene_pairs.bed\n",
      "out: \n",
      "err: \n",
      "/mnt/BioApps/tabix/tabix-0.2.6/tabix -f results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.snp_gene_pairs.bed.gz\n",
      "out: \n",
      "err: \n",
      "Created the gzfn: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.snp_gene_pairs.bed.gz\n",
      "Created the tabix: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.snp_gene_pairs.bed.gz.tbi\n",
      "/mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.snp_gene_loops.bed\n",
      "out: \n",
      "err: \n",
      "/mnt/BioApps/tabix/tabix-0.2.6/tabix -f results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.snp_gene_loops.bed.gz\n",
      "out: \n",
      "err: \n",
      "Created the gzfn: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.snp_gene_loops.bed.gz\n",
      "Created the tabix: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.snp_gene_loops.bed.gz.tbi\n",
      "/mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.anchors_only.bed\n",
      "out: \n",
      "err: \n",
      "/mnt/BioApps/tabix/tabix-0.2.6/tabix -f results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.anchors_only.bed.gz\n",
      "out: \n",
      "err: \n",
      "Created the gzfn: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.anchors_only.bed.gz\n",
      "Created the tabix: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/coloc.anchors_only.bed.gz.tbi\n",
      "# make the hub json file\n",
      "\n",
      "/mnt/BioHome/jreyna/.local/lib/python3.6/site-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  errors=errors,\n",
      "/mnt/BioHome/jreyna/.local/lib/python3.6/site-packages/pandas/core/indexing.py:1596: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[key] = _infer_fill_value(value)\n",
      "/mnt/BioHome/jreyna/.local/lib/python3.6/site-packages/pandas/core/indexing.py:1763: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  isetter(loc, value)\n",
      "/mnt/BioHome/jreyna/.local/lib/python3.6/site-packages/pandas/core/indexing.py:1743: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  isetter(ilocs[0], value)\n",
      "/mnt/BioHome/jreyna/.local/lib/python3.6/site-packages/pandas/core/indexing.py:1743: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  isetter(ilocs[0], value)\n",
      "scripts/loop_analysis/Annotate_Coloc_SNPs_with_Nearby_Genes_Loops.py:776: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  eqtl_snp_gene_pairs['score'] = 0.01\n",
      "/mnt/BioHome/jreyna/.local/lib/python3.6/site-packages/pandas/core/indexing.py:1743: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  isetter(ilocs[0], value)\n",
      "/mnt/BioHome/jreyna/.local/lib/python3.6/site-packages/pandas/core/indexing.py:1763: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  isetter(loc, value)\n",
      "\n",
      "Results can now be found at: results/main/loop_analysis/Coloc_Approach/T1D_34012112_Gaulton/CM/BLUEPRINT_eQTL/Monocyte/\n"
     ]
    }
   ],
   "source": [
    "for gwas in gwas_runs:\n",
    "    for i, sr in eqtl_mapper.iterrows():\n",
    "        \n",
    "        coloc = coloc_tmp.format(gwas=gwas, eqtl=sr.study, cline=sr.eqtl_cline)\n",
    "        fithichip = fithichip_tmp.format(cline=sr.dice_shortname)\n",
    "        outdir = outdir_tmp.format(gwas=gwas, \n",
    "                                   dice_cline=sr.dice_shortname,\n",
    "                                   eqtl_cline=sr.eqtl_cline,\n",
    "                                   eqtl_study=sr.study)\n",
    "        cmd = \"{} {} {} {} {} {} {} {} {}\".format(hic_python, script, coloc, gencode, \n",
    "                                                  fithichip, spp, gs, sr.eqtl_file, outdir)        \n",
    "\n",
    "        \n",
    "        if not os.path.exists(outdir):\n",
    "            \n",
    "            #             script_fn = '{}_{}_{}_{}.qscript'.format(gwas, sr.dice_shortname, sr.eqtl_cline, sr.study)\n",
    "            #             script_fn = os.path.join(qscripts_dir, script_fn)\n",
    "            #             with open(script_fn, 'w') as fw:\n",
    "            #                 fw.write(cmd)\n",
    "                \n",
    "            job = sp.Popen(cmd, shell=True, stderr=sp.PIPE, stdout=sp.PIPE)\n",
    "            out, err = job.communicate()\n",
    "            print('Now running: {}'.format(cmd))\n",
    "            print(out.decode())\n",
    "            print(err.decode())\n",
    "            print('Results can now be found at: {}'.format(outdir))\n",
    "        else:\n",
    "            msg = 'Previously ran {}'.format(outdir)\n",
    "            print(msg)\n",
    "            \n",
    "        break\n",
    "            \n",
    "        print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "475b248f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e058e778",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5111c31a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c02b68c0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "368d4e80",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56fcf006",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cd89a5c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "496e52d2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "223dd4ba",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd2b9274",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2ac27f6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "HiC",
   "language": "python",
   "name": "hic_tls"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
