{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "72b0e9f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Commandline Specification\n",
    "# Input\n",
    "# ----------------------------------------------\n",
    "# 1) colocalization file  x\n",
    "# 2) gencode gene annotation file in bed format x\n",
    "# 3) loop data x \n",
    "# 4) spp network data x \n",
    "# 5) reference chrom sizes x \n",
    "# Output\n",
    "# ----------------------------------------------\n",
    "# 1) snp-gene loop summary file\n",
    "# 2) snp-gene pairs longrange file with index\n",
    "# 3) snp-gene loops longrange file with index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "70cd1d04",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import sys\n",
    "import pybedtools as pbt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess as sp\n",
    "import json\n",
    "os.chdir('/mnt/BioHome/jreyna/jreyna/projects/dchallenge/')\n",
    "pbt.set_bedtools_path('/mnt/BioApps/bedtools/bin/')\n",
    "bgzip = '/mnt/BioApps/tabix/tabix-0.2.6/bgzip'\n",
    "tabix = '/mnt/BioApps/tabix/tabix-0.2.6/tabix'\n",
    "\n",
    "bedpe_6cols = ['chrA', 'startA', 'endA', 'chrB', 'startB', 'endB']\n",
    "bedpe_10cols = ['chrA', 'startA', 'endA', 'chrB', 'startB', 'endB', 'name', 'score', 'strand1', 'strand2']\n",
    "\n",
    "# default values for the command line\n",
    "sys.argv = [0] * 7\n",
    "sys.argv[1] =  'results/main/2021_Nikhil_eQTL/Results/Colocalization/T1D_34012112_Gaulton/'\n",
    "sys.argv[1] += 'BLUEPRINT_eQTL_Monocyte/FINAL_Summary_Coloc_Gene_SNP_Pairs.bed'\n",
    "sys.argv[2] = 'results/refs/ensembl/gencode.v19.annotation.bed'\n",
    "sys.argv[3] = 'results/main/2021_Nikhil_eQTL/Data/FitHiChIP_Loops/CM/FitHiChIP_L/FitHiChIP.interactions_FitHiC_Q0.01.bed'\n",
    "sys.argv[4] = 'results/refs/spp/SPP_D-Challenge_networks.xlsx'\n",
    "sys.argv[5] = 'results/refs/hg19/hg19.chrom.sizes'\n",
    "sys.argv[6] = 'results/main/loop_analysis/washU/'\n",
    "\n",
    "# parsing the commandline arguments\n",
    "coloc_fn = sys.argv[1]\n",
    "genes_fn = sys.argv[2]\n",
    "loop_fn = sys.argv[3]\n",
    "spp_fn = sys.argv[4]\n",
    "gs_fn = sys.argv[5]\n",
    "outdir = sys.argv[6]\n",
    "\n",
    "# setting the output file names\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "summary_fn = os.path.join(outdir, 'sgloop_summary.xlsx')\n",
    "sg_pairs_fn = os.path.join(outdir, 'gs_pairs.longrange.bed')\n",
    "sg_loops_fn = os.path.join(outdir, 'gs_loops.longrange.bed')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "310c39e0",
   "metadata": {},
   "source": [
    "## Load the colocalization data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "55df2d6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the colocalization data\n",
    "coloc = pd.read_table(coloc_fn)\n",
    "\n",
    "# extract the most significant according the H4 \n",
    "coloc_sig_df = coloc[coloc['pp_H4_Coloc_Summary'] > 0.75]\n",
    "coloc_sig_full = coloc_sig_df.copy(deep=True)\n",
    "coloc_sig_df = coloc_sig_df[['chr', 'pos', 'rs_id', 'variant_id', 'geneName']]\n",
    "coloc_sig_df.rename(columns={'pos': 'end'}, inplace=True)\n",
    "coloc_sig_df.loc[:, 'start'] = coloc_sig_df.loc[:, 'end'] - 1\n",
    "coloc_sig_df = coloc_sig_df[['chr', 'start', 'end', 'rs_id', 'variant_id', 'geneName']]\n",
    "coloc_sig_pbt = pbt.BedTool.from_dataframe(coloc_sig_df).sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bf0c49bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>geneName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114426000</td>\n",
       "      <td>114426001</td>\n",
       "      <td>rs11102694</td>\n",
       "      <td>rs11102694:114426001:G:A</td>\n",
       "      <td>AP4B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>192538495</td>\n",
       "      <td>192538496</td>\n",
       "      <td>rs2760530</td>\n",
       "      <td>rs2760530</td>\n",
       "      <td>RGS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114426000</td>\n",
       "      <td>114426001</td>\n",
       "      <td>rs11102694</td>\n",
       "      <td>rs11102694:114426001:G:A</td>\n",
       "      <td>AP4B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6</td>\n",
       "      <td>26383249</td>\n",
       "      <td>26383250</td>\n",
       "      <td>rs9467740</td>\n",
       "      <td>rs9467740:26383250:A:T</td>\n",
       "      <td>BTN2A2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>33548089</td>\n",
       "      <td>33548090</td>\n",
       "      <td>rs3216621</td>\n",
       "      <td>rs3216621:33548090:A:AG</td>\n",
       "      <td>BAK1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr      start        end       rs_id                variant_id geneName\n",
       "0  chr1  114426000  114426001  rs11102694  rs11102694:114426001:G:A    AP4B1\n",
       "1  chr1  192538495  192538496   rs2760530                 rs2760530     RGS1\n",
       "2  chr1  114426000  114426001  rs11102694  rs11102694:114426001:G:A    AP4B1\n",
       "3  chr6   26383249   26383250   rs9467740    rs9467740:26383250:A:T   BTN2A2\n",
       "4  chr6   33548089   33548090   rs3216621   rs3216621:33548090:A:AG     BAK1"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coloc_sig_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "98842c11",
   "metadata": {},
   "outputs": [],
   "source": [
    "# list of SNP-GENE colocalized pairs \n",
    "sg_coloc_set = coloc_sig_full[['rs_id', 'geneName']].values.tolist()\n",
    "sg_coloc_set = set([tuple(x) for x in sg_coloc_set])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e1d73419",
   "metadata": {},
   "outputs": [],
   "source": [
    "csnp_slop_pbt = coloc_sig_pbt.slop(b=500000, g=gs_fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d360665b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In total there are 18 colocalized SNPs.\n"
     ]
    }
   ],
   "source": [
    "print('In total there are {} colocalized SNPs.'.format(len(coloc_sig_df)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55847e28",
   "metadata": {},
   "source": [
    "## Load the gene data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "51e1b172",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/BioHome/jreyna/.local/lib/python3.6/site-packages/pandas/core/indexing.py:1743: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  isetter(ilocs[0], value)\n"
     ]
    }
   ],
   "source": [
    "# load the gencode coords\n",
    "cols = ['chrom', 'start', 'end', 'strand', 'type', 'gene_id', 'gene_name']\n",
    "gencode = genes_df = pd.read_table(genes_fn, header=None, names=cols)\n",
    "\n",
    "# extract just the genes\n",
    "genes_df = gencode.loc[gencode.type.isin(['gene'])]\n",
    "genes_df.loc[:, 'chrom'] = genes_df['chrom'].astype(str)\n",
    "genes_df = genes_df.iloc[:, [0,1,2,6,5]]\n",
    "genes_pbt = pbt.BedTool.from_dataframe(genes_df).sort()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66398501",
   "metadata": {},
   "source": [
    "## Intersect SNPs and Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "cf6665f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bedtools intersect for SNP (with 500kb slop) and genes\n",
    "csnps_gene_pairs = csnp_slop_pbt.intersect(genes_pbt, wa=True, wb=True)\n",
    "\n",
    "# extract the SNP and gene data\n",
    "csnps_gene_pairs_df = csnps_gene_pairs.to_dataframe()\n",
    "csnps_gene_pairs_df = csnps_gene_pairs_df.iloc[:, [0,1,2,6,7,8,3,5,9]]\n",
    "\n",
    "# remove the slop that you added previously. This has to be done before \n",
    "# intersecting the snp-gene pairs with the loops\n",
    "csnps_gene_pairs_df['start'] += 500000\n",
    "csnps_gene_pairs_df['end'] -= 500000\n",
    "\n",
    "# converting to bedpe10 plus data format (allows pairtopair to work)\n",
    "csnps_gene_pairs_df.columns = bedpe_6cols + ['rs_id', 'coloc_gene', 'sg_gene']\n",
    "csnps_gene_pairs_df['name'] = '.'\n",
    "csnps_gene_pairs_df['score'] = '.'\n",
    "csnps_gene_pairs_df['strand1'] = '.'\n",
    "csnps_gene_pairs_df['strand2'] = '.'\n",
    "csnps_gene_pairs_df = csnps_gene_pairs_df[bedpe_10cols + ['rs_id', 'coloc_gene', 'sg_gene']]\n",
    "csnps_gene_pairs_pbt = pbt.BedTool.from_dataframe(csnps_gene_pairs_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ff04554d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrA</th>\n",
       "      <th>startA</th>\n",
       "      <th>endA</th>\n",
       "      <th>chrB</th>\n",
       "      <th>startB</th>\n",
       "      <th>endB</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand1</th>\n",
       "      <th>strand2</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>coloc_gene</th>\n",
       "      <th>sg_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114426000</td>\n",
       "      <td>114426001</td>\n",
       "      <td>chr1</td>\n",
       "      <td>113992322</td>\n",
       "      <td>113993350</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>rs11102694</td>\n",
       "      <td>AP4B1</td>\n",
       "      <td>RP11-512F24.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114426000</td>\n",
       "      <td>114426001</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114119379</td>\n",
       "      <td>114119521</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>rs11102694</td>\n",
       "      <td>AP4B1</td>\n",
       "      <td>MTND5P20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114426000</td>\n",
       "      <td>114426001</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114241506</td>\n",
       "      <td>114242253</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>rs11102694</td>\n",
       "      <td>AP4B1</td>\n",
       "      <td>RP4-730K3.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114426000</td>\n",
       "      <td>114426001</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114447763</td>\n",
       "      <td>114456708</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>rs11102694</td>\n",
       "      <td>AP4B1</td>\n",
       "      <td>DCLRE1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114426000</td>\n",
       "      <td>114426001</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114466622</td>\n",
       "      <td>114472114</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>rs11102694</td>\n",
       "      <td>AP4B1</td>\n",
       "      <td>RP5-1073O3.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrA     startA       endA  chrB     startB       endB name score strand1  \\\n",
       "0  chr1  114426000  114426001  chr1  113992322  113993350    .     .       .   \n",
       "1  chr1  114426000  114426001  chr1  114119379  114119521    .     .       .   \n",
       "2  chr1  114426000  114426001  chr1  114241506  114242253    .     .       .   \n",
       "3  chr1  114426000  114426001  chr1  114447763  114456708    .     .       .   \n",
       "4  chr1  114426000  114426001  chr1  114466622  114472114    .     .       .   \n",
       "\n",
       "  strand2       rs_id coloc_gene        sg_gene  \n",
       "0       .  rs11102694      AP4B1  RP11-512F24.1  \n",
       "1       .  rs11102694      AP4B1       MTND5P20  \n",
       "2       .  rs11102694      AP4B1    RP4-730K3.3  \n",
       "3       .  rs11102694      AP4B1        DCLRE1B  \n",
       "4       .  rs11102694      AP4B1   RP5-1073O3.7  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "csnps_gene_pairs_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5f21b4a",
   "metadata": {},
   "source": [
    "## Loading the Loop Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7919e394",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/BioHome/jreyna/.local/lib/python3.6/site-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  errors=errors,\n",
      "/mnt/BioHome/jreyna/.local/lib/python3.6/site-packages/pandas/core/indexing.py:1596: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[key] = _infer_fill_value(value)\n",
      "/mnt/BioHome/jreyna/.local/lib/python3.6/site-packages/pandas/core/indexing.py:1763: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  isetter(loc, value)\n",
      "/mnt/BioHome/jreyna/.local/lib/python3.6/site-packages/pandas/core/indexing.py:1743: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  isetter(ilocs[0], value)\n"
     ]
    }
   ],
   "source": [
    "# load the loop data\n",
    "loops = pd.read_table(loop_fn)\n",
    "tmp_loops = loops[['chr1', 's1', 'e1', 'chr2', 's2', 'e2']]\n",
    "tmp_loops.rename(columns={'p': 'score'}, inplace=True)\n",
    "tmp_loops.loc[:, 'name'] = '.'\n",
    "tmp_loops.loc[:, 'score'] = loops['p']\n",
    "tmp_loops.loc[:, 'strand1'] = '.'\n",
    "tmp_loops.loc[:, 'strand2'] = '.'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d480f1f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr1</th>\n",
       "      <th>s1</th>\n",
       "      <th>e1</th>\n",
       "      <th>chr2</th>\n",
       "      <th>s2</th>\n",
       "      <th>e2</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand1</th>\n",
       "      <th>strand2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>710000</td>\n",
       "      <td>715000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>735000</td>\n",
       "      <td>740000</td>\n",
       "      <td>.</td>\n",
       "      <td>4.895006e-07</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>710000</td>\n",
       "      <td>715000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>750000</td>\n",
       "      <td>755000</td>\n",
       "      <td>.</td>\n",
       "      <td>3.771037e-07</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>710000</td>\n",
       "      <td>715000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>755000</td>\n",
       "      <td>760000</td>\n",
       "      <td>.</td>\n",
       "      <td>3.431682e-07</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>710000</td>\n",
       "      <td>715000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>760000</td>\n",
       "      <td>765000</td>\n",
       "      <td>.</td>\n",
       "      <td>3.114478e-07</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>710000</td>\n",
       "      <td>715000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>765000</td>\n",
       "      <td>770000</td>\n",
       "      <td>.</td>\n",
       "      <td>2.822039e-07</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498662</th>\n",
       "      <td>chr9</td>\n",
       "      <td>140525000</td>\n",
       "      <td>140530000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>140550000</td>\n",
       "      <td>140555000</td>\n",
       "      <td>.</td>\n",
       "      <td>4.895006e-07</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498663</th>\n",
       "      <td>chr9</td>\n",
       "      <td>140640000</td>\n",
       "      <td>140645000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>140755000</td>\n",
       "      <td>140760000</td>\n",
       "      <td>.</td>\n",
       "      <td>1.109468e-07</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498664</th>\n",
       "      <td>chr9</td>\n",
       "      <td>140955000</td>\n",
       "      <td>140960000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>140990000</td>\n",
       "      <td>140995000</td>\n",
       "      <td>.</td>\n",
       "      <td>4.129931e-07</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498665</th>\n",
       "      <td>chr9</td>\n",
       "      <td>140970000</td>\n",
       "      <td>140975000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>140990000</td>\n",
       "      <td>140995000</td>\n",
       "      <td>.</td>\n",
       "      <td>5.290671e-07</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498666</th>\n",
       "      <td>chr9</td>\n",
       "      <td>141045000</td>\n",
       "      <td>141050000</td>\n",
       "      <td>chr9</td>\n",
       "      <td>141100000</td>\n",
       "      <td>141105000</td>\n",
       "      <td>.</td>\n",
       "      <td>2.822039e-07</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>498667 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr1         s1         e1  chr2         s2         e2 name  \\\n",
       "0       chr1     710000     715000  chr1     735000     740000    .   \n",
       "1       chr1     710000     715000  chr1     750000     755000    .   \n",
       "2       chr1     710000     715000  chr1     755000     760000    .   \n",
       "3       chr1     710000     715000  chr1     760000     765000    .   \n",
       "4       chr1     710000     715000  chr1     765000     770000    .   \n",
       "...      ...        ...        ...   ...        ...        ...  ...   \n",
       "498662  chr9  140525000  140530000  chr9  140550000  140555000    .   \n",
       "498663  chr9  140640000  140645000  chr9  140755000  140760000    .   \n",
       "498664  chr9  140955000  140960000  chr9  140990000  140995000    .   \n",
       "498665  chr9  140970000  140975000  chr9  140990000  140995000    .   \n",
       "498666  chr9  141045000  141050000  chr9  141100000  141105000    .   \n",
       "\n",
       "               score strand1 strand2  \n",
       "0       4.895006e-07       .       .  \n",
       "1       3.771037e-07       .       .  \n",
       "2       3.431682e-07       .       .  \n",
       "3       3.114478e-07       .       .  \n",
       "4       2.822039e-07       .       .  \n",
       "...              ...     ...     ...  \n",
       "498662  4.895006e-07       .       .  \n",
       "498663  1.109468e-07       .       .  \n",
       "498664  4.129931e-07       .       .  \n",
       "498665  5.290671e-07       .       .  \n",
       "498666  2.822039e-07       .       .  \n",
       "\n",
       "[498667 rows x 10 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp_loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7c56b9be",
   "metadata": {},
   "outputs": [],
   "source": [
    "loops = pbt.BedTool.from_dataframe(tmp_loops)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5969e6d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FitHiChIP found 498667 significant loops.\n"
     ]
    }
   ],
   "source": [
    "print('FitHiChIP found {} significant loops.'.format(tmp_loops.shape[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbfd2f9a",
   "metadata": {},
   "source": [
    "## Intersect SNP-Gene Pairs with Loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "af67a10a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrS</th>\n",
       "      <th>startS</th>\n",
       "      <th>endS</th>\n",
       "      <th>chrG</th>\n",
       "      <th>startG</th>\n",
       "      <th>endG</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>coloc_gene</th>\n",
       "      <th>sg_gene</th>\n",
       "      <th>chrLA</th>\n",
       "      <th>startLA</th>\n",
       "      <th>endLA</th>\n",
       "      <th>chrRA</th>\n",
       "      <th>startRA</th>\n",
       "      <th>endRA</th>\n",
       "      <th>p-value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114426000</td>\n",
       "      <td>114426001</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114447763</td>\n",
       "      <td>114456708</td>\n",
       "      <td>rs11102694</td>\n",
       "      <td>AP4B1</td>\n",
       "      <td>DCLRE1B</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114425000</td>\n",
       "      <td>114430000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114445000</td>\n",
       "      <td>114450000</td>\n",
       "      <td>5.290671e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114426000</td>\n",
       "      <td>114426001</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114466622</td>\n",
       "      <td>114472114</td>\n",
       "      <td>rs11102694</td>\n",
       "      <td>AP4B1</td>\n",
       "      <td>RP5-1073O3.7</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114425000</td>\n",
       "      <td>114430000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114470000</td>\n",
       "      <td>114475000</td>\n",
       "      <td>3.431682e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114426000</td>\n",
       "      <td>114426001</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114304454</td>\n",
       "      <td>114355098</td>\n",
       "      <td>rs11102694</td>\n",
       "      <td>AP4B1</td>\n",
       "      <td>RSBN1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114300000</td>\n",
       "      <td>114305000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114425000</td>\n",
       "      <td>114430000</td>\n",
       "      <td>1.011634e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114426000</td>\n",
       "      <td>114426001</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114356433</td>\n",
       "      <td>114414381</td>\n",
       "      <td>rs11102694</td>\n",
       "      <td>AP4B1</td>\n",
       "      <td>PTPN22</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114410000</td>\n",
       "      <td>114415000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114425000</td>\n",
       "      <td>114430000</td>\n",
       "      <td>5.684843e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>114426000</td>\n",
       "      <td>114426001</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114437370</td>\n",
       "      <td>114447823</td>\n",
       "      <td>rs11102694</td>\n",
       "      <td>AP4B1</td>\n",
       "      <td>AP4B1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114425000</td>\n",
       "      <td>114430000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114445000</td>\n",
       "      <td>114450000</td>\n",
       "      <td>5.290671e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>393</th>\n",
       "      <td>chr6</td>\n",
       "      <td>33548089</td>\n",
       "      <td>33548090</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33588142</td>\n",
       "      <td>33664351</td>\n",
       "      <td>rs3216621</td>\n",
       "      <td>BAK1</td>\n",
       "      <td>ITPR3</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33545000</td>\n",
       "      <td>33550000</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33610000</td>\n",
       "      <td>33615000</td>\n",
       "      <td>2.321081e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394</th>\n",
       "      <td>chr6</td>\n",
       "      <td>33548089</td>\n",
       "      <td>33548090</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33662070</td>\n",
       "      <td>33679504</td>\n",
       "      <td>rs3216621</td>\n",
       "      <td>BAK1</td>\n",
       "      <td>UQCC2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33545000</td>\n",
       "      <td>33550000</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33670000</td>\n",
       "      <td>33675000</td>\n",
       "      <td>1.011634e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>chr6</td>\n",
       "      <td>33548089</td>\n",
       "      <td>33548090</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33662070</td>\n",
       "      <td>33679504</td>\n",
       "      <td>rs3216621</td>\n",
       "      <td>BAK1</td>\n",
       "      <td>UQCC2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33545000</td>\n",
       "      <td>33550000</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33675000</td>\n",
       "      <td>33680000</td>\n",
       "      <td>9.701485e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>chr6</td>\n",
       "      <td>33548089</td>\n",
       "      <td>33548090</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33286335</td>\n",
       "      <td>33297046</td>\n",
       "      <td>rs3216621</td>\n",
       "      <td>BAK1</td>\n",
       "      <td>DAXX</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33290000</td>\n",
       "      <td>33295000</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33545000</td>\n",
       "      <td>33550000</td>\n",
       "      <td>4.594176e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>chr6</td>\n",
       "      <td>33548089</td>\n",
       "      <td>33548090</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33422356</td>\n",
       "      <td>33425325</td>\n",
       "      <td>rs3216621</td>\n",
       "      <td>BAK1</td>\n",
       "      <td>ZBTB9</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33420000</td>\n",
       "      <td>33425000</td>\n",
       "      <td>chr6</td>\n",
       "      <td>33545000</td>\n",
       "      <td>33550000</td>\n",
       "      <td>1.011634e-07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>398 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrS     startS       endS  chrG     startG       endG       rs_id  \\\n",
       "0    chr1  114426000  114426001  chr1  114447763  114456708  rs11102694   \n",
       "1    chr1  114426000  114426001  chr1  114466622  114472114  rs11102694   \n",
       "2    chr1  114426000  114426001  chr1  114304454  114355098  rs11102694   \n",
       "3    chr1  114426000  114426001  chr1  114356433  114414381  rs11102694   \n",
       "4    chr1  114426000  114426001  chr1  114437370  114447823  rs11102694   \n",
       "..    ...        ...        ...   ...        ...        ...         ...   \n",
       "393  chr6   33548089   33548090  chr6   33588142   33664351   rs3216621   \n",
       "394  chr6   33548089   33548090  chr6   33662070   33679504   rs3216621   \n",
       "395  chr6   33548089   33548090  chr6   33662070   33679504   rs3216621   \n",
       "396  chr6   33548089   33548090  chr6   33286335   33297046   rs3216621   \n",
       "397  chr6   33548089   33548090  chr6   33422356   33425325   rs3216621   \n",
       "\n",
       "    coloc_gene       sg_gene chrLA    startLA      endLA chrRA    startRA  \\\n",
       "0        AP4B1       DCLRE1B  chr1  114425000  114430000  chr1  114445000   \n",
       "1        AP4B1  RP5-1073O3.7  chr1  114425000  114430000  chr1  114470000   \n",
       "2        AP4B1         RSBN1  chr1  114300000  114305000  chr1  114425000   \n",
       "3        AP4B1        PTPN22  chr1  114410000  114415000  chr1  114425000   \n",
       "4        AP4B1         AP4B1  chr1  114425000  114430000  chr1  114445000   \n",
       "..         ...           ...   ...        ...        ...   ...        ...   \n",
       "393       BAK1         ITPR3  chr6   33545000   33550000  chr6   33610000   \n",
       "394       BAK1         UQCC2  chr6   33545000   33550000  chr6   33670000   \n",
       "395       BAK1         UQCC2  chr6   33545000   33550000  chr6   33675000   \n",
       "396       BAK1          DAXX  chr6   33290000   33295000  chr6   33545000   \n",
       "397       BAK1         ZBTB9  chr6   33420000   33425000  chr6   33545000   \n",
       "\n",
       "         endRA       p-value  \n",
       "0    114450000  5.290671e-07  \n",
       "1    114475000  3.431682e-07  \n",
       "2    114430000  1.011634e-07  \n",
       "3    114430000  5.684843e-07  \n",
       "4    114450000  5.290671e-07  \n",
       "..         ...           ...  \n",
       "393   33615000  2.321081e-07  \n",
       "394   33675000  1.011634e-07  \n",
       "395   33680000  9.701485e-08  \n",
       "396   33550000  4.594176e-08  \n",
       "397   33550000  1.011634e-07  \n",
       "\n",
       "[398 rows x 16 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract loops contain a snp and gene\n",
    "sgloops_both = csnps_gene_pairs_pbt.pair_to_pair(loops, type='both', **{'is':True})\n",
    "sgloops_both_df = sgloops_both.to_dataframe(disable_auto_names=True, header=None)\n",
    "\n",
    "# extract coordinate and meta data columns\n",
    "cols = list(range(0,6)) + list(range(10,19)) + [20]\n",
    "sgloops_both_df = sgloops_both_df[cols]\n",
    "\n",
    "sgloops_both_df.columns = ['chrS', 'startS', 'endS', 'chrG', 'startG', 'endG', 'rs_id', 'coloc_gene', 'sg_gene', \n",
    "                           'chrLA', 'startLA', 'endLA', 'chrRA', 'startRA', 'endRA', 'p-value']\n",
    "                           #'nameL', 'scoreL', 'strandLA', 'strandLB']\n",
    "sgloops_both_df"
   ]
  },
  {
   "cell_type": "raw",
   "id": "cae0ef2b",
   "metadata": {},
   "source": [
    "# extract loops contain a snp and gene\n",
    "sgloops_either = csnps_gene_pairs_pbt.pair_to_pair(loops, type='either', **{'is':True})\n",
    "sgloops_either_df = sgloops_either.to_dataframe(disable_auto_names=True, header=None)\n",
    "\n",
    "# extract coordinate and meta data columns\n",
    "cols = list(range(0,6)) + list(range(10,19))\n",
    "sgloops_either_df = sgloops_either_df[cols]\n",
    "sgloops_either_df.columns = ['chrS', 'startS', 'endS', 'chrG', 'startG', 'endG', 'rs_id', 'coloc_gene', 'sg_gene', \n",
    "                      'chrLA', 'startLA', 'endLA', 'chrRA', 'startRA', 'endRA', 'p-value']\n",
    "sgloops_either_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e18be3a",
   "metadata": {},
   "source": [
    "## Characterize the GSLoops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1f8d2a2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading SPP Differentially expressed genes \n",
    "spp = pd.read_excel(spp_fn, sheet_name='3. Experiment DEGs', usecols=[5])\n",
    "spp_genes = set([x[0] for x in spp.iloc[7:].values])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e2b6afef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# getting the SNP-GENE pairs which overlap a fithic loop\n",
    "sg_hichip_set = sgloops_both_df[['rs_id', 'sg_gene']].values.tolist()\n",
    "sg_hichip_set = set([tuple(x) for x in sg_hichip_set])\n",
    "\n",
    "char_data = []\n",
    "for i, sr in csnps_gene_pairs_df.iterrows():\n",
    "    \n",
    "    # check whether this snp-gene pair has a colocalized record\n",
    "    if (sr.rs_id, sr.sg_gene) in sg_coloc_set:\n",
    "        coloc_sg = 1\n",
    "    else:\n",
    "        coloc_sg = 0\n",
    "        \n",
    "    # check whether this snp-gene pair has a sgloop\n",
    "    if (sr.rs_id, sr.sg_gene) in sg_hichip_set:\n",
    "        fithichip = 1\n",
    "    else:\n",
    "        fithichip = 0\n",
    "        \n",
    "    # check whether this gene is differntially expressed according to SPP\n",
    "    if sr.sg_gene in spp_genes:\n",
    "        spp = 1\n",
    "    else:\n",
    "        spp = 0\n",
    "    \n",
    "    char_data.append([coloc_sg, fithichip, spp])\n",
    "\n",
    "char_data = pd.DataFrame(char_data)\n",
    "char_data = pd.concat([csnps_gene_pairs_df, char_data], axis=1)\n",
    "char_data.columns = bedpe_10cols + ['rs_id', 'coloc_egene', 'sg_gene', 'coloc', 'fithichip', 'spp']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3332df31",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "final_data = char_data[['rs_id', 'coloc_egene', 'sg_gene', 'coloc', 'fithichip', 'spp']]\n",
    "final_data = final_data.sort_values(['rs_id', 'coloc', 'fithichip', 'coloc_egene', 'sg_gene'], \n",
    "                                    ascending=[True, False, False, True, True])\n",
    "final_data = final_data.set_index(['rs_id', 'coloc_egene', 'sg_gene'])\n",
    "\n",
    "# remove duplicate SNP-GENE combination which are tested twice by \n",
    "# the colocalization analysis\n",
    "final_data = final_data[~final_data.reset_index().duplicated(subset=['rs_id', 'sg_gene']).values]\n",
    "final_data.to_excel(summary_fn, index=True, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d47a59da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>coloc</th>\n",
       "      <th>fithichip</th>\n",
       "      <th>spp</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs_id</th>\n",
       "      <th>coloc_egene</th>\n",
       "      <th>sg_gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">rs10085721</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">SKAP2</th>\n",
       "      <th>SKAP2</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC004079.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC004540.4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC004540.5</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC004947.2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">rs9926245</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">SULT1A2</th>\n",
       "      <th>SPNS1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TPRKBP2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TUFM</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XPO6</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snoU13</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>640 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   coloc  fithichip  spp\n",
       "rs_id      coloc_egene sg_gene                          \n",
       "rs10085721 SKAP2       SKAP2           1          0    1\n",
       "                       AC004079.1      0          0    0\n",
       "                       AC004540.4      0          0    0\n",
       "                       AC004540.5      0          0    0\n",
       "                       AC004947.2      0          0    0\n",
       "...                                  ...        ...  ...\n",
       "rs9926245  SULT1A2     SPNS1           0          0    1\n",
       "                       TPRKBP2         0          0    0\n",
       "                       TUFM            0          0    1\n",
       "                       XPO6            0          0    1\n",
       "                       snoU13          0          0    0\n",
       "\n",
       "[640 rows x 3 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50814b0a",
   "metadata": {},
   "source": [
    "## Convert all results to output and WashU visualization results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "df70a6ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bedpe_to_WashU_longrange(fn, df):\n",
    "    \"\"\"\n",
    "        Convert from a loop bedpe file into WashU longrange, \n",
    "        includes bgzip and tabix of the fn. \n",
    "        \n",
    "        Params\n",
    "        -------\n",
    "        fn: str\n",
    "            path to the longrange output file (without gz)\n",
    "            \n",
    "        df: dataframe\n",
    "            columns 1-6 are as expected and column 7 is the p or q-value. \n",
    "            \n",
    "        Output\n",
    "        ------\n",
    "        gzfn: str\n",
    "            path to the longrange with bgzip compression\n",
    "        tabix_fn: str\n",
    "            path to the index of the longrange file\n",
    "            \n",
    "    \"\"\"\n",
    "\n",
    "    # parsing the data into WashU longrage format\n",
    "    data = []\n",
    "    for sr in df.values.tolist():\n",
    "\n",
    "        # calculate the -log(FDR)\n",
    "        qval = -np.log(sr[6])\n",
    "\n",
    "        # get the first pair data\n",
    "        second_pair_str = '{}:{}-{},{:.5f}'.format(*sr[3:6], qval)\n",
    "        first_row = sr[0:3] + [second_pair_str]\n",
    "\n",
    "        # get the second pair data\n",
    "        first_pair_str = '{}:{}-{},{:.5f}'.format(*sr[0:3], qval)\n",
    "        second_row = sr[3:6] + [first_pair_str]\n",
    "\n",
    "        # add each data row\n",
    "        data.append(first_row)\n",
    "        data.append(second_row)\n",
    "\n",
    "    data = sorted(data, key=lambda x: (x[0], x[1], x[2]))\n",
    "\n",
    "    # writing out the data\n",
    "    with open(fn, 'w') as f:\n",
    "        for line in data:\n",
    "            info = [str(x) for x in line]\n",
    "            info = '\\t'.join(info)\n",
    "            f.write(info + '\\n')\n",
    "            \n",
    "    # run bgzip\n",
    "    cmd = '{} {}'.format(bgzip, fn)\n",
    "    print(cmd)\n",
    "    job = sp.Popen(cmd, stderr=sp.PIPE,stdout=sp.PIPE, shell=True)\n",
    "\n",
    "    out, err = job.communicate()\n",
    "    print('out:', out.decode())\n",
    "    print('err:', err.decode())\n",
    "    \n",
    "    # run tabix\n",
    "    lrange_gzfn = fn + '.gz'\n",
    "    cmd = '{} {}'.format(tabix, lrange_gzfn)\n",
    "    print(cmd)\n",
    "    job = sp.Popen(cmd, stderr=sp.PIPE,stdout=sp.PIPE, shell=True)\n",
    "\n",
    "    out, err = job.communicate()\n",
    "    print('out:', out.decode())\n",
    "    print('err:', err.decode())\n",
    "\n",
    "    print('Created the gzfn: {}'.format(fn + '.gz'))\n",
    "    print('Created the tabix: {}'.format(fn + '.gz.tbi'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "e169e3cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bed_WashU_bedgz(fn, df):\n",
    "    \"\"\"\n",
    "        Convert from a bed dataframe into WashU longrange file \n",
    "        includes bgzip and tabix of the fn. \n",
    "        \n",
    "        Params\n",
    "        -------\n",
    "        fn: str\n",
    "            path to the longrange output file (without gz)\n",
    "            \n",
    "        df: dataframe\n",
    "            columns 1-3 are as expected and column 7 is the p or q-value. \n",
    "            \n",
    "        Output\n",
    "        ------\n",
    "        gzfn: str\n",
    "            path to the longrange with bgzip compression\n",
    "        tabix_fn: str\n",
    "            path to the index of the longrange file\n",
    "            \n",
    "    \"\"\"\n",
    "\n",
    "    # parsing the data into WashU longrage format\n",
    "    data = []\n",
    "    for sr in df.values.tolist():\n",
    "        data.append(sr[0:4])\n",
    "    data = sorted(data, key=lambda x: (x[0], x[1], x[2]))\n",
    "\n",
    "    # writing out the data\n",
    "    with open(fn, 'w') as f:\n",
    "        for line in data:\n",
    "            info = [str(x) for x in line]\n",
    "            info = '\\t'.join(info)\n",
    "            f.write(info + '\\n')\n",
    "            \n",
    "    # run bgzip\n",
    "    cmd = '{} {}'.format(bgzip, fn)\n",
    "    print(cmd)\n",
    "    job = sp.Popen(cmd, stderr=sp.PIPE,stdout=sp.PIPE, shell=True)\n",
    "\n",
    "    out, err = job.communicate()\n",
    "    print('out:', out.decode())\n",
    "    print('err:', err.decode())\n",
    "    \n",
    "    # run tabix\n",
    "    gzfn = fn + '.gz'\n",
    "    cmd = '{} {}'.format(tabix, gzfn)\n",
    "    print(cmd)\n",
    "    job = sp.Popen(cmd, stderr=sp.PIPE,stdout=sp.PIPE, shell=True)\n",
    "\n",
    "    out, err = job.communicate()\n",
    "    print('out:', out.decode())\n",
    "    print('err:', err.decode())\n",
    "\n",
    "    print('Created the gzfn: {}'.format(fn + '.gz'))\n",
    "    print('Created the tabix: {}'.format(fn + '.gz.tbi'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c3f980c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/loop_analysis/washU/gs_pairs.longrange.bed\n",
      "out: \n",
      "err: [bgzip] results/main/loop_analysis/washU/gs_pairs.longrange.bed.gz already exists; do you wish to overwrite (y or n)? [bgzip] not overwritten\n",
      "\n",
      "/mnt/BioApps/tabix/tabix-0.2.6/tabix results/main/loop_analysis/washU/gs_pairs.longrange.bed.gz\n",
      "out: \n",
      "err: [tabix] the index file exists. Please use '-f' to overwrite.\n",
      "\n",
      "Created the gzfn: results/main/loop_analysis/washU/gs_pairs.longrange.bed.gz\n",
      "Created the tabix: results/main/loop_analysis/washU/gs_pairs.longrange.bed.gz.tbi\n"
     ]
    }
   ],
   "source": [
    "# make the longrange file\n",
    "final_sg_pairs = sgloops_both_df.iloc[:, [0,1,2,3,4,5,-1]]\n",
    "bedpe_to_WashU_longrange(sg_pairs_fn, final_sg_pairs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "f898f689",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/loop_analysis/washU/gs_loops.longrange.bed\n",
      "out: \n",
      "err: [bgzip] results/main/loop_analysis/washU/gs_loops.longrange.bed.gz already exists; do you wish to overwrite (y or n)? [bgzip] not overwritten\n",
      "\n",
      "/mnt/BioApps/tabix/tabix-0.2.6/tabix results/main/loop_analysis/washU/gs_loops.longrange.bed.gz\n",
      "out: \n",
      "err: [tabix] the index file exists. Please use '-f' to overwrite.\n",
      "\n",
      "Created the gzfn: results/main/loop_analysis/washU/gs_loops.longrange.bed.gz\n",
      "Created the tabix: results/main/loop_analysis/washU/gs_loops.longrange.bed.gz.tbi\n"
     ]
    }
   ],
   "source": [
    "# make the longrange file\n",
    "final_sg_loops = sgloops_both_df.iloc[:, [9,10,11,12,13,14,-1]]\n",
    "bedpe_to_WashU_longrange(sg_loops_fn, final_sg_loops)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "9b2250f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add the snps as elements to plot\n",
    "final_sg_snps = sgloops_both_df.iloc[:, [0,1,2,6]]\n",
    "final_sg_snps = final_sg_snps.loc[~final_sg_snps.duplicated()]\n",
    "final_sg_snps = pbt.BedTool.from_dataframe(final_sg_snps)\n",
    "final_sg_snps = final_sg_snps.slop(b=500, g=gs_fn)\n",
    "final_sg_snps = final_sg_snps.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "5289905e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/loop_analysis/washU/gs_genes.longrange.bed\n",
      "out: \n",
      "err: [bgzip] results/main/loop_analysis/washU/gs_genes.longrange.bed.gz already exists; do you wish to overwrite (y or n)? [bgzip] not overwritten\n",
      "\n",
      "/mnt/BioApps/tabix/tabix-0.2.6/tabix results/main/loop_analysis/washU/gs_genes.longrange.bed.gz\n",
      "out: \n",
      "err: [tabix] the index file exists. Please use '-f' to overwrite.\n",
      "\n",
      "Created the gzfn: results/main/loop_analysis/washU/gs_genes.longrange.bed.gz\n",
      "Created the tabix: results/main/loop_analysis/washU/gs_genes.longrange.bed.gz.tbi\n"
     ]
    }
   ],
   "source": [
    "sg_snps_fn = 'results/main/loop_analysis/washU/gs_snps.bed'\n",
    "bed_WashU_bedgz(sg_snps_fn, final_sg_snps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "cc00dd51",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_sg_genes = csnps_gene_pairs_df.iloc[:, [3,4,5,12]]\n",
    "final_sg_genes = final_sg_genes.loc[~final_sg_genes.duplicated()]\n",
    "final_sg_genes['chr'] = final_sg_genes['chrG'] \n",
    "final_sg_genes['transcript_start'] = final_sg_genes['startG']\n",
    "final_sg_genes['transcript_stop'] = final_sg_genes['endG']\n",
    "final_sg_genes['translation_start'] = final_sg_genes['startG']\n",
    "final_sg_genes['translation_stop'] = final_sg_genes['endG']\n",
    "final_sg_genes['strand'] = '+'\n",
    "final_sg_genes['gene_name'] = final_sg_genes['coloc_gene']\n",
    "final_sg_genes['transcript_id'] = final_sg_genes['coloc_gene']\n",
    "final_sg_genes['type'] = 'coding'\n",
    "final_sg_genes['exon_startG'] = final_sg_genes['startG']\n",
    "final_sg_genes['exon_stops'] = final_sg_genes['endG']\n",
    "refcols = ['chr', 'transcript_start', 'transcript_stop', 'translation_start',\n",
    "           'translation_stop', 'strand', 'gene_name', 'transcript_id',\n",
    "           'type', 'exon_startG', 'exon_stops']\n",
    "final_sg_genes = final_sg_genes.loc[:, refcols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "ab2a0ad3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/BioApps/tabix/tabix-0.2.6/bgzip results/main/loop_analysis/washU/gs_genes.longrange.bed\n",
      "out: \n",
      "err: [bgzip] results/main/loop_analysis/washU/gs_genes.longrange.bed.gz already exists; do you wish to overwrite (y or n)? [bgzip] not overwritten\n",
      "\n",
      "/mnt/BioApps/tabix/tabix-0.2.6/tabix results/main/loop_analysis/washU/gs_genes.longrange.bed.gz\n",
      "out: \n",
      "err: [tabix] the index file exists. Please use '-f' to overwrite.\n",
      "\n",
      "Created the gzfn: results/main/loop_analysis/washU/gs_genes.longrange.bed.gz\n",
      "Created the tabix: results/main/loop_analysis/washU/gs_genes.longrange.bed.gz.tbi\n"
     ]
    }
   ],
   "source": [
    "sg_genes_fn = 'results/main/loop_analysis/washU/gs_genes.refbed'\n",
    "bed_WashU_bedgz(sg_genes_fn, final_sg_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "34700c54",
   "metadata": {},
   "outputs": [],
   "source": [
    "sg_pairs_json = {'type': 'longrange', \n",
    "                 'filename': sg_pairs_fn,\n",
    "                 'name': 'SNP-Gene Pairs'}\n",
    "\n",
    "sg_loops_json = {'type': 'longrange', \n",
    "                 'filename': sg_loops_fn,\n",
    "                 'name': 'SNP-Gene Loops'}\n",
    "\n",
    "hub_json = [sg_pairs_json, sg_loops_json]\n",
    "\n",
    "hub_json_fn = os.path.join(outdir, 'hub.config.json')\n",
    "with open(hub_json_fn, 'w') as f:\n",
    "    f.write(json.dumps(hub_json, indent=4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "331f5aba",
   "metadata": {},
   "outputs": [],
   "source": [
    "loops_link = os.path.join(outdir, os.path.basename(loop_fn + '.gz'))\n",
    "os.link(loop_fn, loops_link)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "3ecc543d",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "encode() missing 1 required positional argument: 'o'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-48-ac8a3704f923>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mjson_encoder\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m: encode() missing 1 required positional argument: 'o'"
     ]
    }
   ],
   "source": [
    "json_encoder.encode()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1d5d5b2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1a4a34b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "180f5a2f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78e21eaf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a9c70e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "[\n",
    "    {\n",
    "    \"type\": \"bigwig\",\n",
    "    \"url\": \"https://vizhub.wustl.edu/public/tmp/TW463_20-5-bonemarrow_MeDIP.bigWig\",\n",
    "    \"name\": \"MeDIP\",\n",
    "    \"options\": {\n",
    "        \"color\": \"red\",\n",
    "        \"backgroundColor\":\"#FFE7AB\"\n",
    "        },\n",
    "    \"metadata\": {\n",
    "        \"sample\": \"bone\",\n",
    "        \"assay\": \"MeDIP\"\n",
    "        }\n",
    "    },\n",
    "    {\n",
    "    \"type\": \"bigwig\",\n",
    "    \"url\": \"https://vizhub.wustl.edu/public/tmp/TW551_20-5-bonemarrow_MRE.CpG.bigWig\",\n",
    "    \"name\": \"MRE\",\n",
    "    \"options\": {\n",
    "        \"color\": \"blue\",\n",
    "        \"backgroundColor\":\"#C0E3CC\"\n",
    "        },\n",
    "    \"metadata\": {\n",
    "        \"sample\": \"bone\",\n",
    "        \"assay\": \"MRE\"\n",
    "        }\n",
    "    }\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "f702192c",
   "metadata": {},
   "outputs": [],
   "source": [
    "s = 'chr, transcript_start, transcript_stop, translation_start, translation_stop, strand, gene_name, transcript_id, type, exon(including UTR bases) starts, exon(including UTR bases) stops, and additional gene info (optional)'.split(',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "e56ae64e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chr',\n",
       " ' transcript_start',\n",
       " ' transcript_stop',\n",
       " ' translation_start',\n",
       " ' translation_stop',\n",
       " ' strand',\n",
       " ' gene_name',\n",
       " ' transcript_id',\n",
       " ' type',\n",
       " ' exon(including UTR bases) starts',\n",
       " ' exon(including UTR bases) stops',\n",
       " ' and additional gene info (optional)']"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41cac261",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "7df751ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c4875d1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e27e541",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3096dd6d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "281e3ae1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9960a179",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "815a09e1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "HiC",
   "language": "python",
   "name": "hic_tls"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
