{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "7b5ad180",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)\n",
    "import subprocess\n",
    "import glob\n",
    "import pybedtools as pbt \n",
    "pbt.set_bedtools_path('/mnt/BioApps/bedtools/bin/')\n",
    "import numpy as np\n",
    "\n",
    "import seaborn as sns \n",
    "import matplotlib.pyplot as plt \n",
    "\n",
    "os.chdir('/mnt/BioHome/jreyna/jreyna-temp/projects/dchallenge/')\n",
    "\n",
    "genome_sizes = 'results/refs/hg19/hg19.chrom.sizes'\n",
    "\n",
    "# make the directory to save our data\n",
    "outdir = 'results/main/pieqtls/sgl_intersect/'\n",
    "os.makedirs(outdir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "c91b4f31",
   "metadata": {},
   "outputs": [],
   "source": [
    "from chromolooper import sgls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a6ef0439",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_fn = 'results/refs/hg19/hg19.chrom.nochr.sizes'\n",
    "gencode_fn = 'results/refs/gencode/v30/gencode.v30.annotation.bed'\n",
    "gencode = pd.read_table(gencode_fn, header=None)\n",
    "gencode = gencode.drop_duplicates(5)\n",
    "gencode_dict = {k:v for k,v in gencode[[5,6]].values.tolist()} \n",
    "gencode_dict.update({v:k for k,v in gencode[[5,6]].values.tolist()})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25ec296e",
   "metadata": {},
   "source": [
    "## Load Significant GWAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "44c1967f",
   "metadata": {},
   "outputs": [],
   "source": [
    "major_gwas = ['T1D_32005708', 'T1D_34594039_GCST90018925', 'T1D_34012112_Gaulton']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "3be9cb20",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_57681/915015123.py:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gwas_bed['start'] = gwas_bed['start'] - 1\n"
     ]
    }
   ],
   "source": [
    "data = []\n",
    "gwas = 'results/main/coloc/Data/T1D_GWAS/*/GRCh37/GWAS_input_colocalization_pval_lt_5eMinus8.txt'\n",
    "for fn in glob.glob(gwas):\n",
    "    path_info = fn.split('/')\n",
    "    df = pd.read_table(fn)\n",
    "    \n",
    "    if path_info[5] not in major_gwas:\n",
    "        continue\n",
    "        \n",
    "    df['gwas_source'] = path_info[5]\n",
    "    data.append(df)    \n",
    "gwas_df = pd.concat(data)\n",
    "gwas_df['sid'] = gwas_df['CHR'].str.replace('chr', '') + ':' + gwas_df['POS'].astype(str)\n",
    "\n",
    "# loading finemap data into bedtools\n",
    "gwas_bed = gwas_df.iloc[:, [0,1,1]]\n",
    "gwas_bed.columns = ['chrom', 'start', 'end']\n",
    "gwas_bed['start'] = gwas_bed['start'] - 1\n",
    "gwas_pbt = pbt.BedTool.from_dataframe(gwas_bed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f764c98f",
   "metadata": {},
   "outputs": [],
   "source": [
    "individual_gwas_studies = gwas_df.groupby('gwas_source').sid.nunique()\n",
    "individual_gwas_studies = individual_gwas_studies.to_frame()\n",
    "individual_gwas_studies.columns = ['Number of Sig. GWAS']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "49f62c4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Number of Sig. GWAS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gwas_source</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>T1D_32005708</th>\n",
       "      <td>21097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34012112_Gaulton</th>\n",
       "      <td>45798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34594039_GCST90018925</th>\n",
       "      <td>20419</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Number of Sig. GWAS\n",
       "gwas_source                                   \n",
       "T1D_32005708                             21097\n",
       "T1D_34012112_Gaulton                     45798\n",
       "T1D_34594039_GCST90018925                20419"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "individual_gwas_studies"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b981dcdf",
   "metadata": {},
   "source": [
    "## Load pieQTLs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "6cd23b57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "results/main/pieqtls/2021_chandra_et_al/CD8_T-cell_naive/proximal.pieqtls.tsv\n",
      "results/main/pieqtls/2021_chandra_et_al/CD4_T-cell_naive/proximal.pieqtls.tsv\n",
      "results/main/pieqtls/2021_chandra_et_al/NK-cell_naive/proximal.pieqtls.tsv\n",
      "results/main/pieqtls/2021_chandra_et_al/monocyte_naive/proximal.pieqtls.tsv\n",
      "results/main/pieqtls/2021_chandra_et_al/B-cell_naive/proximal.pieqtls.tsv\n"
     ]
    }
   ],
   "source": [
    "pieqtls = glob.glob('results/main/pieqtls/2021_chandra_et_al/*/proximal.pieqtls.tsv')\n",
    "\n",
    "pieqtl_data = []\n",
    "for pieqtl in pieqtls:\n",
    "    print(pieqtl)\n",
    "    \n",
    "    cline = pieqtl.split('/')[-2]\n",
    "    df = pd.read_table(pieqtl)    \n",
    "    df['cline'] = cline\n",
    "    pieqtl_data.append(df) \n",
    "    \n",
    "pieqtl_df = pd.concat(pieqtl_data)\n",
    "pieqtl_df['Chromosome'] = pieqtl_df['Chromosome']\n",
    "pieqtl_df.loc[:, 'sid'] = pieqtl_df['Chromosome'].str.replace('chr', '') + ':' + \\\n",
    "                          pieqtl_df['pieQTL.Position'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "458bce03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Number of pieQTL Genes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cline</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B-cell_naive</th>\n",
       "      <td>1233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4_T-cell_naive</th>\n",
       "      <td>1168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8_T-cell_naive</th>\n",
       "      <td>1081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK-cell_naive</th>\n",
       "      <td>872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>monocyte_naive</th>\n",
       "      <td>1065</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Number of pieQTL Genes\n",
       "cline                                   \n",
       "B-cell_naive                        1233\n",
       "CD4_T-cell_naive                    1168\n",
       "CD8_T-cell_naive                    1081\n",
       "NK-cell_naive                        872\n",
       "monocyte_naive                      1065"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "individual_cells = pieqtl_df.groupby('cline').Target_geneID.nunique()\n",
    "individual_cells = individual_cells.to_frame()\n",
    "individual_cells.columns = ['Number of pieQTL Genes']\n",
    "individual_cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e781c384",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Number of pieQTL SNPs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cline</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B-cell_naive</th>\n",
       "      <td>8110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4_T-cell_naive</th>\n",
       "      <td>6800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8_T-cell_naive</th>\n",
       "      <td>5539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK-cell_naive</th>\n",
       "      <td>5332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>monocyte_naive</th>\n",
       "      <td>6006</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Number of pieQTL SNPs\n",
       "cline                                  \n",
       "B-cell_naive                       8110\n",
       "CD4_T-cell_naive                   6800\n",
       "CD8_T-cell_naive                   5539\n",
       "NK-cell_naive                      5332\n",
       "monocyte_naive                     6006"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "individual_cells = pieqtl_df.groupby('cline').sid.nunique()\n",
    "individual_cells = individual_cells.to_frame()\n",
    "individual_cells.columns = ['Number of pieQTL SNPs']\n",
    "individual_cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "4fc67bca",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_57681/2754423286.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  pieqtl_bed.loc[:, 'start'] = pieqtl_bed['start'] - 1\n"
     ]
    }
   ],
   "source": [
    "pieqtl_bed = pieqtl_df.iloc[:, [1,2,2,-1]]\n",
    "pieqtl_bed.columns = ['chrom', 'start', 'end', 'cline']\n",
    "pieqtl_bed.loc[:, 'start'] = pieqtl_bed['start'] - 1\n",
    "pieqtl_pbt = pbt.BedTool.from_dataframe(pieqtl_bed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "647a3492",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(35898, 4)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pieqtl_bed.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4b1e34a",
   "metadata": {},
   "source": [
    "## Intersect pieQTLs and Sig GWAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a27c887c",
   "metadata": {},
   "outputs": [],
   "source": [
    "major_cols = ['ge_source',\n",
    "             'rsid',\n",
    "             'chrom',\n",
    "             'pos',\n",
    "             'geneid',\n",
    "             'genename',\n",
    "             'allele1',\n",
    "             'allele2',\n",
    "             'maf',\n",
    "             'beta_x',\n",
    "             'se',\n",
    "             'z',\n",
    "             'prob',\n",
    "             'log10bf',\n",
    "             'mean',\n",
    "             'sd',\n",
    "             'mean_incl',\n",
    "             'sd_incl',\n",
    "             'pval',\n",
    "             'gwas_source',\n",
    "             'pieQTL.ID',\n",
    "             'Chromosome',\n",
    "             'pieQTL.Position',\n",
    "             'Target_geneName',\n",
    "             'TSS',\n",
    "             'pvalue',\n",
    "             'FDR (DICE)',\n",
    "             'beta_y',\n",
    "             'ref',\n",
    "             'alt',\n",
    "             'Mean.TPM.Homozygous.Reference',\n",
    "             'Mean.TPM.Heterozygous',\n",
    "             'Mean.TPM.Homozygous.Alternative',\n",
    "             'Interaction_type',\n",
    "             'GWAS.Trait',\n",
    "             'cline',\n",
    "             'regionID',\n",
    "             'GWASLoci',\n",
    "             'index']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1f24411",
   "metadata": {},
   "source": [
    "\n",
    "### Intersection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "8ee87b2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "intersect_df = pd.merge(gwas_df, pieqtl_df,\n",
    "                        left_on=['CHR', 'POS'],\n",
    "                        right_on=['Chromosome', 'pieQTL.Position'],\n",
    "                        how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d9864a74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Number of Unique Genes from pieQTL &amp; Sig.GWAS Intersection</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cline</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B-cell_naive</th>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4_T-cell_naive</th>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8_T-cell_naive</th>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK-cell_naive</th>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>monocyte_naive</th>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Number of Unique Genes from pieQTL & Sig.GWAS Intersection\n",
       "cline                                                                       \n",
       "B-cell_naive                                                     39         \n",
       "CD4_T-cell_naive                                                 33         \n",
       "CD8_T-cell_naive                                                 35         \n",
       "NK-cell_naive                                                    45         \n",
       "monocyte_naive                                                   25         "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indiv_cell_intersect = intersect_df.groupby('cline').Target_geneID.nunique().to_frame()\n",
    "indiv_cell_intersect.columns = ['Number of Unique Genes from pieQTL & Sig.GWAS Intersection']\n",
    "indiv_cell_intersect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "6c069f0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Number of Unique SNPs from pieQTL &amp; Sig.GWAS Intersection</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cline</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B-cell_naive</th>\n",
       "      <td>348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4_T-cell_naive</th>\n",
       "      <td>322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8_T-cell_naive</th>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK-cell_naive</th>\n",
       "      <td>319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>monocyte_naive</th>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Number of Unique SNPs from pieQTL & Sig.GWAS Intersection\n",
       "cline                                                                      \n",
       "B-cell_naive                                                    348        \n",
       "CD4_T-cell_naive                                                322        \n",
       "CD8_T-cell_naive                                                269        \n",
       "NK-cell_naive                                                   319        \n",
       "monocyte_naive                                                  116        "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indiv_cell_intersect = intersect_df.groupby('cline').sid_x.nunique().to_frame()\n",
    "indiv_cell_intersect.columns = ['Number of Unique SNPs from pieQTL & Sig.GWAS Intersection']\n",
    "indiv_cell_intersect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "6c82bb6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Number of Unique Genes from pieQTL &amp; Sig.GWAS Intersection</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cline</th>\n",
       "      <th>gwas_source</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">B-cell_naive</th>\n",
       "      <th>T1D_32005708</th>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34012112_Gaulton</th>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34594039_GCST90018925</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">CD4_T-cell_naive</th>\n",
       "      <th>T1D_32005708</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34012112_Gaulton</th>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34594039_GCST90018925</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">CD8_T-cell_naive</th>\n",
       "      <th>T1D_32005708</th>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34012112_Gaulton</th>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34594039_GCST90018925</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">NK-cell_naive</th>\n",
       "      <th>T1D_32005708</th>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34012112_Gaulton</th>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34594039_GCST90018925</th>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">monocyte_naive</th>\n",
       "      <th>T1D_32005708</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34012112_Gaulton</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34594039_GCST90018925</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            Number of Unique Genes from pieQTL & Sig.GWAS Intersection\n",
       "cline            gwas_source                                                                          \n",
       "B-cell_naive     T1D_32005708                                                              28         \n",
       "                 T1D_34012112_Gaulton                                                      34         \n",
       "                 T1D_34594039_GCST90018925                                                 21         \n",
       "CD4_T-cell_naive T1D_32005708                                                              15         \n",
       "                 T1D_34012112_Gaulton                                                      29         \n",
       "                 T1D_34594039_GCST90018925                                                 14         \n",
       "CD8_T-cell_naive T1D_32005708                                                              18         \n",
       "                 T1D_34012112_Gaulton                                                      30         \n",
       "                 T1D_34594039_GCST90018925                                                 15         \n",
       "NK-cell_naive    T1D_32005708                                                              34         \n",
       "                 T1D_34012112_Gaulton                                                      36         \n",
       "                 T1D_34594039_GCST90018925                                                 17         \n",
       "monocyte_naive   T1D_32005708                                                              14         \n",
       "                 T1D_34012112_Gaulton                                                      22         \n",
       "                 T1D_34594039_GCST90018925                                                  8         "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indiv_cell_intersect = intersect_df.groupby(['cline', 'gwas_source']).Target_geneID.nunique().to_frame()\n",
    "indiv_cell_intersect.columns = ['Number of Unique Genes from pieQTL & Sig.GWAS Intersection']\n",
    "indiv_cell_intersect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "2c9fe440",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Number of Unique SNPs from pieQTL &amp; Sig.GWAS Intersection</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cline</th>\n",
       "      <th>gwas_source</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">B-cell_naive</th>\n",
       "      <th>T1D_32005708</th>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34012112_Gaulton</th>\n",
       "      <td>241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34594039_GCST90018925</th>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">CD4_T-cell_naive</th>\n",
       "      <th>T1D_32005708</th>\n",
       "      <td>142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34012112_Gaulton</th>\n",
       "      <td>198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34594039_GCST90018925</th>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">CD8_T-cell_naive</th>\n",
       "      <th>T1D_32005708</th>\n",
       "      <td>136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34012112_Gaulton</th>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34594039_GCST90018925</th>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">NK-cell_naive</th>\n",
       "      <th>T1D_32005708</th>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34012112_Gaulton</th>\n",
       "      <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34594039_GCST90018925</th>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">monocyte_naive</th>\n",
       "      <th>T1D_32005708</th>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34012112_Gaulton</th>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34594039_GCST90018925</th>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            Number of Unique SNPs from pieQTL & Sig.GWAS Intersection\n",
       "cline            gwas_source                                                                         \n",
       "B-cell_naive     T1D_32005708                                                             186        \n",
       "                 T1D_34012112_Gaulton                                                     241        \n",
       "                 T1D_34594039_GCST90018925                                                116        \n",
       "CD4_T-cell_naive T1D_32005708                                                             142        \n",
       "                 T1D_34012112_Gaulton                                                     198        \n",
       "                 T1D_34594039_GCST90018925                                                 53        \n",
       "CD8_T-cell_naive T1D_32005708                                                             136        \n",
       "                 T1D_34012112_Gaulton                                                     155        \n",
       "                 T1D_34594039_GCST90018925                                                 73        \n",
       "NK-cell_naive    T1D_32005708                                                             171        \n",
       "                 T1D_34012112_Gaulton                                                     175        \n",
       "                 T1D_34594039_GCST90018925                                                 77        \n",
       "monocyte_naive   T1D_32005708                                                              82        \n",
       "                 T1D_34012112_Gaulton                                                      76        \n",
       "                 T1D_34594039_GCST90018925                                                 33        "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indiv_cell_intersect = intersect_df.groupby(['cline', 'gwas_source']).sid_x.nunique().to_frame()\n",
    "indiv_cell_intersect.columns = ['Number of Unique SNPs from pieQTL & Sig.GWAS Intersection']\n",
    "indiv_cell_intersect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "25e15fc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genename(x):\n",
    "    if x in gencode_dict:\n",
    "        return(gencode_dict[x])\n",
    "    else:\n",
    "        return(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "7a0895a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "intersect_df.loc[:, 'geneid'] = intersect_df['Target_geneID'].str.replace('\\.[0-9]*', '', regex=True)\n",
    "intersect_df.loc[:, 'genename'] = intersect_df.loc[:, 'geneid'].apply(get_genename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "d18ad120",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "91"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersect_df.loc[:, 'geneid'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "8fdfadf2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCG27\n",
      "ZSCAN9\n",
      "ZSCAN26\n",
      "TRIM27\n",
      "ENSG00000261353\n",
      "BTN2A3P\n",
      "TAP2\n",
      "RPL32P1\n",
      "WDR46\n",
      "AL645929.1\n",
      "BTNL2\n",
      "TSBP1\n",
      "ZNRD1ASP\n",
      "ZKSCAN4\n",
      "BTN2A1\n",
      "BTN2A2\n",
      "AL645933.2\n",
      "LY6G5B\n",
      "BTN3A1\n",
      "TCF19\n",
      "SUOX\n",
      "MICA\n",
      "CCHCR1\n",
      "LY6G5C\n",
      "IFITM4P\n",
      "BTN3A2\n",
      "AL662844.3\n",
      "AL662844.4\n",
      "HMGN4\n",
      "LST1\n",
      "PTPN22\n",
      "TMEM116\n",
      "AL645939.1\n",
      "ZFP57\n",
      "SKIV2L\n",
      "DHX16\n",
      "FLOT1\n",
      "PPP1R18\n",
      "ZNF165\n",
      "RF00019\n",
      "TRIM26\n",
      "HCG18\n",
      "MICB\n",
      "PRRC2A\n",
      "BAG6\n",
      "CSNK2B\n",
      "C6orf47\n",
      "AL671277.2\n",
      "HCG24\n",
      "ENSG00000228962\n",
      "ZDHHC20P1\n",
      "GNL1\n",
      "ZNF192P1\n",
      "ZSCAN23\n",
      "HECTD4\n",
      "HCG4B\n",
      "RNF39\n",
      "PFDN6\n",
      "BACH2\n",
      "C1orf216\n",
      "FHL3\n",
      "SF3A3\n",
      "RGS1\n",
      "NPM1P33\n",
      "KRT18P39\n",
      "CD28\n",
      "SLC22A5\n",
      "VARS2\n",
      "MRPS18B\n",
      "TUBB\n",
      "C6orf136\n",
      "NCR3\n",
      "C6orf48\n",
      "AIF1\n",
      "MSH5\n",
      "PPT2\n",
      "AP003774.4\n",
      "M6PR\n",
      "LINC02390\n",
      "ATXN2\n",
      "MAPKAPK5\n",
      "CTSH\n",
      "AC009121.1\n",
      "RMI2\n",
      "DND1P1\n",
      "KANSL1-AS1\n",
      "AP005482.1\n",
      "ZGLP1\n",
      "PRKD2\n",
      "TMPRSS3\n"
     ]
    }
   ],
   "source": [
    "for x in intersect_df.genename.unique():\n",
    "    print(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6fdf11c",
   "metadata": {},
   "source": [
    "## Summarize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "94f35bf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find the unique SNPs\n",
    "uniq_snps_by_cells = intersect_df.groupby('cline').sid_x.nunique()\n",
    "uniq_snps_by_cells = uniq_snps_by_cells.to_frame()\n",
    "\n",
    "# find the unique genes \n",
    "uniq_genes_by_cells = intersect_df.groupby('cline').geneid.nunique()\n",
    "uniq_genes_by_cells = uniq_genes_by_cells.to_frame()\n",
    "\n",
    "# merge snps and genes\n",
    "uniq_counts_by_cells = pd.merge(uniq_snps_by_cells, uniq_genes_by_cells, left_index=True, right_index=True)\n",
    "uniq_counts_by_cells.columns = ['Number of Unique SNPs', 'Number of Unique Genes']\n",
    "uniq_counts_by_cells.index.name = 'Cell Line'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "85a82a29",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the file\n",
    "excel_analysis = os.path.join(outdir, 'Unique_Counts_By_Cell_Line.xlsx')\n",
    "uniq_counts_by_cells.to_excel(excel_analysis, sheet_name='pieqtls')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "1607399d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cline</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B-cell_naive</th>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4_T-cell_naive</th>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8_T-cell_naive</th>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK-cell_naive</th>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>monocyte_naive</th>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  geneid\n",
       "cline                   \n",
       "B-cell_naive          39\n",
       "CD4_T-cell_naive      33\n",
       "CD8_T-cell_naive      35\n",
       "NK-cell_naive         45\n",
       "monocyte_naive        25"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniq_genes_by_cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "5d735083",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'results/main/pieqtls/sgl_intersect/Unique_Counts_By_Cell_Line.xlsx'"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "excel_analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0785eaf1",
   "metadata": {},
   "source": [
    "#### Write the gene list as well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "619449e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "gh_list = sorted(intersect_df.geneid.unique().tolist())\n",
    "gh_fn = os.path.join(outdir, 'gene_list.txt')\n",
    "with open(gh_fn, 'w') as fw:\n",
    "    for x in gh_list:\n",
    "        fw.write('{}\\n'.format(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "78fca0e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "gh_list = sorted(intersect_df.geneid.unique().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "b1a5a12a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "91"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gh_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "8833dd03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cline</th>\n",
       "      <th>geneid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>ENSG00000206344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NK-cell_naive</td>\n",
       "      <td>ENSG00000137185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B-cell_naive</td>\n",
       "      <td>ENSG00000197062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>ENSG00000204713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>ENSG00000261353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2137</th>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>ENSG00000160183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2138</th>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "      <td>ENSG00000160183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2139</th>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>ENSG00000160183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2140</th>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "      <td>ENSG00000160183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2141</th>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>ENSG00000160183</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2142 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 cline           geneid\n",
       "0       monocyte_naive  ENSG00000206344\n",
       "1        NK-cell_naive  ENSG00000137185\n",
       "2         B-cell_naive  ENSG00000197062\n",
       "3       monocyte_naive  ENSG00000204713\n",
       "4     CD4_T-cell_naive  ENSG00000261353\n",
       "...                ...              ...\n",
       "2137  CD4_T-cell_naive  ENSG00000160183\n",
       "2138  CD8_T-cell_naive  ENSG00000160183\n",
       "2139  CD4_T-cell_naive  ENSG00000160183\n",
       "2140  CD8_T-cell_naive  ENSG00000160183\n",
       "2141  CD4_T-cell_naive  ENSG00000160183\n",
       "\n",
       "[2142 rows x 2 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersect_df[['cline', 'geneid']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67b9a579",
   "metadata": {},
   "source": [
    "#### Get the unique genes per cell type "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "e832fb53",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_by_cell = intersect_df[['cline', 'geneid']].drop_duplicates()\n",
    "genes_by_cell.sort_values(['cline', 'geneid'], inplace=True)\n",
    "genes_by_cell['source'] = 'pieqtls_with_sig_gwas'\n",
    "genes_by_cell.columns = ['cline', 'geneid', 'source']\n",
    "fn = os.path.join(outdir, 'genes_by_cell.xlsx')\n",
    "genes_by_cell.to_excel(fn, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "993c4790",
   "metadata": {},
   "source": [
    "#### Make a figure of uniq SNPs and Genes per cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "fec9e4e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "color_fn = 'results/refs/dice_colors/colors.revised.t1d-project.tsv'\n",
    "colors = pd.read_table(color_fn)\n",
    "cline_pubnames = {'CD4_T-cell_naive': 'CD4 T-cells', 'CD8_T-cell_naive': 'CD8 T-cells',\n",
    "                  'monocyte_naive': 'Monocytes', 'B-cell_naive': 'B-cells', 'NK-cell_naive': 'NK cells'}\n",
    "colors_dict = {sr.studyname: sr.hex for i, sr in colors.iterrows()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "469709d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "uniq_counts_by_cells['pubnames'] = uniq_counts_by_cells.index.map(cline_pubnames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "74806b2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "uniq_counts_by_cells.sort_values('pubnames', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "c31fe385",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Number of Unique SNPs</th>\n",
       "      <th>Number of Unique Genes</th>\n",
       "      <th>pubnames</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell Line</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B-cell_naive</th>\n",
       "      <td>348</td>\n",
       "      <td>39</td>\n",
       "      <td>B-cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4_T-cell_naive</th>\n",
       "      <td>322</td>\n",
       "      <td>33</td>\n",
       "      <td>CD4 T-cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8_T-cell_naive</th>\n",
       "      <td>269</td>\n",
       "      <td>35</td>\n",
       "      <td>CD8 T-cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>monocyte_naive</th>\n",
       "      <td>116</td>\n",
       "      <td>25</td>\n",
       "      <td>Monocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK-cell_naive</th>\n",
       "      <td>319</td>\n",
       "      <td>45</td>\n",
       "      <td>NK cells</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Number of Unique SNPs  Number of Unique Genes     pubnames\n",
       "Cell Line                                                                   \n",
       "B-cell_naive                        348                      39      B-cells\n",
       "CD4_T-cell_naive                    322                      33  CD4 T-cells\n",
       "CD8_T-cell_naive                    269                      35  CD8 T-cells\n",
       "monocyte_naive                      116                      25    Monocytes\n",
       "NK-cell_naive                       319                      45     NK cells"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniq_counts_by_cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "3862c6d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAP8AAADtCAYAAABwFqNaAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAA0pElEQVR4nO2deXhV1dW435U5JCAGmechUEUUCINCFXGoFqyidaxTrf3wa52r1dI64Dzj9LNarVYRh+KMVXHGfohACYqClEESZEYGMQmZs35/7H3lEpObc3Pne/f7POe5ueeec9ZKctc5e6299lqiqjgcjtQjLdYKOByO2OCM3+FIUZzxOxwpijN+hyNFccbvcKQozvgdjhQlI9YKeCEtLU1zc3NjrYbD8SN2796tqpqQD9GEMP7c3FwqKipirYbD8SNEpDLWOrSWhLxjORyJgojkiMhCEVkiIstE5Ea7f6qIbBCRz+02we+cKSKyWkRWiMixkdItIZ78DkcCUw0cqarlIpIJzBWRt+1n96nqPf4Hi8gBwBnAYKAb8L6IDFTV+nAr5p78DkcEUUO5fZtpt0A59ScCL6hqtaqWAKuBUZHQzRm/wxFhRCRdRD4HtgLvqeoC+9HFIvKFiDwpIvvafd2BdX6nr7f7wk7KGf+CBQu4++672bRpU6xVcSQHGSKyyG+b3PgAVa1X1aFAD2CUiBwIPAL0B4YCm4B77eHShIyIrL5LOZ9/+/btLFiwgLlz5/Kzn/2M008/nf322y/WajlaQem/+0ZFTp/DSwJ9XKeqI7xcR1W/E5E5wHH+vr6IPA78y75dD/T0O60HsDEohT2Sck/+CRMm8PjjjzNhwgTef/99Jk+ezBNPPMH3338fa9UcSYiIdBSR9vbnXOBo4L8i0tXvsJOApfbnWcAZIpItIn2BQmBhRHRLhPX8eXl5Gol5/q1bt/Lcc8/x4Ycfkp2dzUknncSkSZNo06ZN2GU5wk88PPlFZLeq5gX4/CDgaSAd87Cdqao3icgzmCG/AqXAhaq6yZ7zF+A3QB1wuaq+3cSlQyaljd/HunXrmDFjBvPmzaNt27aceuqpTJw4kaysrIjJdIROIhh/POOM349Vq1YxY8YMFi9eTEFBAWeeeSZHH300GRkpFxpJCJzxh4Yz/iZYunQp06dPZ/ny5XTp0oWzzjqLww8/nLS0lAuRxDXO+EPDGX8zqCrFxcVMnz6dkpISevfuzTnnnMOoUaMQaWo2xhFtnPGHhjP+FlBV5s6dy4wZM9i4cSODBg3i3HPP5aCDDoqJPo49OOMPDWf8Hqmvr+fDDz/kueeeY9u2bRx88MGcc845DBo0KKZ6pTLO+EPDGX+Q1NTUMHv2bGbOnMmuXbsYPXo055xzDr179461aimHM/7QcMbfSiorK5k1axavvPIKlZWVHHHEEZx55pl07dq15ZMdYcEZf2g44w+RsrIyXnnlFWbNmkV9ff0PKcMdOnSItWpJjzP+0HDGHyZ27NjBP//5T9555x3S09OZOHEip5xyCu3atYu1akmLM/7QaNH4RWQs8LmqVojI2cBw4AFVXRsNBSExjN/H5s2bef755/noo4/Iycn5IWXY1SAMP874Q8OL8X8BHAwcBDwDPAGcrKrjPAkQSQcWARtU9XgRKQD+CfTB5DSfpqo7A10jkYzfxzfffMOzzz7LvHnzaNeuHaeeeioTJkxwKcNhxBl/aHhJWatTc4c4EfPEfwBoG4SMy4Dlfu//BHygqoXAB/Z90tGrVy+mTJnCtGnT6N+/P0888QSTJ09m9uzZ1NXVxVo9h8OT8ZeJyBTgHOBN+yTP9HJxEekBTAT+7rf7RMwqJ+zrJM/aJiCFhYXcdNNN3HbbbXTs2JGHH36Ya6+9lurq6lir5khwROQyEWknhidEZLGI/Mzr+V6M/3RMEcLfqOpmTEmhuz1e/37gaqDBb19n39JF+9rJq7KJzJAhQ7jrrru4/PLL+eqrr7j99tvdCCDKbN1eR/nuhpYPTBx+o6rfAz8DOgLnA3d4PblF47cG/zKQbXdtA15t6TwROR7YqqrFXpVpdP5kX2mkZDESEeGoo47ikksuobi4mHvvvZeGhqT6MsY1/3yznNv+uiPWaoQT3yKTCcA/VHWJ374WaXGtqoj8DzAZKMDUHOsOPAoc1cKpY4ETbD3yHKCdiMwAtohIV1XdZKuZbG3qZFV9DHgMTMDP4++TEBxzzDGUl5fz5JNP0qZNGy6++GK3WCjCqCqrSmoZNji75YMTh2IReRfoC0wRkbbsPcoOiJdh/0UYQ/4eQFVX4WGorqpTVLWHqvbB1CH/UFXPxpQpOs8edh7wuldlk4mTTjqJ008/nXfffZd//OMfJEK+RSKzYXMdFZUNDOrnKVyVKPwGEzAfqaq7gSzM0N8TXqpUVKtqje/JJCIZhFZN9A5gpohcAHwDnBrCtQD42XueYxyt5t1j3g37Nc866ywqKip49dVXyc/P57TTTgu7DIfhv2tqARjYN/GnWkWkELgHMxL/ErgK+E5VtwPbvV7Hi/F/LCJ/BnJF5Bjg98AbwSirqnOAOfbn7bTsMqQEIsLkyZOpqKjgmWeeIS8vj4kTJ8ZaraRkVWkN++2bzn77psdalXDwJDAd+DdwAvAQcHKwF/Fi/H8CLsDcYS4E3mLvqTtHCIgIl112Gbt37+bRRx+lTZs2jB8/PtZqJRWqyoqSWg4sTPynvqWtqj5uf75bRBa35iItGr+qNgCP280RAdLT07n66qu58cYbuf/++2nTpg2jR4+OtVpJw6Zv6ykrb2BQv6Qx/hwRGcaeyH6u/3tV9XQzaDHgJyJjReQ9EVkpImtEpERE1rRabUeTZGVlce211zJgwADuvPNOvvjii1irlDSsKqkBYGDfpAn2bQamYbr83Nvo/T0BztsLL9H+J+yFfwqMBEbYV0eYyc3NZerUqXTr1o2bb76ZFStWxFqlpOC/a2pp3y6NTh2Swt9HVY9Q1fHNbEd6vY4X49+lqm+r6lZV3e7bQtDdEYC2bdty00030b59e6ZOncratVFbPJmUqCqrSmso7JMVk1wKEckRkYUiskRElonIjXZ/gR1Rr7Kv+/qdM0VEVovIChE5tolrHh5gO8yrbl6M/yMRuVtEDhWR4b7NqwBH8BQUFHDLLbeQlZXF9ddf75qKhsC3O+rZuSum8/vVwJGqejCmQ89xInIIzSxwE5EDMHkxg4HjgL/a9TT+/LGJ7SrMqts5XhXzYvyjMUP929jjY3j2Kxyto3Pnztxyyy3U1tZy3XXXsX27G2y1hpUlZn4/VsE+NZTbt5l2862SbWqB24nAC6paraolwGpgVKNr/sJ/A+60191EEAvlvOT2h+RXOFpPz549ufHGG9m1axfXXXcdZWVlsVYp4VhZUkPbvDS6doydvy8i6SLyOSaV/T1VXUDzC9y6A+v8Tl9v9zV13aNs19+bgWmqeoiqes7B8RLt72yXC75t3x9gs/McUaCwsJDrr7+ezZs3c8MNN7B79+5Yq5RQrCipZWDfzEj6+xm+BWh2m9z4AFWtV9WhmHbbo0TkwADXa0rRvTJqRWSiiMzDDPX/Yh/I7wWruJdh/1PAO0A3+34lcHmwghytZ8iQIUyZMoU1a9Zw8803U1NTE2uVEoId39WzbUc9hZFN6a1T1RF+22PNHaiq32F88uOwC9wAGi1wWw/09DutB7Cx0aXesPvrgGtEZJb/5lVxL8a/n6rOxK4WUtU6oN6rAEd4GDlyJFdccQXLli3jjjvucLUAPLBijblJDorh/L6IdBSR9vbnXOBo4L80v8BtFnCGiGSLSF+gEFjY6LLjMcV17mFPHM5/84SX9N4KEemAHXrYSOUurwIc4WPcuHFUVFTwyCOPcP/993PllVe6pcABWFlaS5tcoUeXmHZZ7go8bSP2acBMVf2XiHxKEwvcVHWZiMwEvsI82S9S1b0etqr6sf97EckEDsTUyWxyiXxTePmr/AFzN+ovIp9gKoac4lWAI7xMmDCB3bt38/TTT5OXl8eE6gkRvwH0uqxXRK8fKVaW1FDYO4u0tNjdIFX1C2BYE/ubXeCmqrcCtzZ3TRF5FHjI3ij2AT7FjMYLROQqVX3ei25ecvsXi8g4YBAmGLFCVWu9XNwRGU455RTKy8t5+eWXqWpXxS9H/TLWKsUdu8rq2fxtPYePTMqS6Yep6v/an88HVqrqJBHpArwNhGb8ItIOMx2xSlXrbPJBLjBcRN5R1S0h/gKOEDjvvPNMLYBHXiU3O5cJB0+ItUpxhW9+P8LBvljhH/E9BngRTMm9YEaBgQJ+92Aq+Pi4HZPsczhwo2cJjoggIvzud79jVP9RvLTwJeYsnxNrleKKlSU1ZGcJvbvF1N+PFN+JyPF2Jd9YYDb8UGjH81An0F9mJGb9vo8yVb3UCpkbvL6OcJOWlsZvx/2Wqpoqnpn7DLlZuYzu75YCg5nfH9A7k/T0pAyIXgg8CHQBLrdFdsHEEN70epFAT/4M3buw3Dl+P7f3KsARWTLSM/jd0b9jYJeB/P2jv/PFN24pcFlFAxs21yVFya6mUNWVqnqcqg5V1af89r+jqld6vU4g42+wAQTfhZcCiEh3gqgQ6og82RnZXHrspfTs0JOH33+YFZtSeynw6lI7v59cxTrDTqBh/93AGyJyJfCZ3TccEwvw2rQj6en7WmlU5JRM6hPw89ysXK74+RXcMesOHnjnAa6eeDV9OgY+J1lZUVJLVqbQp4cz/kA0++RX1RnAdcAtmIaaJcBNwPWq+kxUtHMERductlw54UrysvOYNnsaG3c2zgpNDVaW1NCvVyaZGUnp74eNgOm9qjpbVQ9X1Q6qup+qjlPVt6OlnCN4CvILuGrCVaRJGve+dS/byrbFWqWosruygW821jGwT/I/9UNddOclt9+RYHTepzNX/vxKquuqueete9i1O3WysVevrUUVBiZPsc5APEUIi+6c8ScpPTv05IrjrmDX7l3c+/a9VFRXxFqlqLCypIaMdKFfz+R/8hPiojtn/ElM/879ufiYi9n83Wbum30fVbVVsVYp4qwoqaVPjwyys1LC3w9p0Z0r5pHkDO4xmAuPvJCSrSU89O5D1NYn77KM6uoG1m6oTdr5/SZovOhuOnCJ15MjVsyjNVVLHZGhqG8R5487n+Ubl/O3D/9GfUNylmP4+pta6utTZ37fNucYB4zBZP0NtqsIPRHJYh5BVS11RJafDvwpvzr0VywuXczH//245RMSkBUltaSlwYDeqWH8InIu8CugCJODc6bd5wkvxt8qv6IVVUsdEeboA4+ma/uuLFqzKNaqRISVJTX07pZJTnbKhLJG+m2HAVMxjTs9EdFiHrZ6STEwAHhYVReIyF5VS0WkUzPnTgYmg2ll5QgPRX2LePOzN/m+8nva5baLtTpho6ZWWfNNHUeNScr1+02iqnv597awh+cEPC+lu1vtVwRZtbTxuY/5iiJmZCTlssyYMKLPCBTls9LPWj44gShZV0tdvabK/H5z7MbU/PNEi1bVhA8xXERQ1elehajqd7a++A9VS+1T379qqSMK9OzQk45tO1JcWsy4/cfFWp2wsbKkBhEoTBF/H0BE3mBPWe804ABgptfzvTxS/Zty5mDWDC/GTCsEUqwjUGsN31e19E72VC29g72rljqigIhQ1LeI9758j4rqCvKy82KtUlhYUVJLjy4Z5LVJGX8f9u6cVQesVdX1Xk/2UsOvtX5FUFVLHdGjqE8Rs7+YzedrP2fswLEtnxDn1NUpX6+t5bDkrNfXLI2r+AZLa5xpT35Fa6qWOqJDv079KMgroLi0OCmMf+2GWmpqNWXm932ISBmNuvn4PsJMuAWM6Hrx+UPyKxzxh4gwvM9w5vx3DlU1VeRk5cRapZBY4SvW2Sflgn33AZsxI3EBzgLaqupdXk724iD5dwW5HThcVV1iToJT1LeIuvo6vliX+GW/VpbU0K1TBu3y48/fF5GeIvKRiCy3ma6X2f1TRWSDiHxutwl+50wRkdUiskJEjg1w+WNV9a+qWqaq36vqI4DnOu5efP7kTAdLcQo7F9I2py3FJcWM6j+q5RPilPp6ZVVpLYcMjdvRSx1wpe1/0RYoFhFfU837VHWvdve2RP4ZwGBMSv37IjKwcdceS72InAW8gBmdn0k4V/WJSJmIfN/EViYi33sV5Igv0tLSGN5nOF+s+4KausRt/LluUx1V1fE7v6+qm2yuDKpaBiynmZbblhOBF1S1WlVLgNVAc3fnXwGnAVvsdqrd5wkv46T7MPn33THJOtcAt6hq25YCCo74ZkTfEVTXVbN0/dJYq9JqVpbEvhmnV0SkDyYIvsDuulhEvhCRJ/0WuHUH1vmdtp5mbhaqWqqqJ9oqWx1VdZKqlnrVx4vxh+RXOOKXQV0HkZedR3FJcaxVaTUrSmrp1CGd9u3SY6VChogs8tsmN3WQiOQDL2Pq7H8PPAL0xyx628Se7rpNFSLYK6IvIlfb14dE5MHGm2fFPRwTkl/hiF8y0jMY1nsYi0sXU1dfR0Z6YqVRqyqrSmoZNjg7lmrUqeqIQAfYLrovA8+q6isA/u3uRORx4F/27Xqgp9/pPYDGlViX29eQVmh5+W//CnjAbgp8QhB+hSO+KepbxNyVc1m+cTlDeg6JtTpBsWFzHRWVDXE9vy+med4TwHJVnea3v6tvgRtwEuDzvWYBz4nINEzArxBY6H9NVX3Dvj5NCHiJ9pdighCOJOSA7geQm5nLojWLEs74ffP7cV65Zyym29WXIvK53fdnzNr7oZgHaim2NZ5tuz0T+AozU3BRM5F+RGQgcBXQBz9bVtUjvSgWqEvv1ap6l4g8RBNZRL6+fY7EJjM9k4N6HcTitYs5t+Fc0tNi5jsHzcqSGjq0T2e/feNXZ1WdS9N+/FsBzrkVuNXD5V8EHgX+Titc8UBP/rD4FY74p6hvEQu+XsDKzSvZv9v+sVbHE6rKipJaDiyM66d+pKmzAfhW0azxh8uvcMQ/Q3oOISsji0VrFgVt/E+PHNnyQSFy3n/+86N9GzZsoKy8IW7n96PEGyLye+BVTNk8AFR1h5eTveT2h+RXOOKf7IxshvQYwmdrP+PssWdjYlTxzdKlJj6WCPP7EeQ8+/pHv30K9PNyspdof0h+hSMxGNFvBMWlxazesprCLp6LwcSMpUuXsk/bNDp1iF9/P9Koat9Qzvdi/CH5FY7E4KCeB5GRlsGikkVxb/yqytKlSxnYNyshRinhRkRObrRLgW3A5zaF2BNejD8kv8KRGORm5TK4+2AWly7mjEPOiGuj2rJlC9u3b2fcYSk75P9FE/sKgINE5AJV/dDLRbwYf0h+hSNxKOpXxJKPl1C6rZS+HUMaUUYUn78f5/P7EUNVz29qv4j0xtTaGO3lOl6SfOL3W+AIK0N7DSVd0ikuKY5742/Xrh3dOqWuv98UqrrWphJ7IlCST1j8CkfikJ+Tz6BugyguKeaXI38Zt0P/pUuXMnjwYETmxVqVuEJEBuHnmrdEoCd/WPwKR2Ixou8Ips+dzvod6+nZoWfLJ0SZbdu2sWXLFk44wXNjmqSjUWk9HwWYorlne71OoCSfsPgVjsRiWO9hPDP3GYpLi+PS+H3+/oEHHmjWv6Um9zR6r8B2YJWqeq7MEvQazmD9CkdisU+bfRjYZSDFJcVMKpoUa3V+xNKlS8nLy6NPnz58k6LGH67SekFXPAzWr3AkHkV9i9iwcwObv9sca1V+xNKlSznggANIS4u/Yp2JRrN/QRF5Q0RmNdrmYlYj/SF6KjqizfA+wwEoLo2vCj87d+5kw4YNZsjvCJlAw/6w+BWOxKMgv4B+nfpRXFLMxKETY63ODyxbtgwg5Y1fRD5Q1aNE5E5Vvaa11wkU8AvJrxCRnph+fl2ABuAxVX1ARAqAf2IWCpUCp6nqzlBkOcJPUZ8iXlz4ItvKttGLXrFWBzBD/pycHPr1S/n8sq4iMg44QUReoFG9AF+14JaIpOPkq1e+P3AIcJGtSf4n4ANVLQQ+sO8dcUZR3yKAuCruuWzZMvbff39cy3aux9hND2Aae5rq3MuPR+zNEjHjD1Cv/ETAVyPgaWBSpHRwtJ5O7TrRq0OvuPH7y8rKKC0tTfkhP4CqvqSqPwfuUtXxjTbPS+0DBfw+sK93hqpso3rlnX2FC+1rp1Cv74gMRX2KWL1lNdu3b4+1Ks7fbwJVvVlEThCRe+x2fDDnB3ry+/sVw0RkuP/mVUAT9cq9njfZVwu9rq7O62mOMOIb+n/66acx1sT4+1lZWRQWxvdy42giIrcDl2GKfX4FXGb3eSKQ89TYr/BHgRaHF03VKwe2+MoWi0hXYGtT56rqY8BjAHl5eU21IXZEmG77dqNr+67MmzeP448P6qESdpYuXcqgQYPIzHT5ZX5MBIaqagOAiDwNfAZM8XJys0/+UP2K5uqVY+qS+5YJnwe87kVRR2wo6lvE0qVL2bVrV8x0qKioYM2aNW7I3zTt/X7eJ5gTWwz4heBX+OqVH9moDfEdwDEisgo4xr53xCkj+oxAVZk/f37MdFi+fDmq6oz/x9wOfCYiT9mnfjFwm9eTvRTwvB3TJfRZu+syERmrqgGHFgHqlQMc5VVBR2zp2aEnXbp0Yd68eRx7bKBW8ZFj2bJlZGRkMGjQoJjID4XW5LuIyBTgAkzNzEtV9Z2mrq2qz4vIHGAkxtauUVXPOdlepvomAseo6pOq+iRwnN3nSAFEhDFjxrBkyRLKy8tjosOXX35JYWEh2dkx7cnXWoLKd7GfnQEMxtjaX0Wk2aoldkp9lqq+Hozhg/d5/vZ+PwflVzgSnzFjxlBfX8/ChQtbPjjM1DQ0sHr16oQd8rci3+VE4AVVrVbVEmA1ZuQddrwYf0h+hSPxGThwIPvttx/z5kW/cs6mqirq6+sT1vj98Zjv0h1Y53faersv7HgJ+D2PGa68YrdDVfWFSCjjiE9EhEMPPZTFixdTWVkZVdkbKitJS0tj//3jto1Yhi8fxW6TmzooiHyXpuJkP5rqFpE0EVnaxLGe8TTsD8WvcCQHY8aMoba2lkWLotu6cUNlJf379yc3NzeqcoOgTlVH+G2PNT4gUL6L/dw/32U94F9CqQewsfE17dz+EhFp9aorVxHB4YkDDjiAffbZJ6pD/7qGBjZXVSX0kL8V+S6zgDNEJFtE+gKFQHPBlq7AMhH5wL/uhlfdUn55lMMbaWlpHHroocyZM4eamhqysiJfM39zdTV1iT+/78t3+VJEPrf7/ozJb5kpIhcA3wCnAqjqMhGZiUnXrQMuUtXm2uTdGIpiAY1fRNKAL1Q1of/6jvAwduxYZs+ezeLFiznkkEMiLm99ZSWCGXUkKq3Jd1HVW4FbPVz7Y1tQt1BV3xeRNoDnZgYBh/3h8CscycOBBx5Ifn5+1Ib+Gyor2S87m/z8/KjISzRE5H+Al4C/2V3dgde8nu9l2O/zKxYCFb6dqpq6hdNTlIyMDEaPHs38+fOpq6uLaFGNelU2VVVxYLt2EZORBFyEyQFYAKCqq0TE8xJ5L/+9kPwKR3IxduxYPvjgA5YsWUJRUVHE5GypqqK2oYHu8RvljweqVbXG11lJRDJoYlqwObzM83+MyT3OtD//B/BUI8yRfAwdOpTc3Fw++eSTiMrZYPMJuufkRFROgvOxiPwZyBWRY4AXgTe8ntyi8YfqVziSi8zMTEaOHMn8+fOpr28uCB066ysr6ZCVRRtXry8QfwK+Bb4ELsSU1b/W68le/rIh+RWO5GPs2LH8+9///qG0VrhpUGVjVRU/ads2ItdPFlS1wabcL8AM91eoaviG/Vi/wvcmWL/CkXwMHz6crKysiA39v62upqahgR7O3w+IiEwEvgYeBP4fsFpEfu71fC/GH5Jf4Ug+cnJyGDFiBPPnzyeIB41nfvD3nfG3xL3AeFU9QlXHAeOB+7ye7MX4Q/IrHMnJmDFj2LFjB5uqqsJ+7fWVlbTPzCTf+fstsVVVV/u9X0MzNTGbosW/bqh+hSM5GTlyJBkZGawqL6dbGJ/Qav39/nl5YbtmsiEiJ9sfl4nIW8BMjG2eipmN84SXaH9IfoUjOWnTpg3Dhg1jdUVFWIf+22pqqKyvd0P+wPzCbjnAFmAccARmhL6v14t4GVf5/IrVACLSH3gTeDs4fR3JxpgxY3hl2jS2VFfTJUzz8T5/3wX7mkdVzw/HdbwYf0h+hSN5GT16NGkirC4vD6vxt8vIoJ2rz98idsnvJZgioD/YstfU+2aNP1x+hSN5adu2LT1yc1ldXs7YDh3wpZm2FlVlfWUlvdu0CZOGSc9rmFoBb2AqAwdFoCf/L/x+9vkVEKRf4UhuCvPz+WDrVrbV1NAxxOq639XWsru+3g35vVOlqg+29uRmjT9cfoUjuemfl8eHwOry8pCNf52b3w+WB0TkBuBdoNq301ctuCW8NO0Iya9wJDd5GRl0s0P/Qzt0COlaGyorycvIoL3z970yBNsViz3Dfk99NMFbwO81QvArHMlPYX4+c779lp01NezbyvJePn+/R05OyLGDFOIkoJ9/+n0weMnwq1LVB1X1I1X92Le1dJKIPCkiW/3LC4tIgYi8JyKr7KuLHSQBA2xCzqoQOvrsqqujvK7ODfmDYwl7N9QJCi/G/4CI3CAih4rIcN/m4bynMO2G/GmyRZEjsWmbmUmXnBxWh2D8G938fmvoDPxXRN6JVPXeVvkVqvpv26HEnxMxmUhgWhTNAa7xpqojnhmQn8/cbdv4vra2VXP06yoryU1PpyAKVYGTiBtCOdmL8YfkVzRirxZFgeoC2M4nk4GolIl2hEZhXh5zt21jVXk5RfsG781tqKyke26u8/eDwIv7HQgvw/6Q/IrWoqqP+bqgRLJQpCM8tM/KomN2dquG/mW1teyqrXUlu4JERMpE5Hu7VYlIvYgEagW2F16MPyS/ohHNtShyJAED8vPZWFVFeV1dUOdtsMuCeyRhZl8zge+pIrJBRD632wS/z6aIyGoRWSEixwa6tqq2VdV2dssBfolZfOcJL4/UkPyKRvhaFN3B3i2KHElAYX4+n27fzurycoa2b+/5vPWVlWSnpbFfcrp3T2EMcnqj/fep6j3+O0TkAOAMYDDQDXhfRAYG6NizF6r6moh4DqJ7Wc/fKr9CRJ7HBPf2E5H1mJtIky2KHMlBh6wsCrKyWmX83XNzSUtCf7+ZwHdznAi8oKrVQImIrMbUz/y0qYP91t+AGcWPIIgSe14y/Mr8LpgFZAIVqhqwm4KqntnMR022KHIkBwPy8/nPjh3srqvzVHm3oq6OnTU1qdic42IRORdYBFypqjsxlbHn+x2z3u5rDv/1N3WYEvsnelXAy5N/rxKqIjIJczdyOH5EYX4+C3fsYE1FBQfus0+LxydBvb4MEfHvW/5YU226G/EIcDPmoXozpmbGb2i6p1+zT/JQ198EHUYP1q9wpBYds7LYJzOTVeXl3oy/qorMtDQ6hbgoKIbUqeqIYE5Q1S2+n0XkceBf9u16oKffoT2AjY3PF5HrA19eb/aih5dhf0h+hSO1EBEG5Ofz+XffUV1fT3Z64Kax6ysr6ZqTQ3oS+vvNISJdffkumDwa30zALOA5EZmGCfgVAgubuERFE/vygAuADpjRRIt4efKH5Fc4Uo/C/HyKd+5kTUUF+wfw5Svr69lWXc3YEFcDxjPNBL6PEJGhmIdoKaYqNqq6TERmAl9hbO2ipiL9qnqv3/XbApcB5wMvYFwIT3jx+d26fkdQdMnOJt9W9g1k/BsT399vkWYC308EOP5W4NaWrisiBcAfgLMwqfLDbdDQM4HKeIXFr3CkHiJCYX4+X+7aRU1DA1lpTeeSra+sJEOEzonr78cEEbkbOBl4DBiiqq1aURUow6+iiQ2MX+EW4zgCMiA/nzpVSiuack8NG6y/n9HMzcHRLFdiYgLXAhv9UnzLgknvDVTGKyx+hSM16ZaTQ5v0dFaVlzOwiYab1fX1bK2uZlRBQQy0S2xUNSx3y4AXscU3bgG+wNwohqvqNarqcvIdAUkToX9+PiW7d1PX8OMCUJuqqlCgZxL7+/FOs8Zv/Yr/AGUYv2JqsAEFR2pTmJ9PbUMDa3fv/tFn6ysrSRcJW71/R/AEevKHxa9wpC49cnPJsUP/xqyvrKRzdjaZzt+PGYF8fvdfcYREugj98vL4urycetUfEnlqGhrYUl3NiCAW/zjCjzNwR0QpzMujuqGBdX5D/01VVTSoJvX8fiLgjN8RUXq1aUNmWtpeQ/8NlZUIhLW1tyN4nPE7IkpGWpoZ+ldU0GBbeW+orKRzTk6zyT+O6OD++o6IU5iXR2V9PRsqK6lraGBzVZUb8scBrjKmI+L0zssjQ4RV5eWICHWqrlhnHOCM3xFxstLS6GOH/m0yMhCSezFPouCG/Y6oMCAvj/K6OpZ89x37ZWeT08I6f0fkccbviAr98vJIF2F3fb1ryRUnOON3RIXs9HR62br8bsgfHzjjd0SNwe3akZ2W5p78cYIL+DmiRmF+PgPy8lw/vjjBPfkdUcUZfvzgjN/hSFFiYvwicpxtRLja9QBwOGJD1I1fRNKBh4GfAwcAZ9oGhQ6HI4rE4sk/ClitqmtUtQZTE9D1AXA4okwsjL87sM7vfUvNCB0ORwSIxVSfp2aEIjIZmOz7XEQqw6hDBqYjimekSbWjKT+sBCf/8hjKBn4d3hmCoOW38NdP2KSFWBi/p2aEttNpS91OW4WILAq2uaKTn/iy40F+PBGLYf9/gEIR6SsiWcAZmAaFDocjikT9ya+qdSJyMfAOkA48qarLoq2Hw5HqxCS9V1XfAt6KhWxLRNwJJz/uZceD/LhBVH8Ua3M4HCmAS+91OFIUZ/xxjsRwJUwsZTsijzP+OERE0uxrhkbZL4ulbEd0ccYfZ4iIqKqvre1NItKj0ecR+5/FUnassGtNEJFOsdYl2iTdPzNY/P75h4tIYRwMdX1P3juBfVV1vYjki8h4AD/jTDbZeyEiw+yr+L+GWUaaqtaLyD7AWyJSGG4Z8UxKG7/fP78z8A+gXlVVRDqKSEEsdLL6FABHqurvRKQIeAD4p4g8bxOjkk427HUj/g1wjdVJfa8RuAH4rncv8KaqrhKRIhG5U0SGhllW3JHSxs+e3/9G4FFVXSMiZwPTgVvsEyEqiKFQRDKAWmC9iLwP/BF4X1U7AW2Atskk2x9788kGpgB/FpFMEfmLiMwSkb7hjkFYee2ATsDdIvJ74H+BkcDNItI1nPLijZQ2fpttmA7sBBpE5EXgQMwy4zTMlyBanA78BZhgZf8P8D4wTVWfF5FLrc7bk0x2Y/YHFgPlwDSgI2btxzUiEpY2P9aV+YOI9FbV74Evgf8DxgF3quqRVm5ERzoxR1VTbsPc6dcCJ9n3hcBtmC+775jFwMAo6SNAPnARMAO4GRgDpNvPewPLgX7JJLsZfdpgRl6lwEV231jgjTDKGAY8DvwVONXuGwTk2Z/vAp6I9fc00lvKZfjZiLaKyIXAn4ES4FxV/cZ+ngHcB6Cql0RJp3OBHZiU597ABUBfYCHwCfA10ElVVyST7CZ06Q50BlYAWaq6U0R6Aa8Dl6rq/9k4TasDjyKyr71uZ+B4YDRmlPG6qn5so/6PAb9W1e9C/Z3imVQ0/oFAuaputO/vwdQNeAH4vRpX4DTgNTWVhiKtT1vgKqAA2Az8S1WXiMgY4FcYP/tNVZ2ZTLL9dEhX43sfB1wGrMI86ccDZcAIYJKq/sV34w5R3i3AT4DbVbXYlpA7HjP6Wwc8C6xV1SDX/CceqWj8NwG/B+5W1Tvtvi6Yu/1ojM83Lco6ZQBHAocDHYClwGvAJuBsYI2qzk022Y30+BTzfzkVaKOql4vIKGCFqu6yx4T61BdM3chjgKMwv+uDqrpJRA4HJgE1qpoSRWVTzvgB7JfqLmAfzBNgpt1/NHCxqk6Kkh57fZltEs1xwNGYajNfAi+qalUyyW6khwBdgCsxsy7vA79Q1a0iMh14RVVfC7PMNsBg4DTMzeAN4AkgG8hX1c3hlBevpJTx28i++r70InIScBMmuHSdqn4eI70uxwy9+2Mi7d0xT6cRwA2quq75sxNTtu/m4xeDeRI4GbhfVafaG/GdqloUJnk+9yIT6KCqm21Ow2jgJKAfMEVV/xMOeQlBrCOO0drYE70ehLnjnwP0tPumYoI+v42iPmn29efAR8BwYDt+MwxAj2ST7Xf9Lvb1IYx/n41JtvkUeB54jz2R+PQwyp1ut43AWLuvGybu0y0W381YbSn15AcQkfeAuRj/7hm1/r2NNFdpdOay/fWZhZl1OBr4iar+r4iMA4YAD2sE/0Gxkm1jLBcBXYFDgcNUdYeI5GNmHIYBH2uYRh1+o4tfASeo6hki8o2VvQUz1P8uHLISiZRK8hGRM4FvVPVGTALHs3b/acBWzJRXtHTxpZa+jkkuORUT8AKTZZYbQeOLmWwANT71U5ioeyZwlIj0VNVyNSXdloXL8K083+8yGvibiNwGvKCqG4CJwNOShIuWWiLVuvRuB0pE5DFguqpuEZGjMGmsL0byC+/Db7oqw+bKL8Mk19QC3URkItBbVc9MJtlN6FCAGfKvxYw+fikif8OMCL4HfhMB8S8B5wKDVXWM3Xc+8JZGcdFSvJD0dzvfU86+zgMGAEdg0jnB5JE/GA3D96nkJ/cMVZ0PDMUUNH0dM+y9NgllA3s9hcdjEmnmq+oJwHzgt5hh+BUQ+hJi2VOboKOdz/8KqALyROQKEZmBSSb6WyhyEpWk9vn9Isr9MXnrTwMDgV9ihpztgS9V9dIo6ePzPQ/GBLVGq2qZ3+eZqlqbbLKb0UOAO4CdqnqH/SwXQFUrQ53TbyTzWWCldfcQkZMxST3LMP//teGQk2gk9bDf78tzP/CRmkUci0RkHWao61vUEy19fHfan2Lmr8v8pqC6AkeLyAuRMMJYyoa98grS2LN0+gHgCRHJA25W1R+6MoXR8A/HBDPP8o0CVfWVcFw70UmFYf9wzLyuL6qfqapbMF2DvtfYpHEuBsaLyKHsuQFfCgyKwtM3JrL9jPlOMUt0rwXOBB7BZBgOjpRoYIHVQe1NZ6CIzJAwrRJMVJJ62A8gZk3+fcBNqlpq93UFXgZ+pqrlMdApG/gDps9bA8YXPwX4qapGdCQSC9l+I4y2mAfO8cB3mBTbrnbrD5ytqh+HWXY3YDYmZfkJVV1rA76bVfX6cMpKNFLB+DMw1WjGA1djFrD8CViiqjdHSYdfALNVtdYvDtEZk2hUj3E/5toAXNLIbkKXF4DHVPVDv305qlolIqdihudh/5+IyBBMYtfRwLdAHnBMKkb4/Ul64/dhU3n/CCwBylT16ijJ/TUmqn2Efe97Cl4KFKvqJ8ko208HX4DvfMwMw7G+/Zjvny/VehImnXhYiPKGqOqXflOKvv09MTkFnYCvbPwnpUnagJ/45fGLyCBMkGmMiOT6B5YirINgGlxfbN/7agYeiJlu65eMsv3xM8CfAI9aXbJUtUZERoippvMSpq5CSHP7YkqwjWFPwpK/Hr6koTWhyEgmkirgJyLZItIHTH02/yATZj4boDqKKk0F9sNksI3306cNcIGd0kpPQtk/4IuwY+bY/yIiI3RPnYQp2P72qrpEVT8LUdzXQHd7o/N/6vtyPSL++yYSSTPsF5Mv/ghmNFMJnGe/4J0ww8mL7HEhF4TwqE8vTJDpRqAXcBDwDTBTI1wVJ5ay/XRoPOwWTLGOfExwbzMwXlUPaer4VsjLxAQwPwKuUFOo44driki2qkbzxh/3JJPxP4Yx+luBG4A3gWMxC1RW2mPSVbU+SvrcjqkYdKuNOA/H5JZ3wZTIek1Vv0022U3oMglTmWc50A4T5S/A5Fd8aKPvIf1f7O94GvBPzBLtOar6rIj8BDOTMA44BFOzMSouXyKQFMYvIocBd6nqofb915jpnQzgDEwCyT0xVBEAERkAjMIYQyVwdbQiztGU7TercBymIOirmGzKcuAzVX0jzPKOwOTsr8JUXx6FWaiVianVsB54W1XfC6fcRCdZAn69gdH2JpCHSdn0DfP/Dlzkm1KKhjJi+tzV+UW6xeaXrBaRUoxvWh4J44ulbB9+1z4ZuE1VX7Vz/KcBN4rIclVdHQ5Z1t37BFP08xTM/3+b3Xe/2lqNjh+TFAE/VZ2BqbN+DaYK7bt+H3cD9o+W4Vt96uyr+r/6PlPVBWqWrkZEtvgtiImm7CZYA/zaRvTLVPUJzNO5bxhl/B4zuuimqg9hyoG9iXmwnWkfCP6BR4clKYb9/tiprNeARZgy1M8Bf1XVd6IgOw1zs9nsuwFEMcAomCpF5aq6PpqyA+jUFuODb8a4GhswMy8DQx152N83HbMefwjm774KszqxFJMyfBJmTv/hUGQlK0ln/PDDF+MMzCq+/1PVo6Ik9/9hOs48BHyuNp04SrLvwsylHw9cEosvvJ+vf5DVZZZ9HYepzlOBqZswJ5zBVzEtt35qNzDTii9iphFVbfVfx94kpfH7EFOwIl9Vo1KhR0yloD9j/M1tmOIRnYHtqroognILgZdU9WAxy5cvw6SxVmKMLeJLVv0M/wDM770Nk2dwN2Z2YWc4RyIich4mW68Ys0z7Q8zsTgHmRrMb+JPaZiyOH5MUPn9zqGpNFA1fMEUxPsGUny7D+J8vY76ckWQ8ptQ2mODn+ZgGFEXAyyLSO8Ly/YN8F2Cq7h6O8cfPsjqcGkbD743pqnwnptx4R+BBjNtzOmZK81iiWJYtEUlq448mNqK+A2P8w9Q0BMnCGOFYETkjguJfBapsrsM0TGurp9SU41qAmeOPOHZefSDQ06bwzlHVozExmCHhkmNHMn0x8ZzzgYWqepyqXgwchlkt2FdjsGIzkXDGH35eAHLs3PPBmBr48zE3gYhgE3b+gQl2TQe6yp724oMwgbFoUIAJuhUBvxWRIqvfg2qXz0qYCmWq6lpVPRtj/I+KyCIROURVt6tqhbqFOy2S1D5/NBCRfTFBrQHANlV9W8wKwpcwc9zXRVj2/pgU3nWq+olNc30Sk9jSE9hXVSdGUIcfym35LdiZgHn6KmbBzrMa4dLYInI1Zqn2QuB4TYFee6HijD8E7NP175ig3irM9NImTI3AbN1TPCRs9eiakf01xu//FuPzpmPKca8EFqtqRFay+QX59sG0PyvEFMicAqzGZN11VtWpkZDfhD55wIUa5V6LiYoz/hAQkUeASlX9g32fgYlunwD8Rk3L54isJ2hG9p2YZiSnR3J2wU8HXxbhQ5hFNVMxy3IvBR5X1VtkT0vssN8AHaGRLOm9UUdMbcCD1dZ/9w15gStE5FtMKuvHETL85mRfKSLbMSWwI2781vA7YJpd/lFNGbB7ReR1TBpvrt0XtoKcjvDhAn6t5wLgJ3ZeG+vr+m6ms4AO1iePhez2EZSNiOSLyEArezsmnfpEv0PWYObaO0dKB0foOONvPVMwNQGni8jTYtpN+YJMnYFeGrlinLGUDfA34F0RuVpMvYTngENE5HURuQqTXfeuqpa6nPr4xRl/KxCRCzFr5J/BPPG+Bd4SEV9k/1JMxeCkku3HfzEJNAdj+tofqqY23wzMd+olzA0KwBl/nOICfkEipt78wZhqMev99o8ELsFE2eer6vhkkt1Ij0yMcWcCHwC/wpQHe0D9+tvHemGRIzDO+IPArh2fCxyiqtvsPl/EOw1TqeYYTMGKsKxXjwfZfjr8UBNBTDXcezB97p/ELKo5BbOQamok5DvCi4v2B8cETNuvbT5D8HuyDcB0uH0xCWX7il/+Q0SWAXMwT/o/YVZP5qvqIyKyENv+zD314x/n8wfHfKCzmJZfviegr+VTN+D3dkicbLLBFN2chFmfPxZzw3kZM614rYgUqGqxL6HIGX7844zfIzZqXYLpcvOwmC4w6J4KQZdgCkeGvd9dLGX7UFMEtS1mpeLVwBZVHY5Zq38FZhWjI4FwPn+QiCmL/TtMCeodwGeYJpO/UFuGOhllN9KjA/As5mZwiaoujpZsR/hwxt8KbALNMcAITIeYGcAnqvplwBMTXHYTuhwMvI3pBRhStx1H9HHG7wgJ65L0U9WvXZAvsXDG73CkKC7g53CkKM74HY4UxRm/w5GiOON3OFIUZ/wOR4rijN/hSFGc8TscKcr/B8OCvISsXKNAAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 216x216 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(figsize=(3,3))\n",
    "\n",
    "\n",
    "sns.barplot(x='pubnames', y='Number of Unique Genes', palette=colors_dict, data=uniq_counts_by_cells, dodge=False)\n",
    "\n",
    "ax.set_xlabel('')\n",
    "ax.set_xticklabels(uniq_counts_by_cells['pubnames'], rotation = 60)\n",
    "\n",
    "ax2 = ax.twinx()\n",
    "sns.lineplot(x='pubnames', y='Number of Unique SNPs', data=uniq_counts_by_cells, color='#000000B3', ax=ax2)\n",
    "\n",
    "sgl_summary_fn = os.path.join(outdir, 'sgls_pieqtls_with_sig_gwas_summary.svg')\n",
    "fig.savefig(sgl_summary_fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b0cd8aa",
   "metadata": {},
   "source": [
    "## Save a unique list of SNPs within an SGL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "9b2881c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_grps = intersect_df.groupby('gwas_source')\n",
    "for gwas_source, gwas_df in gwas_grps:\n",
    "    \n",
    "    uniq_sids = gwas_df.sid_x.drop_duplicates().sort_values()\n",
    "    uniq_sids = uniq_sids.str.replace(':', '-')\n",
    "    \n",
    "    fn = os.path.join(outdir, 'pieqtls_sgls.{}.tsv'.format(gwas_source))\n",
    "    uniq_sids.to_csv(fn, header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "a3a4d16b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "868"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersect_df.sid_x.nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d99975d",
   "metadata": {},
   "source": [
    "## Check Out SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "2a0c513f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>gwas_source</th>\n",
       "      <th>sid_x</th>\n",
       "      <th>pieQTL.ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>pieQTL.Position</th>\n",
       "      <th>Target_geneID</th>\n",
       "      <th>Target_geneName</th>\n",
       "      <th>TSS</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>FDR (DICE)</th>\n",
       "      <th>beta</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>Mean.TPM.Homozygous.Reference</th>\n",
       "      <th>Mean.TPM.Heterozygous</th>\n",
       "      <th>Mean.TPM.Homozygous.Alternative</th>\n",
       "      <th>Interaction_type</th>\n",
       "      <th>GWAS.Trait</th>\n",
       "      <th>cline</th>\n",
       "      <th>sid_y</th>\n",
       "      <th>geneid</th>\n",
       "      <th>genename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6</td>\n",
       "      <td>31323409</td>\n",
       "      <td>0.2938</td>\n",
       "      <td>0.0381</td>\n",
       "      <td>1.286000e-14</td>\n",
       "      <td>24250.0</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>6:31323409</td>\n",
       "      <td>rs2523605</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31323409</td>\n",
       "      <td>ENSG00000206344.6</td>\n",
       "      <td>HCG27</td>\n",
       "      <td>31165537</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.008290</td>\n",
       "      <td>0.84</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.97</td>\n",
       "      <td>2.70</td>\n",
       "      <td>3.83</td>\n",
       "      <td>Direct_pieQTL</td>\n",
       "      <td></td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>6:31323409</td>\n",
       "      <td>ENSG00000206344</td>\n",
       "      <td>HCG27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>28324397</td>\n",
       "      <td>0.1599</td>\n",
       "      <td>0.0238</td>\n",
       "      <td>1.877000e-11</td>\n",
       "      <td>24250.0</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>6:28324397</td>\n",
       "      <td>rs213236</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28324397</td>\n",
       "      <td>ENSG00000137185.7</td>\n",
       "      <td>ZSCAN9</td>\n",
       "      <td>28192664</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.011826</td>\n",
       "      <td>-0.68</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>50.52</td>\n",
       "      <td>45.47</td>\n",
       "      <td>41.05</td>\n",
       "      <td>Indirect_pieQTL</td>\n",
       "      <td></td>\n",
       "      <td>NK-cell_naive</td>\n",
       "      <td>6:28324397</td>\n",
       "      <td>ENSG00000137185</td>\n",
       "      <td>ZSCAN9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6</td>\n",
       "      <td>28324397</td>\n",
       "      <td>0.1599</td>\n",
       "      <td>0.0238</td>\n",
       "      <td>1.877000e-11</td>\n",
       "      <td>24250.0</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>6:28324397</td>\n",
       "      <td>rs213236</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28324397</td>\n",
       "      <td>ENSG00000197062.7</td>\n",
       "      <td>RP5-874C20.3</td>\n",
       "      <td>28234788</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.017461</td>\n",
       "      <td>-0.64</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>22.50</td>\n",
       "      <td>20.95</td>\n",
       "      <td>18.54</td>\n",
       "      <td>Indirect_pieQTL</td>\n",
       "      <td></td>\n",
       "      <td>B-cell_naive</td>\n",
       "      <td>6:28324397</td>\n",
       "      <td>ENSG00000197062</td>\n",
       "      <td>ZSCAN26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6</td>\n",
       "      <td>28902133</td>\n",
       "      <td>0.1567</td>\n",
       "      <td>0.0259</td>\n",
       "      <td>1.514000e-09</td>\n",
       "      <td>24250.0</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>6:28902133</td>\n",
       "      <td>rs1004062</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28902133</td>\n",
       "      <td>ENSG00000204713.6</td>\n",
       "      <td>TRIM27</td>\n",
       "      <td>28891766</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.009739</td>\n",
       "      <td>0.69</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>8.34</td>\n",
       "      <td>10.23</td>\n",
       "      <td>11.44</td>\n",
       "      <td>Direct_pieQTL</td>\n",
       "      <td>Body height</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>6:28902133</td>\n",
       "      <td>ENSG00000204713</td>\n",
       "      <td>TRIM27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHR       POS    BETA      SE             P        N   gwas_source  \\\n",
       "0  chr6  31323409  0.2938  0.0381  1.286000e-14  24250.0  T1D_32005708   \n",
       "1  chr6  28324397  0.1599  0.0238  1.877000e-11  24250.0  T1D_32005708   \n",
       "2  chr6  28324397  0.1599  0.0238  1.877000e-11  24250.0  T1D_32005708   \n",
       "3  chr6  28902133  0.1567  0.0259  1.514000e-09  24250.0  T1D_32005708   \n",
       "\n",
       "        sid_x  pieQTL.ID Chromosome  pieQTL.Position      Target_geneID  \\\n",
       "0  6:31323409  rs2523605       chr6         31323409  ENSG00000206344.6   \n",
       "1  6:28324397   rs213236       chr6         28324397  ENSG00000137185.7   \n",
       "2  6:28324397   rs213236       chr6         28324397  ENSG00000197062.7   \n",
       "3  6:28902133  rs1004062       chr6         28902133  ENSG00000204713.6   \n",
       "\n",
       "  Target_geneName       TSS    pvalue  FDR (DICE)  beta ref alt  \\\n",
       "0           HCG27  31165537  0.000021    0.008290  0.84   T   C   \n",
       "1          ZSCAN9  28192664  0.000021    0.011826 -0.68   T   C   \n",
       "2    RP5-874C20.3  28234788  0.000056    0.017461 -0.64   T   C   \n",
       "3          TRIM27  28891766  0.000026    0.009739  0.69   G   T   \n",
       "\n",
       "   Mean.TPM.Homozygous.Reference  Mean.TPM.Heterozygous  \\\n",
       "0                           0.97                   2.70   \n",
       "1                          50.52                  45.47   \n",
       "2                          22.50                  20.95   \n",
       "3                           8.34                  10.23   \n",
       "\n",
       "   Mean.TPM.Homozygous.Alternative Interaction_type       GWAS.Trait  \\\n",
       "0                             3.83    Direct_pieQTL                    \n",
       "1                            41.05  Indirect_pieQTL                    \n",
       "2                            18.54  Indirect_pieQTL                    \n",
       "3                            11.44    Direct_pieQTL  Body height       \n",
       "\n",
       "            cline       sid_y           geneid genename  \n",
       "0  monocyte_naive  6:31323409  ENSG00000206344    HCG27  \n",
       "1   NK-cell_naive  6:28324397  ENSG00000137185   ZSCAN9  \n",
       "2    B-cell_naive  6:28324397  ENSG00000197062  ZSCAN26  \n",
       "3  monocyte_naive  6:28902133  ENSG00000204713   TRIM27  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersect_df.head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "b06cea4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>gwas_source</th>\n",
       "      <th>sid_x</th>\n",
       "      <th>pieQTL.ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>pieQTL.Position</th>\n",
       "      <th>Target_geneID</th>\n",
       "      <th>Target_geneName</th>\n",
       "      <th>TSS</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>FDR (DICE)</th>\n",
       "      <th>beta</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>Mean.TPM.Homozygous.Reference</th>\n",
       "      <th>Mean.TPM.Heterozygous</th>\n",
       "      <th>Mean.TPM.Homozygous.Alternative</th>\n",
       "      <th>Interaction_type</th>\n",
       "      <th>GWAS.Trait</th>\n",
       "      <th>cline</th>\n",
       "      <th>sid_y</th>\n",
       "      <th>geneid</th>\n",
       "      <th>genename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [CHR, POS, BETA, SE, P, N, gwas_source, sid_x, pieQTL.ID, Chromosome, pieQTL.Position, Target_geneID, Target_geneName, TSS, pvalue, FDR (DICE), beta, ref, alt, Mean.TPM.Homozygous.Reference, Mean.TPM.Heterozygous, Mean.TPM.Homozygous.Alternative, Interaction_type, GWAS.Trait, cline, sid_y, geneid, genename]\n",
       "Index: []"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersect_df.loc[intersect_df.genename == 'RPS62']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "6ddc87dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>gwas_source</th>\n",
       "      <th>sid_x</th>\n",
       "      <th>pieQTL.ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>pieQTL.Position</th>\n",
       "      <th>Target_geneID</th>\n",
       "      <th>Target_geneName</th>\n",
       "      <th>TSS</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>FDR (DICE)</th>\n",
       "      <th>beta</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>Mean.TPM.Homozygous.Reference</th>\n",
       "      <th>Mean.TPM.Heterozygous</th>\n",
       "      <th>Mean.TPM.Homozygous.Alternative</th>\n",
       "      <th>Interaction_type</th>\n",
       "      <th>GWAS.Trait</th>\n",
       "      <th>cline</th>\n",
       "      <th>sid_y</th>\n",
       "      <th>geneid</th>\n",
       "      <th>genename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [CHR, POS, BETA, SE, P, N, gwas_source, sid_x, pieQTL.ID, Chromosome, pieQTL.Position, Target_geneID, Target_geneName, TSS, pvalue, FDR (DICE), beta, ref, alt, Mean.TPM.Homozygous.Reference, Mean.TPM.Heterozygous, Mean.TPM.Homozygous.Alternative, Interaction_type, GWAS.Trait, cline, sid_y, geneid, genename]\n",
       "Index: []"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersect_df.loc[intersect_df.geneid == 'ENSG00000170485']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "2fa36bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "high = ['SF3A3',\n",
    " 'PTPN22',\n",
    " 'RGS1',\n",
    " 'CTLA4',\n",
    " 'SLC22A5',\n",
    " 'BTN3A2',\n",
    " 'BTN2A2',\n",
    " 'BTN3A1',\n",
    " 'BTN2A1',\n",
    " 'ZKSCAN4',\n",
    " 'TRIM27',\n",
    " 'ZFP57',\n",
    " 'TRIM26',\n",
    " 'MRPS18B',\n",
    " 'FLOT1',\n",
    " 'VARS2',\n",
    " 'MICA',\n",
    " 'NCR3',\n",
    " 'BAG6',\n",
    " 'TSBP1',\n",
    " 'BTNL2',\n",
    " 'TAP2',\n",
    " 'WDR46',\n",
    " 'BACH2',\n",
    " 'IL2RA',\n",
    " 'RBM17',\n",
    " 'M6PR',\n",
    " 'SUOX',\n",
    " 'RPS26',\n",
    " 'CUX2',\n",
    " 'ATXN2',\n",
    " 'NAA25',\n",
    " 'CTSH',\n",
    " 'RMI2',\n",
    " 'ZPBP2',\n",
    " 'GSDMB',\n",
    " 'ORMDL3',\n",
    " 'VASP',\n",
    " 'PRKD2',\n",
    " 'IL2RB',\n",
    " 'RAC2',\n",
    " 'MFNG']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "91f34021",
   "metadata": {},
   "outputs": [],
   "source": [
    "high_df = intersect_df.loc[(intersect_df.genename.isin(high)) & (intersect_df.Interaction_type == 'Direct_pieQTL')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "11fdf469",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rs1004062\n",
      "rs10801129\n",
      "rs10806423\n",
      "rs10842537\n",
      "rs10842660\n",
      "rs10842662\n",
      "rs10858022\n",
      "rs12149160\n",
      "rs1217418\n",
      "rs1217419\n",
      "rs1217420\n",
      "rs12369009\n",
      "rs12592898\n",
      "rs13214027\n",
      "rs1323298\n",
      "rs1345229\n",
      "rs1610625\n",
      "rs1805722\n",
      "rs1805723\n",
      "rs184093\n",
      "rs1977\n",
      "rs1978\n",
      "rs1979\n",
      "rs2071790\n",
      "rs2072806\n",
      "rs2080116\n",
      "rs2080117\n",
      "rs209122\n",
      "rs209137\n",
      "rs209138\n",
      "rs209142\n",
      "rs2394164\n",
      "rs2516708\n",
      "rs2516714\n",
      "rs2516715\n",
      "rs2518030\n",
      "rs261947\n",
      "rs28551159\n",
      "rs3047288\n",
      "rs3130843\n",
      "rs3130889\n",
      "rs3131059\n",
      "rs3135303\n",
      "rs3135315\n",
      "rs34260811\n",
      "rs35260072\n",
      "rs369136\n",
      "rs371337998\n",
      "rs3742003\n",
      "rs3757138\n",
      "rs3803170\n",
      "rs413158\n",
      "rs45553631\n",
      "rs4634439\n",
      "rs4713242\n",
      "rs4713244\n",
      "rs58521088\n",
      "rs60254670\n",
      "rs62408211\n",
      "rs6454805\n",
      "rs6661817\n",
      "rs66823108\n",
      "rs68072215\n",
      "rs6899623\n",
      "rs6920256\n",
      "rs7189239\n",
      "rs72841536\n",
      "rs7309325\n",
      "rs769178\n",
      "rs79166578\n",
      "rs905671\n",
      "rs918738\n",
      "rs9348716\n",
      "rs9366654\n",
      "rs9366655\n",
      "rs9379871\n",
      "rs9393715\n",
      "rs943689\n",
      "rs969577\n",
      "ss1388091585\n",
      "ss1388091598\n"
     ]
    }
   ],
   "source": [
    "for rs in sorted(set(high_df['pieQTL.ID'].tolist())):\n",
    "    print(rs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "da0b6a53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>gwas_source</th>\n",
       "      <th>sid_x</th>\n",
       "      <th>pieQTL.ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>pieQTL.Position</th>\n",
       "      <th>Target_geneID</th>\n",
       "      <th>Target_geneName</th>\n",
       "      <th>TSS</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>FDR (DICE)</th>\n",
       "      <th>beta</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>Mean.TPM.Homozygous.Reference</th>\n",
       "      <th>Mean.TPM.Heterozygous</th>\n",
       "      <th>Mean.TPM.Homozygous.Alternative</th>\n",
       "      <th>Interaction_type</th>\n",
       "      <th>GWAS.Trait</th>\n",
       "      <th>cline</th>\n",
       "      <th>sid_y</th>\n",
       "      <th>geneid</th>\n",
       "      <th>genename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6</td>\n",
       "      <td>31323409</td>\n",
       "      <td>0.293800</td>\n",
       "      <td>0.038100</td>\n",
       "      <td>1.286000e-14</td>\n",
       "      <td>24250.0</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>6:31323409</td>\n",
       "      <td>rs2523605</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31323409</td>\n",
       "      <td>ENSG00000206344.6</td>\n",
       "      <td>HCG27</td>\n",
       "      <td>31165537</td>\n",
       "      <td>2.140000e-05</td>\n",
       "      <td>8.290324e-03</td>\n",
       "      <td>0.84</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.97</td>\n",
       "      <td>2.70</td>\n",
       "      <td>3.83</td>\n",
       "      <td>Direct_pieQTL</td>\n",
       "      <td></td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>6:31323409</td>\n",
       "      <td>ENSG00000206344</td>\n",
       "      <td>HCG27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>28324397</td>\n",
       "      <td>0.159900</td>\n",
       "      <td>0.023800</td>\n",
       "      <td>1.877000e-11</td>\n",
       "      <td>24250.0</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>6:28324397</td>\n",
       "      <td>rs213236</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28324397</td>\n",
       "      <td>ENSG00000137185.7</td>\n",
       "      <td>ZSCAN9</td>\n",
       "      <td>28192664</td>\n",
       "      <td>2.140000e-05</td>\n",
       "      <td>1.182572e-02</td>\n",
       "      <td>-0.68</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>50.52</td>\n",
       "      <td>45.47</td>\n",
       "      <td>41.05</td>\n",
       "      <td>Indirect_pieQTL</td>\n",
       "      <td></td>\n",
       "      <td>NK-cell_naive</td>\n",
       "      <td>6:28324397</td>\n",
       "      <td>ENSG00000137185</td>\n",
       "      <td>ZSCAN9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6</td>\n",
       "      <td>28324397</td>\n",
       "      <td>0.159900</td>\n",
       "      <td>0.023800</td>\n",
       "      <td>1.877000e-11</td>\n",
       "      <td>24250.0</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>6:28324397</td>\n",
       "      <td>rs213236</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28324397</td>\n",
       "      <td>ENSG00000197062.7</td>\n",
       "      <td>RP5-874C20.3</td>\n",
       "      <td>28234788</td>\n",
       "      <td>5.620000e-05</td>\n",
       "      <td>1.746085e-02</td>\n",
       "      <td>-0.64</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>22.50</td>\n",
       "      <td>20.95</td>\n",
       "      <td>18.54</td>\n",
       "      <td>Indirect_pieQTL</td>\n",
       "      <td></td>\n",
       "      <td>B-cell_naive</td>\n",
       "      <td>6:28324397</td>\n",
       "      <td>ENSG00000197062</td>\n",
       "      <td>ZSCAN26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6</td>\n",
       "      <td>28902133</td>\n",
       "      <td>0.156700</td>\n",
       "      <td>0.025900</td>\n",
       "      <td>1.514000e-09</td>\n",
       "      <td>24250.0</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>6:28902133</td>\n",
       "      <td>rs1004062</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28902133</td>\n",
       "      <td>ENSG00000204713.6</td>\n",
       "      <td>TRIM27</td>\n",
       "      <td>28891766</td>\n",
       "      <td>2.580000e-05</td>\n",
       "      <td>9.738841e-03</td>\n",
       "      <td>0.69</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>8.34</td>\n",
       "      <td>10.23</td>\n",
       "      <td>11.44</td>\n",
       "      <td>Direct_pieQTL</td>\n",
       "      <td>Body height</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>6:28902133</td>\n",
       "      <td>ENSG00000204713</td>\n",
       "      <td>TRIM27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>26573218</td>\n",
       "      <td>-0.131900</td>\n",
       "      <td>0.023600</td>\n",
       "      <td>2.134000e-08</td>\n",
       "      <td>24250.0</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>6:26573218</td>\n",
       "      <td>rs1001687</td>\n",
       "      <td>chr6</td>\n",
       "      <td>26573218</td>\n",
       "      <td>ENSG00000261353.1</td>\n",
       "      <td>CTA-14H9.5</td>\n",
       "      <td>26527291</td>\n",
       "      <td>1.170000e-06</td>\n",
       "      <td>7.160810e-04</td>\n",
       "      <td>-0.65</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>3.28</td>\n",
       "      <td>1.98</td>\n",
       "      <td>0.87</td>\n",
       "      <td>Indirect_pieQTL</td>\n",
       "      <td>Education</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>6:26573218</td>\n",
       "      <td>ENSG00000261353</td>\n",
       "      <td>ENSG00000261353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2137</th>\n",
       "      <td>chr21</td>\n",
       "      <td>43845092</td>\n",
       "      <td>0.086722</td>\n",
       "      <td>0.014027</td>\n",
       "      <td>6.310000e-10</td>\n",
       "      <td>520580.0</td>\n",
       "      <td>T1D_34012112_Gaulton</td>\n",
       "      <td>21:43845092</td>\n",
       "      <td>rs883868</td>\n",
       "      <td>chr21</td>\n",
       "      <td>43845092</td>\n",
       "      <td>ENSG00000160183.9</td>\n",
       "      <td>TMPRSS3</td>\n",
       "      <td>43816955</td>\n",
       "      <td>2.770000e-11</td>\n",
       "      <td>2.840000e-07</td>\n",
       "      <td>0.81</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.78</td>\n",
       "      <td>1.29</td>\n",
       "      <td>1.59</td>\n",
       "      <td>Direct_pieQTL</td>\n",
       "      <td>Diabetes mellitus  type 1</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>21:43845092</td>\n",
       "      <td>ENSG00000160183</td>\n",
       "      <td>TMPRSS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2138</th>\n",
       "      <td>chr21</td>\n",
       "      <td>43845294</td>\n",
       "      <td>0.086920</td>\n",
       "      <td>0.014023</td>\n",
       "      <td>5.710000e-10</td>\n",
       "      <td>520580.0</td>\n",
       "      <td>T1D_34012112_Gaulton</td>\n",
       "      <td>21:43845294</td>\n",
       "      <td>rs2839508</td>\n",
       "      <td>chr21</td>\n",
       "      <td>43845294</td>\n",
       "      <td>ENSG00000160183.9</td>\n",
       "      <td>TMPRSS3</td>\n",
       "      <td>43816955</td>\n",
       "      <td>5.700000e-16</td>\n",
       "      <td>7.840000e-12</td>\n",
       "      <td>0.92</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.14</td>\n",
       "      <td>1.78</td>\n",
       "      <td>2.55</td>\n",
       "      <td>Direct_pieQTL</td>\n",
       "      <td>Diabetes mellitus  type 1</td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "      <td>21:43845294</td>\n",
       "      <td>ENSG00000160183</td>\n",
       "      <td>TMPRSS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2139</th>\n",
       "      <td>chr21</td>\n",
       "      <td>43845294</td>\n",
       "      <td>0.086920</td>\n",
       "      <td>0.014023</td>\n",
       "      <td>5.710000e-10</td>\n",
       "      <td>520580.0</td>\n",
       "      <td>T1D_34012112_Gaulton</td>\n",
       "      <td>21:43845294</td>\n",
       "      <td>rs2839508</td>\n",
       "      <td>chr21</td>\n",
       "      <td>43845294</td>\n",
       "      <td>ENSG00000160183.9</td>\n",
       "      <td>TMPRSS3</td>\n",
       "      <td>43816955</td>\n",
       "      <td>2.770000e-11</td>\n",
       "      <td>2.840000e-07</td>\n",
       "      <td>0.81</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.78</td>\n",
       "      <td>1.29</td>\n",
       "      <td>1.59</td>\n",
       "      <td>Direct_pieQTL</td>\n",
       "      <td>Diabetes mellitus  type 1</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>21:43845294</td>\n",
       "      <td>ENSG00000160183</td>\n",
       "      <td>TMPRSS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2140</th>\n",
       "      <td>chr21</td>\n",
       "      <td>43846729</td>\n",
       "      <td>0.100740</td>\n",
       "      <td>0.015422</td>\n",
       "      <td>6.480000e-11</td>\n",
       "      <td>520580.0</td>\n",
       "      <td>T1D_34012112_Gaulton</td>\n",
       "      <td>21:43846729</td>\n",
       "      <td>rs2839509</td>\n",
       "      <td>chr21</td>\n",
       "      <td>43846729</td>\n",
       "      <td>ENSG00000160183.9</td>\n",
       "      <td>TMPRSS3</td>\n",
       "      <td>43816955</td>\n",
       "      <td>6.370000e-08</td>\n",
       "      <td>1.505540e-04</td>\n",
       "      <td>0.81</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.44</td>\n",
       "      <td>2.02</td>\n",
       "      <td>2.91</td>\n",
       "      <td>Direct_pieQTL</td>\n",
       "      <td></td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "      <td>21:43846729</td>\n",
       "      <td>ENSG00000160183</td>\n",
       "      <td>TMPRSS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2141</th>\n",
       "      <td>chr21</td>\n",
       "      <td>43846729</td>\n",
       "      <td>0.100740</td>\n",
       "      <td>0.015422</td>\n",
       "      <td>6.480000e-11</td>\n",
       "      <td>520580.0</td>\n",
       "      <td>T1D_34012112_Gaulton</td>\n",
       "      <td>21:43846729</td>\n",
       "      <td>rs2839509</td>\n",
       "      <td>chr21</td>\n",
       "      <td>43846729</td>\n",
       "      <td>ENSG00000160183.9</td>\n",
       "      <td>TMPRSS3</td>\n",
       "      <td>43816955</td>\n",
       "      <td>1.370000e-07</td>\n",
       "      <td>3.072350e-04</td>\n",
       "      <td>0.82</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.97</td>\n",
       "      <td>1.44</td>\n",
       "      <td>1.79</td>\n",
       "      <td>Direct_pieQTL</td>\n",
       "      <td></td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "      <td>21:43846729</td>\n",
       "      <td>ENSG00000160183</td>\n",
       "      <td>TMPRSS3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2142 rows Ã— 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHR       POS      BETA        SE             P         N  \\\n",
       "0      chr6  31323409  0.293800  0.038100  1.286000e-14   24250.0   \n",
       "1      chr6  28324397  0.159900  0.023800  1.877000e-11   24250.0   \n",
       "2      chr6  28324397  0.159900  0.023800  1.877000e-11   24250.0   \n",
       "3      chr6  28902133  0.156700  0.025900  1.514000e-09   24250.0   \n",
       "4      chr6  26573218 -0.131900  0.023600  2.134000e-08   24250.0   \n",
       "...     ...       ...       ...       ...           ...       ...   \n",
       "2137  chr21  43845092  0.086722  0.014027  6.310000e-10  520580.0   \n",
       "2138  chr21  43845294  0.086920  0.014023  5.710000e-10  520580.0   \n",
       "2139  chr21  43845294  0.086920  0.014023  5.710000e-10  520580.0   \n",
       "2140  chr21  43846729  0.100740  0.015422  6.480000e-11  520580.0   \n",
       "2141  chr21  43846729  0.100740  0.015422  6.480000e-11  520580.0   \n",
       "\n",
       "               gwas_source        sid_x  pieQTL.ID Chromosome  \\\n",
       "0             T1D_32005708   6:31323409  rs2523605       chr6   \n",
       "1             T1D_32005708   6:28324397   rs213236       chr6   \n",
       "2             T1D_32005708   6:28324397   rs213236       chr6   \n",
       "3             T1D_32005708   6:28902133  rs1004062       chr6   \n",
       "4             T1D_32005708   6:26573218  rs1001687       chr6   \n",
       "...                    ...          ...        ...        ...   \n",
       "2137  T1D_34012112_Gaulton  21:43845092   rs883868      chr21   \n",
       "2138  T1D_34012112_Gaulton  21:43845294  rs2839508      chr21   \n",
       "2139  T1D_34012112_Gaulton  21:43845294  rs2839508      chr21   \n",
       "2140  T1D_34012112_Gaulton  21:43846729  rs2839509      chr21   \n",
       "2141  T1D_34012112_Gaulton  21:43846729  rs2839509      chr21   \n",
       "\n",
       "      pieQTL.Position      Target_geneID Target_geneName       TSS  \\\n",
       "0            31323409  ENSG00000206344.6           HCG27  31165537   \n",
       "1            28324397  ENSG00000137185.7          ZSCAN9  28192664   \n",
       "2            28324397  ENSG00000197062.7    RP5-874C20.3  28234788   \n",
       "3            28902133  ENSG00000204713.6          TRIM27  28891766   \n",
       "4            26573218  ENSG00000261353.1      CTA-14H9.5  26527291   \n",
       "...               ...                ...             ...       ...   \n",
       "2137         43845092  ENSG00000160183.9         TMPRSS3  43816955   \n",
       "2138         43845294  ENSG00000160183.9         TMPRSS3  43816955   \n",
       "2139         43845294  ENSG00000160183.9         TMPRSS3  43816955   \n",
       "2140         43846729  ENSG00000160183.9         TMPRSS3  43816955   \n",
       "2141         43846729  ENSG00000160183.9         TMPRSS3  43816955   \n",
       "\n",
       "            pvalue    FDR (DICE)  beta ref alt  Mean.TPM.Homozygous.Reference  \\\n",
       "0     2.140000e-05  8.290324e-03  0.84   T   C                           0.97   \n",
       "1     2.140000e-05  1.182572e-02 -0.68   T   C                          50.52   \n",
       "2     5.620000e-05  1.746085e-02 -0.64   T   C                          22.50   \n",
       "3     2.580000e-05  9.738841e-03  0.69   G   T                           8.34   \n",
       "4     1.170000e-06  7.160810e-04 -0.65   C   G                           3.28   \n",
       "...            ...           ...   ...  ..  ..                            ...   \n",
       "2137  2.770000e-11  2.840000e-07  0.81   T   C                           0.78   \n",
       "2138  5.700000e-16  7.840000e-12  0.92   T   C                           1.14   \n",
       "2139  2.770000e-11  2.840000e-07  0.81   T   C                           0.78   \n",
       "2140  6.370000e-08  1.505540e-04  0.81   A   G                           1.44   \n",
       "2141  1.370000e-07  3.072350e-04  0.82   A   G                           0.97   \n",
       "\n",
       "      Mean.TPM.Heterozygous  Mean.TPM.Homozygous.Alternative Interaction_type  \\\n",
       "0                      2.70                             3.83    Direct_pieQTL   \n",
       "1                     45.47                            41.05  Indirect_pieQTL   \n",
       "2                     20.95                            18.54  Indirect_pieQTL   \n",
       "3                     10.23                            11.44    Direct_pieQTL   \n",
       "4                      1.98                             0.87  Indirect_pieQTL   \n",
       "...                     ...                              ...              ...   \n",
       "2137                   1.29                             1.59    Direct_pieQTL   \n",
       "2138                   1.78                             2.55    Direct_pieQTL   \n",
       "2139                   1.29                             1.59    Direct_pieQTL   \n",
       "2140                   2.02                             2.91    Direct_pieQTL   \n",
       "2141                   1.44                             1.79    Direct_pieQTL   \n",
       "\n",
       "                         GWAS.Trait             cline        sid_y  \\\n",
       "0                                      monocyte_naive   6:31323409   \n",
       "1                                       NK-cell_naive   6:28324397   \n",
       "2                                        B-cell_naive   6:28324397   \n",
       "3                   Body height        monocyte_naive   6:28902133   \n",
       "4                     Education      CD4_T-cell_naive   6:26573218   \n",
       "...                             ...               ...          ...   \n",
       "2137   Diabetes mellitus  type 1     CD4_T-cell_naive  21:43845092   \n",
       "2138  Diabetes mellitus  type 1      CD8_T-cell_naive  21:43845294   \n",
       "2139   Diabetes mellitus  type 1     CD4_T-cell_naive  21:43845294   \n",
       "2140                                 CD8_T-cell_naive  21:43846729   \n",
       "2141                                 CD4_T-cell_naive  21:43846729   \n",
       "\n",
       "               geneid         genename  \n",
       "0     ENSG00000206344            HCG27  \n",
       "1     ENSG00000137185           ZSCAN9  \n",
       "2     ENSG00000197062          ZSCAN26  \n",
       "3     ENSG00000204713           TRIM27  \n",
       "4     ENSG00000261353  ENSG00000261353  \n",
       "...               ...              ...  \n",
       "2137  ENSG00000160183          TMPRSS3  \n",
       "2138  ENSG00000160183          TMPRSS3  \n",
       "2139  ENSG00000160183          TMPRSS3  \n",
       "2140  ENSG00000160183          TMPRSS3  \n",
       "2141  ENSG00000160183          TMPRSS3  \n",
       "\n",
       "[2142 rows x 28 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersect_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "393322d6",
   "metadata": {},
   "source": [
    "## Intersect PC-HiC with HiChIP Loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "5b68ff59",
   "metadata": {},
   "outputs": [],
   "source": [
    "pieqtls_sgls_fn = os.path.join(outdir, 'pieqtls_sgls.tsv')\n",
    "intersect_df.to_csv(pieqtls_sgls_fn, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "eb44e966",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'results/main/pieqtls/sgl_intersect/pieqtls_sgls.tsv'"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pieqtls_sgls_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "cb6bb5e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "      <th>gwas_source</th>\n",
       "      <th>sid_x</th>\n",
       "      <th>pieQTL.ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>pieQTL.Position</th>\n",
       "      <th>Target_geneID</th>\n",
       "      <th>Target_geneName</th>\n",
       "      <th>TSS</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>FDR (DICE)</th>\n",
       "      <th>beta</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>Mean.TPM.Homozygous.Reference</th>\n",
       "      <th>Mean.TPM.Heterozygous</th>\n",
       "      <th>Mean.TPM.Homozygous.Alternative</th>\n",
       "      <th>Interaction_type</th>\n",
       "      <th>GWAS.Trait</th>\n",
       "      <th>cline</th>\n",
       "      <th>sid_y</th>\n",
       "      <th>geneid</th>\n",
       "      <th>genename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>chr6</td>\n",
       "      <td>28863264</td>\n",
       "      <td>0.1607</td>\n",
       "      <td>0.026</td>\n",
       "      <td>6.046000e-10</td>\n",
       "      <td>24250.0</td>\n",
       "      <td>T1D_32005708</td>\n",
       "      <td>6:28863264</td>\n",
       "      <td>rs3135315</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28863264</td>\n",
       "      <td>ENSG00000204713.6</td>\n",
       "      <td>TRIM27</td>\n",
       "      <td>28891766</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>0.01037</td>\n",
       "      <td>0.68</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>8.34</td>\n",
       "      <td>10.22</td>\n",
       "      <td>11.41</td>\n",
       "      <td>Direct_pieQTL</td>\n",
       "      <td>Body height</td>\n",
       "      <td>monocyte_naive</td>\n",
       "      <td>6:28863264</td>\n",
       "      <td>ENSG00000204713</td>\n",
       "      <td>TRIM27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHR       POS    BETA     SE             P        N   gwas_source  \\\n",
       "69  chr6  28863264  0.1607  0.026  6.046000e-10  24250.0  T1D_32005708   \n",
       "\n",
       "         sid_x  pieQTL.ID Chromosome  pieQTL.Position      Target_geneID  \\\n",
       "69  6:28863264  rs3135315       chr6         28863264  ENSG00000204713.6   \n",
       "\n",
       "   Target_geneName       TSS    pvalue  FDR (DICE)  beta ref alt  \\\n",
       "69          TRIM27  28891766  0.000028     0.01037  0.68   C   G   \n",
       "\n",
       "    Mean.TPM.Homozygous.Reference  Mean.TPM.Heterozygous  \\\n",
       "69                           8.34                  10.22   \n",
       "\n",
       "    Mean.TPM.Homozygous.Alternative Interaction_type       GWAS.Trait  \\\n",
       "69                            11.41    Direct_pieQTL  Body height       \n",
       "\n",
       "             cline       sid_y           geneid genename  \n",
       "69  monocyte_naive  6:28863264  ENSG00000204713   TRIM27  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersect_df.loc[intersect_df.sid_x == '6:28863264']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "c9360184",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gwas_source\n",
       "T1D_34012112_Gaulton         945\n",
       "T1D_32005708                 811\n",
       "T1D_34594039_GCST90018925    386\n",
       "dtype: int64"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersect_df.value_counts('gwas_source')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "4fdf4a57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "868"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersect_df.sid_x.nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2286af43",
   "metadata": {},
   "source": [
    "## Characterizing SNPs within an SGL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "3bdf3ae9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#loading Vu et al 2022 chromHMM annotations \n",
    "chromhmm = pd.read_table('results/refs/chromHMM/vu_et_al_2022/hg38lift_genome_100_segments.bed.gz', \n",
    "                         header=None)\n",
    "chromhmm_pbt = pbt.BedTool.from_dataframe(chromhmm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "fd3f47d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10000</td>\n",
       "      <td>10400</td>\n",
       "      <td>1_GapArtf2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10400</td>\n",
       "      <td>10600</td>\n",
       "      <td>26_Acet1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10600</td>\n",
       "      <td>10800</td>\n",
       "      <td>37_EnhWk4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10800</td>\n",
       "      <td>12800</td>\n",
       "      <td>0_GapArtf1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12800</td>\n",
       "      <td>13000</td>\n",
       "      <td>37_EnhWk4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6099633</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56886253</td>\n",
       "      <td>56886453</td>\n",
       "      <td>12_HET5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6099634</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56886453</td>\n",
       "      <td>56886853</td>\n",
       "      <td>1_GapArtf2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6099635</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56886853</td>\n",
       "      <td>56887053</td>\n",
       "      <td>12_HET5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6099636</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887053</td>\n",
       "      <td>56887253</td>\n",
       "      <td>13_HET6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6099637</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887253</td>\n",
       "      <td>57217249</td>\n",
       "      <td>0_GapArtf1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6099638 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0         1         2           3\n",
       "0        chr1     10000     10400  1_GapArtf2\n",
       "1        chr1     10400     10600    26_Acet1\n",
       "2        chr1     10600     10800   37_EnhWk4\n",
       "3        chr1     10800     12800  0_GapArtf1\n",
       "4        chr1     12800     13000   37_EnhWk4\n",
       "...       ...       ...       ...         ...\n",
       "6099633  chrY  56886253  56886453     12_HET5\n",
       "6099634  chrY  56886453  56886853  1_GapArtf2\n",
       "6099635  chrY  56886853  56887053     12_HET5\n",
       "6099636  chrY  56887053  56887253     13_HET6\n",
       "6099637  chrY  56887253  57217249  0_GapArtf1\n",
       "\n",
       "[6099638 rows x 4 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromhmm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "a0a5bc27",
   "metadata": {},
   "outputs": [],
   "source": [
    "finemapping_pbt = intersect_df[['Chromosome', 'pieQTL.Position', 'pieQTL.Position']].drop_duplicates()\n",
    "#finemapping_pbt.iloc[:, 0] = 'chr' + finemapping_pbt.iloc[:, 0].astype(str)\n",
    "finemapping_pbt.iloc[:, 1] = finemapping_pbt.iloc[:, 1] - 1 # convert to bed format \n",
    "finemapping_pbt = pbt.BedTool.from_dataframe(finemapping_pbt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "c6e7461c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_57681/3274623952.py:15: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  finemapping_chromhmm['major_state'] = finemapping_chromhmm.state_name.str.replace('[0-9]+', '')\n"
     ]
    }
   ],
   "source": [
    "finemapping_chromhmm = finemapping_pbt.intersect(chromhmm_pbt, loj=True, wa=True)\n",
    "finemapping_chromhmm = finemapping_chromhmm.to_dataframe()\n",
    "finemapping_chromhmm.columns = ['chr', 'start', 'end', 'chr2', 'chmm_start', 'chmm_end', 'state']\n",
    "\n",
    "# extract the statename without the id number\n",
    "def split_chromhmm(x):\n",
    "    s = x.split('_')\n",
    "    if len(s) > 1:\n",
    "        return(s[1])\n",
    "    else:\n",
    "        return('None')\n",
    "finemapping_chromhmm['state_name'] = finemapping_chromhmm.state.map(split_chromhmm)\n",
    "\n",
    "# add the major state which removes the subtypes\n",
    "finemapping_chromhmm['major_state'] = finemapping_chromhmm.state_name.str.replace('[0-9]+', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "2ddec5ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "state_dict = {'Quies': 'Quiescent',\n",
    "             'ReprPC': 'Repressor',\n",
    "             'GapArtf': 'No signal, unmappable',\n",
    "             'Acet': 'Acetylation',\n",
    "             'TxWk': 'Weak transcription',\n",
    "             'EnhWk': 'Weak enhancer',\n",
    "             'None': 'No overlap',\n",
    "             'Tx': 'Transcription',\n",
    "             'EnhA': 'Enhancer',\n",
    "             'TxEx': 'Transcription & Exons',\n",
    "             'HET': 'H3K9me3 & LTR repeats',\n",
    "             'TxEnh': 'Transcription & Enhancers',\n",
    "             'BivProm': 'Bivalent Promoters',\n",
    "             'znf': 'Zinc Finger',\n",
    "             'PromF': 'Downstream Promoter',\n",
    "             'DNase': 'DNase',\n",
    "             'TSS': 'TSS'}\n",
    "\n",
    "state_colors = {'Quies': 'grey',\n",
    "             'ReprPC': 'grey',\n",
    "             'GapArtf': 'grey',\n",
    "             'Acet': 'grey',\n",
    "             'None': 'grey',\n",
    "             'HET': 'grey',\n",
    "             'EnhWk': 'lightYellow',\n",
    "             'EnhA': 'yellow',\n",
    "             'TxWk': 'lightGreen',\n",
    "             'Tx': 'darkgreen',\n",
    "             'TxEx': 'darkgreen',\n",
    "             'TxEnh': 'darkgreen',\n",
    "             'DNase': 'blue',\n",
    "             'znf': 'blue', \n",
    "             'BivProm': 'red',\n",
    "             'PromF': 'red',\n",
    "             'TSS': 'red'}\n",
    "\n",
    "fm_state_cnts = finemapping_chromhmm.major_state.value_counts()\n",
    "fm_state_cnts = fm_state_cnts.to_frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "16069ab0",
   "metadata": {},
   "outputs": [],
   "source": [
    "fm_state_cnts.columns = ['counts']\n",
    "fm_state_cnts['major_state'] = fm_state_cnts.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "049eae32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>counts</th>\n",
       "      <th>major_state</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ReprPC</th>\n",
       "      <td>233</td>\n",
       "      <td>ReprPC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnhA</th>\n",
       "      <td>151</td>\n",
       "      <td>EnhA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Quies</th>\n",
       "      <td>115</td>\n",
       "      <td>Quies</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnhWk</th>\n",
       "      <td>64</td>\n",
       "      <td>EnhWk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TxEx</th>\n",
       "      <td>47</td>\n",
       "      <td>TxEx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS</th>\n",
       "      <td>45</td>\n",
       "      <td>TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tx</th>\n",
       "      <td>42</td>\n",
       "      <td>Tx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PromF</th>\n",
       "      <td>33</td>\n",
       "      <td>PromF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TxEnh</th>\n",
       "      <td>31</td>\n",
       "      <td>TxEnh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TxWk</th>\n",
       "      <td>31</td>\n",
       "      <td>TxWk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HET</th>\n",
       "      <td>27</td>\n",
       "      <td>HET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Acet</th>\n",
       "      <td>26</td>\n",
       "      <td>Acet</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BivProm</th>\n",
       "      <td>11</td>\n",
       "      <td>BivProm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GapArtf</th>\n",
       "      <td>9</td>\n",
       "      <td>GapArtf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>znf</th>\n",
       "      <td>2</td>\n",
       "      <td>znf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNase</th>\n",
       "      <td>1</td>\n",
       "      <td>DNase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         counts major_state\n",
       "ReprPC      233      ReprPC\n",
       "EnhA        151        EnhA\n",
       "Quies       115       Quies\n",
       "EnhWk        64       EnhWk\n",
       "TxEx         47        TxEx\n",
       "TSS          45         TSS\n",
       "Tx           42          Tx\n",
       "PromF        33       PromF\n",
       "TxEnh        31       TxEnh\n",
       "TxWk         31        TxWk\n",
       "HET          27         HET\n",
       "Acet         26        Acet\n",
       "BivProm      11     BivProm\n",
       "GapArtf       9     GapArtf\n",
       "znf           2         znf\n",
       "DNase         1       DNase"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fm_state_cnts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "6e08b928",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAARsAAADgCAYAAAAzD8JPAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAlnElEQVR4nO2de7hVVbXAf4O38hIEFAQ9aISi10RPpilmakq+QEND0jQt8Fl6NUOtkLiU6fWFioovUFMkn+hVE9FuD/MBlvnKpKRESVErvWUWMu4fY67OZLH22Wufs/fa5xzG7/v2t9djzrXGmmuuMccYc665RFVxHMepNZ3qLYDjOOsHrmwcxykEVzaO4xSCKxvHcQrBlY3jOIXQpd4CtIYBAwZoQ0NDvcVwnPWSpUuXvqWqA/Omb9fKpqGhgSVLltRbDMdZLxGRP1SS3t0ox3EKwZWN4ziF4MrGcZxCcGXjOE4huLJxHKcQXNk4jlMI7brrO2H69Om5006bNq2GkjiOUwq3bBzHKQRXNo7jFIIrG8dxCsGVjeM4heDKxnGcQnBl4zhOIbiycRynEFzZOI5TCK5sHMcpBFc2juMUgisbx3EKwZWN4ziF4MrGcZxCcGXjOE4huLJxHKcQXNk4jlMIrmwcxykEVzaO4xRCzZSNiAwTkUdF5EUReV5Evha29xeRRSLycvjvF+U5S0SWichLIrJfrWRzHKd4amnZrAZOV9VtgF2Ak0RkFDAVWKyqI4DFYZ2wbyKwLTAWmC0inWson+M4BVIzZaOqK1X16bD8HvAisBkwDpgXks0DxoflccB8Vf1AVV8BlgE710o+x3GKpZCYjYg0AKOBJ4BNVHUlmEICBoVkmwGvRtlWhG3pY00WkSUismTVqlU1ldtxnOpRc2UjIr2AO4BTVfXd5pJmbNN1NqjOUdVGVW0cOHBgtcR0HKfG1FTZiEhXTNH8QFXvDJvfEJHBYf9g4M2wfQUwLMo+FHi9lvI5jlMcteyNEuA64EVVvSjatRA4OiwfDdwTbZ8oIt1FZDgwAniyVvI5jlMstfwi5m7AUcCzIvKrsO1s4DxggYgcB/wROAxAVZ8XkQXAC1hP1kmq+mEN5XMcp0BqpmxU9Wdkx2EA9i6RZyYws1YyOY5TP3wEseM4heDKxnGcQnBl4zhOIbiycRynEFzZOI5TCK5sHMcpBFc2juMUgisbx3EKwZWN4ziF4MrGcZxCcGXjOE4huLJxHKcQXNk4jlMIrmwcxykEVzaO4xSCKxvHcQrBlY3jOIXgysZxnEJwZeM4TiG4snEcpxBc2TiOUwiubBzHKQRXNo7jFIIrG8dxCqGWn9+9XkTeFJHnom3nishrIvKr8Ns/2neWiCwTkZdEZL9ayeU4Tn2opWUzFxibsf1iVd0h/O4HEJFRwERg25Bntoh0rqFsjuMUTM2Ujar+BHgnZ/JxwHxV/UBVXwGWATvXSjbHcYqnZt/6boaTReSLwBLgdFX9M7AZ8HiUZkXYtg4iMhmYDLD55pu3SpDly4fnTtvQ8EqrzuU46ztFB4ivBLYCdgBWAheG7ZKRVrMOoKpzVLVRVRsHDhxYEyEdx6k+hSobVX1DVT9U1TXANTS5SiuAYVHSocDrRcrmOE5tKVTZiMjgaPUQIOmpWghMFJHuIjIcGAE8WaRsjuPUlprFbETkVmBPYICIrACmAXuKyA6Yi7QcmAKgqs+LyALgBWA1cJKqflgr2RzHKZ5cykZEvgbcALwHXAuMBqaq6kOl8qjqERmbr2sm/UxgZh55HMdpf+R1o45V1XeBfYGBwJeA82omleM4HY68yibpLdofuEFVnyG7B8lxHCeTvMpmqYg8hCmbH4lIb2BN7cRyHKejkTdAfBw2Nub3qvp3EdkYc6Ucx3FykdeyWaSqT6vqXwBU9W3g4ppJ5ThOh6NZy0ZEegAbYt3X/WiK0/QBhtRYNsdxOhDl3KgpwKmYYllKk7J5F7iidmI5jtPRaFbZqOqlwKUicoqqXlaQTI7jdEByBYhV9TIR+STQEOdR1RtrJJfjOB2MvCOIb8Le1v4VkLxGoIArG8dxcpG367sRGKWqmdM+OI7jlCNv1/dzwKa1FMRxnI5NXstmAPCCiDwJfJBsVNWDayKV4zgdjrzK5txaCuE4Tscnb2/U/9ZaEMdxOjZ5e6Peo2lO4G5AV+BvqtqnVoI5jtOxyGvZ9I7XRWQ8/qkVx3EqoEVzEKvq3cBe1RXFcZyOTF436tBotRM27ma9HXMzffr03GmnTZtWQ0kcp/2QtzfqoGh5NTZZ+biqS+M4Toclb8zGJ8pyHKdV5IrZiMhQEblLRN4UkTdE5A4RGVpr4RzH6TjkDRDfgH1Ibgj2De57wzbHcZxc5FU2A1X1BlVdHX5zsU+6OI7j5CKvsnlLRI4Ukc7hdyTwdi0FcxynY5H7I3XA4cCfgJXABMp8XUFErg8xnueibf1FZJGIvBz++0X7zhKRZSLykojsV/mlOI7TlsmrbGYAR6vqQFUdhCmfc8vkmQuMTW2bCixW1RHA4rCOiIwCJgLbhjyzRaRzTtkcx2kH5FU226vqn5MVVX0H+953SVT1J8A7qc3jgHlheR4wPto+X1U/UNVXgGX46xCO06HIq2w6pVye/uQfEBiziaquBAj/g8L2zYBXo3QrwrZ1EJHJIrJERJasWrWqBSI4jlMP8iqMC4HHROR27DWFw4GZVZQj67vhma9DqOocYA5AY2PjevvKhOO0N/KOIL5RRJZgL18KcKiqvtCC870hIoNVdaWIDAbeDNtXAMOidEOB11twfMdx2ii5XaGgXFqiYGIWAkcD54X/e6Ltt4jIRdjAwRHAk608l+M4bYiWxF1yISK3Antin+5dAUzDlMwCETkO+CNwGICqPi8iCzBltho4SVU/zDyw4zjtkpopG1U9osSuvUukn0l140CO47QhWjR5luM4TqW4snEcpxBc2TiOUwiubBzHKQRXNo7jFIIrG8dxCsGVjeM4heDKxnGcQnBl4zhOIbiycRynEFzZOI5TCK5sHMcpBFc2juMUgisbx3EKwZWN4ziF4MrGcZxCcGXjOE4huLJxHKcQajYtqLMuq1e/ljttly6Zn81ynHaLWzaO4xSCKxvHcQrBlY3jOIXgysZxnEJwZeM4TiHUpTdKRJYD7wEfAqtVtVFE+gO3AQ3AcuBwVf1zPeRzHKf61NOy+bSq7qCqjWF9KrBYVUcAi8O64zgdhLbkRo0D5oXlecD4+oniOE61qZeyUeAhEVkqIpPDtk1UdSVA+B+UlVFEJovIEhFZsmrVqoLEdRyntdRrBPFuqvq6iAwCFonIb/JmVNU5wByAxsZGrZWAjuNUl7ooG1V9Pfy/KSJ3ATsDb4jIYFVdKSKDgTfrIVtbpPcRvXOnfe/W92ooieO0nMKVjYj0BDqp6ntheV/gO8BC4GjgvPB/T9GydSSWDx9eUfqGV16pkSSOY9TDstkEuEtEkvPfoqoPishTwAIROQ74I3BYHWRzHKdGFK5sVPX3wMcytr8N7F20PM7aVOKygbttTn7aUte34zgdGFc2juMUgisbx3EKwZWN4ziF4MrGcZxC8DmInarg43qccrhl4zhOIbhl49SVlr6Kcd2r1+XOd9yw4yqSyakNbtk4jlMIbtk46xXTp0/PnXbatGk1lGT9w5WN4+TAlVTrcTfKcZxCcGXjOE4huLJxHKcQXNk4jlMIrmwcxykEVzaO4xSCKxvHcQrBlY3jOIXgysZxnELwEcSOU0MqmXqjo0+74crGcdogHfH1CHejHMcpBFc2juMUQptTNiIyVkReEpFlIjK13vI4jlMd2pSyEZHOwBXAZ4FRwBEiMqq+UjmOUw3aWoB4Z2BZ+EQvIjIfGAe8UFepHKedMHz48txpX3mloWZyZCGqWugJm0NEJgBjVfXLYf0o4BOqenKUZjIwOayOBF4qcbgBwFstEMPzeT7Ply/fFqo6MO+B2pplIxnb1tKGqjoHmFP2QCJLVLWxYgE8n+fzfK3Ol0WbitkAK4Bh0fpQ4PU6yeI4ThVpa8rmKWCEiAwXkW7ARGBhnWVyHKcKtCk3SlVXi8jJwI+AzsD1qvp8Cw9X1tXyfJ7P89Us3zq0qQCx4zgdl7bmRjmO00FxZbOeISJZPX6OU3Nc2axnqPvN6xWho6VNsN4oGxFZb641CxEZIiITRGSjsC7xf0EyDGplfiknb3KfK7kuETlRRKoylqRoRKShTJKvi0j/ImQpx/r0AJ4jIgeKyIBanSCq6H1bmL9z+N84R9pEWRwsIh/Jcfirgc2BD6DJwlFVLaWII3laXVlF5FDgSxXm6RQtd9NAmYYjsdyuEZEhOc4xEjgWeLES2ULeurqkIrIJcLOI9BSRPhn7TwA+BfxFRPYK2+om83qhbETkGGBP4CVVbcmQ7bKWkYiIqq4J6c4VkRGV5AdQ1Q/D4h0islV5kaQTcBhwUtjQvUTCg4C+qnqRqr4vIpuKyFdF5L9FZGNVXVNGnvkicng4Vtdy15ElKHAO8NOw3pDnOIlcIvI1YI6IzBWRXtH2tcpURHpHbmJXbPhEOWYCd6nq33JfUBPdwnlPFJG+rbGeo4aqkvI9AHgMa0ROyNj/W2AVsATYB5pvXMrI12ol1eGVTSik04HTVfXlaBsi0ilt2ovIKSJygIh8Nqz3gaaK39ypwv/3gR6q+rKIDBKRr4hIQ7n8IjI6Wv0lUHI+SRHppKprwjFnAmuCe7RGRPqJyHapyrELcG3Iu1+Q8XPYCO0nRWQYGUT+/g3AlgCq+q/mrqME5wK/V9XHgtV2K9CruQwiMlJE+ojITsBXsfEePYBXReQbQZY1UfrhwCwROSRsehbYPdq/pYh8NHWOscB4YLiI7CEivfNekIhsB3xBRI4HTlbVv5ZSgjmO1Sk0VL2A80TkNhHZK4eF/EtM4fyC8I5g7Gqq6mLgVWALoLeIfC5YiOXqcpZ8GurWFBG5WUTGJfUjryLq8ONsQmX9uqpODJVAk9YvPGR7AHeGFn8H4GnsYegH9AUeATYD5mOt5bOq+scS59oIuB/YH9gGOAXYCPg7cJSqvl8i31bANeHcs4DtgEGqOjdYSI3AalX9YZRnDHAksDEwJmy+Bfg0MFdVL4nSHhKO+y3g29iDe7uqLhORK7CW/eEo/W7A88A/gU2B/bCXX+/HRnnvCTymqguyrid1bT2BO7EHYzYwA1iuqtNEpHNkPcV5egNXYW/7dwHeUtUrwr6dQlltAnw0sUhEZHds1oCRwBtB9n2A+zDr7z3gVlW9LjrPU8B07P5OAh4HbgZ+U06phus6BzgNuAM4Hlijqn8XkS6qurpc2WQc82rgX8BrwN7YPbgReF5V/5FKK0EBHA98GdgglNnlcSeAiOwCDAJ6E1wq4H5V/XEFciXnugUro08CG6vqZyq6QFXt0D+gD/AM8NloW5fwPxgz7TcJ6z2wyvdAyDcMezjfwh60ucC2Jc7TCTPbZ2Kt6gOY0ukK/BgY1oyM+wC7AecB9wA3Ae9glWcJ9pB+BhgC7B/yNGIWW/L/AqZ0NgF6hTRdQ6UAOC4c+xygc3Tup4GPp+R5EbOGJmFKYkY4/j/D8gxgrwruQSNwEXAb1gJ/LEeeXYELgLtDWX4C2DA+ZrS8J/Cf4R7sCJwd8vwB+G5W2QOjgZnR+hbAZcDDwMlAA6ExzsibNNIbhrK/KpzvKMy1uhP4Us6y6RQtn5FcY7jX54f7MymVJ0nTA7gSs6p3Bn4O/AQYH/YPwhTpwOg+TAVuBz5Z4XO0NfCjsHwvsG9Y/hY2M0P5Y9RLCRT5A74QHtixQJ9o+yzgkrgCheXZwFVh+WfAIWF5aKnKkjzAwEcxzb9NWP8GcEszsu2LuThDwvr+mFJ7AZiWSns38OX4fNG+mcCElCyXAadi8Zr0eTfE3KnrUtvPwiw9MOXZJyz3Bi4FBldQ7pJaH4NZJRdj771tWUGey4DDgZEZefbElNjNQc6ewIGY+3YV1vJ3pklJbIQp9I0yjjUmPIwPAv1KyYcplcHAPmH9YOAhzBL5OZESyVlW07D3AB8BRkXbd8WmcojTfjlc183Aial9XwFWhjJ5NlznPCyu1xtrgD5VoWyCTTVxOtZozI32PQuMyHWcSk7ann7ACKw1PypU7BnhBv0X1tqdgbXaG0QFmiiOTcID8SCwKOf5LgC+gymqrbBWdlioPA1ZD1LYNgS4DvhNUgnC+ScBd4XKuwHmyjwS5bsYi7kkiuXzwAPR/h2xVnFgtG0rgkWAtXKLMXeN6Lx/CuW0P0EBRvuvDfu6ZF1LxrUlsm2DxXyGhrxfCNd1BSkFHuUZFX5bYopiUsgzO50npO8GXI/Fl7aJjnECcFwq7SzM3QAYGPKmldzuJa4pqSPnA5djltCcsK1LOOemOetMcqyx4V4dgjUA92LWWf9m8t6NuVzfS23vEerySVg93xyLS10SyufzFTxDiXzJM3IuFgOaEOrX9YTGOtfxav3Q1+sHPIqZ7fMwbTwasyJOxh7uLwPbJZUk/MfWzWjMhZkbF3yJm3EaZo0cjfnZfbEA6GY0WSxZ+XtFy5Mx5XYU1vr0CDLsFfa/TJPpeiLwcOpY28UPIWZeJ1bQSMx8fjlU0vMxZTgodYzrsYf5WMyS+BbmvvUI+0cBX8xZ/okFsAWwHGuFr8GCvf2wh/zYEnkaQp4bsZb57JB+YyL3BLMsTiS4BFh8aTrwTaBn8qAAXVP39Ylo/X+weNpa97TMtY0MdaM3prATy3d3oHcL6uo5wGFhuT/mVl8L/DCWPZXnImwowaPY1CwTwj19EotDLgR2i8pgu1BWZ1R4/zqFe5eEGo7BLLjrsQa2e+7rLOrhL/KHtWY3heUhmCL4dlyIGXk2wAK6swmmesh7EyFOUiJfcjN6YvGd5DxjgSnN5NsBs1yOx5TTQCzgOxPYNZV2KNby3Q8cirl2m0f7Pw+cG62PwYLAl4Zj34VZEZMwi+ZG1rVaGlIP4V5YS3YJ8DWCX16q/Jq5zlNC/t5YC/t9TOlMykibKO8TQ75emLKbCfwA+Eoq/XQsjvQ81oDsG871NBYY3jrjHPOB+WF5X1KWayi7bmWu6VPAFCwYf09SLuEh3KLC8tkH+D3mujWk7vnWqbSJ1fcZgqsb1sdjCueJUFY3AauxuaD2iNJtRLBScsiV3Iv/BmaH5aQzYsNKrvHfx2xJprb8wx7+Z4D/JbTc4QG9MakUqfSjwv+lWC/UNcCvw0MxGLOEDihzvlOwFvLpaPujNGMFYBbMa1gQ83pM6VyARft/CRwTy4tZOl8KD9brRCZ2SL9rtL4UcyGvCsuPEvnV4Ry7ZcjUP7XeG2sxv4e1tNvkvAfJQ9Efs2QSC0swl+qEZFtG3jHY+JCxUZ6tMEW8Th7gi5iS/yZwJmYhzgfWAHen0nYJ9+oSzGp7mWCVhP37As+UkCu2entjD/bfgB3DtpnAvJzlk66D24TyvRdz+deJFcV5MUV6XKp+dAa2BQ4K5fdpzNJ+Kxx3sxY8S72Bn4blfbBG+21MMfap+Hgtfajb8g+LlUzHlM53MdPyI2Ff9yjdGGxk7eS4YmLxnh+EB3tMxvGPDb9E+w/Eup0TpTEVeDSHnNtjMaQzMUtoH0zZ/ZXgJsUPbnSu/wrXNhVTBJdHxzwb68oGc+MaiAKh4VrvreSBCMfI7etH+R7GXMP3iKw8THHGrs0QLEYkmCv045DnmFSeLtH6gGh5Svgl7l4vLCa3T0qeMeF/O0w5PYjFNnYO2+8FDi1TFkeE/60xi/ZJTFE8RsotLVe2of4dglmmnbDu7geBBZQITof7vxD4P+CEjDS/wGJJN0TbbsNicc1abBnH6hPyPoG533uH7YvIiJuVPV6lGdrLL9zIj2Fa+K9kWCfYwLmvAhdiXYZnpSrxOKKegWh7V8ztuhRzu3qFyncCZsJ+ldDTQioGgLVAe9CkPHYFvk7UHYx1Ozek8j1E9MAD/xEqwr9o8qeTAO/VmPLaNErfDQvqPUeOruf4oaiw3BMFPD6p8OEh+gVRt2wqz92krBbscz6/xSyzvWJ5wgM6C3Pzdg8P4UXYpGs7l5CrEYuBzKApgLwnZsF+F3M17y9zTVOx1j1pAPpiQxY+k9yDHOWT5N0Zi/tcgDVsS7BezB7AF3Ic52Cs82ERYVhHqH9zw/KdNA2TOBs4vJL7l9r2RWxME5gSv6NFz2RLMrXVX1QpjgIWJDc3VPxF4YYOSz9EWPzk61gLdSbmk3cpcY64df1ouKnPVHAzJ2Du2kOh4g/DuhR/TipgGtJ3C/9HAFNT+4S1/fw4wHs56wZ4tyYEmWt8HzoHWeanyus0TOnElkK6l20Wa8ejvoH11MV5GrEu/amYQnoAaxjmhuUsF7EvTTGdOzELdAAWa5uEWRMl3cSQ7kmaXPOkU2GjpHwrLKMfAUdG6xOJrJH4/mNB2c9hFmYXmoYjdAn7nsNcsTE0jamZHMp7Z+DnLZDvxHCMgwjDHbBg/7NUGJf69zFrXfGK/oUbcDMpSyZUtuPDAxpX3D40dU1/AvO9rwyF3bmZ8xxK0+C5z4aK+DMy3K4MOUZjCu0BLJ5yJta6v4HFiDphLeZG2HiYLbFW65dYKzweizscHh23geYDvKMLvAeDMWWwCLP+9o/2pS29zF42miyA9P1KHvK+NFk6UzAldTEWqzk/Si+p//2AV7DA5xxM+XejTNATaxTuxqyheFDkdeSPZSUydMUshM/F5YK5UHul8kwK13Q31u08B1OqMzBFMo7IGo/ybYkFyl8jxL9yyJeU+dcwK/AyTLlcGsqpJzldxczjF1UBC6zoHw8P8bPAuDKF+nUsrnAn1hKPCpV7IvCZZs4xEnMJ5sQ3EuvCPL6ZfOtYS1j8YBIW0F1OGMhHC0bxUjrAex5mta3jElax3Dull4GdMKVzZXi4P5rKU66XbSKpcSTRvoeAidF6N2zE7EdYe+Dm5ql8d4ay7o9ZgHOCbJlDG1jbMvtPbPDdCEy5TyFHbC5dBzDX6UQs0D8Tc8N3CmXRI5W+G+Yinoa57pthbvKfwz1dQDSWKpX324QxQBXI1yPchx7hvLOwuODzwDdbVUdqVfmK/LF2y9c5VOLTMFdiBrBLvD/8b4C1vJ8MlecMTPFcSEaknXVdr16YsroX6/XZMaesG2I9IleT0TUb0rR6FC9VCPC28F58G4t/LMZeEhyG9YycDwzPSN9cL9vTrDsMoKRbWeLav4Ip6AMxl3JhtG8DzPVYx+rDFNcrWOu+AHMnjsAs0Uuw96HmlbqHGcdL6t2xwPfD8kgsprgMs1YyX3HAFOiTWMdAn1BW24Z9/cJ/lrLsTDSWq4x8ybikrbBerGHA49H++4CdWlU3iqiARf2w3pxpoVJtgwUmp2MBuE1TaU/DNHcnzJrpgQWUryMam5BRWQ4Gdoi2b429nPgUqaBuKn9zXexxILdqo3jL7a9iuSeWzAGhHAZg/v4D4TqFVOtL5b1spdzKmZhbeSkl4mbY+JD7MHfkxJzXtHl4wC/ArMM7sRjYbzGr9krC6yEVlNPGWAN4FdErJFhjkDl4L0qzKaawHwWurPL9G4e53FtH2wZgvYIHYa+JZAbPKzpPEZWxoIo+Ljy83wuFdHmolBsRXjTEei6G0RTAXAlMTx0v03fHWpQRWHxlNtZlumnYdxhRnCAjb3Nd7DdjLdv4sK1qo3jrcC8uAc6O1nthll9Jq48cvWxhe6tfDsVc1mWU6FrOSH8o8Efg4Oj67sTcu+PJMXoWa8y2xayY0Zjbtghz5bajRHc02a+27Is1RN+o8n3bBWsA52NWzdCwfWIo9wcp0ctX0XnqXUGrWGCXAZ8Oy12xOMWvaXqvY1OsRYq7tkeHbY8nFSrjuAMwxXU7NgDvESyAdjZmBX0ba40TyyWrkpTrYj8Is8QaqMEo3hqXe2KhDAjleQvWS5eY5bcDR2fkq6SXraovh4aH6J9kjGLOSP/pcI5dsZc9t4rlz5H/fMzC+xNmhfXFXLobQn2YRCrAS1MDOizU4/MwN38oprhup8S7W628lwdh7uFVNAXthTBzQGt/HWI+mzDR1TmYCX++qq4M2x8EvqWqT4W5Qv6sqlNFZHssjjMnpDsaCxJOU9XZqWPfjrVuc7GYwpmYpZEMid8d+K2qXp3M+9GMnDuEvCMxc/wJrFtydZSmv6q+E633xnpQdsJcjQtVteIpLGtBNM/JRth4mAlYV2xf7MEcgrmcY6I8pebKeQBzW/YkmitHbOrLZ7DemIXAr1T19eh412JD808GPmyu/FOyd8feYyo5c2OYFGpDTNmdBvxQVY9JJrvKcY6RwA9UtTFM0jYXqyunik21egx2X09Q1Xcz8v8Q6036U7jGj4Rj7IfNzfPbPNfajHybYQNLO2MWTE/M2p+J3Zs/ANeq6jOtOc+/qbZ2rMcP0/bTsPlapmA+5sHA0rB/D+AXUfrFNDOmJVofTcbwdWxczk8o86o+rexip05B3grKPbFqTiWaDgMbBHYm5gqOTuWpyB2ijbiVmGUxD4vl5LIsMevlEZqsvM0xpRnPy5O2ahID4BPYBGfx/b+M8O5dla7pDiyO9RY2bOAezO29BptOdA028Vx1zlfvCtuKgkpuSm/MbO8SbtAczGK4ATgwpDkyFNxu2JiYu1LHmEXUYxWdo5Go6zCcI3nApmI9HdsllSkjf2IOt6iLPZaxrf6wrtnZWFdsPO5nHQVKhe4QbcitxDoQbqOZl2szyuUIzAWajcUUbwFOTY5XJv+Z2OT0Z0TbRmIuf4tehEzXK6yRnhrkuhyzSPthYYjtSQ1VaO2vXbpRyXSSIrIjZiGsxIKwI7GC2h/roViFWTEPYuMq5oV9J2twl0RkHKa9d884zyCsZbpMVa8O27qofZN8AjbADmz4//upvImMG2Dm/7QgzzjsVYJnsOD0OuZzWye4AKdiwc/3sZded8SUwetYb8mzqTwtcofaklspNhl5T1X9SwV5dsTK5lNYb+kXVPWRMnkE63ofj8VR/o7V4ZHAa6o6q6VTj5Y4325YJ8dHsGELN2iF8xTnOk97VDYJInIv5opsgo1SPUxEtsZiLP2w1u9nqrowyrMdVuGXYD0CtwJXqOqPSpxjb2zum6XAQ6r6a7FJ0H+MTdtwt6q+3YyMp2GBvjOwz4x0xyrNV7G3hH/S4gKoEyJyPdZy/xobZPa78AB8DAt+Hgh8R1UfSOX5B1buo4E3sVb6p6r6DxEZhU3sdWMz55VICTVgls1ttbjGahAU7Hi1eN5ITHkMwZTqbFX9XZQ2mfS8E9aD2lNVXw3zZB+P9RL9GrOsXm2lXFthynpXzK2dizXSYwmvJ2Du2hOtOc861Nr8rPaPJgU5GIuad8fcpu3D9lmUmf8VMyGPwOIEi8uk7YS1SOdjAeHHMQvpqmbytLiLva3/MAvysWj9IMzPT94Z6ofFyOK36xuokjuUJ01b+WFuyguESaawHrsDsLFVw1Npk3p9ORY/eR3rrUp6Uz8eyuwX5Hz9oIRMyVv18zFr+wVstP0BYf8wzLrcvKXnKHnuet+QVt7MS7BxEzOim/siYcBUuYqJtc4lp15Mpe2Nja48JCi6HlnnoIVd7O3lR1P869Cw3jc8AM1OypQuZ1o4V057+wXleznRdBeUeL8IixH+PCw/hvXOvUsI0mKu1f7kHBVc4hzXJc9LtG0KNsH+8WG94hdLc5273jejwoKahgXcLsAGxA0MGnp5aAUWASeFtCVfoqySLKVm/LsaOC8sbw9MjvYdjcVtco1ibas/mkbkPoVZlePD9g3KKfh02dEGe9mqVEYDaRqTcwDwK6IXL6N0nWh6oXc6NpBwAmGSfMx6/4CMyeFbIFPaKo17xQ7HutNrVibt5iN1Yl+13Ad7WP+Bzdm6SlUnYtr+D5iiuQLW+ppjTdBwh1Iy7oG5c1PDposx/zzJM09VBxI+GNdeUdW3VfVALOa1MTBBRPqq6vtZ5ZKRX6MPqS3XNhx3qYQQD0RE9seU8Vkisgj4ELNSzhP7mF7MacCpIrIr1mDeh/VwJmNolmEvnP5ewueQW8EWwC4iciCAhm9ciUgXrCNlAynxwcJq0C4CxKEwlmKjJt8T+7b1ZdgNTL6s+DJws6r+X53ERESOxOb3HYONq5msqodEg99mYS3W4/WSsdoEpTERi2Mdo6q31FmkuhDq5AmqerrYly27YrHBA7FxRW9gU5DuEOXZhabpG1apBcp3wgbXHYC5q0dho9PfLTdoNKecG2P3aggWbH4qbD8Ie9Vk19Ycv1nqbW7mNP+2wrpYz422rcDGB5yG3bCqvi/SClkzX/rDurx/Vm/5anjd3cmYV2V9+mEu0Thsnt4zKfOlBmwqh3hOmx2wAXbLMff0YJo+71PVsEA41zKsN7ZnqLMtDjznOme9b1AFhdMVew9pBdb7cWXYnp4cqaKPg9VQ3uSlv/nhZi4E9qu3XP6ryb3uztrvcu2Mvb/0FNFsfKk8/4ENpfh3ncWGUhyNvSd3NxlTclRZ7rhX9uFanku1HQ7qC2bgHOyGTFDVF+osUklSLsZPVXXvOovkVBkR2RQb69UFs76PUdW/h32HYZb3+1jX8j+ifIINoJuhqkvDttGq+suw/BDWC1Wd95Kav4ZuWJD6nbKJW3Oe9qZsEsIAsgeAB1X12HrL0xxF3UyneERkDqZMZmLjYO7DRjhfq6rPhhHPu6jqoihPEsObgXV3X6iqD0f7T8GmRfligZdSc9qtsoF/tw5bqurvqhE8c5xKEJEx2CwDu4b132GvFXTG3JPvqOqFzeTvjo0k3wJTWE9jLtlUbErbl/K+Yd4eaNfKxnHqSdT7+CksLne8qo4P+z6OjcSdErtPGcfoFvI3Yq8LPAw8paoPdiRFA65sHKdVRF3JWS/4nq2qn2jFsTuUte7KxnGqQMYLvrdgL1tmvuDbzHE6lIKJaTcjiB2nLaOqz2Gv0NyDze3Tq1JFE47TIRUNuGXjOFXHex+zcWXjOE4huBvlOE4huLJxHKcQXNk4jlMIrmwcxykEVzaO4xTC/wP1Q4X1PpgXQAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 288x216 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Ploting the distribution of ALL Fine-mapped SNPs\n",
    "fig, ax = plt.subplots(figsize=(4,3))\n",
    "\n",
    "sns.barplot(x='major_state', y='counts', data=fm_state_cnts, hue='major_state', palette=state_colors, dodge=False)\n",
    "\n",
    "ax.set_xlabel('');\n",
    "ax.set_xticklabels(fm_state_cnts['major_state'], rotation = 60);\n",
    "\n",
    "# ax.legend(loc='upper center', ncol=2, bbox_to_anchor=(1.5, 0.9),\n",
    "#           fancybox=True, shadow=False)\n",
    "ax.legend_.set_visible(False)\n",
    "\n",
    "chromhmm_summary_fn = os.path.join(outdir, 'sgl_pieqtl_snps_characterized_chromhmm.svg')\n",
    "fig.savefig(chromhmm_summary_fn)\n",
    "\n",
    "chromhmm_summary_fn = os.path.join(outdir, 'sgl_pieqtl_snps_characterized_chromhmm.png')\n",
    "fig.savefig(chromhmm_summary_fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "775761c6",
   "metadata": {},
   "source": [
    "## Per study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "2e18578c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "T1D_34012112_Gaulton         945\n",
       "T1D_32005708                 811\n",
       "T1D_34594039_GCST90018925    386\n",
       "Name: gwas_source, dtype: int64"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersect_df.gwas_source.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "0220b05b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sid_y</th>\n",
       "      <th>geneid</th>\n",
       "      <th>cline</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6:31323409</td>\n",
       "      <td>ENSG00000206344</td>\n",
       "      <td>monocyte_naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6:28324397</td>\n",
       "      <td>ENSG00000137185</td>\n",
       "      <td>NK-cell_naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6:28324397</td>\n",
       "      <td>ENSG00000197062</td>\n",
       "      <td>B-cell_naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6:28902133</td>\n",
       "      <td>ENSG00000204713</td>\n",
       "      <td>monocyte_naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6:26573218</td>\n",
       "      <td>ENSG00000261353</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2137</th>\n",
       "      <td>21:43845092</td>\n",
       "      <td>ENSG00000160183</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2138</th>\n",
       "      <td>21:43845294</td>\n",
       "      <td>ENSG00000160183</td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2139</th>\n",
       "      <td>21:43845294</td>\n",
       "      <td>ENSG00000160183</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2140</th>\n",
       "      <td>21:43846729</td>\n",
       "      <td>ENSG00000160183</td>\n",
       "      <td>CD8_T-cell_naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2141</th>\n",
       "      <td>21:43846729</td>\n",
       "      <td>ENSG00000160183</td>\n",
       "      <td>CD4_T-cell_naive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1519 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sid_y           geneid             cline\n",
       "0      6:31323409  ENSG00000206344    monocyte_naive\n",
       "1      6:28324397  ENSG00000137185     NK-cell_naive\n",
       "2      6:28324397  ENSG00000197062      B-cell_naive\n",
       "3      6:28902133  ENSG00000204713    monocyte_naive\n",
       "4      6:26573218  ENSG00000261353  CD4_T-cell_naive\n",
       "...           ...              ...               ...\n",
       "2137  21:43845092  ENSG00000160183  CD4_T-cell_naive\n",
       "2138  21:43845294  ENSG00000160183  CD8_T-cell_naive\n",
       "2139  21:43845294  ENSG00000160183  CD4_T-cell_naive\n",
       "2140  21:43846729  ENSG00000160183  CD8_T-cell_naive\n",
       "2141  21:43846729  ENSG00000160183  CD4_T-cell_naive\n",
       "\n",
       "[1519 rows x 3 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersect_df[['sid_y', 'geneid', 'cline']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cf17542",
   "metadata": {},
   "source": [
    "## Loading the Extra Jurkat Loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "9111a67a",
   "metadata": {},
   "outputs": [],
   "source": [
    "jurkat_loop_glob = '/mnt/BioAdHoc/Groups/vd-vijay/sourya/Projects/2020_IQTL_HiChIP/Data/' + \\\n",
    "                    'Jurkat_FitHiChIP_Loops/HiChIP_Samples_*/Out_FitHiChIP/' + \\\n",
    "                    'FitHiChIP_Peak2ALL_b5000_L10000_U3000000/P2PBckgr_1/Coverage_Bias/FitHiC_BiasCorr/' + \\\n",
    "                    'FitHiChIP.interactions_FitHiC_Q0.01.bed'\n",
    "jurkat_loop_files = glob.glob(jurkat_loop_glob)\n",
    "jurkat_loops = sgls.read_multiple_tables_to_df(jurkat_loop_files, header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "fbff8c53",
   "metadata": {},
   "outputs": [],
   "source": [
    "jurkat_loops.loc[:, 'loop_id'] = sgls.create_loop_id_col(jurkat_loops)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "4e31c347",
   "metadata": {},
   "outputs": [],
   "source": [
    "jurkat_coords = jurkat_loops[jurkat_loops.columns[0:6].tolist() + ['loop_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "9805366c",
   "metadata": {},
   "outputs": [],
   "source": [
    "jurkat_loops_pbt = pbt.BedTool.from_dataframe(jurkat_coords)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13ccd548",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c6bccc27",
   "metadata": {},
   "source": [
    "## Overlap SG pairs with Jurkat loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "03f874e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# copy the intersect data\n",
    "intersect_spg_df = intersect_df.copy()\n",
    "intersect_spg_df.columns = sgls.add_prefix_to_names(intersect_spg_df.columns.tolist(), 'hits_')\n",
    "\n",
    "# create a coordinates only dataframe \n",
    "snp_coords = sgls.snp2bed_df(intersect_spg_df[['hits_CHR', 'hits_POS']])\n",
    "tss_coords = sgls.snp2bed_df(intersect_spg_df[['hits_CHR', 'hits_TSS']])\n",
    "\n",
    "intersect_coords_df = pd.concat([snp_coords, tss_coords], axis=1)\n",
    "intersect_coords_df.columns = ['snp_chr', 'snp_start', 'snp_end', 'gene_chr', 'gene_start', 'gene_end']\n",
    "intersect_coords_df.columns = sgls.add_prefix_to_names(intersect_coords_df.columns.tolist(), 'hits_')\n",
    "intersect_coords_df['hits_sgp_id'] = sgls.create_loop_id_col(intersect_coords_df)\n",
    "\n",
    "# add sgp id to the full dataframe\n",
    "intersect_spg_df.loc[:, 'hits_sgp_id'] = intersect_coords_df['hits_sgp_id'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "95c6dcd4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/bioadhoc-temp/Groups/vd-ay/jreyna/software/mamba/envs/hichip-db/lib/python3.9/site-packages/pybedtools/bedtool.py:3703: UserWarning: Default names for filetype bed are:\n",
      "['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
      "but file has 17 fields; you can supply custom names with the `names` kwarg\n",
      "  warn(\n"
     ]
    }
   ],
   "source": [
    "interset_wJurkat = sgls.pairtopair_dataframe(intersect_coords_df, jurkat_loops)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bac26b7",
   "metadata": {},
   "source": [
    "Remaining steps:\n",
    "- Add back meta from each\n",
    "- Check out genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "0107b4e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hits_snp_chr</th>\n",
       "      <th>hits_snp_start</th>\n",
       "      <th>hits_snp_end</th>\n",
       "      <th>hits_gene_chr</th>\n",
       "      <th>hits_gene_start</th>\n",
       "      <th>hits_gene_end</th>\n",
       "      <th>hits_sgp_id</th>\n",
       "      <th>chr1</th>\n",
       "      <th>s1</th>\n",
       "      <th>e1</th>\n",
       "      <th>chr2</th>\n",
       "      <th>s2</th>\n",
       "      <th>e2</th>\n",
       "      <th>cc</th>\n",
       "      <th>Coverage1</th>\n",
       "      <th>isPeak1</th>\n",
       "      <th>Bias1</th>\n",
       "      <th>Mapp1</th>\n",
       "      <th>GCContent1</th>\n",
       "      <th>RESites1</th>\n",
       "      <th>Coverage2</th>\n",
       "      <th>isPeak2</th>\n",
       "      <th>Bias2</th>\n",
       "      <th>Mapp2</th>\n",
       "      <th>GCContent2</th>\n",
       "      <th>RESites2</th>\n",
       "      <th>Dist</th>\n",
       "      <th>p</th>\n",
       "      <th>exp_cc_Bias</th>\n",
       "      <th>p_Bias</th>\n",
       "      <th>dbinom_Bias</th>\n",
       "      <th>P-Value_Bias</th>\n",
       "      <th>Q-Value_Bias</th>\n",
       "      <th>file</th>\n",
       "      <th>loop_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6</td>\n",
       "      <td>31323408</td>\n",
       "      <td>31323409</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31165536</td>\n",
       "      <td>31165537</td>\n",
       "      <td>chr6.31323408.31323409.chr6.31165536.31165537</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31165000.0</td>\n",
       "      <td>31170000.0</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31320000.0</td>\n",
       "      <td>31325000.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>3889.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.191794</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>676.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.771574</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>155000.0</td>\n",
       "      <td>2.016853e-07</td>\n",
       "      <td>4.005737</td>\n",
       "      <td>2.342296e-07</td>\n",
       "      <td>8.469382e-09</td>\n",
       "      <td>1.044046e-08</td>\n",
       "      <td>7.887956e-05</td>\n",
       "      <td>/mnt/BioAdHoc/Groups/vd-vijay/sourya/Projects/...</td>\n",
       "      <td>chr6.31165000.31170000.chr6.31320000.31325000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>31323408</td>\n",
       "      <td>31323409</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31165536</td>\n",
       "      <td>31165537</td>\n",
       "      <td>chr6.31323408.31323409.chr6.31165536.31165537</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31165000.0</td>\n",
       "      <td>31170000.0</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31320000.0</td>\n",
       "      <td>31325000.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>3899.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.155251</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>657.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.711208</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>155000.0</td>\n",
       "      <td>2.021119e-07</td>\n",
       "      <td>3.996217</td>\n",
       "      <td>2.322386e-07</td>\n",
       "      <td>8.153010e-09</td>\n",
       "      <td>1.004497e-08</td>\n",
       "      <td>7.535110e-05</td>\n",
       "      <td>/mnt/BioAdHoc/Groups/vd-vijay/sourya/Projects/...</td>\n",
       "      <td>chr6.31165000.31170000.chr6.31320000.31325000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6</td>\n",
       "      <td>31323408</td>\n",
       "      <td>31323409</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31165536</td>\n",
       "      <td>31165537</td>\n",
       "      <td>chr6.31323408.31323409.chr6.31165536.31165537</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31165000.0</td>\n",
       "      <td>31170000.0</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31320000.0</td>\n",
       "      <td>31325000.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>4238.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.283181</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>682.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.654820</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>155000.0</td>\n",
       "      <td>2.022006e-07</td>\n",
       "      <td>4.258893</td>\n",
       "      <td>2.285018e-07</td>\n",
       "      <td>2.239721e-08</td>\n",
       "      <td>2.801637e-08</td>\n",
       "      <td>1.718236e-04</td>\n",
       "      <td>/mnt/BioAdHoc/Groups/vd-vijay/sourya/Projects/...</td>\n",
       "      <td>chr6.31165000.31170000.chr6.31320000.31325000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6</td>\n",
       "      <td>31323408</td>\n",
       "      <td>31323409</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31165536</td>\n",
       "      <td>31165537</td>\n",
       "      <td>chr6.31323408.31323409.chr6.31165536.31165537</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31165000.0</td>\n",
       "      <td>31170000.0</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31320000.0</td>\n",
       "      <td>31325000.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>4853.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.035934</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>814.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.683340</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>155000.0</td>\n",
       "      <td>2.026620e-07</td>\n",
       "      <td>4.974057</td>\n",
       "      <td>2.244380e-07</td>\n",
       "      <td>2.231770e-11</td>\n",
       "      <td>2.730960e-11</td>\n",
       "      <td>3.806092e-07</td>\n",
       "      <td>/mnt/BioAdHoc/Groups/vd-vijay/sourya/Projects/...</td>\n",
       "      <td>chr6.31165000.31170000.chr6.31320000.31325000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>31323408</td>\n",
       "      <td>31323409</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31165536</td>\n",
       "      <td>31165537</td>\n",
       "      <td>chr6.31323408.31323409.chr6.31165536.31165537</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31165000.0</td>\n",
       "      <td>31170000.0</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31320000.0</td>\n",
       "      <td>31325000.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4826.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.043671</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>809.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.683657</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>155000.0</td>\n",
       "      <td>2.026249e-07</td>\n",
       "      <td>4.941546</td>\n",
       "      <td>2.244623e-07</td>\n",
       "      <td>1.022904e-10</td>\n",
       "      <td>1.260456e-10</td>\n",
       "      <td>1.449163e-06</td>\n",
       "      <td>/mnt/BioAdHoc/Groups/vd-vijay/sourya/Projects/...</td>\n",
       "      <td>chr6.31165000.31170000.chr6.31320000.31325000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3673</th>\n",
       "      <td>chr21</td>\n",
       "      <td>43845091</td>\n",
       "      <td>43845092</td>\n",
       "      <td>chr21</td>\n",
       "      <td>43816954</td>\n",
       "      <td>43816955</td>\n",
       "      <td>chr21.43845091.43845092.chr21.43816954.43816955</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3674</th>\n",
       "      <td>chr21</td>\n",
       "      <td>43845293</td>\n",
       "      <td>43845294</td>\n",
       "      <td>chr21</td>\n",
       "      <td>43816954</td>\n",
       "      <td>43816955</td>\n",
       "      <td>chr21.43845293.43845294.chr21.43816954.43816955</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3675</th>\n",
       "      <td>chr21</td>\n",
       "      <td>43845293</td>\n",
       "      <td>43845294</td>\n",
       "      <td>chr21</td>\n",
       "      <td>43816954</td>\n",
       "      <td>43816955</td>\n",
       "      <td>chr21.43845293.43845294.chr21.43816954.43816955</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3676</th>\n",
       "      <td>chr21</td>\n",
       "      <td>43846728</td>\n",
       "      <td>43846729</td>\n",
       "      <td>chr21</td>\n",
       "      <td>43816954</td>\n",
       "      <td>43816955</td>\n",
       "      <td>chr21.43846728.43846729.chr21.43816954.43816955</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3677</th>\n",
       "      <td>chr21</td>\n",
       "      <td>43846728</td>\n",
       "      <td>43846729</td>\n",
       "      <td>chr21</td>\n",
       "      <td>43816954</td>\n",
       "      <td>43816955</td>\n",
       "      <td>chr21.43846728.43846729.chr21.43816954.43816955</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3678 rows Ã— 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     hits_snp_chr  hits_snp_start  hits_snp_end hits_gene_chr  \\\n",
       "0            chr6        31323408      31323409          chr6   \n",
       "1            chr6        31323408      31323409          chr6   \n",
       "2            chr6        31323408      31323409          chr6   \n",
       "3            chr6        31323408      31323409          chr6   \n",
       "4            chr6        31323408      31323409          chr6   \n",
       "...           ...             ...           ...           ...   \n",
       "3673        chr21        43845091      43845092         chr21   \n",
       "3674        chr21        43845293      43845294         chr21   \n",
       "3675        chr21        43845293      43845294         chr21   \n",
       "3676        chr21        43846728      43846729         chr21   \n",
       "3677        chr21        43846728      43846729         chr21   \n",
       "\n",
       "      hits_gene_start  hits_gene_end  \\\n",
       "0            31165536       31165537   \n",
       "1            31165536       31165537   \n",
       "2            31165536       31165537   \n",
       "3            31165536       31165537   \n",
       "4            31165536       31165537   \n",
       "...               ...            ...   \n",
       "3673         43816954       43816955   \n",
       "3674         43816954       43816955   \n",
       "3675         43816954       43816955   \n",
       "3676         43816954       43816955   \n",
       "3677         43816954       43816955   \n",
       "\n",
       "                                          hits_sgp_id  chr1          s1  \\\n",
       "0       chr6.31323408.31323409.chr6.31165536.31165537  chr6  31165000.0   \n",
       "1       chr6.31323408.31323409.chr6.31165536.31165537  chr6  31165000.0   \n",
       "2       chr6.31323408.31323409.chr6.31165536.31165537  chr6  31165000.0   \n",
       "3       chr6.31323408.31323409.chr6.31165536.31165537  chr6  31165000.0   \n",
       "4       chr6.31323408.31323409.chr6.31165536.31165537  chr6  31165000.0   \n",
       "...                                               ...   ...         ...   \n",
       "3673  chr21.43845091.43845092.chr21.43816954.43816955   NaN         NaN   \n",
       "3674  chr21.43845293.43845294.chr21.43816954.43816955   NaN         NaN   \n",
       "3675  chr21.43845293.43845294.chr21.43816954.43816955   NaN         NaN   \n",
       "3676  chr21.43846728.43846729.chr21.43816954.43816955   NaN         NaN   \n",
       "3677  chr21.43846728.43846729.chr21.43816954.43816955   NaN         NaN   \n",
       "\n",
       "              e1  chr2          s2          e2    cc  Coverage1  isPeak1  \\\n",
       "0     31170000.0  chr6  31320000.0  31325000.0  20.0     3889.0      1.0   \n",
       "1     31170000.0  chr6  31320000.0  31325000.0  20.0     3899.0      1.0   \n",
       "2     31170000.0  chr6  31320000.0  31325000.0  20.0     4238.0      1.0   \n",
       "3     31170000.0  chr6  31320000.0  31325000.0  26.0     4853.0      1.0   \n",
       "4     31170000.0  chr6  31320000.0  31325000.0  25.0     4826.0      1.0   \n",
       "...          ...   ...         ...         ...   ...        ...      ...   \n",
       "3673         NaN   NaN         NaN         NaN   NaN        NaN      NaN   \n",
       "3674         NaN   NaN         NaN         NaN   NaN        NaN      NaN   \n",
       "3675         NaN   NaN         NaN         NaN   NaN        NaN      NaN   \n",
       "3676         NaN   NaN         NaN         NaN   NaN        NaN      NaN   \n",
       "3677         NaN   NaN         NaN         NaN   NaN        NaN      NaN   \n",
       "\n",
       "          Bias1  Mapp1  GCContent1  RESites1  Coverage2  isPeak2     Bias2  \\\n",
       "0     10.191794    0.0         0.0       0.0      676.0      1.0  1.771574   \n",
       "1     10.155251    0.0         0.0       0.0      657.0      1.0  1.711208   \n",
       "2     10.283181    0.0         0.0       0.0      682.0      1.0  1.654820   \n",
       "3     10.035934    0.0         0.0       0.0      814.0      1.0  1.683340   \n",
       "4     10.043671    0.0         0.0       0.0      809.0      1.0  1.683657   \n",
       "...         ...    ...         ...       ...        ...      ...       ...   \n",
       "3673        NaN    NaN         NaN       NaN        NaN      NaN       NaN   \n",
       "3674        NaN    NaN         NaN       NaN        NaN      NaN       NaN   \n",
       "3675        NaN    NaN         NaN       NaN        NaN      NaN       NaN   \n",
       "3676        NaN    NaN         NaN       NaN        NaN      NaN       NaN   \n",
       "3677        NaN    NaN         NaN       NaN        NaN      NaN       NaN   \n",
       "\n",
       "      Mapp2  GCContent2  RESites2      Dist             p  exp_cc_Bias  \\\n",
       "0       0.0         0.0       0.0  155000.0  2.016853e-07     4.005737   \n",
       "1       0.0         0.0       0.0  155000.0  2.021119e-07     3.996217   \n",
       "2       0.0         0.0       0.0  155000.0  2.022006e-07     4.258893   \n",
       "3       0.0         0.0       0.0  155000.0  2.026620e-07     4.974057   \n",
       "4       0.0         0.0       0.0  155000.0  2.026249e-07     4.941546   \n",
       "...     ...         ...       ...       ...           ...          ...   \n",
       "3673    NaN         NaN       NaN       NaN           NaN          NaN   \n",
       "3674    NaN         NaN       NaN       NaN           NaN          NaN   \n",
       "3675    NaN         NaN       NaN       NaN           NaN          NaN   \n",
       "3676    NaN         NaN       NaN       NaN           NaN          NaN   \n",
       "3677    NaN         NaN       NaN       NaN           NaN          NaN   \n",
       "\n",
       "            p_Bias   dbinom_Bias  P-Value_Bias  Q-Value_Bias  \\\n",
       "0     2.342296e-07  8.469382e-09  1.044046e-08  7.887956e-05   \n",
       "1     2.322386e-07  8.153010e-09  1.004497e-08  7.535110e-05   \n",
       "2     2.285018e-07  2.239721e-08  2.801637e-08  1.718236e-04   \n",
       "3     2.244380e-07  2.231770e-11  2.730960e-11  3.806092e-07   \n",
       "4     2.244623e-07  1.022904e-10  1.260456e-10  1.449163e-06   \n",
       "...            ...           ...           ...           ...   \n",
       "3673           NaN           NaN           NaN           NaN   \n",
       "3674           NaN           NaN           NaN           NaN   \n",
       "3675           NaN           NaN           NaN           NaN   \n",
       "3676           NaN           NaN           NaN           NaN   \n",
       "3677           NaN           NaN           NaN           NaN   \n",
       "\n",
       "                                                   file  \\\n",
       "0     /mnt/BioAdHoc/Groups/vd-vijay/sourya/Projects/...   \n",
       "1     /mnt/BioAdHoc/Groups/vd-vijay/sourya/Projects/...   \n",
       "2     /mnt/BioAdHoc/Groups/vd-vijay/sourya/Projects/...   \n",
       "3     /mnt/BioAdHoc/Groups/vd-vijay/sourya/Projects/...   \n",
       "4     /mnt/BioAdHoc/Groups/vd-vijay/sourya/Projects/...   \n",
       "...                                                 ...   \n",
       "3673                                                NaN   \n",
       "3674                                                NaN   \n",
       "3675                                                NaN   \n",
       "3676                                                NaN   \n",
       "3677                                                NaN   \n",
       "\n",
       "                                            loop_id  \n",
       "0     chr6.31165000.31170000.chr6.31320000.31325000  \n",
       "1     chr6.31165000.31170000.chr6.31320000.31325000  \n",
       "2     chr6.31165000.31170000.chr6.31320000.31325000  \n",
       "3     chr6.31165000.31170000.chr6.31320000.31325000  \n",
       "4     chr6.31165000.31170000.chr6.31320000.31325000  \n",
       "...                                             ...  \n",
       "3673                                            NaN  \n",
       "3674                                            NaN  \n",
       "3675                                            NaN  \n",
       "3676                                            NaN  \n",
       "3677                                            NaN  \n",
       "\n",
       "[3678 rows x 35 columns]"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interset_wJurkat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "678f8de0",
   "metadata": {},
   "outputs": [],
   "source": [
    "tdf = intersect_df.merge(interset_wJurkat,\n",
    "                   left_on=['CHR', 'POS', 'TSS'], \n",
    "                   right_on=['hits_snp_chr', 'hits_snp_end', 'hits_gene_end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "d1a3564d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCG27\n",
      "ZSCAN9\n",
      "RP5-874C20.3\n",
      "TRIM27\n",
      "CTA-14H9.5\n",
      "BTN2A3P\n",
      "TAP2\n",
      "RPL32P1\n",
      "WDR46\n",
      "HCG4P7\n",
      "BTNL2\n",
      "C6orf10\n",
      "ZNRD1-AS1\n",
      "ZKSCAN4\n",
      "BTN2A1\n",
      "BTN2A2\n",
      "XXbac-BPG181B23.7\n",
      "LY6G5B\n",
      "BTN3A1\n",
      "TCF19\n",
      "SUOX\n",
      "MICA\n",
      "CCHCR1\n",
      "LY6G5C\n",
      "IFITM4P\n",
      "BTN3A2\n",
      "XXbac-BPG299F13.14\n",
      "XXbac-BPG299F13.17\n",
      "HMGN4\n",
      "LST1\n",
      "PTPN22\n",
      "TMEM116\n",
      "HCG4P11\n",
      "ZFP57\n",
      "SKIV2L\n",
      "DHX16\n",
      "FLOT1\n",
      "PPP1R18\n",
      "ZNF165\n",
      "Y_RNA\n",
      "TRIM26\n",
      "HCG18\n",
      "MICB\n",
      "PRRC2A\n",
      "BAG6\n",
      "CSNK2B\n",
      "C6orf47\n",
      "HCG4P3\n",
      "HCG24\n",
      "HCG23\n",
      "ZDHHC20P1\n",
      "GNL1\n",
      "ZNF192P1\n",
      "ZSCAN23\n",
      "HECTD4\n",
      "HCG4B\n",
      "RNF39\n",
      "PFDN6\n",
      "BACH2\n",
      "C1orf216\n",
      "FHL3\n",
      "SF3A3\n",
      "RGS1\n",
      "NPM1P33\n",
      "KRT18P39\n",
      "CD28\n",
      "SLC22A5\n",
      "VARS2\n",
      "MRPS18B\n",
      "TUBB\n",
      "C6orf136\n",
      "NCR3\n",
      "C6orf48\n",
      "AIF1\n",
      "MSH5\n",
      "PPT2\n",
      "AP003774.1\n",
      "M6PR\n",
      "RP11-75L1.1\n",
      "ATXN2\n",
      "MAPKAPK5\n",
      "CTSH\n",
      "RP11-485G7.6\n",
      "RMI2\n",
      "DND1P1\n",
      "KANSL1-AS1\n",
      "RP11-973H7.1\n",
      "ZGLP1\n",
      "PRKD2\n",
      "TMPRSS3\n"
     ]
    }
   ],
   "source": [
    "for g in tdf.Target_geneName.unique():\n",
    "    print(g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f1fe3cf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdd325f7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11f3c0b2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6b31239",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1e4a43d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcdcb99d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "(hichip-db)",
   "language": "python",
   "name": "hichip-db"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
