{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c5fcf266",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import json\n",
    "import glob\n",
    "import gzip\n",
    "import pandas as pd \n",
    "\n",
    "outdir = 'results/main/hubs/gwas_hub/'\n",
    "link_path = 'https://informaticsdata.liai.org/BioAdHoc/Groups/vd-ay/jreyna/projects/dchallenge/results/main/hubs/gwas_hub/'\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "\n",
    "os.chdir('/mnt/BioHome/jreyna/jreyna/projects/dchallenge/')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f4ef2c2",
   "metadata": {},
   "source": [
    "## Get all WashU Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2311c874",
   "metadata": {},
   "outputs": [],
   "source": [
    "# setting template names\n",
    "gwas_tpl = 'results/main/gwas/source/{gwas_source}/GRCh37/GWAS_input_colocalization_pval_lt_5eMinus8.onekg.anno.vcf.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e1ce9f65",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_data = []\n",
    "for gwas in glob.glob(gwas_tpl.format(gwas_source='*')):\n",
    "        \n",
    "    run_info = gwas.split('/')\n",
    "    gwas_source = run_info[4]    \n",
    "    gwas_data.append([gwas_source, gwas])\n",
    "    \n",
    "gwas_df = pd.DataFrame(gwas_data)\n",
    "gwas_df.columns = ['gwas_source', 'file']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af6e6395",
   "metadata": {},
   "source": [
    "## Get GWAS Meta Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b5297c6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>authorship</th>\n",
       "      <th>link</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gwas_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>T1D_34012112_Gaulton</th>\n",
       "      <td>Chiou et al., 2021</td>\n",
       "      <td>https://www.ebi.ac.uk/gwas/publications/34012112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_34594039_GCST90018925</th>\n",
       "      <td>Sakaue et al., 2021</td>\n",
       "      <td>https://www.ebi.ac.uk/gwas/publications/34594039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_32005708</th>\n",
       "      <td>Forgetta et al., 2020</td>\n",
       "      <td>https://www.ebi.ac.uk/gwas/publications/32005708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1D_25751624</th>\n",
       "      <td>Onengut-Gumuscu et al., 2015</td>\n",
       "      <td>https://www.ebi.ac.uk/gwas/publications/25751624</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             authorship  \\\n",
       "gwas_id                                                   \n",
       "T1D_34012112_Gaulton                 Chiou et al., 2021   \n",
       "T1D_34594039_GCST90018925           Sakaue et al., 2021   \n",
       "T1D_32005708                      Forgetta et al., 2020   \n",
       "T1D_25751624               Onengut-Gumuscu et al., 2015   \n",
       "\n",
       "                                                                        link  \n",
       "gwas_id                                                                       \n",
       "T1D_34012112_Gaulton        https://www.ebi.ac.uk/gwas/publications/34012112  \n",
       "T1D_34594039_GCST90018925  https://www.ebi.ac.uk/gwas/publications/34594039   \n",
       "T1D_32005708                https://www.ebi.ac.uk/gwas/publications/32005708  \n",
       "T1D_25751624               https://www.ebi.ac.uk/gwas/publications/25751624   "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_meta = pd.read_table('config/gwas_samplesheets/gwas.samplesheet.tsv')\n",
    "gwas_meta.set_index('gwas_id', inplace=True)\n",
    "gwas_meta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3380136",
   "metadata": {},
   "source": [
    "## Making the Hub"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e540b513",
   "metadata": {},
   "source": [
    "### Creating a remote hub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "f90657b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_washu_longrange_dict(name, file):\n",
    "    d = {'type': 'longrange',\n",
    "         'showOnHubLoad': True,\n",
    "         'name': name,\n",
    "         'filename': file,\n",
    "         'options': {'displayMode': 'arc'}}\n",
    "    return(d)\n",
    "\n",
    "def create_washu_bed_dict(name, file):\n",
    "    d = {'type': 'bed',\n",
    "         'showOnHubLoad': True,\n",
    "         'name': name,\n",
    "         'filename': file}\n",
    "    return(d)\n",
    "\n",
    "def create_washu_vcf_dict(name, file):\n",
    "    d = {'type': 'vcf',\n",
    "         'showOnHubLoad': True,\n",
    "         'name': name,\n",
    "         'filename': file,\n",
    "         'options': {'ensemblStyle': True, 'height': 100, 'displayMode': 'auto'}}\n",
    "    return(d)\n",
    "\n",
    "def create_washu_geneAnnotation_dict(name, file):\n",
    "    d = {'type': 'refbed',\n",
    "         'showOnHubLoad': True,\n",
    "         'name': name,\n",
    "         'filename': file}\n",
    "    return(d)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "c54a5239",
   "metadata": {},
   "source": [
    "outdir = 'results/main/hubs/gwas_hub/'\n",
    "link_path = 'https://informaticsdata.liai.org/BioAdHoc/Groups/vd-ay/jreyna/projects/dchallenge/results/main/hubs/gwas_hub/'\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "\n",
    "json_files = []\n",
    "\n",
    "for i, sr in gwas_df.iterrows():\n",
    "    \n",
    "    print(sr.file)\n",
    "    \n",
    "    basename = os.path.basename(sr.file)\n",
    "    \n",
    "    old_path = sr.file\n",
    "    old_index_path = sr.file + '.tbi'\n",
    "    \n",
    "    new_basename = '{}.{}'.format(sr.gwas_source, basename)\n",
    "    new_path = os.path.join(outdir, new_basename)\n",
    "    new_index_path = new_path + '.tbi'\n",
    "    \n",
    "    print('new_path:', new_path)\n",
    "    print()\n",
    "\n",
    "    if os.path.exists(old_path) and os.path.exists(old_index_path):\n",
    "        if not os.path.exists(new_path) and not os.path.exists(new_index_path):\n",
    "                \n",
    "            os.link(old_path, new_path)\n",
    "            os.link(old_index_path, new_index_path)\n",
    "\n",
    "        # add the file to the json hub config\n",
    "        authorship = gwas_meta.loc[sr.gwas_source, 'authorship']\n",
    "        name = 'GWAS SNPs (Sigs Only) - {}'.format(authorship)\n",
    "        d = create_washu_vcf_dict(name, new_basename)\n",
    "        json_files.append(d)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5538169",
   "metadata": {},
   "source": [
    "### Creating a Remote Hub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a00fe786",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_remote_washu_vcf_dict(name, url):\n",
    "    d = {'type': 'vcf',\n",
    "         'showOnHubLoad': True,\n",
    "         'name': name,\n",
    "         'url': url,\n",
    "         'options': {'ensemblStyle': True, 'height': 100, 'displayMode': 'auto'}}\n",
    "    return(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "13245355",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "results/main/gwas/source/T1D_34012112_Gaulton/GRCh37/GWAS_input_colocalization_pval_lt_5eMinus8.onekg.anno.vcf.gz\n",
      "new_path: results/main/hubs/gwas_hub/T1D_34012112_Gaulton.GWAS_input_colocalization_pval_lt_5eMinus8.onekg.anno.vcf.gz\n",
      "\n",
      "results/main/gwas/source/T1D_25751624/GRCh37/GWAS_input_colocalization_pval_lt_5eMinus8.onekg.anno.vcf.gz\n",
      "new_path: results/main/hubs/gwas_hub/T1D_25751624.GWAS_input_colocalization_pval_lt_5eMinus8.onekg.anno.vcf.gz\n",
      "\n",
      "results/main/gwas/source/T1D_32005708/GRCh37/GWAS_input_colocalization_pval_lt_5eMinus8.onekg.anno.vcf.gz\n",
      "new_path: results/main/hubs/gwas_hub/T1D_32005708.GWAS_input_colocalization_pval_lt_5eMinus8.onekg.anno.vcf.gz\n",
      "\n",
      "results/main/gwas/source/T1D_34594039_GCST90018925/GRCh37/GWAS_input_colocalization_pval_lt_5eMinus8.onekg.anno.vcf.gz\n",
      "new_path: results/main/hubs/gwas_hub/T1D_34594039_GCST90018925.GWAS_input_colocalization_pval_lt_5eMinus8.onekg.anno.vcf.gz\n",
      "\n"
     ]
    }
   ],
   "source": [
    "json_files = []\n",
    "\n",
    "for i, sr in gwas_df.iterrows():\n",
    "    \n",
    "    print(sr.file)\n",
    "    \n",
    "    basename = os.path.basename(sr.file)\n",
    "    \n",
    "    old_path = sr.file\n",
    "    old_index_path = sr.file + '.tbi'\n",
    "    \n",
    "    new_basename = '{}.{}'.format(sr.gwas_source, basename)\n",
    "    new_path = os.path.join(outdir, new_basename)\n",
    "    new_index_path = new_path + '.tbi'\n",
    "    \n",
    "    print('new_path:', new_path)\n",
    "    print()\n",
    "\n",
    "    if os.path.exists(old_path) and os.path.exists(old_index_path):\n",
    "        if not os.path.exists(new_path) and not os.path.exists(new_index_path):\n",
    "                \n",
    "            os.link(old_path, new_path)\n",
    "            os.link(old_index_path, new_index_path)\n",
    "\n",
    "        # add the file to the json hub config\n",
    "        authorship = gwas_meta.loc[sr.gwas_source, 'authorship']\n",
    "        name = 'OneKG GWAS SNPs (Sigs Only) - {}'.format(authorship)\n",
    "        url = os.path.join(link_path, new_basename)\n",
    "        d = create_remote_washu_vcf_dict(name, url)\n",
    "        json_files.append(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "08281b27",
   "metadata": {},
   "outputs": [],
   "source": [
    "hub_path = os.path.join(outdir, 'hub.config.json')\n",
    "with open(hub_path, 'w') as fw:\n",
    "    print(json.dumps(json_files, indent='\\t'), file=fw)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "8628272a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.1\r\n",
      "##FILTER=<ID=PASS,Description=\"All filters passed\">\r\n",
      "##fileDate=20210319\r\n",
      "##source=ensembl;version=104;url=https://e104.ensembl.org/homo_sapiens\r\n",
      "##reference=ftp://ftp.ensembl.org/pub/release-104/fasta/homo_sapiens/dna/\r\n",
      "##INFO=<ID=COSMIC_92,Number=0,Type=Flag,Description=\"Somatic mutations found in human cancers from the COSMIC catalogue\">\r\n",
      "##INFO=<ID=ClinVar_202012,Number=0,Type=Flag,Description=\"Variants of clinical significance imported from ClinVar\">\r\n",
      "##INFO=<ID=dbSNP_154,Number=0,Type=Flag,Description=\"Variants (including SNPs and indels) imported from dbSNP\">\r\n",
      "##INFO=<ID=HGMD-PUBLIC_20204,Number=0,Type=Flag,Description=\"Variants from HGMD-PUBLIC dataset December 2020\">\r\n",
      "##INFO=<ID=TSA,Number=1,Type=String,Description=\"Type of sequence alteration. Child of term sequence_alteration as defined by the sequence ontology project.\">\r\n",
      "##INFO=<ID=E_Cited,Number=0,Type=Flag,Description=\"Cited.https://www.ensembl.org/info/genome/variation/prediction/variant_quality.html#evidence_status\">\r\n",
      "##INFO=<ID=E_Multiple_observations,Number=0,Type=Flag,Description=\"Multiple_observations.https://www.ensembl.org/info/genome/variation/prediction/variant_quality.html#evidence_status\">\r\n",
      "##INFO=<ID=E_Freq,Number=0,Type=Flag,Description=\"Frequency.https://www.ensembl.org/info/genome/variation/prediction/variant_quality.html#evidence_status\">\r\n",
      "##INFO=<ID=E_TOPMed,Number=0,Type=Flag,Description=\"TOPMed.https://www.ensembl.org/info/genome/variation/prediction/variant_quality.html#evidence_status\">\r\n",
      "##INFO=<ID=E_Hapmap,Number=0,Type=Flag,Description=\"HapMap.https://www.ensembl.org/info/genome/variation/prediction/variant_quality.html#evidence_status\">\r\n",
      "##INFO=<ID=E_Phenotype_or_Disease,Number=0,Type=Flag,Description=\"Phenotype_or_Disease.https://www.ensembl.org/info/genome/variation/prediction/variant_quality.html#evidence_status\">\r\n",
      "##INFO=<ID=E_ESP,Number=0,Type=Flag,Description=\"ESP.https://www.ensembl.org/info/genome/variation/prediction/variant_quality.html#evidence_status\">\r\n",
      "##INFO=<ID=E_gnomAD,Number=0,Type=Flag,Description=\"gnomAD.https://www.ensembl.org/info/genome/variation/prediction/variant_quality.html#evidence_status\">\r\n",
      "##INFO=<ID=E_1000G,Number=0,Type=Flag,Description=\"1000Genomes.https://www.ensembl.org/info/genome/variation/prediction/variant_quality.html#evidence_status\">\r\n",
      "##INFO=<ID=E_ExAC,Number=0,Type=Flag,Description=\"ExAC.https://www.ensembl.org/info/genome/variation/prediction/variant_quality.html#evidence_status\">\r\n",
      "##INFO=<ID=CLIN_risk_factor,Number=0,Type=Flag,Description=\"risk factor.https://www.ensembl.org/info/genome/variation/phenotype/phenotype_annotation.html#clin_significance\">\r\n",
      "##INFO=<ID=CLIN_protective,Number=0,Type=Flag,Description=\"protective.https://www.ensembl.org/info/genome/variation/phenotype/phenotype_annotation.html#clin_significance\">\r\n",
      "##INFO=<ID=CLIN_confers_sensitivity,Number=0,Type=Flag,Description=\"confers sensitivity.https://www.ensembl.org/info/genome/variation/phenotype/phenotype_annotation.html#clin_significance\">\r\n",
      "##INFO=<ID=CLIN_other,Number=0,Type=Flag,Description=\"other.https://www.ensembl.org/info/genome/variation/phenotype/phenotype_annotation.html#clin_significance\">\r\n",
      "##INFO=<ID=CLIN_drug_response,Number=0,Type=Flag,Description=\"drug response.https://www.ensembl.org/info/genome/variation/phenotype/phenotype_annotation.html#clin_significance\">\r\n",
      "##INFO=<ID=CLIN_uncertain_significance,Number=0,Type=Flag,Description=\"uncertain significance.https://www.ensembl.org/info/genome/variation/phenotype/phenotype_annotation.html#clin_significance\">\r\n",
      "##INFO=<ID=CLIN_benign,Number=0,Type=Flag,Description=\"benign.https://www.ensembl.org/info/genome/variation/phenotype/phenotype_annotation.html#clin_significance\">\r\n",
      "##INFO=<ID=CLIN_likely_pathogenic,Number=0,Type=Flag,Description=\"likely pathogenic.https://www.ensembl.org/info/genome/variation/phenotype/phenotype_annotation.html#clin_significance\">\r\n",
      "##INFO=<ID=CLIN_pathogenic,Number=0,Type=Flag,Description=\"pathogenic.https://www.ensembl.org/info/genome/variation/phenotype/phenotype_annotation.html#clin_significance\">\r\n",
      "##INFO=<ID=CLIN_likely_benign,Number=0,Type=Flag,Description=\"likely benign.https://www.ensembl.org/info/genome/variation/phenotype/phenotype_annotation.html#clin_significance\">\r\n",
      "##INFO=<ID=CLIN_histocompatibility,Number=0,Type=Flag,Description=\"histocompatibility.https://www.ensembl.org/info/genome/variation/phenotype/phenotype_annotation.html#clin_significance\">\r\n",
      "##INFO=<ID=CLIN_not_provided,Number=0,Type=Flag,Description=\"not provided.https://www.ensembl.org/info/genome/variation/phenotype/phenotype_annotation.html#clin_significance\">\r\n",
      "##INFO=<ID=CLIN_association,Number=0,Type=Flag,Description=\"association.https://www.ensembl.org/info/genome/variation/phenotype/phenotype_annotation.html#clin_significance\">\r\n",
      "##INFO=<ID=MA,Number=1,Type=String,Description=\"Minor Allele\">\r\n",
      "##INFO=<ID=MAF,Number=1,Type=Float,Description=\"Minor Allele Frequency\">\r\n",
      "##INFO=<ID=MAC,Number=1,Type=Integer,Description=\"Minor Alelele Count\">\r\n",
      "##INFO=<ID=AA,Number=1,Type=String,Description=\"Ancestral Allele\">\r\n",
      "##contig=<ID=1,length=249250621>\r\n",
      "##contig=<ID=2,length=243199373>\r\n",
      "##contig=<ID=3,length=198022430>\r\n",
      "##contig=<ID=4,length=191154276>\r\n",
      "##contig=<ID=5,length=180915260>\r\n",
      "##contig=<ID=6,length=171115067>\r\n",
      "##contig=<ID=7,length=159138663>\r\n",
      "##contig=<ID=8,length=146364022>\r\n",
      "##contig=<ID=9,length=141213431>\r\n",
      "##contig=<ID=10,length=135534747>\r\n",
      "##contig=<ID=11,length=135006516>\r\n",
      "##contig=<ID=12,length=133851895>\r\n",
      "##contig=<ID=13,length=115169878>\r\n",
      "##contig=<ID=14,length=107349540>\r\n",
      "##contig=<ID=15,length=102531392>\r\n",
      "##contig=<ID=16,length=90354753>\r\n",
      "##contig=<ID=17,length=81195210>\r\n",
      "##contig=<ID=18,length=78077248>\r\n",
      "##contig=<ID=19,length=59128983>\r\n",
      "##contig=<ID=20,length=63025520>\r\n",
      "##contig=<ID=21,length=48129895>\r\n",
      "##contig=<ID=22,length=51304566>\r\n",
      "##contig=<ID=MT,length=16569>\r\n",
      "##contig=<ID=X,length=155270560>\r\n",
      "##contig=<ID=Y,length=59373566>\r\n",
      "##INFO=<ID=AFR,Number=A,Type=Float,Description=\"Allele frequency for African populations in the 1000 Genomes Project Phase 3\">\r\n",
      "##INFO=<ID=AMR,Number=A,Type=Float,Description=\"Allele frequency for populations from the Americas in the 1000 Genomes Project Phase 3\">\r\n",
      "##INFO=<ID=EAS,Number=A,Type=Float,Description=\"Allele frequency for East Asian populations in the 1000 Genomes Project Phase 3\">\r\n",
      "##INFO=<ID=EUR,Number=A,Type=Float,Description=\"Allele frequency for European populations in the 1000 Genomes Project Phase 3\">\r\n",
      "##INFO=<ID=SAS,Number=A,Type=Float,Description=\"Allele frequency for South Asian populations in the 1000 Genomes Project Phase 3\">\r\n",
      "##bcftools_viewVersion=1.9+htslib-1.9\r\n",
      "##bcftools_viewCommand=view -R results/main/gwas/source/T1D_34012112_Gaulton/GRCh37/GWAS_input_colocalization_pval_lt_5eMinus8.regions.txt --types snps results/refs/ensembl/variation/1000GENOMES-phase_3.vcf.gz; Date=Sat Apr 30 16:50:41 2022\r\n",
      "##INFO=<ID=GWAS_BETA,Number=1,Type=Float,Description=\"GWAS BETA\",Source=\"GWAS Catalog\">\r\n",
      "##INFO=<ID=GWAS_SE,Number=1,Type=Float,Description=\"GWAS SE\",Source=\"GWAS Catalog\">\r\n",
      "##INFO=<ID=GWAS_P,Number=1,Type=Float,Description=\"GWAS P\",Source=\"GWAS Catalog\">\r\n",
      "##INFO=<ID=GWAS_N,Number=1,Type=Integer,Description=\"GWAS N\",Source=\"GWAS Catalog\">\r\n",
      "##bcftools_annotateVersion=1.9+htslib-1.9\r\n",
      "##bcftools_annotateCommand=annotate -O z -h results/main/gwas/source/T1D_34012112_Gaulton/GRCh37/GWAS_input_colocalization_pval_lt_5eMinus8.vcf.header.txt -a results/main/gwas/source/T1D_34012112_Gaulton/GRCh37/GWAS_input_colocalization_pval_lt_5eMinus8.anno.bed.gz -c CHROM,FROM,TO,GWAS_BETA,GWAS_SE,GWAS_P,GWAS_N results/main/gwas/source/T1D_34012112_Gaulton/GRCh37/GWAS_input_colocalization_pval_lt_5eMinus8.onekg.vcf.gz; Date=Sat Apr 30 16:51:22 2022\r\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\r\n",
      "1\t25293941\trs4648890\tG\tA,C\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_TOPMed;E_gnomAD;MA=G;MAF=0.4042;MAC=2024;AA=A;AFR=0.438,0;AMR=0.7262,0;EAS=0.6855,0;EUR=0.5447,0;SAS=0.6769,0;GWAS_BETA=0.076575;GWAS_SE=0.01396;GWAS_P=4.13e-08;GWAS_N=520580\r\n",
      "1\t25294607\trs7542079\tT\tA,C\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_TOPMed;E_gnomAD;MA=T;MAF=0.4627;MAC=2317;AA=T;AFR=0,0.2247;AMR=0,0.7104;EAS=0,0.6855;EUR=0,0.5447;SAS=0,0.6769;GWAS_BETA=0.077268;GWAS_SE=0.01404;GWAS_P=3.73e-08;GWAS_N=520580\r\n",
      "1\t25294878\trs10903118\tT\tC\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_Cited;E_TOPMed;E_gnomAD;MA=T;MAF=0.4599;MAC=2303;AA=C;AFR=0.2413;AMR=0.7089;EAS=0.6855;EUR=0.5437;SAS=0.6708;GWAS_BETA=0.076429;GWAS_SE=0.013922;GWAS_P=4.03e-08;GWAS_N=520580\r\n",
      "1\t25295580\trs7529070\tG\tA,C\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_Cited;E_TOPMed;E_gnomAD;MA=G;MAF=0.4083;MAC=2045;AA=G;AFR=0.4213,0;AMR=0.7262,0;EAS=0.6855,0;EUR=0.5457,0;SAS=0.6769,0;GWAS_BETA=0.076999;GWAS_SE=0.013951;GWAS_P=3.4e-08;GWAS_N=520580\r\n",
      "1\t25296478\trs10751775\tT\tA,C,G\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_TOPMed;E_gnomAD;MA=T;MAF=0.3233;MAC=1619;AA=C;AFR=0,0.7269,0;AMR=0,0.755,0;EAS=0,0.6865,0;EUR=0,0.5467,0;SAS=0,0.6769,0;GWAS_BETA=0.077267;GWAS_SE=0.014047;GWAS_P=3.78e-08;GWAS_N=520580\r\n",
      "1\t25296743\trs10751776\tA\tC,T\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_Cited;E_Phenotype_or_Disease;E_TOPMed;E_gnomAD;MA=A;MAF=0.3001;MAC=1503;AA=C;AFR=0.8094,0;AMR=0.7651,0;EAS=0.6865,0;EUR=0.5467,0;SAS=0.6769,0;GWAS_BETA=0.078145;GWAS_SE=0.01405;GWAS_P=2.67e-08;GWAS_N=520580\r\n",
      "1\t25299602\trs7550552\tC\tA,G,T\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_TOPMed;E_gnomAD;MA=C;MAF=0.3632;MAC=1819;AA=C;AFR=0,0,0.5908;AMR=0,0,0.745;EAS=0,0,0.6746;EUR=0,0,0.5437;SAS=0,0,0.6789;GWAS_BETA=0.076765;GWAS_SE=0.014049;GWAS_P=4.66e-08;GWAS_N=520580\r\n",
      "1\t25301477\trs7545113\tT\tA,C,G\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_TOPMed;E_gnomAD;MA=T;MAF=0.3624;MAC=1815;AA=A;AFR=0.5923,0,0;AMR=0.745,0,0;EAS=0.6756,0,0;EUR=0.5437,0,0;SAS=0.68,0,0;GWAS_BETA=0.076626;GWAS_SE=0.01405;GWAS_P=4.93e-08;GWAS_N=520580\r\n",
      "1\t25301869\trs1609996\tT\tC,G\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_TOPMed;E_gnomAD;MA=T;MAF=0.3624;MAC=1815;AA=C;AFR=0.5923,0;AMR=0.745,0;EAS=0.6756,0;EUR=0.5437,0;SAS=0.68,0;GWAS_BETA=0.07761;GWAS_SE=0.013956;GWAS_P=2.68e-08;GWAS_N=520580\r\n",
      "1\t36037222\trs6425948\tA\tG,T\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_TOPMed;E_gnomAD;MA=A;MAF=0.4451;MAC=2229;AA=G;AFR=0.9463,0;AMR=0.4438,0;EAS=0.7748,0;EUR=0.1352,0;SAS=0.3098,0;GWAS_BETA=0.126663;GWAS_SE=0.022935;GWAS_P=3.34e-08;GWAS_N=520580\r\n",
      "1\t36087661\trs574384\tC\tA,G,T\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_TOPMed;E_gnomAD;MA=A;MAF=0.4792;MAC=2400;AA=C;AFR=0.1634,0,0;AMR=0.5663,0,0;EAS=0.2331,0,0;EUR=0.8857,0,0;SAS=0.68,0,0;GWAS_BETA=-0.133602;GWAS_SE=0.023876;GWAS_P=2.2e-08;GWAS_N=520580\r\n",
      "1\t36107510\trs676614\tC\tA\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_Cited;E_TOPMed;E_gnomAD;MA=A;MAF=0.4804;MAC=2406;AA=C;AFR=0.1649;AMR=0.5706;EAS=0.2331;EUR=0.8867;SAS=0.68;GWAS_BETA=-0.129921;GWAS_SE=0.023253;GWAS_P=2.31e-08;GWAS_N=520580\r\n",
      "1\t38345021\trs4585948\tT\tA,C\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_TOPMed;E_gnomAD;MA=C;MAF=0.2869;MAC=1437;AA=T;AFR=0,0.0779;AMR=0,0.2968;EAS=0,0.2619;EUR=0,0.4433;SAS=0,0.4274;GWAS_BETA=-0.082878;GWAS_SE=0.015083;GWAS_P=3.91e-08;GWAS_N=520580\r\n",
      "1\t38346396\trs9701812\tC\tT\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_TOPMed;E_gnomAD;MA=T;MAF=0.2869;MAC=1437;AA=C;AFR=0.0787;AMR=0.2968;EAS=0.2619;EUR=0.4433;SAS=0.4264;GWAS_BETA=-0.082924;GWAS_SE=0.015081;GWAS_P=3.82e-08;GWAS_N=520580\r\n",
      "1\t38347417\trs12742756\tA\tG\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_TOPMed;E_gnomAD;MA=G;MAF=0.2869;MAC=1437;AA=A;AFR=0.0787;AMR=0.2968;EAS=0.2619;EUR=0.4433;SAS=0.4264;GWAS_BETA=-0.083107;GWAS_SE=0.015077;GWAS_P=3.54e-08;GWAS_N=520580\r\n",
      "1\t38397341\trs34655914\tG\tA,C\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_ExAC;E_TOPMed;E_gnomAD;MA=A;MAF=0.355;MAC=1778;AA=G;AFR=0.084,0;AMR=0.4366,0;EAS=0.3185,0;EUR=0.5825,0;SAS=0.4673,0;GWAS_BETA=-0.083213;GWAS_SE=0.015137;GWAS_P=3.85e-08;GWAS_N=520580\r\n",
      "1\t38397369\trs35267671\tC\tA,T\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_ExAC;E_TOPMed;E_gnomAD;MA=T;MAF=0.3552;MAC=1779;AA=C;AFR=0,0.084;AMR=0,0.4366;EAS=0,0.3185;EUR=0,0.5835;SAS=0,0.4673;GWAS_BETA=-0.082736;GWAS_SE=0.015137;GWAS_P=4.61e-08;GWAS_N=520580\r\n",
      "1\t38398588\trs17465420\tT\tC,G\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_TOPMed;E_gnomAD;MA=C;MAF=0.3562;MAC=1784;AA=C;AFR=0.0877,0;AMR=0.4395,0;EAS=0.3185,0;EUR=0.5815,0;SAS=0.4673,0;GWAS_BETA=-0.082543;GWAS_SE=0.015127;GWAS_P=4.85e-08;GWAS_N=520580\r\n",
      "1\t38405257\trs28730330\tT\tC\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_TOPMed;E_gnomAD;MA=C;MAF=0.3572;MAC=1789;AFR=0.0908;AMR=0.4395;EAS=0.3175;EUR=0.5815;SAS=0.4693;GWAS_BETA=-0.082805;GWAS_SE=0.015135;GWAS_P=4.48e-08;GWAS_N=520580\r\n",
      "1\t38405358\trs28626364\tA\tG,T\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_TOPMed;E_gnomAD;MA=G;MAF=0.357;MAC=1788;AFR=0.0915,0;AMR=0.4395,0;EAS=0.3175,0;EUR=0.5815,0;SAS=0.4673,0;GWAS_BETA=-0.08297;GWAS_SE=0.015137;GWAS_P=4.22e-08;GWAS_N=520580\r\n",
      "1\t38408868\trs28645205\tC\tG\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_TOPMed;E_gnomAD;MA=G;MAF=0.3588;MAC=1797;AA=C;AFR=0.0991;AMR=0.4395;EAS=0.3175;EUR=0.5815;SAS=0.4663;GWAS_BETA=-0.082685;GWAS_SE=0.015144;GWAS_P=4.76e-08;GWAS_N=520580\r\n",
      "1\t64106817\trs2301055\tT\tC,G\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_TOPMed;E_gnomAD;MA=C;MAF=0.2035;MAC=1019;AA=T;AFR=0.0976,0;AMR=0.3948,0;EAS=0.2093,0;EUR=0.2336,0;SAS=0.1738,0;GWAS_BETA=0.103201;GWAS_SE=0.017724;GWAS_P=5.79e-09;GWAS_N=520580\r\n",
      "1\t64107028\trs2301054\tA\tC,G\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_TOPMed;E_gnomAD;MA=G;MAF=0.2214;MAC=1109;AA=A;AFR=0,0.1589;AMR=0,0.3991;EAS=0,0.2113;EUR=0,0.2346;SAS=0,0.1769;GWAS_BETA=0.100041;GWAS_SE=0.017802;GWAS_P=1.91e-08;GWAS_N=520580\r\n",
      "1\t64107284\trs2269247\tC\tG,T\t.\t.\tdbSNP_154;TSA=SNV;E_Freq;E_1000G;E_Cited;E_Phenotype_or_Disease;E_TOPMed;MA=T;MAF=0.2216;MAC=1110;AA=C;AFR=0,0.1936;AMR=0,0.389;EAS=0,0.2093;EUR=0,0.2058;SAS=0,0.1697;GWAS_BETA=0.106889;GWAS_SE=0.018154;GWAS_P=3.91e-09;GWAS_N=520580\r\n",
      "\r\n",
      "gzip: stdout: Broken pipe\r\n"
     ]
    }
   ],
   "source": [
    "!zcat results/main/hubs/gwas_hub/T1D_34012112_Gaulton.GWAS_input_colocalization_pval_lt_5eMinus8.onekg.anno.vcf.gz | head -n 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0099821",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "HiChIP-DB",
   "language": "python",
   "name": "hichip-db"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
