{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "153750ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import requests\n",
    "\n",
    "os.chdir('/mnt/BioHome/jreyna/jreyna/projects/dchallenge/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3f7e2ced",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_meta = 'results/refs/gwas_catalog/gwas-catalog-v1.0.3-studies-r2022-02-02.tsv'\n",
    "gwas_meta = pd.read_table(gwas_meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3088d4e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATE ADDED TO CATALOG</th>\n",
       "      <th>PUBMED ID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>LINK</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>...</th>\n",
       "      <th>MAPPED_TRAIT</th>\n",
       "      <th>MAPPED_TRAIT_URI</th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>GENOTYPING TECHNOLOGY</th>\n",
       "      <th>SUMMARY STATS LOCATION</th>\n",
       "      <th>SUBMISSION DATE</th>\n",
       "      <th>STATISTICAL MODEL</th>\n",
       "      <th>BACKGROUND TRAIT</th>\n",
       "      <th>MAPPED BACKGROUND TRAIT</th>\n",
       "      <th>MAPPED BACKGROUND TRAIT URI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2017-08-04</td>\n",
       "      <td>27989323</td>\n",
       "      <td>Ahola-Olli AV</td>\n",
       "      <td>2016-12-13</td>\n",
       "      <td>Am J Hum Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/27989323</td>\n",
       "      <td>Genome-wide Association Study Identifies 27 Lo...</td>\n",
       "      <td>beta-nerve growth factor levels</td>\n",
       "      <td>3,531 Finnish ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>beta-nerve growth factor measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0008035</td>\n",
       "      <td>GCST004421</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2017-08-04</td>\n",
       "      <td>27989323</td>\n",
       "      <td>Ahola-Olli AV</td>\n",
       "      <td>2016-12-13</td>\n",
       "      <td>Am J Hum Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/27989323</td>\n",
       "      <td>Genome-wide Association Study Identifies 27 Lo...</td>\n",
       "      <td>Vascular endothelial growth factor levels</td>\n",
       "      <td>7,118 Finnish ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>vascular endothelial growth factor measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004762</td>\n",
       "      <td>GCST004422</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2017-08-04</td>\n",
       "      <td>27989323</td>\n",
       "      <td>Ahola-Olli AV</td>\n",
       "      <td>2016-12-13</td>\n",
       "      <td>Am J Hum Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/27989323</td>\n",
       "      <td>Genome-wide Association Study Identifies 27 Lo...</td>\n",
       "      <td>CTACK levels</td>\n",
       "      <td>3,631 Finnish ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>chemokine (C-C motif) ligand 27 measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0008082</td>\n",
       "      <td>GCST004420</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2017-08-04</td>\n",
       "      <td>27989323</td>\n",
       "      <td>Ahola-Olli AV</td>\n",
       "      <td>2016-12-13</td>\n",
       "      <td>Am J Hum Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/27989323</td>\n",
       "      <td>Genome-wide Association Study Identifies 27 Lo...</td>\n",
       "      <td>Eotaxin levels</td>\n",
       "      <td>8,153 Finnish ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>eotaxin measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0008122</td>\n",
       "      <td>GCST004460</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2017-08-04</td>\n",
       "      <td>27989323</td>\n",
       "      <td>Ahola-Olli AV</td>\n",
       "      <td>2016-12-13</td>\n",
       "      <td>Am J Hum Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/27989323</td>\n",
       "      <td>Genome-wide Association Study Identifies 27 Lo...</td>\n",
       "      <td>Fibroblast growth factor basic levels</td>\n",
       "      <td>7,565 Finnish ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>fibroblast growth factor basic measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0008130</td>\n",
       "      <td>GCST004459</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATE ADDED TO CATALOG  PUBMED ID   FIRST AUTHOR        DATE         JOURNAL  \\\n",
       "0            2017-08-04   27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
       "1            2017-08-04   27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
       "2            2017-08-04   27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
       "3            2017-08-04   27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
       "4            2017-08-04   27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
       "\n",
       "                                   LINK  \\\n",
       "0  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
       "1  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
       "2  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
       "3  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
       "4  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
       "\n",
       "                                               STUDY  \\\n",
       "0  Genome-wide Association Study Identifies 27 Lo...   \n",
       "1  Genome-wide Association Study Identifies 27 Lo...   \n",
       "2  Genome-wide Association Study Identifies 27 Lo...   \n",
       "3  Genome-wide Association Study Identifies 27 Lo...   \n",
       "4  Genome-wide Association Study Identifies 27 Lo...   \n",
       "\n",
       "                               DISEASE/TRAIT  \\\n",
       "0            beta-nerve growth factor levels   \n",
       "1  Vascular endothelial growth factor levels   \n",
       "2                               CTACK levels   \n",
       "3                             Eotaxin levels   \n",
       "4      Fibroblast growth factor basic levels   \n",
       "\n",
       "                  INITIAL SAMPLE SIZE REPLICATION SAMPLE SIZE  ...  \\\n",
       "0  3,531 Finnish ancestry individuals                     NaN  ...   \n",
       "1  7,118 Finnish ancestry individuals                     NaN  ...   \n",
       "2  3,631 Finnish ancestry individuals                     NaN  ...   \n",
       "3  8,153 Finnish ancestry individuals                     NaN  ...   \n",
       "4  7,565 Finnish ancestry individuals                     NaN  ...   \n",
       "\n",
       "                                     MAPPED_TRAIT  \\\n",
       "0            beta-nerve growth factor measurement   \n",
       "1  vascular endothelial growth factor measurement   \n",
       "2     chemokine (C-C motif) ligand 27 measurement   \n",
       "3                             eotaxin measurement   \n",
       "4      fibroblast growth factor basic measurement   \n",
       "\n",
       "                       MAPPED_TRAIT_URI STUDY ACCESSION  \\\n",
       "0  http://www.ebi.ac.uk/efo/EFO_0008035      GCST004421   \n",
       "1  http://www.ebi.ac.uk/efo/EFO_0004762      GCST004422   \n",
       "2  http://www.ebi.ac.uk/efo/EFO_0008082      GCST004420   \n",
       "3  http://www.ebi.ac.uk/efo/EFO_0008122      GCST004460   \n",
       "4  http://www.ebi.ac.uk/efo/EFO_0008130      GCST004459   \n",
       "\n",
       "          GENOTYPING TECHNOLOGY SUMMARY STATS LOCATION SUBMISSION DATE  \\\n",
       "0  Genome-wide genotyping array                    NaN             NaN   \n",
       "1  Genome-wide genotyping array                    NaN             NaN   \n",
       "2  Genome-wide genotyping array                    NaN             NaN   \n",
       "3  Genome-wide genotyping array                    NaN             NaN   \n",
       "4  Genome-wide genotyping array                    NaN             NaN   \n",
       "\n",
       "   STATISTICAL MODEL  BACKGROUND TRAIT  MAPPED BACKGROUND TRAIT  \\\n",
       "0                NaN               NaN                      NaN   \n",
       "1                NaN               NaN                      NaN   \n",
       "2                NaN               NaN                      NaN   \n",
       "3                NaN               NaN                      NaN   \n",
       "4                NaN               NaN                      NaN   \n",
       "\n",
       "   MAPPED BACKGROUND TRAIT URI  \n",
       "0                          NaN  \n",
       "1                          NaN  \n",
       "2                          NaN  \n",
       "3                          NaN  \n",
       "4                          NaN  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a42567e",
   "metadata": {},
   "source": [
    "## Getting all the OLS Link ID/Disease Information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "ebf425c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "disease_info = []\n",
    "for i in range(0, 37):\n",
    "    \n",
    "    split_fn = 'results/notebooks/gwas_catalog_disease_traits/processed/split.ols_id.{}.txt'\n",
    "    split_fn = split_fn.format(i)\n",
    "    split_info = open(split_fn).readlines()\n",
    "    split_info = [x.strip().split() for x in split_info]\n",
    "    disease_info.extend(split_info)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "742596cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "disease_index_dict = {k:v for k,v in disease_info}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "3869238b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3687"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(disease_index_dict.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a29bca46",
   "metadata": {},
   "source": [
    "## Add the final disease_or_disorder column to the gwas_meta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "9796d33c",
   "metadata": {},
   "outputs": [],
   "source": [
    "dd_list = []\n",
    "for i, uris in enumerate(gwas_meta['MAPPED_TRAIT_URI']):\n",
    "    \n",
    "    # default dd will be other\n",
    "    dd_list.append('other')\n",
    "    \n",
    "    # assign disease or disorder when possible\n",
    "    if type(uris) == str:\n",
    "        for uri in uris.split(','):\n",
    "            uri = uri.strip()\n",
    "            \n",
    "            if uri in disease_index_dict:\n",
    "                \n",
    "                if disease_index_dict[uri] == 'disease':\n",
    "                    dd_list[i] = ('disease')\n",
    "                    break\n",
    "\n",
    "                elif disease_index_dict[uri] == 'disorder':\n",
    "                    dd_list[i] = ('disorder')\n",
    "\n",
    "\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "00ce5d0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_meta['disease_or_disorder'] = dd_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "0d23b85c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATE ADDED TO CATALOG</th>\n",
       "      <th>PUBMED ID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>LINK</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>...</th>\n",
       "      <th>MAPPED_TRAIT_URI</th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>GENOTYPING TECHNOLOGY</th>\n",
       "      <th>SUMMARY STATS LOCATION</th>\n",
       "      <th>SUBMISSION DATE</th>\n",
       "      <th>STATISTICAL MODEL</th>\n",
       "      <th>BACKGROUND TRAIT</th>\n",
       "      <th>MAPPED BACKGROUND TRAIT</th>\n",
       "      <th>MAPPED BACKGROUND TRAIT URI</th>\n",
       "      <th>disease_or_disorder</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>2017-08-17</td>\n",
       "      <td>28548082</td>\n",
       "      <td>Southam L</td>\n",
       "      <td>2017-05-26</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/28548082</td>\n",
       "      <td>Whole genome sequencing and imputation in isol...</td>\n",
       "      <td>C-reactive protein levels</td>\n",
       "      <td>up to 1,476 Mylopotamos (founder/genetic isola...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004458</td>\n",
       "      <td>GCST004530</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>2017-08-17</td>\n",
       "      <td>28548082</td>\n",
       "      <td>Southam L</td>\n",
       "      <td>2017-05-26</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/28548082</td>\n",
       "      <td>Whole genome sequencing and imputation in isol...</td>\n",
       "      <td>Hematocrit</td>\n",
       "      <td>up to 1,476 Mylopotamos (founder/genetic isola...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004348</td>\n",
       "      <td>GCST004535</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>2017-08-18</td>\n",
       "      <td>28548082</td>\n",
       "      <td>Southam L</td>\n",
       "      <td>2017-05-26</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/28548082</td>\n",
       "      <td>Whole genome sequencing and imputation in isol...</td>\n",
       "      <td>Waist-hip ratio</td>\n",
       "      <td>up to 1,476 Mylopotamos (founder/genetic isola...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004343</td>\n",
       "      <td>GCST004555</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>2017-08-17</td>\n",
       "      <td>28548082</td>\n",
       "      <td>Southam L</td>\n",
       "      <td>2017-05-26</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/28548082</td>\n",
       "      <td>Whole genome sequencing and imputation in isol...</td>\n",
       "      <td>Low density lipoprotein cholesterol levels</td>\n",
       "      <td>up to 1,476 Mylopotamos (founder/genetic isola...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004611</td>\n",
       "      <td>GCST004541</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>2017-08-18</td>\n",
       "      <td>28548082</td>\n",
       "      <td>Southam L</td>\n",
       "      <td>2017-05-26</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/28548082</td>\n",
       "      <td>Whole genome sequencing and imputation in isol...</td>\n",
       "      <td>Triglyceride levels</td>\n",
       "      <td>up to 1,476 Mylopotamos (founder/genetic isola...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004530</td>\n",
       "      <td>GCST004550</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33094</th>\n",
       "      <td>2019-06-21</td>\n",
       "      <td>31015401</td>\n",
       "      <td>Wu Y</td>\n",
       "      <td>2019-04-23</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/31015401</td>\n",
       "      <td>Genome-wide association study of medication-us...</td>\n",
       "      <td>Medication use (drugs for peptic ulcer and gas...</td>\n",
       "      <td>53,137 European ancestry cases, 79,230 Europea...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0009923</td>\n",
       "      <td>GCST007922</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33097</th>\n",
       "      <td>2019-06-21</td>\n",
       "      <td>31015401</td>\n",
       "      <td>Wu Y</td>\n",
       "      <td>2019-04-23</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/31015401</td>\n",
       "      <td>Genome-wide association study of medication-us...</td>\n",
       "      <td>Medication use (vasodilators used in cardiac d...</td>\n",
       "      <td>5,546 European ancestry cases, 237,113 Europea...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0009926</td>\n",
       "      <td>GCST007925</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33127</th>\n",
       "      <td>2020-07-28</td>\n",
       "      <td>32450446</td>\n",
       "      <td>Wang H</td>\n",
       "      <td>2020-04-29</td>\n",
       "      <td>Neurobiol Aging</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32450446</td>\n",
       "      <td>Genome-wide interaction analysis of pathologic...</td>\n",
       "      <td>Neuritic plaques (SNP x SNP interaction)</td>\n",
       "      <td>1,316 individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0006798</td>\n",
       "      <td>GCST010342</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33130</th>\n",
       "      <td>2020-07-28</td>\n",
       "      <td>32450446</td>\n",
       "      <td>Wang H</td>\n",
       "      <td>2020-04-29</td>\n",
       "      <td>Neurobiol Aging</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32450446</td>\n",
       "      <td>Genome-wide interaction analysis of pathologic...</td>\n",
       "      <td>Total amyloid (SNP x SNP interaction)</td>\n",
       "      <td>1,276 individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0005194</td>\n",
       "      <td>GCST010339</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33141</th>\n",
       "      <td>2020-09-18</td>\n",
       "      <td>32888494</td>\n",
       "      <td>Vuckovic D</td>\n",
       "      <td>2020-09-01</td>\n",
       "      <td>Cell</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32888494</td>\n",
       "      <td>The Polygenic and Monogenic Basis of Blood Tra...</td>\n",
       "      <td>Hematocrit</td>\n",
       "      <td>408,112 British individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004348</td>\n",
       "      <td>GCST90002383</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>disease</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10031 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      DATE ADDED TO CATALOG  PUBMED ID FIRST AUTHOR        DATE  \\\n",
       "26               2017-08-17   28548082    Southam L  2017-05-26   \n",
       "33               2017-08-17   28548082    Southam L  2017-05-26   \n",
       "38               2017-08-18   28548082    Southam L  2017-05-26   \n",
       "43               2017-08-17   28548082    Southam L  2017-05-26   \n",
       "49               2017-08-18   28548082    Southam L  2017-05-26   \n",
       "...                     ...        ...          ...         ...   \n",
       "33094            2019-06-21   31015401         Wu Y  2019-04-23   \n",
       "33097            2019-06-21   31015401         Wu Y  2019-04-23   \n",
       "33127            2020-07-28   32450446       Wang H  2020-04-29   \n",
       "33130            2020-07-28   32450446       Wang H  2020-04-29   \n",
       "33141            2020-09-18   32888494   Vuckovic D  2020-09-01   \n",
       "\n",
       "               JOURNAL                                  LINK  \\\n",
       "26          Nat Commun  www.ncbi.nlm.nih.gov/pubmed/28548082   \n",
       "33          Nat Commun  www.ncbi.nlm.nih.gov/pubmed/28548082   \n",
       "38          Nat Commun  www.ncbi.nlm.nih.gov/pubmed/28548082   \n",
       "43          Nat Commun  www.ncbi.nlm.nih.gov/pubmed/28548082   \n",
       "49          Nat Commun  www.ncbi.nlm.nih.gov/pubmed/28548082   \n",
       "...                ...                                   ...   \n",
       "33094       Nat Commun  www.ncbi.nlm.nih.gov/pubmed/31015401   \n",
       "33097       Nat Commun  www.ncbi.nlm.nih.gov/pubmed/31015401   \n",
       "33127  Neurobiol Aging  www.ncbi.nlm.nih.gov/pubmed/32450446   \n",
       "33130  Neurobiol Aging  www.ncbi.nlm.nih.gov/pubmed/32450446   \n",
       "33141             Cell  www.ncbi.nlm.nih.gov/pubmed/32888494   \n",
       "\n",
       "                                                   STUDY  \\\n",
       "26     Whole genome sequencing and imputation in isol...   \n",
       "33     Whole genome sequencing and imputation in isol...   \n",
       "38     Whole genome sequencing and imputation in isol...   \n",
       "43     Whole genome sequencing and imputation in isol...   \n",
       "49     Whole genome sequencing and imputation in isol...   \n",
       "...                                                  ...   \n",
       "33094  Genome-wide association study of medication-us...   \n",
       "33097  Genome-wide association study of medication-us...   \n",
       "33127  Genome-wide interaction analysis of pathologic...   \n",
       "33130  Genome-wide interaction analysis of pathologic...   \n",
       "33141  The Polygenic and Monogenic Basis of Blood Tra...   \n",
       "\n",
       "                                           DISEASE/TRAIT  \\\n",
       "26                             C-reactive protein levels   \n",
       "33                                            Hematocrit   \n",
       "38                                       Waist-hip ratio   \n",
       "43            Low density lipoprotein cholesterol levels   \n",
       "49                                   Triglyceride levels   \n",
       "...                                                  ...   \n",
       "33094  Medication use (drugs for peptic ulcer and gas...   \n",
       "33097  Medication use (vasodilators used in cardiac d...   \n",
       "33127           Neuritic plaques (SNP x SNP interaction)   \n",
       "33130              Total amyloid (SNP x SNP interaction)   \n",
       "33141                                         Hematocrit   \n",
       "\n",
       "                                     INITIAL SAMPLE SIZE  \\\n",
       "26     up to 1,476 Mylopotamos (founder/genetic isola...   \n",
       "33     up to 1,476 Mylopotamos (founder/genetic isola...   \n",
       "38     up to 1,476 Mylopotamos (founder/genetic isola...   \n",
       "43     up to 1,476 Mylopotamos (founder/genetic isola...   \n",
       "49     up to 1,476 Mylopotamos (founder/genetic isola...   \n",
       "...                                                  ...   \n",
       "33094  53,137 European ancestry cases, 79,230 Europea...   \n",
       "33097  5,546 European ancestry cases, 237,113 Europea...   \n",
       "33127                                  1,316 individuals   \n",
       "33130                                  1,276 individuals   \n",
       "33141                        408,112 British individuals   \n",
       "\n",
       "      REPLICATION SAMPLE SIZE  ...                      MAPPED_TRAIT_URI  \\\n",
       "26                        NaN  ...  http://www.ebi.ac.uk/efo/EFO_0004458   \n",
       "33                        NaN  ...  http://www.ebi.ac.uk/efo/EFO_0004348   \n",
       "38                        NaN  ...  http://www.ebi.ac.uk/efo/EFO_0004343   \n",
       "43                        NaN  ...  http://www.ebi.ac.uk/efo/EFO_0004611   \n",
       "49                        NaN  ...  http://www.ebi.ac.uk/efo/EFO_0004530   \n",
       "...                       ...  ...                                   ...   \n",
       "33094                     NaN  ...  http://www.ebi.ac.uk/efo/EFO_0009923   \n",
       "33097                     NaN  ...  http://www.ebi.ac.uk/efo/EFO_0009926   \n",
       "33127                     NaN  ...  http://www.ebi.ac.uk/efo/EFO_0006798   \n",
       "33130                     NaN  ...  http://www.ebi.ac.uk/efo/EFO_0005194   \n",
       "33141                     NaN  ...  http://www.ebi.ac.uk/efo/EFO_0004348   \n",
       "\n",
       "       STUDY ACCESSION         GENOTYPING TECHNOLOGY SUMMARY STATS LOCATION  \\\n",
       "26          GCST004530  Genome-wide genotyping array                    NaN   \n",
       "33          GCST004535  Genome-wide genotyping array                    NaN   \n",
       "38          GCST004555  Genome-wide genotyping array                    NaN   \n",
       "43          GCST004541  Genome-wide genotyping array                    NaN   \n",
       "49          GCST004550  Genome-wide genotyping array                    NaN   \n",
       "...                ...                           ...                    ...   \n",
       "33094       GCST007922  Genome-wide genotyping array                    NaN   \n",
       "33097       GCST007925  Genome-wide genotyping array                    NaN   \n",
       "33127       GCST010342  Genome-wide genotyping array                    NaN   \n",
       "33130       GCST010339  Genome-wide genotyping array                    NaN   \n",
       "33141     GCST90002383  Genome-wide genotyping array                    NaN   \n",
       "\n",
       "      SUBMISSION DATE STATISTICAL MODEL  BACKGROUND TRAIT  \\\n",
       "26                NaN               NaN               NaN   \n",
       "33                NaN               NaN               NaN   \n",
       "38                NaN               NaN               NaN   \n",
       "43                NaN               NaN               NaN   \n",
       "49                NaN               NaN               NaN   \n",
       "...               ...               ...               ...   \n",
       "33094             NaN               NaN               NaN   \n",
       "33097             NaN               NaN               NaN   \n",
       "33127             NaN               NaN               NaN   \n",
       "33130             NaN               NaN               NaN   \n",
       "33141             NaN               NaN               NaN   \n",
       "\n",
       "       MAPPED BACKGROUND TRAIT  MAPPED BACKGROUND TRAIT URI  \\\n",
       "26                         NaN                          NaN   \n",
       "33                         NaN                          NaN   \n",
       "38                         NaN                          NaN   \n",
       "43                         NaN                          NaN   \n",
       "49                         NaN                          NaN   \n",
       "...                        ...                          ...   \n",
       "33094                      NaN                          NaN   \n",
       "33097                      NaN                          NaN   \n",
       "33127                      NaN                          NaN   \n",
       "33130                      NaN                          NaN   \n",
       "33141                      NaN                          NaN   \n",
       "\n",
       "       disease_or_disorder  \n",
       "26                 disease  \n",
       "33                 disease  \n",
       "38                 disease  \n",
       "43                 disease  \n",
       "49                 disease  \n",
       "...                    ...  \n",
       "33094              disease  \n",
       "33097              disease  \n",
       "33127              disease  \n",
       "33130              disease  \n",
       "33141              disease  \n",
       "\n",
       "[10031 rows x 23 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_meta[gwas_meta.disease_or_disorder.isin(['disease', 'disorder'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "fd49575f",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_gwas_fn = 'results/refs/gwas_catalog/gwas-catalog-v1.0.3-studies-r2022-02-02.withDDColumn.tsv'\n",
    "gwas_meta.to_csv(new_gwas_fn, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70c1430d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "HiChIP DB",
   "language": "python",
   "name": "hichip-db"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
